<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>8.3 Gene expression analysis using high-throughput sequencing technologies | Computational Genomics with R</title>
  <meta name="description" content="A guide to computationa genomics using R. The book covers fundemental topics with practical examples for an interdisciplinery audience" />
  <meta name="generator" content="bookdown 0.19 and GitBook 2.6.7" />

  <meta property="og:title" content="8.3 Gene expression analysis using high-throughput sequencing technologies | Computational Genomics with R" />
  <meta property="og:type" content="book" />
  <meta property="og:url" content="https://compmgenomr.github.io/book/" />
  <meta property="og:image" content="https://compmgenomr.github.io/book/images/cover.jpg" />
  <meta property="og:description" content="A guide to computationa genomics using R. The book covers fundemental topics with practical examples for an interdisciplinery audience" />
  <meta name="github-repo" content="compgenomr/book" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="8.3 Gene expression analysis using high-throughput sequencing technologies | Computational Genomics with R" />
  
  <meta name="twitter:description" content="A guide to computationa genomics using R. The book covers fundemental topics with practical examples for an interdisciplinery audience" />
  <meta name="twitter:image" content="https://compmgenomr.github.io/book/images/cover.jpg" />

<meta name="author" content="Altuna Akalin" />


<meta name="date" content="2020-09-30" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="methods-to-detect-gene-expression.html"/>
<link rel="next" href="other-applications-of-rna-seq.html"/>
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />









<!-- Global site tag (gtag.js) - Google Analytics -->
<script async src="https://www.googletagmanager.com/gtag/js?id=UA-83786243-1"></script>
<script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-83786243-1');
</script>


<style type="text/css">
code.sourceCode > span { display: inline-block; line-height: 1.25; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<link rel="stylesheet" href="css/style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">Computational Genomics with R</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Preface</a><ul>
<li class="chapter" data-level="" data-path="who-is-this-book-for.html"><a href="who-is-this-book-for.html"><i class="fa fa-check"></i>Who is this book for?</a><ul>
<li class="chapter" data-level="" data-path="who-is-this-book-for.html"><a href="who-is-this-book-for.html#what-will-you-get-out-of-this"><i class="fa fa-check"></i>What will you get out of this?</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="structure-of-the-book.html"><a href="structure-of-the-book.html"><i class="fa fa-check"></i>Structure of the book</a></li>
<li class="chapter" data-level="" data-path="software-information-and-conventions.html"><a href="software-information-and-conventions.html"><i class="fa fa-check"></i>Software information and conventions</a><ul>
<li class="chapter" data-level="" data-path="software-information-and-conventions.html"><a href="software-information-and-conventions.html#assignment-operator-convention"><i class="fa fa-check"></i>Assignment operator convention</a></li>
<li class="chapter" data-level="" data-path="software-information-and-conventions.html"><a href="software-information-and-conventions.html#packages-needed-to-run-the-book-code"><i class="fa fa-check"></i>Packages needed to run the book code</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="data-for-the-book.html"><a href="data-for-the-book.html"><i class="fa fa-check"></i>Data for the book</a></li>
<li class="chapter" data-level="" data-path="exercises-in-the-book.html"><a href="exercises-in-the-book.html"><i class="fa fa-check"></i>Exercises in the book</a></li>
<li class="chapter" data-level="" data-path="reproducibility-statement.html"><a href="reproducibility-statement.html"><i class="fa fa-check"></i>Reproducibility statement</a></li>
<li class="chapter" data-level="" data-path="acknowledgements.html"><a href="acknowledgements.html"><i class="fa fa-check"></i>Acknowledgements</a></li>
<li class="chapter" data-level="" data-path="how-to-contribute.html"><a href="how-to-contribute.html"><i class="fa fa-check"></i>How to contribute</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="about-the-authors.html"><a href="about-the-authors.html"><i class="fa fa-check"></i>About the Authors</a></li>
<li class="chapter" data-level="1" data-path="intro.html"><a href="intro.html"><i class="fa fa-check"></i><b>1</b> Introduction to Genomics</a><ul>
<li class="chapter" data-level="1.1" data-path="genes-dna-and-central-dogma.html"><a href="genes-dna-and-central-dogma.html"><i class="fa fa-check"></i><b>1.1</b> Genes, DNA and central dogma</a><ul>
<li class="chapter" data-level="1.1.1" data-path="genes-dna-and-central-dogma.html"><a href="genes-dna-and-central-dogma.html#what-is-a-genome"><i class="fa fa-check"></i><b>1.1.1</b> What is a genome?</a></li>
<li class="chapter" data-level="1.1.2" data-path="genes-dna-and-central-dogma.html"><a href="genes-dna-and-central-dogma.html#what-is-a-gene"><i class="fa fa-check"></i><b>1.1.2</b> What is a gene?</a></li>
<li class="chapter" data-level="1.1.3" data-path="genes-dna-and-central-dogma.html"><a href="genes-dna-and-central-dogma.html#how-are-genes-controlled-transcriptional-and-post-transcriptional-regulation"><i class="fa fa-check"></i><b>1.1.3</b> How are genes controlled? Transcriptional and post-transcriptional regulation</a></li>
<li class="chapter" data-level="1.1.4" data-path="genes-dna-and-central-dogma.html"><a href="genes-dna-and-central-dogma.html#what-does-a-gene-look-like"><i class="fa fa-check"></i><b>1.1.4</b> What does a gene look like?</a></li>
</ul></li>
<li class="chapter" data-level="1.2" data-path="elements-of-gene-regulation.html"><a href="elements-of-gene-regulation.html"><i class="fa fa-check"></i><b>1.2</b> Elements of gene regulation</a><ul>
<li class="chapter" data-level="1.2.1" data-path="elements-of-gene-regulation.html"><a href="elements-of-gene-regulation.html#transcriptional-regulation"><i class="fa fa-check"></i><b>1.2.1</b> Transcriptional regulation</a></li>
<li class="chapter" data-level="1.2.2" data-path="elements-of-gene-regulation.html"><a href="elements-of-gene-regulation.html#post-transcriptional-regulation"><i class="fa fa-check"></i><b>1.2.2</b> Post-transcriptional regulation</a></li>
</ul></li>
<li class="chapter" data-level="1.3" data-path="shaping-the-genome-dna-mutation.html"><a href="shaping-the-genome-dna-mutation.html"><i class="fa fa-check"></i><b>1.3</b> Shaping the genome: DNA mutation</a></li>
<li class="chapter" data-level="1.4" data-path="high-throughput-experimental-methods-in-genomics.html"><a href="high-throughput-experimental-methods-in-genomics.html"><i class="fa fa-check"></i><b>1.4</b> High-throughput experimental methods in genomics</a><ul>
<li class="chapter" data-level="1.4.1" data-path="high-throughput-experimental-methods-in-genomics.html"><a href="high-throughput-experimental-methods-in-genomics.html#the-general-idea-behind-high-throughput-techniques"><i class="fa fa-check"></i><b>1.4.1</b> The general idea behind high-throughput techniques</a></li>
<li class="chapter" data-level="1.4.2" data-path="high-throughput-experimental-methods-in-genomics.html"><a href="high-throughput-experimental-methods-in-genomics.html#high-throughput-sequencing"><i class="fa fa-check"></i><b>1.4.2</b> High-throughput sequencing</a></li>
</ul></li>
<li class="chapter" data-level="1.5" data-path="visualization-and-data-repositories-for-genomics.html"><a href="visualization-and-data-repositories-for-genomics.html"><i class="fa fa-check"></i><b>1.5</b> Visualization and data repositories for genomics</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="Rintro.html"><a href="Rintro.html"><i class="fa fa-check"></i><b>2</b> Introduction to R for Genomic Data Analysis</a><ul>
<li class="chapter" data-level="2.1" data-path="steps-of-genomic-data-analysis.html"><a href="steps-of-genomic-data-analysis.html"><i class="fa fa-check"></i><b>2.1</b> Steps of (genomic) data analysis</a><ul>
<li class="chapter" data-level="2.1.1" data-path="steps-of-genomic-data-analysis.html"><a href="steps-of-genomic-data-analysis.html#data-collection"><i class="fa fa-check"></i><b>2.1.1</b> Data collection</a></li>
<li class="chapter" data-level="2.1.2" data-path="steps-of-genomic-data-analysis.html"><a href="steps-of-genomic-data-analysis.html#data-quality-check-and-cleaning"><i class="fa fa-check"></i><b>2.1.2</b> Data quality check and cleaning</a></li>
<li class="chapter" data-level="2.1.3" data-path="steps-of-genomic-data-analysis.html"><a href="steps-of-genomic-data-analysis.html#data-processing"><i class="fa fa-check"></i><b>2.1.3</b> Data processing</a></li>
<li class="chapter" data-level="2.1.4" data-path="steps-of-genomic-data-analysis.html"><a href="steps-of-genomic-data-analysis.html#exploratory-data-analysis-and-modeling"><i class="fa fa-check"></i><b>2.1.4</b> Exploratory data analysis and modeling</a></li>
<li class="chapter" data-level="2.1.5" data-path="steps-of-genomic-data-analysis.html"><a href="steps-of-genomic-data-analysis.html#visualization-and-reporting"><i class="fa fa-check"></i><b>2.1.5</b> Visualization and reporting</a></li>
<li class="chapter" data-level="2.1.6" data-path="steps-of-genomic-data-analysis.html"><a href="steps-of-genomic-data-analysis.html#why-use-r-for-genomics"><i class="fa fa-check"></i><b>2.1.6</b> Why use R for genomics ?</a></li>
</ul></li>
<li class="chapter" data-level="2.2" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html"><i class="fa fa-check"></i><b>2.2</b> Getting started with R</a><ul>
<li class="chapter" data-level="2.2.1" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#installing-packages"><i class="fa fa-check"></i><b>2.2.1</b> Installing packages</a></li>
<li class="chapter" data-level="2.2.2" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#installing-packages-in-custom-locations"><i class="fa fa-check"></i><b>2.2.2</b> Installing packages in custom locations</a></li>
<li class="chapter" data-level="2.2.3" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#getting-help-on-functions-and-packages"><i class="fa fa-check"></i><b>2.2.3</b> Getting help on functions and packages</a></li>
</ul></li>
<li class="chapter" data-level="2.3" data-path="computations-in-r.html"><a href="computations-in-r.html"><i class="fa fa-check"></i><b>2.3</b> Computations in R</a></li>
<li class="chapter" data-level="2.4" data-path="data-structures.html"><a href="data-structures.html"><i class="fa fa-check"></i><b>2.4</b> Data structures</a><ul>
<li class="chapter" data-level="2.4.1" data-path="data-structures.html"><a href="data-structures.html#vectors"><i class="fa fa-check"></i><b>2.4.1</b> Vectors</a></li>
<li class="chapter" data-level="2.4.2" data-path="data-structures.html"><a href="data-structures.html#matrices"><i class="fa fa-check"></i><b>2.4.2</b> Matrices</a></li>
<li class="chapter" data-level="2.4.3" data-path="data-structures.html"><a href="data-structures.html#data-frames"><i class="fa fa-check"></i><b>2.4.3</b> Data frames</a></li>
<li class="chapter" data-level="2.4.4" data-path="data-structures.html"><a href="data-structures.html#lists"><i class="fa fa-check"></i><b>2.4.4</b> Lists</a></li>
<li class="chapter" data-level="2.4.5" data-path="data-structures.html"><a href="data-structures.html#factors"><i class="fa fa-check"></i><b>2.4.5</b> Factors</a></li>
</ul></li>
<li class="chapter" data-level="2.5" data-path="data-types.html"><a href="data-types.html"><i class="fa fa-check"></i><b>2.5</b> Data types</a></li>
<li class="chapter" data-level="2.6" data-path="reading-and-writing-data.html"><a href="reading-and-writing-data.html"><i class="fa fa-check"></i><b>2.6</b> Reading and writing data</a><ul>
<li class="chapter" data-level="2.6.1" data-path="reading-and-writing-data.html"><a href="reading-and-writing-data.html#reading-large-files"><i class="fa fa-check"></i><b>2.6.1</b> Reading large files</a></li>
</ul></li>
<li class="chapter" data-level="2.7" data-path="plotting-in-r-with-base-graphics.html"><a href="plotting-in-r-with-base-graphics.html"><i class="fa fa-check"></i><b>2.7</b> Plotting in R with base graphics</a><ul>
<li class="chapter" data-level="2.7.1" data-path="plotting-in-r-with-base-graphics.html"><a href="plotting-in-r-with-base-graphics.html#combining-multiple-plots"><i class="fa fa-check"></i><b>2.7.1</b> Combining multiple plots</a></li>
<li class="chapter" data-level="2.7.2" data-path="plotting-in-r-with-base-graphics.html"><a href="plotting-in-r-with-base-graphics.html#saving-plots"><i class="fa fa-check"></i><b>2.7.2</b> Saving plots</a></li>
</ul></li>
<li class="chapter" data-level="2.8" data-path="plotting-in-r-with-ggplot2.html"><a href="plotting-in-r-with-ggplot2.html"><i class="fa fa-check"></i><b>2.8</b> Plotting in R with ggplot2</a><ul>
<li class="chapter" data-level="2.8.1" data-path="plotting-in-r-with-ggplot2.html"><a href="plotting-in-r-with-ggplot2.html#combining-multiple-plots-1"><i class="fa fa-check"></i><b>2.8.1</b> Combining multiple plots</a></li>
<li class="chapter" data-level="2.8.2" data-path="plotting-in-r-with-ggplot2.html"><a href="plotting-in-r-with-ggplot2.html#ggplot2-and-tidyverse"><i class="fa fa-check"></i><b>2.8.2</b> ggplot2 and tidyverse</a></li>
</ul></li>
<li class="chapter" data-level="2.9" data-path="functions-and-control-structures-for-ifelse-etc-.html"><a href="functions-and-control-structures-for-ifelse-etc-.html"><i class="fa fa-check"></i><b>2.9</b> Functions and control structures (for, if/else etc.)</a><ul>
<li class="chapter" data-level="2.9.1" data-path="functions-and-control-structures-for-ifelse-etc-.html"><a href="functions-and-control-structures-for-ifelse-etc-.html#user-defined-functions"><i class="fa fa-check"></i><b>2.9.1</b> User-defined functions</a></li>
<li class="chapter" data-level="2.9.2" data-path="functions-and-control-structures-for-ifelse-etc-.html"><a href="functions-and-control-structures-for-ifelse-etc-.html#loops-and-looping-structures-in-r"><i class="fa fa-check"></i><b>2.9.2</b> Loops and looping structures in R</a></li>
</ul></li>
<li class="chapter" data-level="2.10" data-path="exercises.html"><a href="exercises.html"><i class="fa fa-check"></i><b>2.10</b> Exercises</a><ul>
<li class="chapter" data-level="2.10.1" data-path="exercises.html"><a href="exercises.html#computations-in-r-1"><i class="fa fa-check"></i><b>2.10.1</b> Computations in R</a></li>
<li class="chapter" data-level="2.10.2" data-path="exercises.html"><a href="exercises.html#data-structures-in-r"><i class="fa fa-check"></i><b>2.10.2</b> Data structures in R</a></li>
<li class="chapter" data-level="2.10.3" data-path="exercises.html"><a href="exercises.html#reading-in-and-writing-data-out-in-r"><i class="fa fa-check"></i><b>2.10.3</b> Reading in and writing data out in R</a></li>
<li class="chapter" data-level="2.10.4" data-path="exercises.html"><a href="exercises.html#plotting-in-r"><i class="fa fa-check"></i><b>2.10.4</b> Plotting in R</a></li>
<li class="chapter" data-level="2.10.5" data-path="exercises.html"><a href="exercises.html#functions-and-control-structures-for-ifelse-etc.-1"><i class="fa fa-check"></i><b>2.10.5</b> Functions and control structures (for, if/else, etc.)</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="3" data-path="stats.html"><a href="stats.html"><i class="fa fa-check"></i><b>3</b> Statistics for Genomics</a><ul>
<li class="chapter" data-level="3.1" data-path="how-to-summarize-collection-of-data-points-the-idea-behind-statistical-distributions.html"><a href="how-to-summarize-collection-of-data-points-the-idea-behind-statistical-distributions.html"><i class="fa fa-check"></i><b>3.1</b> How to summarize collection of data points: The idea behind statistical distributions</a><ul>
<li class="chapter" data-level="3.1.1" data-path="how-to-summarize-collection-of-data-points-the-idea-behind-statistical-distributions.html"><a href="how-to-summarize-collection-of-data-points-the-idea-behind-statistical-distributions.html#describing-the-central-tendency-mean-and-median"><i class="fa fa-check"></i><b>3.1.1</b> Describing the central tendency: Mean and median</a></li>
<li class="chapter" data-level="3.1.2" data-path="how-to-summarize-collection-of-data-points-the-idea-behind-statistical-distributions.html"><a href="how-to-summarize-collection-of-data-points-the-idea-behind-statistical-distributions.html#describing-the-spread-measurements-of-variation"><i class="fa fa-check"></i><b>3.1.2</b> Describing the spread: Measurements of variation</a></li>
<li class="chapter" data-level="3.1.3" data-path="how-to-summarize-collection-of-data-points-the-idea-behind-statistical-distributions.html"><a href="how-to-summarize-collection-of-data-points-the-idea-behind-statistical-distributions.html#precision-of-estimates-confidence-intervals"><i class="fa fa-check"></i><b>3.1.3</b> Precision of estimates: Confidence intervals</a></li>
</ul></li>
<li class="chapter" data-level="3.2" data-path="how-to-test-for-differences-between-samples.html"><a href="how-to-test-for-differences-between-samples.html"><i class="fa fa-check"></i><b>3.2</b> How to test for differences between samples</a><ul>
<li class="chapter" data-level="3.2.1" data-path="how-to-test-for-differences-between-samples.html"><a href="how-to-test-for-differences-between-samples.html#randomization-based-testing-for-difference-of-the-means"><i class="fa fa-check"></i><b>3.2.1</b> Randomization-based testing for difference of the means</a></li>
<li class="chapter" data-level="3.2.2" data-path="how-to-test-for-differences-between-samples.html"><a href="how-to-test-for-differences-between-samples.html#using-t-test-for-difference-of-the-means-between-two-samples"><i class="fa fa-check"></i><b>3.2.2</b> Using t-test for difference of the means between two samples</a></li>
<li class="chapter" data-level="3.2.3" data-path="how-to-test-for-differences-between-samples.html"><a href="how-to-test-for-differences-between-samples.html#multiple-testing-correction"><i class="fa fa-check"></i><b>3.2.3</b> Multiple testing correction</a></li>
<li class="chapter" data-level="3.2.4" data-path="how-to-test-for-differences-between-samples.html"><a href="how-to-test-for-differences-between-samples.html#moderated-t-tests-using-information-from-multiple-comparisons"><i class="fa fa-check"></i><b>3.2.4</b> Moderated t-tests: Using information from multiple comparisons</a></li>
</ul></li>
<li class="chapter" data-level="3.3" data-path="relationship-between-variables-linear-models-and-correlation.html"><a href="relationship-between-variables-linear-models-and-correlation.html"><i class="fa fa-check"></i><b>3.3</b> Relationship between variables: Linear models and correlation</a><ul>
<li class="chapter" data-level="3.3.1" data-path="relationship-between-variables-linear-models-and-correlation.html"><a href="relationship-between-variables-linear-models-and-correlation.html#how-to-fit-a-line"><i class="fa fa-check"></i><b>3.3.1</b> How to fit a line</a></li>
<li class="chapter" data-level="3.3.2" data-path="relationship-between-variables-linear-models-and-correlation.html"><a href="relationship-between-variables-linear-models-and-correlation.html#how-to-estimate-the-error-of-the-coefficients"><i class="fa fa-check"></i><b>3.3.2</b> How to estimate the error of the coefficients</a></li>
<li class="chapter" data-level="3.3.3" data-path="relationship-between-variables-linear-models-and-correlation.html"><a href="relationship-between-variables-linear-models-and-correlation.html#accuracy-of-the-model"><i class="fa fa-check"></i><b>3.3.3</b> Accuracy of the model</a></li>
<li class="chapter" data-level="3.3.4" data-path="relationship-between-variables-linear-models-and-correlation.html"><a href="relationship-between-variables-linear-models-and-correlation.html#regression-with-categorical-variables"><i class="fa fa-check"></i><b>3.3.4</b> Regression with categorical variables</a></li>
<li class="chapter" data-level="3.3.5" data-path="relationship-between-variables-linear-models-and-correlation.html"><a href="relationship-between-variables-linear-models-and-correlation.html#regression-pitfalls"><i class="fa fa-check"></i><b>3.3.5</b> Regression pitfalls</a></li>
</ul></li>
<li class="chapter" data-level="3.4" data-path="exercises-1.html"><a href="exercises-1.html"><i class="fa fa-check"></i><b>3.4</b> Exercises</a><ul>
<li class="chapter" data-level="3.4.1" data-path="exercises-1.html"><a href="exercises-1.html#how-to-summarize-collection-of-data-points-the-idea-behind-statistical-distributions-1"><i class="fa fa-check"></i><b>3.4.1</b> How to summarize collection of data points: The idea behind statistical distributions</a></li>
<li class="chapter" data-level="3.4.2" data-path="exercises-1.html"><a href="exercises-1.html#how-to-test-for-differences-in-samples"><i class="fa fa-check"></i><b>3.4.2</b> How to test for differences in samples</a></li>
<li class="chapter" data-level="3.4.3" data-path="exercises-1.html"><a href="exercises-1.html#relationship-between-variables-linear-models-and-correlation-1"><i class="fa fa-check"></i><b>3.4.3</b> Relationship between variables: Linear models and correlation</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="4" data-path="unsupervisedLearning.html"><a href="unsupervisedLearning.html"><i class="fa fa-check"></i><b>4</b> Exploratory Data Analysis with Unsupervised Machine Learning</a><ul>
<li class="chapter" data-level="4.1" data-path="clustering-grouping-samples-based-on-their-similarity.html"><a href="clustering-grouping-samples-based-on-their-similarity.html"><i class="fa fa-check"></i><b>4.1</b> Clustering: Grouping samples based on their similarity</a><ul>
<li class="chapter" data-level="4.1.1" data-path="clustering-grouping-samples-based-on-their-similarity.html"><a href="clustering-grouping-samples-based-on-their-similarity.html#distance-metrics"><i class="fa fa-check"></i><b>4.1.1</b> Distance metrics</a></li>
<li class="chapter" data-level="4.1.2" data-path="clustering-grouping-samples-based-on-their-similarity.html"><a href="clustering-grouping-samples-based-on-their-similarity.html#hiearchical-clustering"><i class="fa fa-check"></i><b>4.1.2</b> Hiearchical clustering</a></li>
<li class="chapter" data-level="4.1.3" data-path="clustering-grouping-samples-based-on-their-similarity.html"><a href="clustering-grouping-samples-based-on-their-similarity.html#k-means-clustering"><i class="fa fa-check"></i><b>4.1.3</b> K-means clustering</a></li>
<li class="chapter" data-level="4.1.4" data-path="clustering-grouping-samples-based-on-their-similarity.html"><a href="clustering-grouping-samples-based-on-their-similarity.html#how-to-choose-k-the-number-of-clusters"><i class="fa fa-check"></i><b>4.1.4</b> How to choose “k”, the number of clusters</a></li>
</ul></li>
<li class="chapter" data-level="4.2" data-path="dimensionality-reduction-techniques-visualizing-complex-data-sets-in-2d.html"><a href="dimensionality-reduction-techniques-visualizing-complex-data-sets-in-2d.html"><i class="fa fa-check"></i><b>4.2</b> Dimensionality reduction techniques: Visualizing complex data sets in 2D</a><ul>
<li class="chapter" data-level="4.2.1" data-path="dimensionality-reduction-techniques-visualizing-complex-data-sets-in-2d.html"><a href="dimensionality-reduction-techniques-visualizing-complex-data-sets-in-2d.html#principal-component-analysis"><i class="fa fa-check"></i><b>4.2.1</b> Principal component analysis</a></li>
<li class="chapter" data-level="4.2.2" data-path="dimensionality-reduction-techniques-visualizing-complex-data-sets-in-2d.html"><a href="dimensionality-reduction-techniques-visualizing-complex-data-sets-in-2d.html#other-matrix-factorization-methods-for-dimensionality-reduction"><i class="fa fa-check"></i><b>4.2.2</b> Other matrix factorization methods for dimensionality reduction</a></li>
<li class="chapter" data-level="4.2.3" data-path="dimensionality-reduction-techniques-visualizing-complex-data-sets-in-2d.html"><a href="dimensionality-reduction-techniques-visualizing-complex-data-sets-in-2d.html#multi-dimensional-scaling"><i class="fa fa-check"></i><b>4.2.3</b> Multi-dimensional scaling</a></li>
<li class="chapter" data-level="4.2.4" data-path="dimensionality-reduction-techniques-visualizing-complex-data-sets-in-2d.html"><a href="dimensionality-reduction-techniques-visualizing-complex-data-sets-in-2d.html#t-distributed-stochastic-neighbor-embedding-t-sne"><i class="fa fa-check"></i><b>4.2.4</b> t-Distributed Stochastic Neighbor Embedding (t-SNE)</a></li>
</ul></li>
<li class="chapter" data-level="4.3" data-path="exercises-2.html"><a href="exercises-2.html"><i class="fa fa-check"></i><b>4.3</b> Exercises</a><ul>
<li class="chapter" data-level="4.3.1" data-path="exercises-2.html"><a href="exercises-2.html#clustering"><i class="fa fa-check"></i><b>4.3.1</b> Clustering</a></li>
<li class="chapter" data-level="4.3.2" data-path="exercises-2.html"><a href="exercises-2.html#dimension-reduction"><i class="fa fa-check"></i><b>4.3.2</b> Dimension reduction</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="5" data-path="supervisedLearning.html"><a href="supervisedLearning.html"><i class="fa fa-check"></i><b>5</b> Predictive Modeling with Supervised Machine Learning</a><ul>
<li class="chapter" data-level="5.1" data-path="how-are-machine-learning-models-fit.html"><a href="how-are-machine-learning-models-fit.html"><i class="fa fa-check"></i><b>5.1</b> How are machine learning models fit?</a><ul>
<li class="chapter" data-level="5.1.1" data-path="how-are-machine-learning-models-fit.html"><a href="how-are-machine-learning-models-fit.html#machine-learning-vs.-statistics"><i class="fa fa-check"></i><b>5.1.1</b> Machine learning vs. statistics</a></li>
</ul></li>
<li class="chapter" data-level="5.2" data-path="steps-in-supervised-machine-learning.html"><a href="steps-in-supervised-machine-learning.html"><i class="fa fa-check"></i><b>5.2</b> Steps in supervised machine learning</a></li>
<li class="chapter" data-level="5.3" data-path="use-case-disease-subtype-from-genomics-data.html"><a href="use-case-disease-subtype-from-genomics-data.html"><i class="fa fa-check"></i><b>5.3</b> Use case: Disease subtype from genomics data</a></li>
<li class="chapter" data-level="5.4" data-path="data-preprocessing.html"><a href="data-preprocessing.html"><i class="fa fa-check"></i><b>5.4</b> Data preprocessing</a><ul>
<li class="chapter" data-level="5.4.1" data-path="data-preprocessing.html"><a href="data-preprocessing.html#data-transformation"><i class="fa fa-check"></i><b>5.4.1</b> Data transformation</a></li>
<li class="chapter" data-level="5.4.2" data-path="data-preprocessing.html"><a href="data-preprocessing.html#filtering-data-and-scaling"><i class="fa fa-check"></i><b>5.4.2</b> Filtering data and scaling</a></li>
<li class="chapter" data-level="5.4.3" data-path="data-preprocessing.html"><a href="data-preprocessing.html#dealing-with-missing-values"><i class="fa fa-check"></i><b>5.4.3</b> Dealing with missing values</a></li>
</ul></li>
<li class="chapter" data-level="5.5" data-path="splitting-the-data.html"><a href="splitting-the-data.html"><i class="fa fa-check"></i><b>5.5</b> Splitting the data</a><ul>
<li class="chapter" data-level="5.5.1" data-path="splitting-the-data.html"><a href="splitting-the-data.html#holdout-test-dataset"><i class="fa fa-check"></i><b>5.5.1</b> Holdout test dataset</a></li>
<li class="chapter" data-level="5.5.2" data-path="splitting-the-data.html"><a href="splitting-the-data.html#cross-validation"><i class="fa fa-check"></i><b>5.5.2</b> Cross-validation</a></li>
<li class="chapter" data-level="5.5.3" data-path="splitting-the-data.html"><a href="splitting-the-data.html#bootstrap-resampling"><i class="fa fa-check"></i><b>5.5.3</b> Bootstrap resampling</a></li>
</ul></li>
<li class="chapter" data-level="5.6" data-path="predicting-the-subtype-with-k-nearest-neighbors.html"><a href="predicting-the-subtype-with-k-nearest-neighbors.html"><i class="fa fa-check"></i><b>5.6</b> Predicting the subtype with k-nearest neighbors</a></li>
<li class="chapter" data-level="5.7" data-path="assessing-the-performance-of-our-model.html"><a href="assessing-the-performance-of-our-model.html"><i class="fa fa-check"></i><b>5.7</b> Assessing the performance of our model</a><ul>
<li class="chapter" data-level="5.7.1" data-path="assessing-the-performance-of-our-model.html"><a href="assessing-the-performance-of-our-model.html#receiver-operating-characteristic-roc-curves"><i class="fa fa-check"></i><b>5.7.1</b> Receiver Operating Characteristic (ROC) curves</a></li>
</ul></li>
<li class="chapter" data-level="5.8" data-path="model-tuning-and-avoiding-overfitting.html"><a href="model-tuning-and-avoiding-overfitting.html"><i class="fa fa-check"></i><b>5.8</b> Model tuning and avoiding overfitting</a><ul>
<li class="chapter" data-level="5.8.1" data-path="model-tuning-and-avoiding-overfitting.html"><a href="model-tuning-and-avoiding-overfitting.html#model-complexity-and-bias-variance-trade-off"><i class="fa fa-check"></i><b>5.8.1</b> Model complexity and bias variance trade-off</a></li>
<li class="chapter" data-level="5.8.2" data-path="model-tuning-and-avoiding-overfitting.html"><a href="model-tuning-and-avoiding-overfitting.html#data-split-strategies-for-model-tuning-and-testing"><i class="fa fa-check"></i><b>5.8.2</b> Data split strategies for model tuning and testing</a></li>
</ul></li>
<li class="chapter" data-level="5.9" data-path="variable-importance.html"><a href="variable-importance.html"><i class="fa fa-check"></i><b>5.9</b> Variable importance</a></li>
<li class="chapter" data-level="5.10" data-path="how-to-deal-with-class-imbalance.html"><a href="how-to-deal-with-class-imbalance.html"><i class="fa fa-check"></i><b>5.10</b> How to deal with class imbalance</a><ul>
<li class="chapter" data-level="5.10.1" data-path="how-to-deal-with-class-imbalance.html"><a href="how-to-deal-with-class-imbalance.html#sampling-for-class-balance"><i class="fa fa-check"></i><b>5.10.1</b> Sampling for class balance</a></li>
<li class="chapter" data-level="5.10.2" data-path="how-to-deal-with-class-imbalance.html"><a href="how-to-deal-with-class-imbalance.html#altering-case-weights"><i class="fa fa-check"></i><b>5.10.2</b> Altering case weights</a></li>
<li class="chapter" data-level="5.10.3" data-path="how-to-deal-with-class-imbalance.html"><a href="how-to-deal-with-class-imbalance.html#selecting-different-classification-score-cutoffs"><i class="fa fa-check"></i><b>5.10.3</b> Selecting different classification score cutoffs</a></li>
</ul></li>
<li class="chapter" data-level="5.11" data-path="dealing-with-correlated-predictors.html"><a href="dealing-with-correlated-predictors.html"><i class="fa fa-check"></i><b>5.11</b> Dealing with correlated predictors</a></li>
<li class="chapter" data-level="5.12" data-path="trees-and-forests-random-forests-in-action.html"><a href="trees-and-forests-random-forests-in-action.html"><i class="fa fa-check"></i><b>5.12</b> Trees and forests: Random forests in action</a><ul>
<li class="chapter" data-level="5.12.1" data-path="trees-and-forests-random-forests-in-action.html"><a href="trees-and-forests-random-forests-in-action.html#decision-trees"><i class="fa fa-check"></i><b>5.12.1</b> Decision trees</a></li>
<li class="chapter" data-level="5.12.2" data-path="trees-and-forests-random-forests-in-action.html"><a href="trees-and-forests-random-forests-in-action.html#trees-to-forests"><i class="fa fa-check"></i><b>5.12.2</b> Trees to forests</a></li>
<li class="chapter" data-level="5.12.3" data-path="trees-and-forests-random-forests-in-action.html"><a href="trees-and-forests-random-forests-in-action.html#variable-importance-1"><i class="fa fa-check"></i><b>5.12.3</b> Variable importance</a></li>
</ul></li>
<li class="chapter" data-level="5.13" data-path="logistic-regression-and-regularization.html"><a href="logistic-regression-and-regularization.html"><i class="fa fa-check"></i><b>5.13</b> Logistic regression and regularization</a><ul>
<li class="chapter" data-level="5.13.1" data-path="logistic-regression-and-regularization.html"><a href="logistic-regression-and-regularization.html#regularization-in-order-to-avoid-overfitting"><i class="fa fa-check"></i><b>5.13.1</b> Regularization in order to avoid overfitting</a></li>
<li class="chapter" data-level="5.13.2" data-path="logistic-regression-and-regularization.html"><a href="logistic-regression-and-regularization.html#variable-importance-2"><i class="fa fa-check"></i><b>5.13.2</b> Variable importance</a></li>
</ul></li>
<li class="chapter" data-level="5.14" data-path="other-supervised-algorithms.html"><a href="other-supervised-algorithms.html"><i class="fa fa-check"></i><b>5.14</b> Other supervised algorithms</a><ul>
<li class="chapter" data-level="5.14.1" data-path="other-supervised-algorithms.html"><a href="other-supervised-algorithms.html#gradient-boosting"><i class="fa fa-check"></i><b>5.14.1</b> Gradient boosting</a></li>
<li class="chapter" data-level="5.14.2" data-path="other-supervised-algorithms.html"><a href="other-supervised-algorithms.html#support-vector-machines-svm"><i class="fa fa-check"></i><b>5.14.2</b> Support Vector Machines (SVM)</a></li>
<li class="chapter" data-level="5.14.3" data-path="other-supervised-algorithms.html"><a href="other-supervised-algorithms.html#neural-networks-and-deep-versions-of-it"><i class="fa fa-check"></i><b>5.14.3</b> Neural networks and deep versions of it</a></li>
<li class="chapter" data-level="5.14.4" data-path="other-supervised-algorithms.html"><a href="other-supervised-algorithms.html#ensemble-learning"><i class="fa fa-check"></i><b>5.14.4</b> Ensemble learning</a></li>
</ul></li>
<li class="chapter" data-level="5.15" data-path="predicting-continuous-variables-regression-with-machine-learning.html"><a href="predicting-continuous-variables-regression-with-machine-learning.html"><i class="fa fa-check"></i><b>5.15</b> Predicting continuous variables: Regression with machine learning</a><ul>
<li class="chapter" data-level="5.15.1" data-path="predicting-continuous-variables-regression-with-machine-learning.html"><a href="predicting-continuous-variables-regression-with-machine-learning.html#use-case-predicting-age-from-dna-methylation"><i class="fa fa-check"></i><b>5.15.1</b> Use case: Predicting age from DNA methylation</a></li>
<li class="chapter" data-level="5.15.2" data-path="predicting-continuous-variables-regression-with-machine-learning.html"><a href="predicting-continuous-variables-regression-with-machine-learning.html#reading-and-processing-the-data"><i class="fa fa-check"></i><b>5.15.2</b> Reading and processing the data</a></li>
<li class="chapter" data-level="5.15.3" data-path="predicting-continuous-variables-regression-with-machine-learning.html"><a href="predicting-continuous-variables-regression-with-machine-learning.html#running-random-forest-regression"><i class="fa fa-check"></i><b>5.15.3</b> Running random forest regression</a></li>
</ul></li>
<li class="chapter" data-level="5.16" data-path="exercises-3.html"><a href="exercises-3.html"><i class="fa fa-check"></i><b>5.16</b> Exercises</a><ul>
<li class="chapter" data-level="5.16.1" data-path="exercises-3.html"><a href="exercises-3.html#classification"><i class="fa fa-check"></i><b>5.16.1</b> Classification</a></li>
<li class="chapter" data-level="5.16.2" data-path="exercises-3.html"><a href="exercises-3.html#regression"><i class="fa fa-check"></i><b>5.16.2</b> Regression</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="6" data-path="genomicIntervals.html"><a href="genomicIntervals.html"><i class="fa fa-check"></i><b>6</b> Operations on Genomic Intervals and Genome Arithmetic</a><ul>
<li class="chapter" data-level="6.1" data-path="operations-on-genomic-intervals-with-genomicranges-package.html"><a href="operations-on-genomic-intervals-with-genomicranges-package.html"><i class="fa fa-check"></i><b>6.1</b> Operations on genomic intervals with <code>GenomicRanges</code> package</a><ul>
<li class="chapter" data-level="6.1.1" data-path="operations-on-genomic-intervals-with-genomicranges-package.html"><a href="operations-on-genomic-intervals-with-genomicranges-package.html#how-to-create-and-manipulate-a-granges-object"><i class="fa fa-check"></i><b>6.1.1</b> How to create and manipulate a GRanges object</a></li>
<li class="chapter" data-level="6.1.2" data-path="operations-on-genomic-intervals-with-genomicranges-package.html"><a href="operations-on-genomic-intervals-with-genomicranges-package.html#getting-genomic-regions-into-r-as-granges-objects"><i class="fa fa-check"></i><b>6.1.2</b> Getting genomic regions into R as GRanges objects</a></li>
<li class="chapter" data-level="6.1.3" data-path="operations-on-genomic-intervals-with-genomicranges-package.html"><a href="operations-on-genomic-intervals-with-genomicranges-package.html#finding-regions-that-dodo-not-overlap-with-another-set-of-regions"><i class="fa fa-check"></i><b>6.1.3</b> Finding regions that do/do not overlap with another set of regions</a></li>
</ul></li>
<li class="chapter" data-level="6.2" data-path="dealing-with-mapped-high-throughput-sequencing-reads.html"><a href="dealing-with-mapped-high-throughput-sequencing-reads.html"><i class="fa fa-check"></i><b>6.2</b> Dealing with mapped high-throughput sequencing reads</a><ul>
<li class="chapter" data-level="6.2.1" data-path="dealing-with-mapped-high-throughput-sequencing-reads.html"><a href="dealing-with-mapped-high-throughput-sequencing-reads.html#counting-mapped-reads-for-a-set-of-regions"><i class="fa fa-check"></i><b>6.2.1</b> Counting mapped reads for a set of regions</a></li>
</ul></li>
<li class="chapter" data-level="6.3" data-path="dealing-with-continuous-scores-over-the-genome.html"><a href="dealing-with-continuous-scores-over-the-genome.html"><i class="fa fa-check"></i><b>6.3</b> Dealing with continuous scores over the genome</a><ul>
<li class="chapter" data-level="6.3.1" data-path="dealing-with-continuous-scores-over-the-genome.html"><a href="dealing-with-continuous-scores-over-the-genome.html#extracting-subsections-of-rle-and-rlelist-objects"><i class="fa fa-check"></i><b>6.3.1</b> Extracting subsections of Rle and RleList objects</a></li>
</ul></li>
<li class="chapter" data-level="6.4" data-path="genomic-intervals-with-more-information-summarizedexperiment-class.html"><a href="genomic-intervals-with-more-information-summarizedexperiment-class.html"><i class="fa fa-check"></i><b>6.4</b> Genomic intervals with more information: SummarizedExperiment class</a><ul>
<li class="chapter" data-level="6.4.1" data-path="genomic-intervals-with-more-information-summarizedexperiment-class.html"><a href="genomic-intervals-with-more-information-summarizedexperiment-class.html#create-a-summarizedexperiment-object"><i class="fa fa-check"></i><b>6.4.1</b> Create a SummarizedExperiment object</a></li>
<li class="chapter" data-level="6.4.2" data-path="genomic-intervals-with-more-information-summarizedexperiment-class.html"><a href="genomic-intervals-with-more-information-summarizedexperiment-class.html#subset-and-manipulate-the-summarizedexperiment-object"><i class="fa fa-check"></i><b>6.4.2</b> Subset and manipulate the SummarizedExperiment object</a></li>
</ul></li>
<li class="chapter" data-level="6.5" data-path="visualizing-and-summarizing-genomic-intervals.html"><a href="visualizing-and-summarizing-genomic-intervals.html"><i class="fa fa-check"></i><b>6.5</b> Visualizing and summarizing genomic intervals</a><ul>
<li class="chapter" data-level="6.5.1" data-path="visualizing-and-summarizing-genomic-intervals.html"><a href="visualizing-and-summarizing-genomic-intervals.html#visualizing-intervals-on-a-locus-of-interest"><i class="fa fa-check"></i><b>6.5.1</b> Visualizing intervals on a locus of interest</a></li>
<li class="chapter" data-level="6.5.2" data-path="visualizing-and-summarizing-genomic-intervals.html"><a href="visualizing-and-summarizing-genomic-intervals.html#summaries-of-genomic-intervals-on-multiple-loci"><i class="fa fa-check"></i><b>6.5.2</b> Summaries of genomic intervals on multiple loci</a></li>
<li class="chapter" data-level="6.5.3" data-path="visualizing-and-summarizing-genomic-intervals.html"><a href="visualizing-and-summarizing-genomic-intervals.html#making-karyograms-and-circos-plots"><i class="fa fa-check"></i><b>6.5.3</b> Making karyograms and circos plots</a></li>
</ul></li>
<li class="chapter" data-level="6.6" data-path="exercises-4.html"><a href="exercises-4.html"><i class="fa fa-check"></i><b>6.6</b> Exercises</a><ul>
<li class="chapter" data-level="6.6.1" data-path="exercises-4.html"><a href="exercises-4.html#operations-on-genomic-intervals-with-the-genomicranges-package"><i class="fa fa-check"></i><b>6.6.1</b> Operations on genomic intervals with the <code>GenomicRanges</code> package</a></li>
<li class="chapter" data-level="6.6.2" data-path="exercises-4.html"><a href="exercises-4.html#dealing-with-mapped-high-throughput-sequencing-reads-1"><i class="fa fa-check"></i><b>6.6.2</b> Dealing with mapped high-throughput sequencing reads</a></li>
<li class="chapter" data-level="6.6.3" data-path="exercises-4.html"><a href="exercises-4.html#dealing-with-contiguous-scores-over-the-genome"><i class="fa fa-check"></i><b>6.6.3</b> Dealing with contiguous scores over the genome</a></li>
<li class="chapter" data-level="6.6.4" data-path="exercises-4.html"><a href="exercises-4.html#visualizing-and-summarizing-genomic-intervals-1"><i class="fa fa-check"></i><b>6.6.4</b> Visualizing and summarizing genomic intervals</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="7" data-path="processingReads.html"><a href="processingReads.html"><i class="fa fa-check"></i><b>7</b> Quality Check, Processing and Alignment of High-throughput Sequencing Reads</a><ul>
<li class="chapter" data-level="7.1" data-path="fasta-and-fastq-formats.html"><a href="fasta-and-fastq-formats.html"><i class="fa fa-check"></i><b>7.1</b> FASTA and FASTQ formats</a></li>
<li class="chapter" data-level="7.2" data-path="quality-check-on-sequencing-reads.html"><a href="quality-check-on-sequencing-reads.html"><i class="fa fa-check"></i><b>7.2</b> Quality check on sequencing reads</a><ul>
<li class="chapter" data-level="7.2.1" data-path="quality-check-on-sequencing-reads.html"><a href="quality-check-on-sequencing-reads.html#sequence-quality-per-basecycle"><i class="fa fa-check"></i><b>7.2.1</b> Sequence quality per base/cycle</a></li>
<li class="chapter" data-level="7.2.2" data-path="quality-check-on-sequencing-reads.html"><a href="quality-check-on-sequencing-reads.html#sequence-content-per-basecycle"><i class="fa fa-check"></i><b>7.2.2</b> Sequence content per base/cycle</a></li>
<li class="chapter" data-level="7.2.3" data-path="quality-check-on-sequencing-reads.html"><a href="quality-check-on-sequencing-reads.html#read-frequency-plot"><i class="fa fa-check"></i><b>7.2.3</b> Read frequency plot</a></li>
<li class="chapter" data-level="7.2.4" data-path="quality-check-on-sequencing-reads.html"><a href="quality-check-on-sequencing-reads.html#other-quality-metrics-and-qc-tools"><i class="fa fa-check"></i><b>7.2.4</b> Other quality metrics and QC tools</a></li>
</ul></li>
<li class="chapter" data-level="7.3" data-path="filtering-and-trimming-reads.html"><a href="filtering-and-trimming-reads.html"><i class="fa fa-check"></i><b>7.3</b> Filtering and trimming reads</a></li>
<li class="chapter" data-level="7.4" data-path="mappingaligning-reads-to-the-genome.html"><a href="mappingaligning-reads-to-the-genome.html"><i class="fa fa-check"></i><b>7.4</b> Mapping/aligning reads to the genome</a></li>
<li class="chapter" data-level="7.5" data-path="further-processing-of-aligned-reads.html"><a href="further-processing-of-aligned-reads.html"><i class="fa fa-check"></i><b>7.5</b> Further processing of aligned reads</a></li>
<li class="chapter" data-level="7.6" data-path="exercises-5.html"><a href="exercises-5.html"><i class="fa fa-check"></i><b>7.6</b> Exercises</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="rnaseqanalysis.html"><a href="rnaseqanalysis.html"><i class="fa fa-check"></i><b>8</b> RNA-seq Analysis</a><ul>
<li class="chapter" data-level="8.1" data-path="what-is-gene-expression.html"><a href="what-is-gene-expression.html"><i class="fa fa-check"></i><b>8.1</b> What is gene expression?</a></li>
<li class="chapter" data-level="8.2" data-path="methods-to-detect-gene-expression.html"><a href="methods-to-detect-gene-expression.html"><i class="fa fa-check"></i><b>8.2</b> Methods to detect gene expression</a></li>
<li class="chapter" data-level="8.3" data-path="gene-expression-analysis-using-high-throughput-sequencing-technologies.html"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html"><i class="fa fa-check"></i><b>8.3</b> Gene expression analysis using high-throughput sequencing technologies</a><ul>
<li class="chapter" data-level="8.3.1" data-path="gene-expression-analysis-using-high-throughput-sequencing-technologies.html"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#processing-raw-data"><i class="fa fa-check"></i><b>8.3.1</b> Processing raw data</a></li>
<li class="chapter" data-level="8.3.2" data-path="gene-expression-analysis-using-high-throughput-sequencing-technologies.html"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#alignment"><i class="fa fa-check"></i><b>8.3.2</b> Alignment</a></li>
<li class="chapter" data-level="8.3.3" data-path="gene-expression-analysis-using-high-throughput-sequencing-technologies.html"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#quantification"><i class="fa fa-check"></i><b>8.3.3</b> Quantification</a></li>
<li class="chapter" data-level="8.3.4" data-path="gene-expression-analysis-using-high-throughput-sequencing-technologies.html"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#within-sample-normalization-of-the-read-counts"><i class="fa fa-check"></i><b>8.3.4</b> Within sample normalization of the read counts</a></li>
<li class="chapter" data-level="8.3.5" data-path="gene-expression-analysis-using-high-throughput-sequencing-technologies.html"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#computing-different-normalization-schemes-in-r"><i class="fa fa-check"></i><b>8.3.5</b> Computing different normalization schemes in R</a></li>
<li class="chapter" data-level="8.3.6" data-path="gene-expression-analysis-using-high-throughput-sequencing-technologies.html"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#exploratory-analysis-of-the-read-count-table"><i class="fa fa-check"></i><b>8.3.6</b> Exploratory analysis of the read count table</a></li>
<li class="chapter" data-level="8.3.7" data-path="gene-expression-analysis-using-high-throughput-sequencing-technologies.html"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#differential-expression-analysis"><i class="fa fa-check"></i><b>8.3.7</b> Differential expression analysis</a></li>
<li class="chapter" data-level="8.3.8" data-path="gene-expression-analysis-using-high-throughput-sequencing-technologies.html"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#functional-enrichment-analysis"><i class="fa fa-check"></i><b>8.3.8</b> Functional enrichment analysis</a></li>
<li class="chapter" data-level="8.3.9" data-path="gene-expression-analysis-using-high-throughput-sequencing-technologies.html"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#accounting-for-additional-sources-of-variation"><i class="fa fa-check"></i><b>8.3.9</b> Accounting for additional sources of variation</a></li>
</ul></li>
<li class="chapter" data-level="8.4" data-path="other-applications-of-rna-seq.html"><a href="other-applications-of-rna-seq.html"><i class="fa fa-check"></i><b>8.4</b> Other applications of RNA-seq</a></li>
<li class="chapter" data-level="8.5" data-path="exercises-6.html"><a href="exercises-6.html"><i class="fa fa-check"></i><b>8.5</b> Exercises</a><ul>
<li class="chapter" data-level="8.5.1" data-path="exercises-6.html"><a href="exercises-6.html#exploring-the-count-tables"><i class="fa fa-check"></i><b>8.5.1</b> Exploring the count tables</a></li>
<li class="chapter" data-level="8.5.2" data-path="exercises-6.html"><a href="exercises-6.html#differential-expression-analysis-1"><i class="fa fa-check"></i><b>8.5.2</b> Differential expression analysis</a></li>
<li class="chapter" data-level="8.5.3" data-path="exercises-6.html"><a href="exercises-6.html#functional-enrichment-analysis-1"><i class="fa fa-check"></i><b>8.5.3</b> Functional enrichment analysis</a></li>
<li class="chapter" data-level="8.5.4" data-path="exercises-6.html"><a href="exercises-6.html#removing-unwanted-variation-from-the-expression-data"><i class="fa fa-check"></i><b>8.5.4</b> Removing unwanted variation from the expression data</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="9" data-path="chipseq.html"><a href="chipseq.html"><i class="fa fa-check"></i><b>9</b> ChIP-seq analysis</a><ul>
<li class="chapter" data-level="9.1" data-path="regulatory-protein-dna-interactions.html"><a href="regulatory-protein-dna-interactions.html"><i class="fa fa-check"></i><b>9.1</b> Regulatory protein-DNA interactions</a></li>
<li class="chapter" data-level="9.2" data-path="measuring-protein-dna-interactions-with-chip-seq.html"><a href="measuring-protein-dna-interactions-with-chip-seq.html"><i class="fa fa-check"></i><b>9.2</b> Measuring protein-DNA interactions with ChIP-seq</a></li>
<li class="chapter" data-level="9.3" data-path="factors-that-affect-chip-seq-experiment-and-analysis-quality.html"><a href="factors-that-affect-chip-seq-experiment-and-analysis-quality.html"><i class="fa fa-check"></i><b>9.3</b> Factors that affect ChIP-seq experiment and analysis quality</a><ul>
<li class="chapter" data-level="9.3.1" data-path="factors-that-affect-chip-seq-experiment-and-analysis-quality.html"><a href="factors-that-affect-chip-seq-experiment-and-analysis-quality.html#antibody-specificity"><i class="fa fa-check"></i><b>9.3.1</b> Antibody specificity</a></li>
<li class="chapter" data-level="9.3.2" data-path="factors-that-affect-chip-seq-experiment-and-analysis-quality.html"><a href="factors-that-affect-chip-seq-experiment-and-analysis-quality.html#sequencing-depth"><i class="fa fa-check"></i><b>9.3.2</b> Sequencing depth</a></li>
<li class="chapter" data-level="9.3.3" data-path="factors-that-affect-chip-seq-experiment-and-analysis-quality.html"><a href="factors-that-affect-chip-seq-experiment-and-analysis-quality.html#pcr-duplication"><i class="fa fa-check"></i><b>9.3.3</b> PCR duplication</a></li>
<li class="chapter" data-level="9.3.4" data-path="factors-that-affect-chip-seq-experiment-and-analysis-quality.html"><a href="factors-that-affect-chip-seq-experiment-and-analysis-quality.html#biological-replicates"><i class="fa fa-check"></i><b>9.3.4</b> Biological replicates</a></li>
<li class="chapter" data-level="9.3.5" data-path="factors-that-affect-chip-seq-experiment-and-analysis-quality.html"><a href="factors-that-affect-chip-seq-experiment-and-analysis-quality.html#control-experiments"><i class="fa fa-check"></i><b>9.3.5</b> Control experiments</a></li>
<li class="chapter" data-level="9.3.6" data-path="factors-that-affect-chip-seq-experiment-and-analysis-quality.html"><a href="factors-that-affect-chip-seq-experiment-and-analysis-quality.html#using-tagged-proteins"><i class="fa fa-check"></i><b>9.3.6</b> Using tagged proteins</a></li>
</ul></li>
<li class="chapter" data-level="9.4" data-path="pre-processing-chip-data.html"><a href="pre-processing-chip-data.html"><i class="fa fa-check"></i><b>9.4</b> Pre-processing ChIP data</a><ul>
<li class="chapter" data-level="9.4.1" data-path="pre-processing-chip-data.html"><a href="pre-processing-chip-data.html#mapping-of-chip-seq-data"><i class="fa fa-check"></i><b>9.4.1</b> Mapping of ChIP-seq data</a></li>
</ul></li>
<li class="chapter" data-level="9.5" data-path="chip-quality-control.html"><a href="chip-quality-control.html"><i class="fa fa-check"></i><b>9.5</b> ChIP quality control</a><ul>
<li class="chapter" data-level="9.5.1" data-path="chip-quality-control.html"><a href="chip-quality-control.html#the-data"><i class="fa fa-check"></i><b>9.5.1</b> The data</a></li>
<li class="chapter" data-level="9.5.2" data-path="chip-quality-control.html"><a href="chip-quality-control.html#sample-clustering"><i class="fa fa-check"></i><b>9.5.2</b> Sample clustering</a></li>
<li class="chapter" data-level="9.5.3" data-path="chip-quality-control.html"><a href="chip-quality-control.html#visualization-in-the-genome-browser"><i class="fa fa-check"></i><b>9.5.3</b> Visualization in the genome browser</a></li>
<li class="chapter" data-level="9.5.4" data-path="chip-quality-control.html"><a href="chip-quality-control.html#plus-and-minus-strand-cross-correlation"><i class="fa fa-check"></i><b>9.5.4</b> Plus and minus strand cross-correlation</a></li>
<li class="chapter" data-level="9.5.5" data-path="chip-quality-control.html"><a href="chip-quality-control.html#gc-bias-quantification"><i class="fa fa-check"></i><b>9.5.5</b> GC bias quantification</a></li>
<li class="chapter" data-level="9.5.6" data-path="chip-quality-control.html"><a href="chip-quality-control.html#sequence-read-genomic-distribution"><i class="fa fa-check"></i><b>9.5.6</b> Sequence read genomic distribution</a></li>
</ul></li>
<li class="chapter" data-level="9.6" data-path="peak-calling.html"><a href="peak-calling.html"><i class="fa fa-check"></i><b>9.6</b> Peak calling</a><ul>
<li class="chapter" data-level="9.6.1" data-path="peak-calling.html"><a href="peak-calling.html#types-of-chip-seq-experiments"><i class="fa fa-check"></i><b>9.6.1</b> Types of ChIP-seq experiments</a></li>
<li class="chapter" data-level="9.6.2" data-path="peak-calling.html"><a href="peak-calling.html#peak-calling-sharp-peaks"><i class="fa fa-check"></i><b>9.6.2</b> Peak calling: Sharp peaks</a></li>
<li class="chapter" data-level="9.6.3" data-path="peak-calling.html"><a href="peak-calling.html#peak-calling-broad-regions"><i class="fa fa-check"></i><b>9.6.3</b> Peak calling: Broad regions</a></li>
<li class="chapter" data-level="9.6.4" data-path="peak-calling.html"><a href="peak-calling.html#peak-quality-control"><i class="fa fa-check"></i><b>9.6.4</b> Peak quality control</a></li>
<li class="chapter" data-level="9.6.5" data-path="peak-calling.html"><a href="peak-calling.html#peak-annotation"><i class="fa fa-check"></i><b>9.6.5</b> Peak annotation</a></li>
</ul></li>
<li class="chapter" data-level="9.7" data-path="motif-discovery.html"><a href="motif-discovery.html"><i class="fa fa-check"></i><b>9.7</b> Motif discovery</a><ul>
<li class="chapter" data-level="9.7.1" data-path="motif-discovery.html"><a href="motif-discovery.html#motif-comparison"><i class="fa fa-check"></i><b>9.7.1</b> Motif comparison</a></li>
</ul></li>
<li class="chapter" data-level="9.8" data-path="what-to-do-next.html"><a href="what-to-do-next.html"><i class="fa fa-check"></i><b>9.8</b> What to do next?</a></li>
<li class="chapter" data-level="9.9" data-path="exercises-7.html"><a href="exercises-7.html"><i class="fa fa-check"></i><b>9.9</b> Exercises</a><ul>
<li class="chapter" data-level="9.9.1" data-path="exercises-7.html"><a href="exercises-7.html#quality-control"><i class="fa fa-check"></i><b>9.9.1</b> Quality control</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="10" data-path="bsseq.html"><a href="bsseq.html"><i class="fa fa-check"></i><b>10</b> DNA methylation analysis using bisulfite sequencing data</a><ul>
<li class="chapter" data-level="10.1" data-path="what-is-dna-methylation.html"><a href="what-is-dna-methylation.html"><i class="fa fa-check"></i><b>10.1</b> What is DNA methylation?</a><ul>
<li class="chapter" data-level="10.1.1" data-path="what-is-dna-methylation.html"><a href="what-is-dna-methylation.html#how-dna-methylation-is-set"><i class="fa fa-check"></i><b>10.1.1</b> How DNA methylation is set ?</a></li>
<li class="chapter" data-level="10.1.2" data-path="what-is-dna-methylation.html"><a href="what-is-dna-methylation.html#how-to-measure-dna-methylation-with-bisulfite-sequencing"><i class="fa fa-check"></i><b>10.1.2</b> How to measure DNA methylation with bisulfite sequencing</a></li>
</ul></li>
<li class="chapter" data-level="10.2" data-path="analyzing-dna-methylation-data.html"><a href="analyzing-dna-methylation-data.html"><i class="fa fa-check"></i><b>10.2</b> Analyzing DNA methylation data</a></li>
<li class="chapter" data-level="10.3" data-path="processing-raw-data-and-getting-data-into-r.html"><a href="processing-raw-data-and-getting-data-into-r.html"><i class="fa fa-check"></i><b>10.3</b> Processing raw data and getting data into R</a></li>
<li class="chapter" data-level="10.4" data-path="data-filtering-and-exploratory-analysis.html"><a href="data-filtering-and-exploratory-analysis.html"><i class="fa fa-check"></i><b>10.4</b> Data filtering and exploratory analysis</a><ul>
<li class="chapter" data-level="10.4.1" data-path="data-filtering-and-exploratory-analysis.html"><a href="data-filtering-and-exploratory-analysis.html#reading-methylation-call-files"><i class="fa fa-check"></i><b>10.4.1</b> Reading methylation call files</a></li>
<li class="chapter" data-level="10.4.2" data-path="data-filtering-and-exploratory-analysis.html"><a href="data-filtering-and-exploratory-analysis.html#further-quality-check"><i class="fa fa-check"></i><b>10.4.2</b> Further quality check</a></li>
<li class="chapter" data-level="10.4.3" data-path="data-filtering-and-exploratory-analysis.html"><a href="data-filtering-and-exploratory-analysis.html#merging-samples-into-a-single-table"><i class="fa fa-check"></i><b>10.4.3</b> Merging samples into a single table</a></li>
<li class="chapter" data-level="10.4.4" data-path="data-filtering-and-exploratory-analysis.html"><a href="data-filtering-and-exploratory-analysis.html#filtering-cpgs"><i class="fa fa-check"></i><b>10.4.4</b> Filtering CpGs</a></li>
<li class="chapter" data-level="10.4.5" data-path="data-filtering-and-exploratory-analysis.html"><a href="data-filtering-and-exploratory-analysis.html#clustering-samples"><i class="fa fa-check"></i><b>10.4.5</b> Clustering samples</a></li>
<li class="chapter" data-level="10.4.6" data-path="data-filtering-and-exploratory-analysis.html"><a href="data-filtering-and-exploratory-analysis.html#principal-component-analysis-1"><i class="fa fa-check"></i><b>10.4.6</b> Principal component analysis</a></li>
</ul></li>
<li class="chapter" data-level="10.5" data-path="extracting-interesting-regions-differential-methylation-and-segmentation.html"><a href="extracting-interesting-regions-differential-methylation-and-segmentation.html"><i class="fa fa-check"></i><b>10.5</b> Extracting interesting regions: Differential methylation and segmentation</a><ul>
<li class="chapter" data-level="10.5.1" data-path="extracting-interesting-regions-differential-methylation-and-segmentation.html"><a href="extracting-interesting-regions-differential-methylation-and-segmentation.html#differential-methylation"><i class="fa fa-check"></i><b>10.5.1</b> Differential methylation</a></li>
<li class="chapter" data-level="10.5.2" data-path="extracting-interesting-regions-differential-methylation-and-segmentation.html"><a href="extracting-interesting-regions-differential-methylation-and-segmentation.html#methylation-segmentation"><i class="fa fa-check"></i><b>10.5.2</b> Methylation segmentation</a></li>
<li class="chapter" data-level="10.5.3" data-path="extracting-interesting-regions-differential-methylation-and-segmentation.html"><a href="extracting-interesting-regions-differential-methylation-and-segmentation.html#working-with-large-files"><i class="fa fa-check"></i><b>10.5.3</b> Working with large files</a></li>
</ul></li>
<li class="chapter" data-level="10.6" data-path="annotation-of-dmrsdmcs-and-segments.html"><a href="annotation-of-dmrsdmcs-and-segments.html"><i class="fa fa-check"></i><b>10.6</b> Annotation of DMRs/DMCs and segments</a><ul>
<li class="chapter" data-level="10.6.1" data-path="annotation-of-dmrsdmcs-and-segments.html"><a href="annotation-of-dmrsdmcs-and-segments.html#further-annotation-with-genes-or-gene-sets"><i class="fa fa-check"></i><b>10.6.1</b> Further annotation with genes or gene sets</a></li>
</ul></li>
<li class="chapter" data-level="10.7" data-path="other-r-packages-that-can-be-used-for-methylation-analysis.html"><a href="other-r-packages-that-can-be-used-for-methylation-analysis.html"><i class="fa fa-check"></i><b>10.7</b> Other R packages that can be used for methylation analysis</a></li>
<li class="chapter" data-level="10.8" data-path="exercises-8.html"><a href="exercises-8.html"><i class="fa fa-check"></i><b>10.8</b> Exercises</a><ul>
<li class="chapter" data-level="10.8.1" data-path="exercises-8.html"><a href="exercises-8.html#differential-methylation-1"><i class="fa fa-check"></i><b>10.8.1</b> Differential methylation</a></li>
<li class="chapter" data-level="10.8.2" data-path="exercises-8.html"><a href="exercises-8.html#methylome-segmentation"><i class="fa fa-check"></i><b>10.8.2</b> Methylome segmentation</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="11" data-path="multiomics.html"><a href="multiomics.html"><i class="fa fa-check"></i><b>11</b> Multi-omics Analysis</a><ul>
<li class="chapter" data-level="11.1" data-path="use-case-multi-omics-data-from-colorectal-cancer.html"><a href="use-case-multi-omics-data-from-colorectal-cancer.html"><i class="fa fa-check"></i><b>11.1</b> Use case: Multi-omics data from colorectal cancer</a></li>
<li class="chapter" data-level="11.2" data-path="latent-variable-models-for-multi-omics-integration.html"><a href="latent-variable-models-for-multi-omics-integration.html"><i class="fa fa-check"></i><b>11.2</b> Latent variable models for multi-omics integration</a></li>
<li class="chapter" data-level="11.3" data-path="matrix-factorization-methods-for-unsupervised-multi-omics-data-integration.html"><a href="matrix-factorization-methods-for-unsupervised-multi-omics-data-integration.html"><i class="fa fa-check"></i><b>11.3</b> Matrix factorization methods for unsupervised multi-omics data integration</a><ul>
<li class="chapter" data-level="11.3.1" data-path="matrix-factorization-methods-for-unsupervised-multi-omics-data-integration.html"><a href="matrix-factorization-methods-for-unsupervised-multi-omics-data-integration.html#multiple-factor-analysis"><i class="fa fa-check"></i><b>11.3.1</b> Multiple factor analysis</a></li>
<li class="chapter" data-level="11.3.2" data-path="matrix-factorization-methods-for-unsupervised-multi-omics-data-integration.html"><a href="matrix-factorization-methods-for-unsupervised-multi-omics-data-integration.html#joint-non-negative-matrix-factorization"><i class="fa fa-check"></i><b>11.3.2</b> Joint non-negative matrix factorization</a></li>
<li class="chapter" data-level="11.3.3" data-path="matrix-factorization-methods-for-unsupervised-multi-omics-data-integration.html"><a href="matrix-factorization-methods-for-unsupervised-multi-omics-data-integration.html#icluster"><i class="fa fa-check"></i><b>11.3.3</b> iCluster</a></li>
</ul></li>
<li class="chapter" data-level="11.4" data-path="clustering-using-latent-factors.html"><a href="clustering-using-latent-factors.html"><i class="fa fa-check"></i><b>11.4</b> Clustering using latent factors</a><ul>
<li class="chapter" data-level="11.4.1" data-path="clustering-using-latent-factors.html"><a href="clustering-using-latent-factors.html#one-hot-clustering"><i class="fa fa-check"></i><b>11.4.1</b> One-hot clustering</a></li>
<li class="chapter" data-level="11.4.2" data-path="clustering-using-latent-factors.html"><a href="clustering-using-latent-factors.html#k-means-clustering-1"><i class="fa fa-check"></i><b>11.4.2</b> K-means clustering</a></li>
</ul></li>
<li class="chapter" data-level="11.5" data-path="biological-interpretation-of-latent-factors.html"><a href="biological-interpretation-of-latent-factors.html"><i class="fa fa-check"></i><b>11.5</b> Biological interpretation of latent factors</a><ul>
<li class="chapter" data-level="11.5.1" data-path="biological-interpretation-of-latent-factors.html"><a href="biological-interpretation-of-latent-factors.html#inspection-of-feature-weights-in-loading-vectors"><i class="fa fa-check"></i><b>11.5.1</b> Inspection of feature weights in loading vectors</a></li>
<li class="chapter" data-level="11.5.2" data-path="biological-interpretation-of-latent-factors.html"><a href="biological-interpretation-of-latent-factors.html#making-sense-of-factors-using-enrichment-analysis"><i class="fa fa-check"></i><b>11.5.2</b> Making sense of factors using enrichment analysis</a></li>
<li class="chapter" data-level="11.5.3" data-path="biological-interpretation-of-latent-factors.html"><a href="biological-interpretation-of-latent-factors.html#interpretation-using-additional-covariates"><i class="fa fa-check"></i><b>11.5.3</b> Interpretation using additional covariates</a></li>
</ul></li>
<li class="chapter" data-level="11.6" data-path="exercises-9.html"><a href="exercises-9.html"><i class="fa fa-check"></i><b>11.6</b> Exercises</a><ul>
<li class="chapter" data-level="11.6.1" data-path="exercises-9.html"><a href="exercises-9.html#matrix-factorization-methods"><i class="fa fa-check"></i><b>11.6.1</b> Matrix factorization methods</a></li>
<li class="chapter" data-level="11.6.2" data-path="exercises-9.html"><a href="exercises-9.html#clustering-using-latent-factors-1"><i class="fa fa-check"></i><b>11.6.2</b> Clustering using latent factors</a></li>
<li class="chapter" data-level="11.6.3" data-path="exercises-9.html"><a href="exercises-9.html#biological-interpretation-of-latent-factors-1"><i class="fa fa-check"></i><b>11.6.3</b> Biological interpretation of latent factors</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i>References</a></li>
<li class="divider"></li>
<li><a href="https://bookdown.org" target="_blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Computational Genomics with R</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="gene-expression-analysis-using-high-throughput-sequencing-technologies" class="section level2">
<h2><span class="header-section-number">8.3</span> Gene expression analysis using high-throughput sequencing technologies</h2>
<p>With the advent of the second-generation (a.k.a next-generation or high-throughput) sequencing technologies,
the number of genes that can be profiled for expression levels with a single experiment has increased to the order of tens of thousands of genes. Therefore, the bottleneck in this process has become the data analysis rather than the data generation. Many statistical methods and computational tools are required for getting meaningful results from the data, which comes with a lot of valuable information along with a lot of sources of noise. Fortunately, most of the steps of RNA-seq analysis have become quite mature over the years. Below we will first describe how to reach a read count table from raw fastq reads obtained from an Illumina sequencing run. We will then demonstrate in R how to process the count table, make a case-control differential expression analysis, and do some downstream functional enrichment analysis.</p>
<div id="processing-raw-data" class="section level3">
<h3><span class="header-section-number">8.3.1</span> Processing raw data</h3>
<div id="quality-check-and-read-processing" class="section level4">
<h4><span class="header-section-number">8.3.1.1</span> Quality check and read processing</h4>
<p>The first step in any experiment that involves high-throughput short-read sequencing should be to check the sequencing quality of the reads before starting to do any downstream analysis. The quality of the input sequences holds fundamental importance in the confidence for the biological conclusions drawn from the experiment. We have introduced quality check and processing in Chapter <a href="processingReads.html#processingReads">7</a>, and those tools and workflows also apply in RNA-seq analysis.</p>
</div>
<div id="improving-the-quality" class="section level4">
<h4><span class="header-section-number">8.3.1.2</span> Improving the quality</h4>
<p>The second step in the RNA-seq analysis workflow is to improve the quality of the input reads. This step could be regarded as an optional step when the sequencing quality is very good. However, even with the highest-quality sequencing datasets, this step may still improve the quality of the input sequences. The most common technical artifacts that can be filtered out are the adapter sequences that contaminate the sequenced reads, and the low-quality bases that are usually found at the ends of the sequences. Commonly used tools in the field (trimmomatic <span class="citation">(Bolger, Lohse, and Usadel <a href="#ref-bolger_trimmomatic:_2014" role="doc-biblioref">2014</a>)</span>, trimGalore <span class="citation">(Andrews <a href="#ref-noauthor_babraham_nodate" role="doc-biblioref">2010</a>)</span>) are again not written in R, however there are alternative R libraries for carrying out the same functionality, for instance, QuasR <span class="citation">(Gaidatzis, Lerch, Hahne, et al. <a href="#ref-gaidatzis_quasr:_2015" role="doc-biblioref">2015</a>)</span> (see <code>QuasR::preprocessReads</code> function) and ShortRead <span class="citation">(Morgan, Anders, Lawrence, et al. <a href="#ref-morgan_shortread:_2009" role="doc-biblioref">2009</a>)</span> (see <code>ShortRead::filterFastq</code> function). Some of these approaches are introduced in Chapter <a href="processingReads.html#processingReads">7</a>.</p>
<p>The sequencing quality control and read pre-processing steps can be visited multiple times until achieving a satisfactory level of quality in the sequence data before moving on to the downstream analysis steps.</p>
</div>
</div>
<div id="alignment" class="section level3">
<h3><span class="header-section-number">8.3.2</span> Alignment</h3>
<p>Once a decent level of quality in the sequences is reached, the expression level of the genes can be quantified by first mapping the sequences to a reference genome, and secondly matching the aligned reads to the gene annotations, in order to count the number of reads mapping to each gene. If the species under study has a well-annotated transcriptome, the reads can be aligned to the transcript sequences instead of the reference genome. In cases where there is no good quality reference genome or transcriptome, it is possible to de novo assemble the transcriptome from the sequences and then quantify the expression levels of genes/transcripts.</p>
<p>For RNA-seq read alignments, apart from the availability of reference genomes and annotations, probably the most important factor to consider when choosing an alignment tool is whether the alignment method considers the absence of intronic regions in the sequenced reads, while the target genome may contain introns. Therefore, it is important to choose alignment tools that take into account alternative splicing. In the basic setting where a read, which originates from a cDNA sequence corresponding to an exon-exon junction, needs to be split into two parts when aligned against the genome. There are various tools that consider this factor such as STAR <span class="citation">(Dobin, Davis, Schlesinger, et al. <a href="#ref-dobin_star:_2013" role="doc-biblioref">2013</a>)</span>, Tophat2 <span class="citation">(Kim, Pertea, Trapnell, et al. <a href="#ref-kim_tophat2:_2013" role="doc-biblioref">2013</a>)</span>, Hisat2 <span class="citation">(Kim, Langmead, and Salzberg <a href="#ref-kim_hisat:_2015" role="doc-biblioref">2015</a>)</span>, and GSNAP <span class="citation">(Wu, Reeder, Lawrence, et al. <a href="#ref-wu_gmap_2016" role="doc-biblioref">2016</a>)</span>. Most alignment tools are written in C/C++ languages because of performance concerns. There are also R libraries that can do short read alignments; these are discussed in Chapter <a href="processingReads.html#processingReads">7</a>.</p>
</div>
<div id="quantification" class="section level3">
<h3><span class="header-section-number">8.3.3</span> Quantification</h3>
<p>After the reads are aligned to the target, a SAM/BAM file sorted by coordinates should have been obtained. The BAM file contains all alignment-related information of all the reads that have been attempted to be aligned to the target sequence. This information consists of - most basically - the genomic coordinates (chromosome, start, end, strand) of where a sequence was matched (if at all) in the target, specific insertions/deletions/mismatches that describe the differences between the input and target sequences. These pieces of information are used along with the genomic coordinates of genome annotations such as gene/transcript models in order to count how many reads have been sequenced from a gene/transcript. As simple as it may sound, it is not a trivial task to assign reads to a gene/transcript just by comparing the genomic coordinates of the annotations and the sequences, because of confounding factors such as overlapping gene annotations, overlapping exon annotations from different transcript isoforms of a gene, and overlapping annotations from opposite DNA strands in the absence of a strand-specific sequencing protocol. Therefore, for read counting, it is important to consider:</p>
<ol style="list-style-type: decimal">
<li>Strand specificity of the sequencing protocol: Are the reads expected to originate from the forward strand, reverse strand, or unspecific?</li>
<li>Counting mode:
- when counting at the gene-level: When there are overlapping annotations, which features should the read be assigned to? Tools usually have a parameter that lets the user select a counting mode.
- when counting at the transcript-level: When there are multiple isoforms of a gene, which isoform should the read be assigned to? This consideration is usually an algorithmic consideration that is not modifiable by the end-user.</li>
</ol>
<p>Some tools can couple alignment to quantification (e.g. STAR), while some assume the alignments are already calculated and require BAM files as input. On the other hand, in the presence of good transcriptome annotations, alignment-free methods (Salmon <span class="citation">(Patro, Duggal, Love, et al. <a href="#ref-patro_salmon:_2017" role="doc-biblioref">2017</a>)</span>, Kallisto <span class="citation">(Bray, Pimentel, Melsted, et al. <a href="#ref-bray_near-optimal_2016" role="doc-biblioref">2016</a>)</span>, Sailfish <span class="citation">(Patro, Mount, and Kingsford <a href="#ref-patro_sailfish_2014" role="doc-biblioref">2014</a>)</span>) can also be used to estimate the expression levels of transcripts/genes. There are also reference-free quantification methods that can first de novo assemble the transcriptome and estimate the expression levels based on this assembly. Such a strategy can be useful in discovering novel transcripts or may be required in cases when a good reference does not exist. If a reference transcriptome exists but of low quality, a reference-based transcriptome assembler such as Cufflinks <span class="citation">(Trapnell, Williams, Pertea, et al. <a href="#ref-trapnell_transcript_2010" role="doc-biblioref">2010</a>)</span> can be used to improve the transcriptome. In case there is no available transcriptome annotation, a de novo assembler such as Trinity <span class="citation">(Haas, Papanicolaou, Yassour, et al. <a href="#ref-haas_novo_2013" role="doc-biblioref">2013</a>)</span> or Trans-ABySS <span class="citation">(Robertson, Schein, Chiu, et al. <a href="#ref-robertson_novo_2010" role="doc-biblioref">2010</a>)</span> can be used to assemble the transcriptome from scratch.</p>
<p>Within R, quantification can be done using:
- <code>Rsubread::featureCounts</code>
- <code>QuasR::qCount</code>
- <code>GenomicAlignments::summarizeOverlaps</code></p>
</div>
<div id="within-sample-normalization-of-the-read-counts" class="section level3">
<h3><span class="header-section-number">8.3.4</span> Within sample normalization of the read counts</h3>
<p>The most common application after a gene’s expression is quantified (as the number of reads aligned to the gene), is to compare the gene’s expression in different conditions, for instance, in a case-control setting (e.g. disease versus normal) or in a time-series (e.g. along different developmental stages). Making such comparisons helps identify the genes that might be responsible for a disease or an impaired developmental trajectory. However, there are multiple caveats that needs to be addressed before making a comparison between the read counts of a gene in different conditions <span class="citation">(Maza, Frasse, Senin, et al. <a href="#ref-maza_comparison_2013" role="doc-biblioref">2013</a>)</span>.</p>
<ul>
<li>Library size (i.e. sequencing depth) varies between samples coming from different lanes of the flow cell of the sequencing machine.</li>
<li>Longer genes will have a higher number of reads.</li>
<li>Library composition (i.e. relative size of the studied transcriptome) can be different in two different biological conditions.</li>
<li>GC content biases across different samples may lead to a biased sampling of genes <span class="citation">(Risso, Schwartz, Sherlock, et al. <a href="#ref-risso_gc-content_2011" role="doc-biblioref">2011</a>)</span>.</li>
<li>Read coverage of a transcript can be biased and non-uniformly distributed along the transcript <span class="citation">(Mortazavi, Williams, McCue, et al. <a href="#ref-mortazavi_mapping_2008" role="doc-biblioref">2008</a>)</span>.</li>
</ul>
<p>Therefore these factors need to be taken into account before making comparisons.</p>
<p>The most basic normalization approaches address the sequencing depth bias. Such procedures normalize the read counts per gene by dividing each gene’s read count by a certain value and multiplying it by 10^6. These normalized values are usually referred to as CPM (counts per million reads):</p>
<ul>
<li>Total Counts Normalization (divide counts by the <strong>sum</strong> of all counts)</li>
<li>Upper Quartile Normalization (divide counts by the <strong>upper quartile</strong> value of the counts)</li>
<li>Median Normalization (divide counts by the <strong>median</strong> of all counts)</li>
</ul>
<p>Popular metrics that improve upon CPM are RPKM/FPKM (reads/fragments per kilobase of million reads) and TPM (transcripts per million). RPKM is obtained by dividing the CPM value by another factor, which is the length of the gene per kilobase. FPKM is the same as RPKM, but is used for paired-end reads. Thus, RPKM/FPKM methods account for, firstly, the <strong>library size</strong>, and secondly, the <strong>gene lengths</strong>.</p>
<p>TPM also controls for both the library size and the gene lengths, however, with the TPM method, the read counts are first normalized by the gene length (per kilobase), and then gene-length normalized values are divided by the sum of the gene-length normalized values and multiplied by 10^6. Thus, the sum of normalized values for TPM will always be equal to 10^6 for each library, while the sum of RPKM/FPKM values do not sum to 10^6. Therefore, it is easier to interpret TPM values than RPKM/FPKM values.</p>
</div>
<div id="computing-different-normalization-schemes-in-r" class="section level3">
<h3><span class="header-section-number">8.3.5</span> Computing different normalization schemes in R</h3>
<p>Here we will assume that there is an RNA-seq count table comprising raw counts, meaning the number of reads counted for each gene has not been exposed to any kind of normalization and consists of integers. The rows of the count table correspond to the genes and the columns represent different samples. Here we will use a subset of the RNA-seq count table from a colorectal cancer study. We have filtered the original count table for only protein-coding genes (to improve the speed of calculation) and also selected only five metastasized colorectal cancer samples along with five normal colon samples. There is an additional column <code>width</code> that contains the length of the corresponding gene in the unit of base pairs. The length of the genes are important to compute RPKM and TPM values. The original count tables can be found from the recount2 database (<a href="https://jhubiostatistics.shinyapps.io/recount/" class="uri">https://jhubiostatistics.shinyapps.io/recount/</a>) using the SRA project code <em>SRP029880</em>, and the experimental setup along with other accessory information can be found from the NCBI Trace archive using the SRA project code <a href="https://trace.ncbi.nlm.nih.gov/Traces/sra/?study=SRP029880">SRP029880`</a>.</p>
<div class="sourceCode" id="cb383"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb383-1"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb383-1"></a><span class="co">#colorectal cancer</span></span>
<span id="cb383-2"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb383-2"></a>counts_file &lt;-<span class="st"> </span><span class="kw">system.file</span>(<span class="st">&quot;extdata/rna-seq/SRP029880.raw_counts.tsv&quot;</span>,</span>
<span id="cb383-3"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb383-3"></a>                           <span class="dt">package =</span> <span class="st">&quot;compGenomRData&quot;</span>)</span>
<span id="cb383-4"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb383-4"></a>coldata_file &lt;-<span class="st"> </span><span class="kw">system.file</span>(<span class="st">&quot;extdata/rna-seq/SRP029880.colData.tsv&quot;</span>,</span>
<span id="cb383-5"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb383-5"></a>                            <span class="dt">package =</span> <span class="st">&quot;compGenomRData&quot;</span>)</span>
<span id="cb383-6"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb383-6"></a></span>
<span id="cb383-7"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb383-7"></a>counts &lt;-<span class="st"> </span><span class="kw">as.matrix</span>(<span class="kw">read.table</span>(counts_file, <span class="dt">header =</span> T, <span class="dt">sep =</span> <span class="st">&#39;</span><span class="ch">\t</span><span class="st">&#39;</span>))</span></code></pre></div>
<div id="computing-cpm" class="section level4">
<h4><span class="header-section-number">8.3.5.1</span> Computing CPM</h4>
<p>Let’s do a summary of the counts table. Due to space limitations, the summary for only the first three columns is displayed.</p>
<div class="sourceCode" id="cb384"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb384-1"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb384-1"></a><span class="kw">summary</span>(counts[,<span class="dv">1</span><span class="op">:</span><span class="dv">3</span>])</span></code></pre></div>
<pre><code>##      CASE_1              CASE_2              CASE_3         
##  Min.   :        0   Min.   :        0   Min.   :        0  
##  1st Qu.:     5155   1st Qu.:     6464   1st Qu.:     3972  
##  Median :    80023   Median :    85064   Median :    64145  
##  Mean   :   295932   Mean   :   273099   Mean   :   263045  
##  3rd Qu.:   252164   3rd Qu.:   245484   3rd Qu.:   210788  
##  Max.   :205067466   Max.   :105248041   Max.   :222511278</code></pre>
<p>To compute the CPM values for each sample (excluding the <code>width</code> column):</p>
<div class="sourceCode" id="cb386"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb386-1"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb386-1"></a>cpm &lt;-<span class="st"> </span><span class="kw">apply</span>(<span class="kw">subset</span>(counts, <span class="dt">select =</span> <span class="kw">c</span>(<span class="op">-</span>width)), <span class="dv">2</span>, </span>
<span id="cb386-2"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb386-2"></a>             <span class="cf">function</span>(x) x<span class="op">/</span><span class="kw">sum</span>(<span class="kw">as.numeric</span>(x)) <span class="op">*</span><span class="st"> </span><span class="dv">10</span><span class="op">^</span><span class="dv">6</span>)</span></code></pre></div>
<p>Check that the sum of each column after normalization equals to 10^6 (except the width column).</p>
<div class="sourceCode" id="cb387"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb387-1"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb387-1"></a><span class="kw">colSums</span>(cpm)</span></code></pre></div>
<pre><code>## CASE_1 CASE_2 CASE_3 CASE_4 CASE_5 CTRL_1 CTRL_2 CTRL_3 CTRL_4 CTRL_5 
##  1e+06  1e+06  1e+06  1e+06  1e+06  1e+06  1e+06  1e+06  1e+06  1e+06</code></pre>
</div>
<div id="computing-rpkm" class="section level4">
<h4><span class="header-section-number">8.3.5.2</span> Computing RPKM</h4>
<div class="sourceCode" id="cb389"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb389-1"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb389-1"></a><span class="co"># create a vector of gene lengths </span></span>
<span id="cb389-2"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb389-2"></a>geneLengths &lt;-<span class="st"> </span><span class="kw">as.vector</span>(<span class="kw">subset</span>(counts, <span class="dt">select =</span> <span class="kw">c</span>(width)))</span>
<span id="cb389-3"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb389-3"></a></span>
<span id="cb389-4"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb389-4"></a><span class="co"># compute rpkm </span></span>
<span id="cb389-5"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb389-5"></a>rpkm &lt;-<span class="st"> </span><span class="kw">apply</span>(<span class="dt">X =</span> <span class="kw">subset</span>(counts, <span class="dt">select =</span> <span class="kw">c</span>(<span class="op">-</span>width)),</span>
<span id="cb389-6"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb389-6"></a>              <span class="dt">MARGIN =</span> <span class="dv">2</span>, </span>
<span id="cb389-7"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb389-7"></a>              <span class="dt">FUN =</span> <span class="cf">function</span>(x) {</span>
<span id="cb389-8"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb389-8"></a>                <span class="dv">10</span><span class="op">^</span><span class="dv">9</span> <span class="op">*</span><span class="st"> </span>x <span class="op">/</span><span class="st"> </span>geneLengths <span class="op">/</span><span class="st"> </span><span class="kw">sum</span>(<span class="kw">as.numeric</span>(x))</span>
<span id="cb389-9"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb389-9"></a>               })</span></code></pre></div>
<p>Check the sample sizes of RPKM. Notice that the sums of samples are all different.</p>
<div class="sourceCode" id="cb390"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb390-1"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb390-1"></a><span class="kw">colSums</span>(rpkm)</span></code></pre></div>
<pre><code>##   CASE_1   CASE_2   CASE_3   CASE_4   CASE_5   CTRL_1   CTRL_2   CTRL_3 
## 158291.0 153324.2 161775.4 173047.4 172761.4 210032.6 301764.2 241418.3 
##   CTRL_4   CTRL_5 
## 291674.5 252005.7</code></pre>
</div>
<div id="computing-tpm" class="section level4">
<h4><span class="header-section-number">8.3.5.3</span> Computing TPM</h4>
<div class="sourceCode" id="cb392"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb392-1"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb392-1"></a><span class="co">#find gene length normalized values </span></span>
<span id="cb392-2"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb392-2"></a>rpk &lt;-<span class="st"> </span><span class="kw">apply</span>( <span class="kw">subset</span>(counts, <span class="dt">select =</span> <span class="kw">c</span>(<span class="op">-</span>width)), <span class="dv">2</span>, </span>
<span id="cb392-3"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb392-3"></a>              <span class="cf">function</span>(x) x<span class="op">/</span>(geneLengths<span class="op">/</span><span class="dv">1000</span>))</span>
<span id="cb392-4"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb392-4"></a><span class="co">#normalize by the sample size using rpk values</span></span>
<span id="cb392-5"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb392-5"></a>tpm &lt;-<span class="st"> </span><span class="kw">apply</span>(rpk, <span class="dv">2</span>, <span class="cf">function</span>(x) x <span class="op">/</span><span class="st"> </span><span class="kw">sum</span>(<span class="kw">as.numeric</span>(x)) <span class="op">*</span><span class="st"> </span><span class="dv">10</span><span class="op">^</span><span class="dv">6</span>)</span></code></pre></div>
<p>Check the sample sizes of <code>tpm</code>. Notice that the sums of samples are all equal to 10^6.</p>
<div class="sourceCode" id="cb393"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb393-1"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb393-1"></a><span class="kw">colSums</span>(tpm)</span></code></pre></div>
<pre><code>## CASE_1 CASE_2 CASE_3 CASE_4 CASE_5 CTRL_1 CTRL_2 CTRL_3 CTRL_4 CTRL_5 
##  1e+06  1e+06  1e+06  1e+06  1e+06  1e+06  1e+06  1e+06  1e+06  1e+06</code></pre>
<p>None of these metrics (CPM, RPKM/FPKM, TPM) account for the other important confounding factor when comparing expression levels of genes across samples: the <strong>library composition</strong>, which may also be referred to as the <strong>relative size of the compared transcriptomes</strong>. This factor is not dependent on the sequencing technology, it is rather biological. For instance, when comparing transcriptomes of different tissues, there can be sets of genes in one tissue that consume a big chunk of the reads, while in the other tissues they are not expressed at all. This kind of imbalance in the composition of compared transcriptomes can lead to wrong conclusions about which genes are actually differentially expressed. This consideration is addressed in two popular R packages: <code>DESeq2</code> <span class="citation">(Love, Huber, and Anders <a href="#ref-love_moderated_2014" role="doc-biblioref">2014</a>)</span> and edgeR <span class="citation">(Robinson, McCarthy, and Smyth <a href="#ref-robinson_edger:_2010" role="doc-biblioref">2010</a>)</span> each with a different algorithm. <code>edgeR</code> uses a normalization procedure called Trimmed Mean of M-values (TMM). <code>DESeq2</code> implements a normalization procedure using median of Ratios, which is obtained by finding the ratio of the log-transformed count of a gene divided by the average of log-transformed values of the gene in all samples (geometric mean), and then taking the median of these values for all genes. The raw read count of the gene is finally divided by this value (median of ratios) to obtain the normalized counts.</p>
</div>
</div>
<div id="exploratory-analysis-of-the-read-count-table" class="section level3">
<h3><span class="header-section-number">8.3.6</span> Exploratory analysis of the read count table</h3>
<p>A typical quality control, in this case interrogating the RNA-seq experiment design, is to measure the similarity of the samples with each other in terms of the quantified expression level profiles across a set of genes. One important observation to make is to see whether the most similar samples to any given sample are the biological replicates of that sample. This can be computed using unsupervised clustering techniques such as hierarchical clustering and visualized as a heatmap with dendrograms. Another most commonly applied technique is a dimensionality reduction technique called Principal Component Analysis (PCA) and visualized as a two-dimensional (or in some cases three-dimensional) scatter plot. In order to find out more about the clustering methods and PCA, please refer to Chapter <a href="unsupervisedLearning.html#unsupervisedLearning">4</a>.</p>
<div id="clustering-1" class="section level4">
<h4><span class="header-section-number">8.3.6.1</span> Clustering</h4>
<p>We can combine clustering and visualization of the clustering results by using heatmap functions that are available in a variety of R libraries. The basic R installation comes with the <code>stats::heatmap</code> function. However, there are other libraries available in CRAN (e.g. <code>pheatmap</code> <span class="citation">(Kolde <a href="#ref-pheatmap" role="doc-biblioref">2019</a>)</span>) or Bioconductor (e.g. <code>ComplexHeatmap</code> <span class="citation">(Z. Gu, Eils, and Schlesner <a href="#ref-gu_complex_2016" role="doc-biblioref">2016</a><a href="#ref-gu_complex_2016" role="doc-biblioref">a</a>)</span>) that come with more flexibility and more appealing visualizations.</p>
<p>Here we demonstrate a heatmap using the <code>pheatmap</code> package and the previously calculated <code>tpm</code> matrix.
As these matrices can be quite large, both computing the clustering and rendering the heatmaps can take a lot of resources and time. Therefore, a quick and informative way to compare samples is to select a subset of genes that are, for instance, most variable across samples, and use that subset to do the clustering and visualization.</p>
<p>Let’s select the top 100 most variable genes among the samples.</p>
<div class="sourceCode" id="cb395"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb395-1"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb395-1"></a><span class="co">#compute the variance of each gene across samples</span></span>
<span id="cb395-2"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb395-2"></a>V &lt;-<span class="st"> </span><span class="kw">apply</span>(tpm, <span class="dv">1</span>, var)</span>
<span id="cb395-3"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb395-3"></a><span class="co">#sort the results by variance in decreasing order </span></span>
<span id="cb395-4"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb395-4"></a><span class="co">#and select the top 100 genes </span></span>
<span id="cb395-5"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb395-5"></a>selectedGenes &lt;-<span class="st"> </span><span class="kw">names</span>(V[<span class="kw">order</span>(V, <span class="dt">decreasing =</span> T)][<span class="dv">1</span><span class="op">:</span><span class="dv">100</span>])</span></code></pre></div>
<p>Now we can quickly produce a heatmap where samples and genes are clustered (see Figure <a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#fig:tpmhierClust1">8.1</a> ).</p>
<div class="sourceCode" id="cb396"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb396-1"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb396-1"></a><span class="kw">library</span>(pheatmap)</span>
<span id="cb396-2"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb396-2"></a><span class="kw">pheatmap</span>(tpm[selectedGenes,], <span class="dt">scale =</span> <span class="st">&#39;row&#39;</span>, <span class="dt">show_rownames =</span> <span class="ot">FALSE</span>)</span></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:tpmhierClust1"></span>
<img src="08-rna-seq-analysis_files/figure-html/tpmhierClust1-1.png" alt="Clustering and visualization of the topmost variable genes as a heatmap." width="50%" />
<p class="caption">
FIGURE 8.1: Clustering and visualization of the topmost variable genes as a heatmap.
</p>
</div>
<p>We can also overlay some annotation tracks to observe the clusters.
Here it is important to observe whether the replicates of the same sample cluster most closely with each other, or not. Overlaying the heatmap with such annotation and displaying sample groups with distinct colors helps quickly see if there are samples that don’t cluster as expected (see Figure <a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#fig:tpmhierclust2">8.2</a> ).</p>
<div class="sourceCode" id="cb397"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb397-1"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb397-1"></a>colData &lt;-<span class="st"> </span><span class="kw">read.table</span>(coldata_file, <span class="dt">header =</span> T, <span class="dt">sep =</span> <span class="st">&#39;</span><span class="ch">\t</span><span class="st">&#39;</span>, </span>
<span id="cb397-2"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb397-2"></a>                      <span class="dt">stringsAsFactors =</span> <span class="ot">TRUE</span>)</span>
<span id="cb397-3"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb397-3"></a><span class="kw">pheatmap</span>(tpm[selectedGenes,], <span class="dt">scale =</span> <span class="st">&#39;row&#39;</span>, </span>
<span id="cb397-4"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb397-4"></a>         <span class="dt">show_rownames =</span> <span class="ot">FALSE</span>, </span>
<span id="cb397-5"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb397-5"></a>         <span class="dt">annotation_col =</span> colData)</span></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:tpmhierclust2"></span>
<img src="08-rna-seq-analysis_files/figure-html/tpmhierclust2-1.png" alt="Clustering samples as a heatmap with sample annotations." width="60%" />
<p class="caption">
FIGURE 8.2: Clustering samples as a heatmap with sample annotations.
</p>
</div>
</div>
<div id="pca" class="section level4">
<h4><span class="header-section-number">8.3.6.2</span> PCA</h4>
<p>Let’s make a PCA plot  to see the clustering of replicates as a scatter plot in two dimensions (Figure <a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#fig:pca1">8.3</a>).</p>
<div class="sourceCode" id="cb398"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb398-1"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb398-1"></a><span class="kw">library</span>(stats)</span>
<span id="cb398-2"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb398-2"></a><span class="kw">library</span>(ggplot2)</span>
<span id="cb398-3"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb398-3"></a><span class="co">#transpose the matrix </span></span>
<span id="cb398-4"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb398-4"></a>M &lt;-<span class="st"> </span><span class="kw">t</span>(tpm[selectedGenes,])</span>
<span id="cb398-5"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb398-5"></a><span class="co"># transform the counts to log2 scale </span></span>
<span id="cb398-6"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb398-6"></a>M &lt;-<span class="st"> </span><span class="kw">log2</span>(M <span class="op">+</span><span class="st"> </span><span class="dv">1</span>)</span>
<span id="cb398-7"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb398-7"></a><span class="co">#compute PCA </span></span>
<span id="cb398-8"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb398-8"></a>pcaResults &lt;-<span class="st"> </span><span class="kw">prcomp</span>(M)</span>
<span id="cb398-9"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb398-9"></a></span>
<span id="cb398-10"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb398-10"></a><span class="co">#plot PCA results making use of ggplot2&#39;s autoplot function</span></span>
<span id="cb398-11"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb398-11"></a><span class="co">#ggfortify is needed to let ggplot2 know about PCA data structure. </span></span>
<span id="cb398-12"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb398-12"></a><span class="kw">autoplot</span>(pcaResults, <span class="dt">data =</span> colData, <span class="dt">colour =</span> <span class="st">&#39;group&#39;</span>)</span></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:pca1"></span>
<img src="08-rna-seq-analysis_files/figure-html/pca1-1.png" alt="PCA plot of samples using TPM counts." width="55%" />
<p class="caption">
FIGURE 8.3: PCA plot of samples using TPM counts.
</p>
</div>
<p>We should observe here whether the samples from the case group (CASE) and samples from the control group (CTRL) can be split into two distinct clusters on the scatter plot of the first two largest principal components.</p>
<p>We can use the <code>summary</code> function to summarize the PCA results to observe the contribution of the principal components in the explained variation.</p>
<div class="sourceCode" id="cb399"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb399-1"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb399-1"></a><span class="kw">summary</span>(pcaResults)</span></code></pre></div>
<pre><code>## Importance of components:
##                           PC1     PC2     PC3     PC4     PC5    PC6     PC7
## Standard deviation     24.396 2.50514 2.39327 1.93841 1.79193 1.6357 1.46059
## Proportion of Variance  0.957 0.01009 0.00921 0.00604 0.00516 0.0043 0.00343
## Cumulative Proportion   0.957 0.96706 0.97627 0.98231 0.98747 0.9918 0.99520
##                            PC8     PC9      PC10
## Standard deviation     1.30902 1.12657 4.616e-15
## Proportion of Variance 0.00276 0.00204 0.000e+00
## Cumulative Proportion  0.99796 1.00000 1.000e+00</code></pre>
</div>
<div id="correlation-plots" class="section level4">
<h4><span class="header-section-number">8.3.6.3</span> Correlation plots</h4>
<p>Another complementary approach to see the reproducibility of the experiments is to compute the correlation scores between each pair of samples and draw a correlation plot. </p>
<p>Let’s first compute pairwise correlation scores between every pair of samples.</p>
<div class="sourceCode" id="cb401"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb401-1"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb401-1"></a><span class="kw">library</span>(stats)</span>
<span id="cb401-2"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb401-2"></a>correlationMatrix &lt;-<span class="st"> </span><span class="kw">cor</span>(tpm)</span></code></pre></div>
<p>Let’s have a look at how the correlation matrix looks (<a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#tab:corrplot2">8.1</a>)
(showing only two samples each of case and control samples):</p>
<table>
<caption><span id="tab:corrplot2">TABLE 8.1: </span>Correlation scores between samples</caption>
<thead>
<tr class="header">
<th></th>
<th align="right">CASE_1</th>
<th align="right">CASE_2</th>
<th align="right">CTRL_1</th>
<th align="right">CTRL_2</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>CASE_1</td>
<td align="right">1.0000000</td>
<td align="right">0.9924606</td>
<td align="right">0.9594011</td>
<td align="right">0.9635760</td>
</tr>
<tr class="even">
<td>CASE_2</td>
<td align="right">0.9924606</td>
<td align="right">1.0000000</td>
<td align="right">0.9725646</td>
<td align="right">0.9793835</td>
</tr>
<tr class="odd">
<td>CTRL_1</td>
<td align="right">0.9594011</td>
<td align="right">0.9725646</td>
<td align="right">1.0000000</td>
<td align="right">0.9879862</td>
</tr>
<tr class="even">
<td>CTRL_2</td>
<td align="right">0.9635760</td>
<td align="right">0.9793835</td>
<td align="right">0.9879862</td>
<td align="right">1.0000000</td>
</tr>
</tbody>
</table>
<p>We can also draw more visually appealing correlation plots using the <code>corrplot</code> package (Figure <a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#fig:corrplot3">8.4</a>).
Using the <code>addrect</code> argument, we can split clusters into groups and surround them with rectangles.
By setting the <code>addCoef.col</code> argument to ‘white’, we can display the correlation coefficients as numbers in white color.</p>
<div class="sourceCode" id="cb402"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb402-1"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb402-1"></a><span class="kw">library</span>(corrplot)</span>
<span id="cb402-2"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb402-2"></a><span class="kw">corrplot</span>(correlationMatrix, <span class="dt">order =</span> <span class="st">&#39;hclust&#39;</span>, </span>
<span id="cb402-3"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb402-3"></a>         <span class="dt">addrect =</span> <span class="dv">2</span>, <span class="dt">addCoef.col =</span> <span class="st">&#39;white&#39;</span>, </span>
<span id="cb402-4"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb402-4"></a>         <span class="dt">number.cex =</span> <span class="fl">0.7</span>) </span></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:corrplot3"></span>
<img src="08-rna-seq-analysis_files/figure-html/corrplot3-1.png" alt="Correlation plot of samples ordered by hierarchical clustering." width="55%" />
<p class="caption">
FIGURE 8.4: Correlation plot of samples ordered by hierarchical clustering.
</p>
</div>
<p>Here pairwise correlation levels are visualized as colored circles. <code>Blue</code> indicates positive correlation, while <code>Red</code> indicates negative correlation.</p>
<p>We could also plot this correlation matrix as a heatmap (Figure <a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#fig:corrplot4">8.5</a>). As all the samples have a high pairwise
correlation score, using a heatmap instead of a corrplot helps to see the differences between samples more easily. The
<code>annotation_col</code> argument helps to display sample annotations and the <code>cutree_cols</code> argument is set to 2 to split the clusters into two groups based on the hierarchical clustering results.</p>
<div class="sourceCode" id="cb403"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb403-1"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb403-1"></a><span class="kw">library</span>(pheatmap)</span>
<span id="cb403-2"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb403-2"></a><span class="co"># split the clusters into two based on the clustering similarity </span></span>
<span id="cb403-3"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb403-3"></a><span class="kw">pheatmap</span>(correlationMatrix,  </span>
<span id="cb403-4"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb403-4"></a>         <span class="dt">annotation_col =</span> colData, </span>
<span id="cb403-5"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb403-5"></a>         <span class="dt">cutree_cols =</span> <span class="dv">2</span>)</span></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:corrplot4"></span>
<img src="08-rna-seq-analysis_files/figure-html/corrplot4-1.png" alt="Pairwise correlation of samples displayed as a heatmap." width="55%" />
<p class="caption">
FIGURE 8.5: Pairwise correlation of samples displayed as a heatmap.
</p>
</div>
</div>
</div>
<div id="differential-expression-analysis" class="section level3">
<h3><span class="header-section-number">8.3.7</span> Differential expression analysis</h3>
<p>Differential expression analysis allows us to test tens of thousands of hypotheses (one test for each gene) against the null hypothesis that the activity of the gene stays the same in two different conditions. There are multiple limiting factors that influence the power of detecting genes that have real changes between two biological conditions. Among these are the limited number of biological replicates, non-normality of the distribution of the read counts, and higher uncertainty of measurements for lowly expressed genes than highly expressed genes <span class="citation">(Love, Huber, and Anders <a href="#ref-love_moderated_2014" role="doc-biblioref">2014</a>)</span>. Tools such as <code>edgeR</code> and <code>DESeq2</code> address these limitations using sophisticated statistical models in order to maximize the amount of knowledge that can be extracted from such noisy datasets. In essence, these models assume that for each gene, the read counts are generated by a negative binomial distribution. This is a popular distribution that is used for modeling count data. This distribution can be specified with a mean parameter, <span class="math inline">\(m\)</span>, and a dispersion parameter, <span class="math inline">\(\alpha\)</span>. The dispersion parameter <span class="math inline">\(\alpha\)</span> is directly related to the variance as the variance of this distribution is formulated as: <span class="math inline">\(m+\alpha m^{2}\)</span>. Therefore, estimating these parameters is crucial for differential expression tests. The methods used in <code>edgeR</code> and <code>DESeq2</code> use dispersion estimates from other genes with similar counts to precisely estimate the per-gene dispersion values. With accurate dispersion parameter estimates, one can estimate the variance more precisely, which in turn
improves the result of the differential expression test. Although statistical models are different, the process here is similar to the moderated t-test and qualifies as an empirical Bayes method  which we introduced in Chapter <a href="stats.html#stats">3</a>. There, we calculated gene-wise variability and shrunk each gene-wise variability towards the median variability of all genes. In the case of RNA-seq the dispersion coefficient <span class="math inline">\(\alpha\)</span> is shrunk towards the value of dispersion from other genes with similar read counts.</p>
<p>Now let us take a closer look at the <code>DESeq2</code> workflow and how it calculates differential expression:</p>
<ol style="list-style-type: decimal">
<li>The read counts are normalized by computing size factors, which addresses the differences not only in the library sizes, but also the library compositions.</li>
<li>For each gene, a dispersion estimate is calculated. The dispersion value computed by <code>DESeq2</code> is equal to the squared coefficient of variation (variation divided by the mean).</li>
<li>A line is fit across the dispersion estimates of all genes computed in step 2 versus the mean normalized counts of the genes.</li>
<li>Dispersion values of each gene are shrunk towards the fitted line in step 3.</li>
<li>A Generalized Linear Model is fitted which considers additional confounding variables related to the experimental design such as sequencing batches, treatment, temperature, patient’s age, sequencing technology, etc., and uses negative binomial distribution for fitting count data.</li>
<li>For a given contrast (e.g. treatment type: drug-A versus untreated), a test for differential expression is carried out against the null hypothesis that the log fold change of the normalized counts of the gene in the given pair of groups is exactly zero.</li>
<li>It adjusts p-values for multiple-testing.</li>
</ol>
<p>In order to carry out a differential expression analysis using <code>DESeq2</code>, three kinds of inputs are necessary:</p>
<ol style="list-style-type: decimal">
<li>The <strong>read count table</strong>: This table must be raw read counts as integers that are not processed in any form by a normalization technique. The rows represent features (e.g. genes, transcripts, genomic intervals) and columns represent samples.</li>
<li>A <strong>colData</strong> table: This table describes the experimental design.</li>
<li>A <strong>design formula</strong>: This formula is needed to describe the variable of interest in the analysis (e.g. treatment status) along with (optionally) other covariates (e.g. batch, temperature, sequencing technology).</li>
</ol>
<p>Let’s define these inputs:</p>
<div class="sourceCode" id="cb404"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb404-1"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb404-1"></a><span class="co">#remove the &#39;width&#39; column</span></span>
<span id="cb404-2"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb404-2"></a>countData &lt;-<span class="st"> </span><span class="kw">as.matrix</span>(<span class="kw">subset</span>(counts, <span class="dt">select =</span> <span class="kw">c</span>(<span class="op">-</span>width)))</span>
<span id="cb404-3"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb404-3"></a><span class="co">#define the experimental setup </span></span>
<span id="cb404-4"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb404-4"></a>colData &lt;-<span class="st"> </span><span class="kw">read.table</span>(coldata_file, <span class="dt">header =</span> T, <span class="dt">sep =</span> <span class="st">&#39;</span><span class="ch">\t</span><span class="st">&#39;</span>, </span>
<span id="cb404-5"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb404-5"></a>                      <span class="dt">stringsAsFactors =</span> <span class="ot">TRUE</span>)</span>
<span id="cb404-6"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb404-6"></a><span class="co">#define the design formula</span></span>
<span id="cb404-7"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb404-7"></a>designFormula &lt;-<span class="st"> &quot;~ group&quot;</span></span></code></pre></div>
<p>Now, we are ready to run <code>DESeq2</code>.</p>
<div class="sourceCode" id="cb405"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb405-1"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb405-1"></a><span class="kw">library</span>(DESeq2)</span>
<span id="cb405-2"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb405-2"></a><span class="kw">library</span>(stats)</span>
<span id="cb405-3"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb405-3"></a><span class="co">#create a DESeq dataset object from the count matrix and the colData </span></span>
<span id="cb405-4"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb405-4"></a>dds &lt;-<span class="st"> </span><span class="kw">DESeqDataSetFromMatrix</span>(<span class="dt">countData =</span> countData, </span>
<span id="cb405-5"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb405-5"></a>                              <span class="dt">colData =</span> colData, </span>
<span id="cb405-6"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb405-6"></a>                              <span class="dt">design =</span> <span class="kw">as.formula</span>(designFormula))</span>
<span id="cb405-7"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb405-7"></a><span class="co">#print dds object to see the contents</span></span>
<span id="cb405-8"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb405-8"></a><span class="kw">print</span>(dds)</span></code></pre></div>
<pre><code>## class: DESeqDataSet 
## dim: 19719 10 
## metadata(1): version
## assays(1): counts
## rownames(19719): TSPAN6 TNMD ... MYOCOS HSFX3
## rowData names(0):
## colnames(10): CASE_1 CASE_2 ... CTRL_4 CTRL_5
## colData names(2): source_name group</code></pre>
<p>The <code>DESeqDataSet</code> object contains all the information about the experimental setup, the read counts, and the design formulas. Certain functions can be used to access this information separately: <code>rownames(dds)</code> shows which features are used in the study (e.g. genes), <code>colnames(dds)</code> displays the studied samples, <code>counts(dds)</code> displays the count table, and <code>colData(dds)</code> displays the experimental setup.</p>
<p>Remove genes that have almost no information in any of the given samples.</p>
<div class="sourceCode" id="cb407"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb407-1"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb407-1"></a><span class="co">#For each gene, we count the total number of reads for that gene in all samples </span></span>
<span id="cb407-2"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb407-2"></a><span class="co">#and remove those that don&#39;t have at least 1 read. </span></span>
<span id="cb407-3"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb407-3"></a>dds &lt;-<span class="st"> </span>dds[ <span class="kw">rowSums</span>(DESeq2<span class="op">::</span><span class="kw">counts</span>(dds)) <span class="op">&gt;</span><span class="st"> </span><span class="dv">1</span>, ]</span></code></pre></div>
<p>Now, we can use the <code>DESeq()</code> function of <code>DESeq2</code>, which is a wrapper function that implements estimation of size factors to normalize the counts, estimation of dispersion values, and computing a GLM model based on the experimental design formula. This function returns a <code>DESeqDataSet</code> object, which is an updated version of the <code>dds</code> variable that we pass to the function as input.</p>
<div class="sourceCode" id="cb408"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb408-1"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb408-1"></a>dds &lt;-<span class="st"> </span><span class="kw">DESeq</span>(dds)</span></code></pre></div>
<p>Now, we can compare and contrast the samples based on different variables of interest. In this case, we currently have only one variable, which is the <code>group</code> variable that determines if a sample belongs to the CASE group or the CTRL group.</p>
<div class="sourceCode" id="cb409"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb409-1"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb409-1"></a><span class="co">#compute the contrast for the &#39;group&#39; variable where &#39;CTRL&#39; </span></span>
<span id="cb409-2"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb409-2"></a><span class="co">#samples are used as the control group. </span></span>
<span id="cb409-3"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb409-3"></a>DEresults =<span class="st"> </span><span class="kw">results</span>(dds, <span class="dt">contrast =</span> <span class="kw">c</span>(<span class="st">&quot;group&quot;</span>, <span class="st">&#39;CASE&#39;</span>, <span class="st">&#39;CTRL&#39;</span>))</span>
<span id="cb409-4"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb409-4"></a><span class="co">#sort results by increasing p-value</span></span>
<span id="cb409-5"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb409-5"></a>DEresults &lt;-<span class="st"> </span>DEresults[<span class="kw">order</span>(DEresults<span class="op">$</span>pvalue),]</span></code></pre></div>
<p>Thus we have obtained a table containing the differential expression status of case samples compared to the control samples.</p>
<p>It is important to note that the sequence of the elements provided in the <code>contrast</code> argument determines which group of samples are to be used as the control. This impacts the way the results are interpreted, for instance, if a gene is found up-regulated (has a positive log2 fold change), the up-regulation status is only relative to the factor that is provided as control. In this case, we used samples from the “CTRL” group as control and contrasted the samples from the “CASE” group with respect to the “CTRL” samples. Thus genes with a positive log2 fold change are called up-regulated in the case samples with respect to the control, while genes with a negative log2 fold change are down-regulated in the case samples. Whether the deregulation is significant or not, warrants assessment of the adjusted p-values.</p>
<p>Let’s have a look at the contents of the <code>DEresults</code> table.</p>
<div class="sourceCode" id="cb410"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb410-1"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb410-1"></a><span class="co">#shows a summary of the results</span></span>
<span id="cb410-2"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb410-2"></a><span class="kw">print</span>(DEresults)</span></code></pre></div>
<pre><code>## log2 fold change (MLE): group CASE vs CTRL 
## Wald test p-value: group CASE vs CTRL 
## DataFrame with 19097 rows and 6 columns
##             baseMean log2FoldChange     lfcSE       stat       pvalue
##            &lt;numeric&gt;      &lt;numeric&gt; &lt;numeric&gt;  &lt;numeric&gt;    &lt;numeric&gt;
## CYP2E1       4829889        9.36024  0.215223    43.4909  0.00000e+00
## FCGBP       10349993       -7.57579  0.186433   -40.6355  0.00000e+00
## ASGR2         426422        8.01830  0.216207    37.0863 4.67898e-301
## GCKR          100183        7.82841  0.233376    33.5442 1.09479e-246
## APOA5         438054       10.20248  0.312503    32.6477 8.64906e-234
## ...              ...            ...       ...        ...          ...
## CCDC195      20.4981      -0.215607   2.89255 -0.0745386           NA
## SPEM3        23.6370     -22.154765   3.02785 -7.3170030           NA
## AC022167.5   21.8451      -2.056240   2.89545 -0.7101618           NA
## BX276092.9   29.9636       0.407326   2.89048  0.1409199           NA
## ETDC         22.5675      -1.795274   2.89421 -0.6202983           NA
##                    padj
##               &lt;numeric&gt;
## CYP2E1      0.00000e+00
## FCGBP       0.00000e+00
## ASGR2      2.87741e-297
## GCKR       5.04945e-243
## APOA5      3.19133e-230
## ...                 ...
## CCDC195              NA
## SPEM3                NA
## AC022167.5           NA
## BX276092.9           NA
## ETDC                 NA</code></pre>
<p>The first three lines in this output show the contrast and the statistical test that were used to compute these results, along with the dimensions of the resulting table (number of columns and rows). Below these lines is the actual table with 6 columns: <code>baseMean</code> represents the average normalized expression of the gene across all considered samples. <code>log2FoldChange</code> represents the base-2 logarithm of the fold change of the normalized expression of the gene in the given contrast. <code>lfcSE</code> represents the standard error of log2 fold change estimate, and <code>stat</code> is the statistic calculated in the contrast which is translated into a <code>pvalue</code> and adjusted for multiple testing in the <code>padj</code> column. To find out about the importance of adjusting for multiple testing, see Chapter <a href="stats.html#stats">3</a>.</p>
<div id="diagnostic-plots" class="section level4">
<h4><span class="header-section-number">8.3.7.1</span> Diagnostic plots</h4>
<p>At this point, before proceeding to do any downstream analysis and jumping to conclusions about the biological insights that are reachable with the experimental data at hand, it is important to do some more diagnostic tests to improve our confidence about the quality of the data and the experimental setup.</p>
<div id="ma-plot" class="section level5">
<h5><span class="header-section-number">8.3.7.1.1</span> MA plot</h5>
<p>An MA plot is useful to observe if the data normalization worked well (Figure <a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#fig:DEmaplot">8.6</a>). The MA plot is a scatter plot where the x-axis denotes the average of normalized counts across samples and the y-axis denotes the log fold change in the given contrast. Most points are expected to be on the horizontal 0 line (most genes are not expected to be differentially expressed).</p>
<div class="sourceCode" id="cb412"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb412-1"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb412-1"></a><span class="kw">library</span>(DESeq2)</span>
<span id="cb412-2"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb412-2"></a>DESeq2<span class="op">::</span><span class="kw">plotMA</span>(<span class="dt">object =</span> dds, <span class="dt">ylim =</span> <span class="kw">c</span>(<span class="op">-</span><span class="dv">5</span>, <span class="dv">5</span>))</span></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:DEmaplot"></span>
<img src="08-rna-seq-analysis_files/figure-html/DEmaplot-1.png" alt="MA plot of differential expression results." width="55%" />
<p class="caption">
FIGURE 8.6: MA plot of differential expression results.
</p>
</div>
</div>
<div id="p-value-distribution" class="section level5">
<h5><span class="header-section-number">8.3.7.1.2</span> P-value distribution</h5>
<p>It is also important to observe the distribution of raw p-values (Figure <a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#fig:DEpvaldist">8.7</a>). We expect to see a peak around low p-values and a uniform distribution at P-values above 0.1. Otherwise, adjustment for multiple testing does not work and the results are not meaningful.</p>
<div class="sourceCode" id="cb413"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb413-1"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb413-1"></a><span class="kw">library</span>(ggplot2)</span>
<span id="cb413-2"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb413-2"></a><span class="kw">ggplot</span>(<span class="dt">data =</span> <span class="kw">as.data.frame</span>(DEresults), <span class="kw">aes</span>(<span class="dt">x =</span> pvalue)) <span class="op">+</span><span class="st"> </span></span>
<span id="cb413-3"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb413-3"></a><span class="st">  </span><span class="kw">geom_histogram</span>(<span class="dt">bins =</span> <span class="dv">100</span>)</span></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:DEpvaldist"></span>
<img src="08-rna-seq-analysis_files/figure-html/DEpvaldist-1.png" alt="P-value distribution genes before adjusting for multiple testing." width="55%" />
<p class="caption">
FIGURE 8.7: P-value distribution genes before adjusting for multiple testing.
</p>
</div>
</div>
<div id="pca-plot" class="section level5">
<h5><span class="header-section-number">8.3.7.1.3</span> PCA plot</h5>
<p>A final diagnosis is to check the biological reproducibility of the sample replicates in a PCA plot or a heatmap. To plot the PCA results, we need to extract the normalized counts from the DESeqDataSet object. It is possible to color the points in the scatter plot by the variable of interest, which helps to see if the replicates cluster well (Figure <a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#fig:DEpca">8.8</a>).</p>
<div class="sourceCode" id="cb414"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb414-1"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb414-1"></a><span class="kw">library</span>(DESeq2)</span>
<span id="cb414-2"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb414-2"></a><span class="co"># extract normalized counts from the DESeqDataSet object</span></span>
<span id="cb414-3"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb414-3"></a>countsNormalized &lt;-<span class="st"> </span>DESeq2<span class="op">::</span><span class="kw">counts</span>(dds, <span class="dt">normalized =</span> <span class="ot">TRUE</span>)</span>
<span id="cb414-4"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb414-4"></a></span>
<span id="cb414-5"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb414-5"></a><span class="co"># select top 500 most variable genes</span></span>
<span id="cb414-6"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb414-6"></a>selectedGenes &lt;-<span class="st"> </span><span class="kw">names</span>(<span class="kw">sort</span>(<span class="kw">apply</span>(countsNormalized, <span class="dv">1</span>, var), </span>
<span id="cb414-7"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb414-7"></a>                            <span class="dt">decreasing =</span> <span class="ot">TRUE</span>)[<span class="dv">1</span><span class="op">:</span><span class="dv">500</span>])</span>
<span id="cb414-8"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb414-8"></a></span>
<span id="cb414-9"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb414-9"></a><span class="kw">plotPCA</span>(countsNormalized[selectedGenes,], </span>
<span id="cb414-10"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb414-10"></a>        <span class="dt">col =</span> <span class="kw">as.numeric</span>(colData<span class="op">$</span>group), <span class="dt">adj =</span> <span class="fl">0.5</span>, </span>
<span id="cb414-11"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb414-11"></a>        <span class="dt">xlim =</span> <span class="kw">c</span>(<span class="op">-</span><span class="fl">0.5</span>, <span class="fl">0.5</span>), <span class="dt">ylim =</span> <span class="kw">c</span>(<span class="op">-</span><span class="fl">0.5</span>, <span class="fl">0.6</span>))</span></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:DEpca"></span>
<img src="08-rna-seq-analysis_files/figure-html/DEpca-1.png" alt="Principle component analysis plot based on top 500 most variable genes." width="55%" />
<p class="caption">
FIGURE 8.8: Principle component analysis plot based on top 500 most variable genes.
</p>
</div>
<p>Alternatively, the normalized counts can be transformed using the <code>DESeq2::rlog</code> function and <code>DESeq2::plotPCA()</code> can be readily used to plot the PCA results (Figure <a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#fig:DErldnorm">8.9</a>).</p>
<div class="sourceCode" id="cb415"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb415-1"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb415-1"></a>rld &lt;-<span class="st"> </span><span class="kw">rlog</span>(dds)</span>
<span id="cb415-2"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb415-2"></a>DESeq2<span class="op">::</span><span class="kw">plotPCA</span>(rld, <span class="dt">ntop =</span> <span class="dv">500</span>, <span class="dt">intgroup =</span> <span class="st">&#39;group&#39;</span>) <span class="op">+</span><span class="st"> </span></span>
<span id="cb415-3"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb415-3"></a><span class="st">  </span><span class="kw">ylim</span>(<span class="op">-</span><span class="dv">50</span>, <span class="dv">50</span>) <span class="op">+</span><span class="st"> </span><span class="kw">theme_bw</span>()</span></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:DErldnorm"></span>
<img src="08-rna-seq-analysis_files/figure-html/DErldnorm-1.png" alt="PCA plot of top 500 most variable genes." width="55%" />
<p class="caption">
FIGURE 8.9: PCA plot of top 500 most variable genes.
</p>
</div>
</div>
<div id="relative-log-expression-rle-plot" class="section level5">
<h5><span class="header-section-number">8.3.7.1.4</span> Relative Log Expression (RLE) plot</h5>
<p>A similar plot to the MA plot is the RLE (Relative Log Expression) plot that is useful in finding out if the data at hand needs normalization <span class="citation">(Gandolfo and Speed <a href="#ref-gandolfo_rle_2018" role="doc-biblioref">2018</a>)</span>. Sometimes, even the datasets normalized using the explained methods above may need further normalization due to unforeseen sources of variation that might stem from the library preparation, the person who carries out the experiment, the date of sequencing, the temperature changes in the laboratory at the time of library preparation, and so on and so forth. The RLE plot is a quick diagnostic that can be applied on the raw or normalized count matrices to see if further processing is required.</p>
<p>Let’s do RLE plots on the raw counts and normalized counts using the <code>EDASeq</code> package <span class="citation">(Risso, Schwartz, Sherlock, et al. <a href="#ref-risso_gc-content_2011" role="doc-biblioref">2011</a>)</span> (see Figure <a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#fig:DErleplot">8.10</a>).</p>
<div class="sourceCode" id="cb416"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb416-1"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb416-1"></a><span class="kw">library</span>(EDASeq)</span>
<span id="cb416-2"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb416-2"></a><span class="kw">par</span>(<span class="dt">mfrow =</span> <span class="kw">c</span>(<span class="dv">1</span>, <span class="dv">2</span>))</span>
<span id="cb416-3"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb416-3"></a><span class="kw">plotRLE</span>(countData, <span class="dt">outline=</span><span class="ot">FALSE</span>, <span class="dt">ylim=</span><span class="kw">c</span>(<span class="op">-</span><span class="dv">4</span>, <span class="dv">4</span>), </span>
<span id="cb416-4"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb416-4"></a>        <span class="dt">col=</span><span class="kw">as.numeric</span>(colData<span class="op">$</span>group), </span>
<span id="cb416-5"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb416-5"></a>        <span class="dt">main =</span> <span class="st">&#39;Raw Counts&#39;</span>)</span>
<span id="cb416-6"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb416-6"></a><span class="kw">plotRLE</span>(DESeq2<span class="op">::</span><span class="kw">counts</span>(dds, <span class="dt">normalized =</span> <span class="ot">TRUE</span>), </span>
<span id="cb416-7"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb416-7"></a>        <span class="dt">outline=</span><span class="ot">FALSE</span>, <span class="dt">ylim=</span><span class="kw">c</span>(<span class="op">-</span><span class="dv">4</span>, <span class="dv">4</span>), </span>
<span id="cb416-8"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb416-8"></a>        <span class="dt">col =</span> <span class="kw">as.numeric</span>(colData<span class="op">$</span>group), </span>
<span id="cb416-9"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb416-9"></a>        <span class="dt">main =</span> <span class="st">&#39;Normalized Counts&#39;</span>)</span></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:DErleplot"></span>
<img src="08-rna-seq-analysis_files/figure-html/DErleplot-1.png" alt="Relative log expression plots based on raw and normalized count matrices" width="55%" />
<p class="caption">
FIGURE 8.10: Relative log expression plots based on raw and normalized count matrices
</p>
</div>
<p>Here the RLE plot is comprised of boxplots, where each box-plot represents the distribution of the relative log expression of the genes expressed in the corresponding sample. Each gene’s expression is divided by the median expression value of that gene across all samples. Then this is transformed to log scale, which gives the relative log expression value for a single gene. The RLE values for all the genes from a sample are visualized as a boxplot.</p>
<p>Ideally the boxplots are centered around the horizontal zero line and are as tightly distributed as possible <span class="citation">(Risso, Ngai, Speed, et al. <a href="#ref-risso_normalization_2014" role="doc-biblioref">2014</a>)</span>. From the plots that we have made for the raw and normalized count data, we can observe how the normalized dataset has improved upon the raw count data for all the samples. However, in some cases, it is important to visualize RLE plots in combination with other diagnostic plots such as PCA plots, heatmaps, and correlation plots to see if there is more unwanted variation in the data, which can be further accounted for using packages such as <code>RUVSeq</code> <span class="citation">(Risso, Ngai, Speed, et al. <a href="#ref-risso_normalization_2014" role="doc-biblioref">2014</a>)</span> and <code>sva</code> <span class="citation">(Leek, Johnson, Parker, et al. <a href="#ref-leek_sva_2012" role="doc-biblioref">2012</a>)</span>. We will cover details about the <code>RUVSeq</code> package to account for unwanted sources of noise in RNA-seq datasets in later sections.</p>
</div>
</div>
</div>
<div id="functional-enrichment-analysis" class="section level3">
<h3><span class="header-section-number">8.3.8</span> Functional enrichment analysis</h3>
<div id="go-term-analysis" class="section level4">
<h4><span class="header-section-number">8.3.8.1</span> GO term analysis</h4>
<p>In a typical differential expression analysis, thousands of genes are found differentially expressed between two groups of samples. While prior knowledge of the functions of individual genes can give some clues about what kind of cellular processes have been affected, e.g. by a drug treatment, manually going through the whole list of thousands of genes would be very cumbersome and not be very informative in the end. Therefore a commonly used tool to address this problem is to do enrichment analyses of functional terms that appear associated to the given set of differentially expressed genes more often than expected by chance. The functional terms are usually associated to multiple genes. Thus, genes can be grouped into sets by shared functional terms. However, it is important to have an agreed upon controlled vocabulary on the list of terms used to describe the functions of genes. Otherwise, it would be impossible to exchange scientific results globally. That’s why initiatives such as the Gene Ontology Consortium have collated a list of Gene Ontology (GO) terms for each gene. GO term analysis is probably the most common analysis applied after a differential expression analysis. GO term analysis helps quickly find out systematic changes that can describe differences between groups of samples.</p>
<p>In R, one of the simplest ways to do functional enrichment analysis for a set of genes is via the <code>gProfileR</code> package. </p>
<p>Let’s select the genes that are significantly differentially expressed between the case and control samples.
Let’s extract genes that have an adjusted p-value below 0.1 and that show a 2-fold change (either negative or positive) in the case compared to control. We will then feed this gene set into the <code>gProfileR</code> function. The top 10 detected GO terms are displayed in Table <a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#tab:GOanalysistable">8.2</a>.</p>
<div class="sourceCode" id="cb417"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb417-1"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb417-1"></a><span class="kw">library</span>(DESeq2)</span>
<span id="cb417-2"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb417-2"></a><span class="kw">library</span>(gProfileR)</span>
<span id="cb417-3"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb417-3"></a><span class="kw">library</span>(knitr)</span>
<span id="cb417-4"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb417-4"></a><span class="co"># extract differential expression results</span></span>
<span id="cb417-5"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb417-5"></a>DEresults &lt;-<span class="st"> </span><span class="kw">results</span>(dds, <span class="dt">contrast =</span> <span class="kw">c</span>(<span class="st">&#39;group&#39;</span>, <span class="st">&#39;CASE&#39;</span>, <span class="st">&#39;CTRL&#39;</span>))</span>
<span id="cb417-6"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb417-6"></a></span>
<span id="cb417-7"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb417-7"></a><span class="co">#remove genes with NA values </span></span>
<span id="cb417-8"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb417-8"></a>DE &lt;-<span class="st"> </span>DEresults[<span class="op">!</span><span class="kw">is.na</span>(DEresults<span class="op">$</span>padj),]</span>
<span id="cb417-9"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb417-9"></a><span class="co">#select genes with adjusted p-values below 0.1</span></span>
<span id="cb417-10"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb417-10"></a>DE &lt;-<span class="st"> </span>DE[DE<span class="op">$</span>padj <span class="op">&lt;</span><span class="st"> </span><span class="fl">0.1</span>,]</span>
<span id="cb417-11"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb417-11"></a><span class="co">#select genes with absolute log2 fold change above 1 (two-fold change)</span></span>
<span id="cb417-12"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb417-12"></a>DE &lt;-<span class="st"> </span>DE[<span class="kw">abs</span>(DE<span class="op">$</span>log2FoldChange) <span class="op">&gt;</span><span class="st"> </span><span class="dv">1</span>,]</span>
<span id="cb417-13"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb417-13"></a></span>
<span id="cb417-14"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb417-14"></a><span class="co">#get the list of genes of interest</span></span>
<span id="cb417-15"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb417-15"></a>genesOfInterest &lt;-<span class="st"> </span><span class="kw">rownames</span>(DE)</span>
<span id="cb417-16"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb417-16"></a></span>
<span id="cb417-17"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb417-17"></a><span class="co">#calculate enriched GO terms</span></span>
<span id="cb417-18"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb417-18"></a>goResults &lt;-<span class="st"> </span><span class="kw">gprofiler</span>(<span class="dt">query =</span> genesOfInterest, </span>
<span id="cb417-19"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb417-19"></a>                     <span class="dt">organism =</span> <span class="st">&#39;hsapiens&#39;</span>, </span>
<span id="cb417-20"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb417-20"></a>                     <span class="dt">src_filter =</span> <span class="st">&#39;GO&#39;</span>, </span>
<span id="cb417-21"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb417-21"></a>                     <span class="dt">hier_filtering =</span> <span class="st">&#39;moderate&#39;</span>)</span></code></pre></div>
<table>
<caption><span id="tab:GOanalysistable">TABLE 8.2: </span>Top GO terms sorted by p-value.</caption>
<thead>
<tr class="header">
<th></th>
<th align="right">p.value</th>
<th align="right">term.size</th>
<th align="right">precision</th>
<th align="left">domain</th>
<th align="left">term.name</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>64</td>
<td align="right">0</td>
<td align="right">2740</td>
<td align="right">0.223</td>
<td align="left">CC</td>
<td align="left">plasma membrane part</td>
</tr>
<tr class="even">
<td>23</td>
<td align="right">0</td>
<td align="right">1609</td>
<td align="right">0.136</td>
<td align="left">BP</td>
<td align="left">ion transport</td>
</tr>
<tr class="odd">
<td>16</td>
<td align="right">0</td>
<td align="right">3656</td>
<td align="right">0.258</td>
<td align="left">BP</td>
<td align="left">regulation of biological quality</td>
</tr>
<tr class="even">
<td>30</td>
<td align="right">0</td>
<td align="right">385</td>
<td align="right">0.042</td>
<td align="left">BP</td>
<td align="left">extracellular structure organization</td>
</tr>
<tr class="odd">
<td>34</td>
<td align="right">0</td>
<td align="right">7414</td>
<td align="right">0.452</td>
<td align="left">BP</td>
<td align="left">multicellular organismal process</td>
</tr>
<tr class="even">
<td>78</td>
<td align="right">0</td>
<td align="right">1069</td>
<td align="right">0.090</td>
<td align="left">MF</td>
<td align="left">transmembrane transporter activity</td>
</tr>
<tr class="odd">
<td>47</td>
<td align="right">0</td>
<td align="right">1073</td>
<td align="right">0.090</td>
<td align="left">BP</td>
<td align="left">organic acid metabolic process</td>
</tr>
<tr class="even">
<td>5</td>
<td align="right">0</td>
<td align="right">975</td>
<td align="right">0.083</td>
<td align="left">BP</td>
<td align="left">response to drug</td>
</tr>
<tr class="odd">
<td>18</td>
<td align="right">0</td>
<td align="right">1351</td>
<td align="right">0.107</td>
<td align="left">BP</td>
<td align="left">biological adhesion</td>
</tr>
<tr class="even">
<td>31</td>
<td align="right">0</td>
<td align="right">4760</td>
<td align="right">0.302</td>
<td align="left">BP</td>
<td align="left">system development</td>
</tr>
</tbody>
</table>
</div>
<div id="gene-set-enrichment-analysis" class="section level4">
<h4><span class="header-section-number">8.3.8.2</span> Gene set enrichment analysis</h4>
<p>A gene set is a collection of genes with some common property. This shared property among a set of genes could be a GO term, a common biological pathway, a shared interaction partner, or any biologically relevant commonality that is meaningful in the context of the pursued experiment. Gene set enrichment analysis (GSEA) is a valuable exploratory analysis tool that can associate systematic changes to a high-level function rather than individual genes. Analysis of coordinated changes of expression levels of gene sets can provide complementary benefits on top of per-gene-based differential expression analyses. For instance, consider a gene set belonging to a biological pathway where each member of the pathway displays a slight deregulation in a disease sample compared to a normal sample. In such a case, individual genes might not be picked up by the per-gene-based differential expression analysis. Thus, the GO/Pathway enrichment on the differentially expressed list of genes would not show an enrichment of this pathway. However, the additive effect of slight changes of the genes could amount to a large effect at the level of the gene set, thus the pathway could be detected as a significant pathway that could explain the mechanistic problems in the disease sample.</p>
<p>We use the bioconductor package <code>gage</code> <span class="citation">(Luo, Friedman, Shedden, et al. <a href="#ref-luo_gage:_2009" role="doc-biblioref">2009</a>)</span> to demonstrate how to do GSEA using normalized expression data of the samples as input. Here we are using only two gene sets: one from the top GO term discovered from the previous GO analysis, one that we compile by randomly selecting a list of genes. However, annotated gene sets can be used from databases such as MSIGDB <span class="citation">(Subramanian, Tamayo, Mootha, et al. <a href="#ref-subramanian_gene_2005" role="doc-biblioref">2005</a>)</span>, which compile gene sets from a variety of resources such as KEGG <span class="citation">(Kanehisa, Sato, Kawashima, et al. <a href="#ref-kanehisa_kegg_2016" role="doc-biblioref">2016</a>)</span> and REACTOME <span class="citation">(Antonio Fabregat, Jupe, Matthews, et al. <a href="#ref-fabregat_reactome_2018" role="doc-biblioref">2018</a>)</span>.</p>
<div class="sourceCode" id="cb418"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb418-1"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb418-1"></a><span class="co">#Let&#39;s define the first gene set as the list of genes from one of the</span></span>
<span id="cb418-2"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb418-2"></a><span class="co">#significant GO terms found in the GO analysis. order go results by pvalue</span></span>
<span id="cb418-3"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb418-3"></a>goResults &lt;-<span class="st"> </span>goResults[<span class="kw">order</span>(goResults<span class="op">$</span>p.value),]</span>
<span id="cb418-4"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb418-4"></a><span class="co">#restrict the terms that have at most 100 genes overlapping with the query</span></span>
<span id="cb418-5"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb418-5"></a>go &lt;-<span class="st"> </span>goResults[goResults<span class="op">$</span>overlap.size <span class="op">&lt;</span><span class="st"> </span><span class="dv">100</span>,]</span>
<span id="cb418-6"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb418-6"></a><span class="co"># use the top term from this table to create a gene set </span></span>
<span id="cb418-7"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb418-7"></a>geneSet1 &lt;-<span class="st"> </span><span class="kw">unlist</span>(<span class="kw">strsplit</span>(go[<span class="dv">1</span>,]<span class="op">$</span>intersection, <span class="st">&#39;,&#39;</span>))</span>
<span id="cb418-8"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb418-8"></a></span>
<span id="cb418-9"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb418-9"></a><span class="co">#Define another gene set by just randomly selecting 25 genes from the counts</span></span>
<span id="cb418-10"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb418-10"></a><span class="co">#table get normalized counts from DESeq2 results</span></span>
<span id="cb418-11"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb418-11"></a>normalizedCounts &lt;-<span class="st"> </span>DESeq2<span class="op">::</span><span class="kw">counts</span>(dds, <span class="dt">normalized =</span> <span class="ot">TRUE</span>)</span>
<span id="cb418-12"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb418-12"></a>geneSet2 &lt;-<span class="st"> </span><span class="kw">sample</span>(<span class="kw">rownames</span>(normalizedCounts), <span class="dv">25</span>)</span>
<span id="cb418-13"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb418-13"></a></span>
<span id="cb418-14"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb418-14"></a>geneSets &lt;-<span class="st"> </span><span class="kw">list</span>(<span class="st">&#39;top_GO_term&#39;</span> =<span class="st"> </span>geneSet1,</span>
<span id="cb418-15"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb418-15"></a>                 <span class="st">&#39;random_set&#39;</span> =<span class="st"> </span>geneSet2)</span></code></pre></div>
<p>Using the defined gene sets, we’d like to do a group comparison between the case samples with respect to the control samples.</p>
<div class="sourceCode" id="cb419"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb419-1"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb419-1"></a><span class="kw">library</span>(gage)</span>
<span id="cb419-2"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb419-2"></a><span class="co">#use the normalized counts to carry out a GSEA. </span></span>
<span id="cb419-3"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb419-3"></a>gseaResults &lt;-<span class="st"> </span><span class="kw">gage</span>(<span class="dt">exprs =</span> <span class="kw">log2</span>(normalizedCounts<span class="op">+</span><span class="dv">1</span>), </span>
<span id="cb419-4"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb419-4"></a>           <span class="dt">ref =</span> <span class="kw">match</span>(<span class="kw">rownames</span>(colData[colData<span class="op">$</span>group <span class="op">==</span><span class="st"> &#39;CTRL&#39;</span>,]), </span>
<span id="cb419-5"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb419-5"></a>                       <span class="kw">colnames</span>(normalizedCounts)), </span>
<span id="cb419-6"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb419-6"></a>           <span class="dt">samp =</span> <span class="kw">match</span>(<span class="kw">rownames</span>(colData[colData<span class="op">$</span>group <span class="op">==</span><span class="st"> &#39;CASE&#39;</span>,]), </span>
<span id="cb419-7"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb419-7"></a>                        <span class="kw">colnames</span>(normalizedCounts)),</span>
<span id="cb419-8"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb419-8"></a>           <span class="dt">gsets =</span> geneSets, <span class="dt">compare =</span> <span class="st">&#39;as.group&#39;</span>)</span></code></pre></div>
<p>We can observe if there is a significant up-regulation or down-regulation of the gene set in the case group compared to the controls by accessing <code>gseaResults$greater</code> as in Table <a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#tab:gseaPost1">8.3</a> or <code>gseaResults$less</code> as in Table <a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#tab:gseaPost2">8.4</a>.</p>
<table>
<caption><span id="tab:gseaPost1">TABLE 8.3: </span>Up-regulation statistics</caption>
<thead>
<tr class="header">
<th></th>
<th align="right">p.geomean</th>
<th align="right">stat.mean</th>
<th align="right">p.val</th>
<th align="right">q.val</th>
<th align="right">set.size</th>
<th align="right">exp1</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>top_GO_term</td>
<td align="right">0.0000</td>
<td align="right">7.1994</td>
<td align="right">0.0000</td>
<td align="right">0.0000</td>
<td align="right">32</td>
<td align="right">0.0000</td>
</tr>
<tr class="even">
<td>random_set</td>
<td align="right">0.5832</td>
<td align="right">-0.2113</td>
<td align="right">0.5832</td>
<td align="right">0.5832</td>
<td align="right">25</td>
<td align="right">0.5832</td>
</tr>
</tbody>
</table>
<table>
<caption><span id="tab:gseaPost2">TABLE 8.4: </span>Down-regulation statistics</caption>
<thead>
<tr class="header">
<th></th>
<th align="right">p.geomean</th>
<th align="right">stat.mean</th>
<th align="right">p.val</th>
<th align="right">q.val</th>
<th align="right">set.size</th>
<th align="right">exp1</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>random_set</td>
<td align="right">0.4168</td>
<td align="right">-0.2113</td>
<td align="right">0.4168</td>
<td align="right">0.8336</td>
<td align="right">25</td>
<td align="right">0.4168</td>
</tr>
<tr class="even">
<td>top_GO_term</td>
<td align="right">1.0000</td>
<td align="right">7.1994</td>
<td align="right">1.0000</td>
<td align="right">1.0000</td>
<td align="right">32</td>
<td align="right">1.0000</td>
</tr>
</tbody>
</table>
<p>We can see that the random gene set shows no significant up- or down-regulation (Tables <a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#tab:gseaPost1">8.3</a> and (<a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#tab:gseaPost2">8.4</a>), while the gene set we defined using the top GO term shows a significant up-regulation (adjusted p-value &lt; 0.0007) (<a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#tab:gseaPost1">8.3</a>). It is worthwhile to visualize these systematic changes in a heatmap as in Figure <a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#fig:gseaPost3">8.11</a>.</p>
<div class="sourceCode" id="cb420"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb420-1"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb420-1"></a><span class="kw">library</span>(pheatmap)</span>
<span id="cb420-2"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb420-2"></a><span class="co"># get the expression data for the gene set of interest</span></span>
<span id="cb420-3"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb420-3"></a>M &lt;-<span class="st"> </span>normalizedCounts[<span class="kw">rownames</span>(normalizedCounts) <span class="op">%in%</span><span class="st"> </span>geneSet1, ]</span>
<span id="cb420-4"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb420-4"></a><span class="co"># log transform the counts for visualization scaling by row helps visualizing</span></span>
<span id="cb420-5"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb420-5"></a><span class="co"># relative change of expression of a gene in multiple conditions</span></span>
<span id="cb420-6"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb420-6"></a><span class="kw">pheatmap</span>(<span class="kw">log2</span>(M<span class="op">+</span><span class="dv">1</span>), </span>
<span id="cb420-7"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb420-7"></a>         <span class="dt">annotation_col =</span> colData, </span>
<span id="cb420-8"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb420-8"></a>         <span class="dt">show_rownames =</span> <span class="ot">TRUE</span>, </span>
<span id="cb420-9"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb420-9"></a>         <span class="dt">fontsize_row =</span> <span class="dv">8</span>,</span>
<span id="cb420-10"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb420-10"></a>         <span class="dt">scale =</span> <span class="st">&#39;row&#39;</span>, </span>
<span id="cb420-11"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb420-11"></a>         <span class="dt">cutree_cols =</span> <span class="dv">2</span>, </span>
<span id="cb420-12"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb420-12"></a>         <span class="dt">cutree_rows =</span> <span class="dv">2</span>)</span></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:gseaPost3"></span>
<img src="08-rna-seq-analysis_files/figure-html/gseaPost3-1.png" alt="Heatmap of expression value from the genes with the top GO term." width="55%" />
<p class="caption">
FIGURE 8.11: Heatmap of expression value from the genes with the top GO term.
</p>
</div>
<p>We can see that almost all genes from this gene set display an increased level of expression in the case samples
compared to the controls.</p>
</div>
</div>
<div id="accounting-for-additional-sources-of-variation" class="section level3">
<h3><span class="header-section-number">8.3.9</span> Accounting for additional sources of variation</h3>
<p>When doing a differential expression analysis in a case-control setting, the variable of interest, i.e. the variable that explains the separation of the case samples from the control, is usually the treatment, genotypic differences, a certain phenotype, etc. However, in reality, depending on how the experiment and the sequencing were designed, there may be additional factors that might contribute to the variation between the compared samples. Sometimes, such variables are known, for instance, the date of the sequencing for each sample (batch information), or the temperature under which samples were kept. Such variables are not necessarily biological but rather technical, however, they still impact the measurements obtained from an RNA-seq experiment. Such variables can introduce systematic shifts in the obtained measurements. Here, we will demonstrate: firstly how to account for such variables using DESeq2, when the possible sources of variation are actually known; secondly, how to account for such variables when all we have is just a count table but we observe that the variable of interest only explains a small proportion of the differences between case and control samples.</p>
<div id="accounting-for-covariates-using-deseq2" class="section level4">
<h4><span class="header-section-number">8.3.9.1</span> Accounting for covariates using DESeq2</h4>
<p>For demonstration purposes, we will use a subset of the count table obtained for a heart disease study, where there are RNA-seq samples from subjects with normal and failing hearts. We again use a subset of the samples, focusing on 6 case and 6 control samples and we only consider protein-coding genes (for speed concerns).</p>
<p>Let’s import count and colData for this experiment.</p>
<div class="sourceCode" id="cb421"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb421-1"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb421-1"></a>counts_file &lt;-<span class="st"> </span><span class="kw">system.file</span>(<span class="st">&#39;extdata/rna-seq/SRP021193.raw_counts.tsv&#39;</span>, </span>
<span id="cb421-2"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb421-2"></a>                           <span class="dt">package =</span> <span class="st">&#39;compGenomRData&#39;</span>)</span>
<span id="cb421-3"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb421-3"></a>colData_file &lt;-<span class="st"> </span><span class="kw">system.file</span>(<span class="st">&#39;extdata/rna-seq/SRP021193.colData.tsv&#39;</span>, </span>
<span id="cb421-4"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb421-4"></a>                            <span class="dt">package =</span> <span class="st">&#39;compGenomRData&#39;</span>)</span>
<span id="cb421-5"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb421-5"></a></span>
<span id="cb421-6"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb421-6"></a>counts &lt;-<span class="st"> </span><span class="kw">read.table</span>(counts_file)</span>
<span id="cb421-7"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb421-7"></a>colData &lt;-<span class="st"> </span><span class="kw">read.table</span>(colData_file, <span class="dt">header =</span> T, <span class="dt">sep =</span> <span class="st">&#39;</span><span class="ch">\t</span><span class="st">&#39;</span>, </span>
<span id="cb421-8"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb421-8"></a>                      <span class="dt">stringsAsFactors =</span> <span class="ot">TRUE</span>)</span></code></pre></div>
<p>Let’s take a look at how the samples cluster by calculating the TPM counts as displayed as a heatmap in Figure <a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#fig:batcheffects2">8.12</a>.</p>
<div class="sourceCode" id="cb422"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb422-1"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb422-1"></a><span class="kw">library</span>(pheatmap)</span>
<span id="cb422-2"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb422-2"></a><span class="co">#find gene length normalized values </span></span>
<span id="cb422-3"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb422-3"></a>geneLengths &lt;-<span class="st"> </span>counts<span class="op">$</span>width</span>
<span id="cb422-4"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb422-4"></a>rpk &lt;-<span class="st"> </span><span class="kw">apply</span>( <span class="kw">subset</span>(counts, <span class="dt">select =</span> <span class="kw">c</span>(<span class="op">-</span>width)), <span class="dv">2</span>, </span>
<span id="cb422-5"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb422-5"></a>              <span class="cf">function</span>(x) x<span class="op">/</span>(geneLengths<span class="op">/</span><span class="dv">1000</span>))</span>
<span id="cb422-6"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb422-6"></a><span class="co">#normalize by the sample size using rpk values</span></span>
<span id="cb422-7"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb422-7"></a>tpm &lt;-<span class="st"> </span><span class="kw">apply</span>(rpk, <span class="dv">2</span>, <span class="cf">function</span>(x) x <span class="op">/</span><span class="st"> </span><span class="kw">sum</span>(<span class="kw">as.numeric</span>(x)) <span class="op">*</span><span class="st"> </span><span class="dv">10</span><span class="op">^</span><span class="dv">6</span>)</span>
<span id="cb422-8"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb422-8"></a></span>
<span id="cb422-9"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb422-9"></a>selectedGenes &lt;-<span class="st"> </span><span class="kw">names</span>(<span class="kw">sort</span>(<span class="kw">apply</span>(tpm, <span class="dv">1</span>, var), </span>
<span id="cb422-10"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb422-10"></a>                            <span class="dt">decreasing =</span> T)[<span class="dv">1</span><span class="op">:</span><span class="dv">100</span>])</span>
<span id="cb422-11"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb422-11"></a><span class="kw">pheatmap</span>(tpm[selectedGenes,], </span>
<span id="cb422-12"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb422-12"></a>             <span class="dt">scale =</span> <span class="st">&#39;row&#39;</span>,</span>
<span id="cb422-13"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb422-13"></a>             <span class="dt">annotation_col =</span> colData, </span>
<span id="cb422-14"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb422-14"></a>             <span class="dt">show_rownames =</span> <span class="ot">FALSE</span>)</span></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:batcheffects2"></span>
<img src="08-rna-seq-analysis_files/figure-html/batcheffects2-1.png" alt="Visualizing batch effects in an experiment." width="55%" />
<p class="caption">
FIGURE 8.12: Visualizing batch effects in an experiment.
</p>
</div>
<p>Here we can see from the clusters that the dominating variable is the ‘Library Selection’ variable rather than the ‘diagnosis’ variable, which determines the state of the organ from which the sample was taken. Case and control samples are all mixed in both two major clusters. However, ideally, we’d like to see a separation of the case and control samples regardless of the additional covariates. When testing for differential gene expression between conditions, such confounding variables can be accounted for using <code>DESeq2</code>. Below is a demonstration of how we instruct <code>DESeq2</code> to account for the ‘library selection’ variable:</p>
<div class="sourceCode" id="cb423"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb423-1"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb423-1"></a><span class="kw">library</span>(DESeq2)</span>
<span id="cb423-2"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb423-2"></a><span class="co"># remove the &#39;width&#39; column from the counts matrix</span></span>
<span id="cb423-3"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb423-3"></a>countData &lt;-<span class="st"> </span><span class="kw">as.matrix</span>(<span class="kw">subset</span>(counts, <span class="dt">select =</span> <span class="kw">c</span>(<span class="op">-</span>width)))</span>
<span id="cb423-4"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb423-4"></a><span class="co"># set up a DESeqDataSet object</span></span>
<span id="cb423-5"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb423-5"></a>dds &lt;-<span class="st"> </span><span class="kw">DESeqDataSetFromMatrix</span>(<span class="dt">countData =</span> countData, </span>
<span id="cb423-6"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb423-6"></a>                              <span class="dt">colData =</span> colData, </span>
<span id="cb423-7"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb423-7"></a>                              <span class="dt">design =</span> <span class="op">~</span><span class="st"> </span>LibrarySelection <span class="op">+</span><span class="st"> </span>group)</span></code></pre></div>
<p>When constructing the design formula, it is very important to pay attention to the sequence of variables. We leave the variable of interest to the last and we can add as many covariates as we want to the beginning of the design formula. Please refer to the <code>DESeq2</code> vignette if you’d like to learn more about how to construct design formulas.</p>
<p>Now, we can run the differential expression analysis as has been demonstrated previously.</p>
<div class="sourceCode" id="cb424"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb424-1"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb424-1"></a><span class="co"># run DESeq</span></span>
<span id="cb424-2"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb424-2"></a>dds &lt;-<span class="st"> </span><span class="kw">DESeq</span>(dds)</span>
<span id="cb424-3"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb424-3"></a><span class="co"># extract results</span></span>
<span id="cb424-4"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb424-4"></a>DEresults &lt;-<span class="st"> </span><span class="kw">results</span>(dds, <span class="dt">contrast =</span> <span class="kw">c</span>(<span class="st">&#39;group&#39;</span>, <span class="st">&#39;CASE&#39;</span>, <span class="st">&#39;CTRL&#39;</span>))</span></code></pre></div>
</div>
<div id="accounting-for-estimated-covariates-using-ruvseq" class="section level4">
<h4><span class="header-section-number">8.3.9.2</span> Accounting for estimated covariates using RUVSeq</h4>
<p>In cases when the sources of potential variation are not known, it is worthwhile to use tools such as <code>RUVSeq</code> or <code>sva</code> that can estimate potential sources of variation and clean up the counts table from those sources of variation. Later on, the estimated covariates can be integrated into DESeq2’s design formula.</p>
<p>Let’s see how to utilize the <code>RUVseq</code> package to first diagnose the problem and then solve it. Here, for demonstration purposes, we’ll use a count table from a lung carcinoma study in which a transcription factor (Ets homologous factor - EHF) is overexpressed and compared to the control samples with baseline EHF expression. Again, we only consider protein coding genes and use only five case and five control samples. The original data can be found on the <code>recount2</code> database with the accession ‘SRP049988’.</p>
<div class="sourceCode" id="cb425"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb425-1"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb425-1"></a>counts_file &lt;-<span class="st"> </span><span class="kw">system.file</span>(<span class="st">&#39;extdata/rna-seq/SRP049988.raw_counts.tsv&#39;</span>, </span>
<span id="cb425-2"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb425-2"></a>                           <span class="dt">package =</span> <span class="st">&#39;compGenomRData&#39;</span>)</span>
<span id="cb425-3"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb425-3"></a>colData_file &lt;-<span class="st"> </span><span class="kw">system.file</span>(<span class="st">&#39;extdata/rna-seq/SRP049988.colData.tsv&#39;</span>, </span>
<span id="cb425-4"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb425-4"></a>                            <span class="dt">package =</span> <span class="st">&#39;compGenomRData&#39;</span>)</span>
<span id="cb425-5"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb425-5"></a></span>
<span id="cb425-6"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb425-6"></a>counts &lt;-<span class="st"> </span><span class="kw">read.table</span>(counts_file)</span>
<span id="cb425-7"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb425-7"></a>colData &lt;-<span class="st"> </span><span class="kw">read.table</span>(colData_file, <span class="dt">header =</span> T, </span>
<span id="cb425-8"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb425-8"></a>                      <span class="dt">sep =</span> <span class="st">&#39;</span><span class="ch">\t</span><span class="st">&#39;</span>, <span class="dt">stringsAsFactors =</span> <span class="ot">TRUE</span>)</span>
<span id="cb425-9"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb425-9"></a><span class="co"># simplify condition descriptions</span></span>
<span id="cb425-10"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb425-10"></a>colData<span class="op">$</span>source_name &lt;-<span class="st"> </span><span class="kw">ifelse</span>(colData<span class="op">$</span>group <span class="op">==</span><span class="st"> &#39;CASE&#39;</span>, </span>
<span id="cb425-11"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb425-11"></a>                              <span class="st">&#39;EHF_overexpression&#39;</span>, <span class="st">&#39;Empty_Vector&#39;</span>)</span></code></pre></div>
<p>Let’s start by making heatmaps of the samples using TPM counts (see Figure <a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#fig:ruvdiagnose1">8.13</a>).</p>
<div class="sourceCode" id="cb426"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb426-1"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb426-1"></a><span class="co">#find gene length normalized values </span></span>
<span id="cb426-2"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb426-2"></a>geneLengths &lt;-<span class="st"> </span>counts<span class="op">$</span>width</span>
<span id="cb426-3"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb426-3"></a>rpk &lt;-<span class="st"> </span><span class="kw">apply</span>( <span class="kw">subset</span>(counts, <span class="dt">select =</span> <span class="kw">c</span>(<span class="op">-</span>width)), <span class="dv">2</span>, </span>
<span id="cb426-4"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb426-4"></a>              <span class="cf">function</span>(x) x<span class="op">/</span>(geneLengths<span class="op">/</span><span class="dv">1000</span>))</span>
<span id="cb426-5"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb426-5"></a><span class="co">#normalize by the sample size using rpk values</span></span>
<span id="cb426-6"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb426-6"></a>tpm &lt;-<span class="st"> </span><span class="kw">apply</span>(rpk, <span class="dv">2</span>, <span class="cf">function</span>(x) x <span class="op">/</span><span class="st"> </span><span class="kw">sum</span>(<span class="kw">as.numeric</span>(x)) <span class="op">*</span><span class="st"> </span><span class="dv">10</span><span class="op">^</span><span class="dv">6</span>)</span>
<span id="cb426-7"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb426-7"></a>selectedGenes &lt;-<span class="st"> </span><span class="kw">names</span>(<span class="kw">sort</span>(<span class="kw">apply</span>(tpm, <span class="dv">1</span>, var), </span>
<span id="cb426-8"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb426-8"></a>                            <span class="dt">decreasing =</span> T)[<span class="dv">1</span><span class="op">:</span><span class="dv">100</span>])</span>
<span id="cb426-9"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb426-9"></a><span class="kw">pheatmap</span>(tpm[selectedGenes,], </span>
<span id="cb426-10"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb426-10"></a>             <span class="dt">scale =</span> <span class="st">&#39;row&#39;</span>,</span>
<span id="cb426-11"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb426-11"></a>             <span class="dt">annotation_col =</span> colData, </span>
<span id="cb426-12"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb426-12"></a>             <span class="dt">cutree_cols =</span> <span class="dv">2</span>, </span>
<span id="cb426-13"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb426-13"></a>             <span class="dt">show_rownames =</span> <span class="ot">FALSE</span>)</span></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:ruvdiagnose1"></span>
<img src="08-rna-seq-analysis_files/figure-html/ruvdiagnose1-1.png" alt="Diagnostic plot to observe." width="55%" />
<p class="caption">
FIGURE 8.13: Diagnostic plot to observe.
</p>
</div>
<p>We can see that the overall clusters look fine, except that one of the case samples (CASE_5) clusters more closely with the control samples than the other case samples. This mis-clustering could be a result of some batch effect, or any other technical preparation steps. However, the <code>colData</code> object doesn’t contain any variables that we can use to pinpoint the exact cause of this. So, let’s use <code>RUVSeq</code> to estimate potential covariates to see if the clustering results can be improved.</p>
<p>First, we set up the experiment:</p>
<div class="sourceCode" id="cb427"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb427-1"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb427-1"></a><span class="kw">library</span>(EDASeq)</span>
<span id="cb427-2"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb427-2"></a><span class="co"># remove &#39;width&#39; column from counts</span></span>
<span id="cb427-3"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb427-3"></a>countData &lt;-<span class="st"> </span><span class="kw">as.matrix</span>(<span class="kw">subset</span>(counts, <span class="dt">select =</span> <span class="kw">c</span>(<span class="op">-</span>width)))</span>
<span id="cb427-4"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb427-4"></a><span class="co"># create a seqExpressionSet object using EDASeq package </span></span>
<span id="cb427-5"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb427-5"></a>set &lt;-<span class="st"> </span><span class="kw">newSeqExpressionSet</span>(<span class="dt">counts =</span> countData,</span>
<span id="cb427-6"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb427-6"></a>                           <span class="dt">phenoData =</span> colData)</span></code></pre></div>
<p>Next, let’s make a diagnostic RLE plot on the raw count table.</p>
<div class="sourceCode" id="cb428"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb428-1"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb428-1"></a><span class="co"># make an RLE plot and a PCA plot on raw count data and color samples by group</span></span>
<span id="cb428-2"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb428-2"></a><span class="kw">par</span>(<span class="dt">mfrow =</span> <span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">2</span>))</span>
<span id="cb428-3"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb428-3"></a><span class="kw">plotRLE</span>(set, <span class="dt">outline=</span><span class="ot">FALSE</span>, <span class="dt">ylim=</span><span class="kw">c</span>(<span class="op">-</span><span class="dv">4</span>, <span class="dv">4</span>), <span class="dt">col=</span><span class="kw">as.numeric</span>(colData<span class="op">$</span>group))</span>
<span id="cb428-4"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb428-4"></a><span class="kw">plotPCA</span>(set, <span class="dt">col =</span> <span class="kw">as.numeric</span>(colData<span class="op">$</span>group), <span class="dt">adj =</span> <span class="fl">0.5</span>, </span>
<span id="cb428-5"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb428-5"></a>        <span class="dt">ylim =</span> <span class="kw">c</span>(<span class="op">-</span><span class="fl">0.7</span>, <span class="fl">0.5</span>), <span class="dt">xlim =</span> <span class="kw">c</span>(<span class="op">-</span><span class="fl">0.5</span>, <span class="fl">0.5</span>))</span></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:ruvdiagnose2p1"></span>
<img src="08-rna-seq-analysis_files/figure-html/ruvdiagnose2p1-1.png" alt="Diagnostic RLE and PCA plots based on raw count table." width="55%" />
<p class="caption">
FIGURE 8.14: Diagnostic RLE and PCA plots based on raw count table.
</p>
</div>
<div class="sourceCode" id="cb429"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb429-1"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb429-1"></a><span class="co">## make RLE and PCA plots on TPM matrix </span></span>
<span id="cb429-2"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb429-2"></a><span class="kw">par</span>(<span class="dt">mfrow =</span> <span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">2</span>))</span>
<span id="cb429-3"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb429-3"></a><span class="kw">plotRLE</span>(tpm, <span class="dt">outline=</span><span class="ot">FALSE</span>, <span class="dt">ylim=</span><span class="kw">c</span>(<span class="op">-</span><span class="dv">4</span>, <span class="dv">4</span>), <span class="dt">col=</span><span class="kw">as.numeric</span>(colData<span class="op">$</span>group))</span>
<span id="cb429-4"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb429-4"></a><span class="kw">plotPCA</span>(tpm, <span class="dt">col=</span><span class="kw">as.numeric</span>(colData<span class="op">$</span>group), <span class="dt">adj =</span> <span class="fl">0.5</span>, </span>
<span id="cb429-5"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb429-5"></a>        <span class="dt">ylim =</span> <span class="kw">c</span>(<span class="op">-</span><span class="fl">0.3</span>, <span class="dv">1</span>), <span class="dt">xlim =</span> <span class="kw">c</span>(<span class="op">-</span><span class="fl">0.5</span>, <span class="fl">0.5</span>))</span></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:ruvdiagnose2p2"></span>
<img src="08-rna-seq-analysis_files/figure-html/ruvdiagnose2p2-1.png" alt="Diagnostic RLE and PCA plots based on TPM normalized count table." width="55%" />
<p class="caption">
FIGURE 8.15: Diagnostic RLE and PCA plots based on TPM normalized count table.
</p>
</div>
<p>Both RLE and PCA plots look better on normalized data (Figure <a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#fig:ruvdiagnose2p2">8.15</a>) compared to raw data (Figure <a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#fig:ruvdiagnose2p1">8.14</a>), but still suggest the necessity of further improvement, because the CASE_5 sample still clusters with the control samples. We haven’t yet accounted for the source of unwanted variation.</p>
</div>
<div id="removing-unwanted-variation-from-the-data" class="section level4">
<h4><span class="header-section-number">8.3.9.3</span> Removing unwanted variation from the data</h4>
<p><code>RUVSeq</code> has three main functions for removing unwanted variation: <code>RUVg()</code>, <code>RUVs()</code>, and <code>RUVr()</code>. Here, we will demonstrate how to use <code>RUVg</code> and <code>RUVs</code>. <code>RUVr</code> will be left as an exercise for the reader.</p>
<div id="using-ruvg" class="section level5">
<h5><span class="header-section-number">8.3.9.3.1</span> Using RUVg</h5>
<p>One way of removing unwanted variation depends on using a set of reference genes that are not expected to change by the sources of technical variation. One strategy along this line is to use spike-in genes, which are artificially introduced into the sequencing run <span class="citation">(Jiang, Schlesinger, Davis, et al. <a href="#ref-jiang_synthetic_2011" role="doc-biblioref">2011</a>)</span>. However, there are many sequencing datasets that don’t have this spike-in data available. In such cases, an empirical set of genes can be collected from the expression data by doing a differential expression analysis and discovering genes that are unchanged in the given conditions. These unchanged genes are used to clean up the data from systematic shifts in expression due to the unwanted sources of variation. Another strategy could be to use a set of house-keeping genes as negative controls, and use them as a reference to correct the systematic biases in the data. Let’s use a list of ~500 house-keeping genes compiled here: <a href="https://www.tau.ac.il/~elieis/HKG/HK_genes.txt" class="uri">https://www.tau.ac.il/~elieis/HKG/HK_genes.txt</a>.</p>
<div class="sourceCode" id="cb430"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb430-1"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb430-1"></a><span class="kw">library</span>(RUVSeq)</span>
<span id="cb430-2"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb430-2"></a> </span>
<span id="cb430-3"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb430-3"></a><span class="co">#source for house-keeping genes collection:</span></span>
<span id="cb430-4"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb430-4"></a><span class="co">#https://m.tau.ac.il/~elieis/HKG/HK_genes.txt</span></span>
<span id="cb430-5"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb430-5"></a>HK_genes &lt;-<span class="st"> </span><span class="kw">read.table</span>(<span class="dt">file =</span> <span class="kw">system.file</span>(<span class="st">&quot;extdata/rna-seq/HK_genes.txt&quot;</span>, </span>
<span id="cb430-6"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb430-6"></a>                                          <span class="dt">package =</span> <span class="st">&#39;compGenomRData&#39;</span>), </span>
<span id="cb430-7"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb430-7"></a>                       <span class="dt">header =</span> <span class="ot">FALSE</span>)</span>
<span id="cb430-8"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb430-8"></a><span class="co"># let&#39;s take an intersection of the house-keeping genes with the genes available</span></span>
<span id="cb430-9"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb430-9"></a><span class="co"># in the count table</span></span>
<span id="cb430-10"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb430-10"></a>house_keeping_genes &lt;-<span class="st"> </span><span class="kw">intersect</span>(<span class="kw">rownames</span>(set), HK_genes<span class="op">$</span>V1)</span></code></pre></div>
<p>We will now run <code>RUVg()</code> with the different number of factors of unwanted variation. We will plot the PCA after removing the unwanted variation. We should be able to see which <code>k</code> values, number of factors, produce better separation between sample groups.</p>
<div class="sourceCode" id="cb431"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb431-1"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb431-1"></a><span class="co"># now, we use these genes as the empirical set of genes as input to RUVg.</span></span>
<span id="cb431-2"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb431-2"></a><span class="co"># we try different values of k and see how the PCA plots look </span></span>
<span id="cb431-3"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb431-3"></a> </span>
<span id="cb431-4"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb431-4"></a><span class="kw">par</span>(<span class="dt">mfrow =</span> <span class="kw">c</span>(<span class="dv">2</span>, <span class="dv">2</span>))</span>
<span id="cb431-5"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb431-5"></a><span class="cf">for</span>(k <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span><span class="dv">4</span>) {</span>
<span id="cb431-6"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb431-6"></a>  set_g &lt;-<span class="st"> </span><span class="kw">RUVg</span>(<span class="dt">x =</span> set, <span class="dt">cIdx =</span> house_keeping_genes, <span class="dt">k =</span> k)</span>
<span id="cb431-7"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb431-7"></a>  <span class="kw">plotPCA</span>(set_g, <span class="dt">col=</span><span class="kw">as.numeric</span>(colData<span class="op">$</span>group), <span class="dt">cex =</span> <span class="fl">0.9</span>, <span class="dt">adj =</span> <span class="fl">0.5</span>, </span>
<span id="cb431-8"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb431-8"></a>          <span class="dt">main =</span> <span class="kw">paste0</span>(<span class="st">&#39;with RUVg, k = &#39;</span>,k), </span>
<span id="cb431-9"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb431-9"></a>          <span class="dt">ylim =</span> <span class="kw">c</span>(<span class="op">-</span><span class="dv">1</span>, <span class="dv">1</span>), <span class="dt">xlim =</span> <span class="kw">c</span>(<span class="op">-</span><span class="dv">1</span>, <span class="dv">1</span>), )</span>
<span id="cb431-10"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb431-10"></a>}</span></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:ruvgf1"></span>
<img src="08-rna-seq-analysis_files/figure-html/ruvgf1-1.png" alt="PCA plots on RUVg normalized data with varying number of covariates (k)." width="55%" />
<p class="caption">
FIGURE 8.16: PCA plots on RUVg normalized data with varying number of covariates (k).
</p>
</div>
<p>Based on the separation of case and control samples in the PCA plots in Figure <a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#fig:ruvgf1">8.16</a>,
we choose k = 1 and re-run the <code>RUVg()</code> function with the house-keeping genes to do more diagnostic plots.</p>
<div class="sourceCode" id="cb432"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb432-1"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb432-1"></a><span class="co"># choose k = 1</span></span>
<span id="cb432-2"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb432-2"></a> </span>
<span id="cb432-3"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb432-3"></a>set_g &lt;-<span class="st"> </span><span class="kw">RUVg</span>(<span class="dt">x =</span> set, <span class="dt">cIdx =</span> house_keeping_genes, <span class="dt">k =</span> <span class="dv">1</span>)</span></code></pre></div>
<p>Now let’s do diagnostics: compare the count matrices with or without RUVg processing, comparing RLE plots (Figure <a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#fig:ruvgf2">8.17</a>) and PCA plots (Figure <a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#fig:ruvgf3">8.18</a>) to see the effect of RUVg on the normalization and separation of case and control samples.</p>
<div class="sourceCode" id="cb433"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb433-1"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb433-1"></a><span class="co"># RLE plots</span></span>
<span id="cb433-2"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb433-2"></a> </span>
<span id="cb433-3"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb433-3"></a><span class="kw">par</span>(<span class="dt">mfrow =</span> <span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">2</span>))</span>
<span id="cb433-4"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb433-4"></a><span class="kw">plotRLE</span>(set, <span class="dt">outline=</span><span class="ot">FALSE</span>, <span class="dt">ylim=</span><span class="kw">c</span>(<span class="op">-</span><span class="dv">4</span>, <span class="dv">4</span>), </span>
<span id="cb433-5"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb433-5"></a>        <span class="dt">col=</span><span class="kw">as.numeric</span>(colData<span class="op">$</span>group), <span class="dt">main =</span> <span class="st">&#39;without RUVg&#39;</span>)</span>
<span id="cb433-6"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb433-6"></a><span class="kw">plotRLE</span>(set_g, <span class="dt">outline=</span><span class="ot">FALSE</span>, <span class="dt">ylim=</span><span class="kw">c</span>(<span class="op">-</span><span class="dv">4</span>, <span class="dv">4</span>), </span>
<span id="cb433-7"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb433-7"></a>        <span class="dt">col=</span><span class="kw">as.numeric</span>(colData<span class="op">$</span>group), <span class="dt">main =</span> <span class="st">&#39;with RUVg&#39;</span>)</span></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:ruvgf2"></span>
<img src="08-rna-seq-analysis_files/figure-html/ruvgf2-1.png" alt="RLE plots to observe the effect of RUVg." width="55%" />
<p class="caption">
FIGURE 8.17: RLE plots to observe the effect of RUVg.
</p>
</div>
<div class="sourceCode" id="cb434"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb434-1"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb434-1"></a><span class="co"># PCA plots</span></span>
<span id="cb434-2"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb434-2"></a> </span>
<span id="cb434-3"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb434-3"></a><span class="kw">par</span>(<span class="dt">mfrow =</span> <span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">2</span>))</span>
<span id="cb434-4"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb434-4"></a><span class="kw">plotPCA</span>(set, <span class="dt">col=</span><span class="kw">as.numeric</span>(colData<span class="op">$</span>group), <span class="dt">adj =</span> <span class="fl">0.5</span>,</span>
<span id="cb434-5"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb434-5"></a>        <span class="dt">main =</span> <span class="st">&#39;without RUVg&#39;</span>, </span>
<span id="cb434-6"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb434-6"></a>        <span class="dt">ylim =</span> <span class="kw">c</span>(<span class="op">-</span><span class="dv">1</span>, <span class="fl">0.5</span>), <span class="dt">xlim =</span> <span class="kw">c</span>(<span class="op">-</span><span class="fl">0.5</span>, <span class="fl">0.5</span>))</span>
<span id="cb434-7"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb434-7"></a><span class="kw">plotPCA</span>(set_g, <span class="dt">col=</span><span class="kw">as.numeric</span>(colData<span class="op">$</span>group), <span class="dt">adj =</span> <span class="fl">0.5</span>, </span>
<span id="cb434-8"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb434-8"></a>        <span class="dt">main =</span> <span class="st">&#39;with RUVg&#39;</span>,</span>
<span id="cb434-9"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb434-9"></a>        <span class="dt">ylim =</span> <span class="kw">c</span>(<span class="op">-</span><span class="dv">1</span>, <span class="fl">0.5</span>), <span class="dt">xlim =</span> <span class="kw">c</span>(<span class="op">-</span><span class="fl">0.5</span>, <span class="fl">0.5</span>))</span></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:ruvgf3"></span>
<img src="08-rna-seq-analysis_files/figure-html/ruvgf3-1.png" alt="PCA plots to observe the effect of RUVg." width="55%" />
<p class="caption">
FIGURE 8.18: PCA plots to observe the effect of RUVg.
</p>
</div>
<p>We can observe that using <code>RUVg()</code> with house-keeping genes as reference has improved the clusters, however not yielded ideal separation. Probably the effect that is causing the ‘CASE_5’ to cluster with the control samples still hasn’t been completely eliminated.</p>
</div>
<div id="using-ruvs" class="section level5">
<h5><span class="header-section-number">8.3.9.3.2</span> Using RUVs</h5>
<p>There is another strategy of <code>RUVSeq</code> that works better in the presence of replicates in the absence of a confounded experimental design, which is the <code>RUVs()</code> function. Let’s see how that performs with this data. This time we don’t use the house-keeping genes. We rather use all genes as input to <code>RUVs()</code>. This function estimates the correction factor by assuming that replicates should have constant biological variation, rather, the variation in the replicates are the unwanted variation.</p>
<div class="sourceCode" id="cb435"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb435-1"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb435-1"></a><span class="co"># make a table of sample groups from colData </span></span>
<span id="cb435-2"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb435-2"></a>differences &lt;-<span class="st"> </span><span class="kw">makeGroups</span>(colData<span class="op">$</span>group)</span>
<span id="cb435-3"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb435-3"></a><span class="co">## looking for two different sources of unwanted variation (k = 2)</span></span>
<span id="cb435-4"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb435-4"></a><span class="co">## use information from all genes in the expression object</span></span>
<span id="cb435-5"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb435-5"></a><span class="kw">par</span>(<span class="dt">mfrow =</span> <span class="kw">c</span>(<span class="dv">2</span>, <span class="dv">2</span>))</span>
<span id="cb435-6"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb435-6"></a><span class="cf">for</span>(k <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span><span class="dv">4</span>) {</span>
<span id="cb435-7"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb435-7"></a>  set_s &lt;-<span class="st"> </span><span class="kw">RUVs</span>(set, <span class="kw">unique</span>(<span class="kw">rownames</span>(set)), </span>
<span id="cb435-8"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb435-8"></a>                <span class="dt">k=</span>k, differences) <span class="co">#all genes</span></span>
<span id="cb435-9"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb435-9"></a>  <span class="kw">plotPCA</span>(set_s, <span class="dt">col=</span><span class="kw">as.numeric</span>(colData<span class="op">$</span>group), </span>
<span id="cb435-10"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb435-10"></a>          <span class="dt">cex =</span> <span class="fl">0.9</span>, <span class="dt">adj =</span> <span class="fl">0.5</span>, </span>
<span id="cb435-11"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb435-11"></a>        <span class="dt">main =</span> <span class="kw">paste0</span>(<span class="st">&#39;with RUVs, k = &#39;</span>,k), </span>
<span id="cb435-12"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb435-12"></a>        <span class="dt">ylim =</span> <span class="kw">c</span>(<span class="op">-</span><span class="dv">1</span>, <span class="dv">1</span>), <span class="dt">xlim =</span> <span class="kw">c</span>(<span class="op">-</span><span class="fl">0.6</span>, <span class="fl">0.6</span>))</span>
<span id="cb435-13"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb435-13"></a>}</span></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:ruvsf1"></span>
<img src="08-rna-seq-analysis_files/figure-html/ruvsf1-1.png" alt="PCA plots on RUVs normalized data with varying number of covariates (k)." width="55%" />
<p class="caption">
FIGURE 8.19: PCA plots on RUVs normalized data with varying number of covariates (k).
</p>
</div>
<p>Based on the separation of case and control samples in the PCA plots in Figure <a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#fig:ruvsf1">8.19</a>,
we can see that the samples are better separated even at k = 2 when using <code>RUVs()</code>. Here, we re-run the <code>RUVs()</code> function using k = 2, in order to do more diagnostic plots. We try to pick a value of k that is good enough to distinguish the samples by condition of interest. While setting the value of k to higher values could improve the percentage of explained variation by the first principle component to up to 61%, we try to avoid setting the value unnecessarily high to avoid removing factors that might also correlate with important biological differences between conditions.</p>
<div class="sourceCode" id="cb436"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb436-1"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb436-1"></a><span class="co"># choose k = 2</span></span>
<span id="cb436-2"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb436-2"></a>set_s &lt;-<span class="st"> </span><span class="kw">RUVs</span>(set, <span class="kw">unique</span>(<span class="kw">rownames</span>(set)), <span class="dt">k=</span><span class="dv">2</span>, differences) <span class="co">#</span></span></code></pre></div>
<p>Now let’s do diagnostics again: compare the count matrices with or without RUVs processing, comparing RLE plots (Figure <a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#fig:ruvsf2">8.20</a>) and PCA plots (Figure <a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#fig:ruvsf3">8.21</a>) to see the effect of RUVg on the normalization and separation of case and control samples.</p>
<div class="sourceCode" id="cb437"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb437-1"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb437-1"></a><span class="co">## compare the initial and processed objects</span></span>
<span id="cb437-2"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb437-2"></a><span class="co">## RLE plots</span></span>
<span id="cb437-3"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb437-3"></a> </span>
<span id="cb437-4"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb437-4"></a><span class="kw">par</span>(<span class="dt">mfrow =</span> <span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">2</span>))</span>
<span id="cb437-5"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb437-5"></a><span class="kw">plotRLE</span>(set, <span class="dt">outline=</span><span class="ot">FALSE</span>, <span class="dt">ylim=</span><span class="kw">c</span>(<span class="op">-</span><span class="dv">4</span>, <span class="dv">4</span>), </span>
<span id="cb437-6"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb437-6"></a>        <span class="dt">col=</span><span class="kw">as.numeric</span>(colData<span class="op">$</span>group), </span>
<span id="cb437-7"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb437-7"></a>        <span class="dt">main =</span> <span class="st">&#39;without RUVs&#39;</span>)</span>
<span id="cb437-8"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb437-8"></a><span class="kw">plotRLE</span>(set_s, <span class="dt">outline=</span><span class="ot">FALSE</span>, <span class="dt">ylim=</span><span class="kw">c</span>(<span class="op">-</span><span class="dv">4</span>, <span class="dv">4</span>), </span>
<span id="cb437-9"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb437-9"></a>        <span class="dt">col=</span><span class="kw">as.numeric</span>(colData<span class="op">$</span>group),</span>
<span id="cb437-10"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb437-10"></a>        <span class="dt">main =</span> <span class="st">&#39;with RUVs&#39;</span>)</span></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:ruvsf2"></span>
<img src="08-rna-seq-analysis_files/figure-html/ruvsf2-1.png" alt="RLE plots to observe the effect of RUVs." width="55%" />
<p class="caption">
FIGURE 8.20: RLE plots to observe the effect of RUVs.
</p>
</div>
<div class="sourceCode" id="cb438"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb438-1"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb438-1"></a><span class="co">## PCA plots</span></span>
<span id="cb438-2"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb438-2"></a> </span>
<span id="cb438-3"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb438-3"></a><span class="kw">par</span>(<span class="dt">mfrow =</span> <span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">2</span>))</span>
<span id="cb438-4"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb438-4"></a><span class="kw">plotPCA</span>(set, <span class="dt">col=</span><span class="kw">as.numeric</span>(colData<span class="op">$</span>group), </span>
<span id="cb438-5"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb438-5"></a>        <span class="dt">main =</span> <span class="st">&#39;without RUVs&#39;</span>, <span class="dt">adj =</span> <span class="fl">0.5</span>,</span>
<span id="cb438-6"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb438-6"></a>        <span class="dt">ylim =</span> <span class="kw">c</span>(<span class="op">-</span><span class="fl">0.75</span>, <span class="fl">0.75</span>), <span class="dt">xlim =</span> <span class="kw">c</span>(<span class="op">-</span><span class="fl">0.75</span>, <span class="fl">0.75</span>))</span>
<span id="cb438-7"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb438-7"></a><span class="kw">plotPCA</span>(set_s, <span class="dt">col=</span><span class="kw">as.numeric</span>(colData<span class="op">$</span>group), </span>
<span id="cb438-8"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb438-8"></a>        <span class="dt">main =</span> <span class="st">&#39;with RUVs&#39;</span>, <span class="dt">adj =</span> <span class="fl">0.5</span>,</span>
<span id="cb438-9"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb438-9"></a>        <span class="dt">ylim =</span> <span class="kw">c</span>(<span class="op">-</span><span class="fl">0.75</span>, <span class="fl">0.75</span>), <span class="dt">xlim =</span> <span class="kw">c</span>(<span class="op">-</span><span class="fl">0.75</span>, <span class="fl">0.75</span>))</span></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:ruvsf3"></span>
<img src="08-rna-seq-analysis_files/figure-html/ruvsf3-1.png" alt="PCA plots to observe the effect of RUVs." width="55%" />
<p class="caption">
FIGURE 8.21: PCA plots to observe the effect of RUVs.
</p>
</div>
<p>Let’s compare PCA results from RUVs and RUVg with the initial raw counts matrix. We will simply run the <code>plotPCA()</code> function on different normalization schemes. The resulting plots are in Figure <a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#fig:ruvcompare">8.22</a>:</p>
<div class="sourceCode" id="cb439"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb439-1"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb439-1"></a><span class="kw">par</span>(<span class="dt">mfrow =</span> <span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">3</span>))</span>
<span id="cb439-2"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb439-2"></a><span class="kw">plotPCA</span>(countData, <span class="dt">col=</span><span class="kw">as.numeric</span>(colData<span class="op">$</span>group), </span>
<span id="cb439-3"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb439-3"></a>        <span class="dt">main =</span> <span class="st">&#39;without RUV - raw counts&#39;</span>, <span class="dt">adj =</span> <span class="fl">0.5</span>,</span>
<span id="cb439-4"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb439-4"></a>        <span class="dt">ylim =</span> <span class="kw">c</span>(<span class="op">-</span><span class="fl">0.75</span>, <span class="fl">0.75</span>), <span class="dt">xlim =</span> <span class="kw">c</span>(<span class="op">-</span><span class="fl">0.75</span>, <span class="fl">0.75</span>))</span>
<span id="cb439-5"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb439-5"></a><span class="kw">plotPCA</span>(set_g, <span class="dt">col=</span><span class="kw">as.numeric</span>(colData<span class="op">$</span>group), </span>
<span id="cb439-6"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb439-6"></a>        <span class="dt">main =</span> <span class="st">&#39;with RUVg&#39;</span>, <span class="dt">adj =</span> <span class="fl">0.5</span>,</span>
<span id="cb439-7"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb439-7"></a>        <span class="dt">ylim =</span> <span class="kw">c</span>(<span class="op">-</span><span class="fl">0.75</span>, <span class="fl">0.75</span>), <span class="dt">xlim =</span> <span class="kw">c</span>(<span class="op">-</span><span class="fl">0.75</span>, <span class="fl">0.75</span>))</span>
<span id="cb439-8"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb439-8"></a><span class="kw">plotPCA</span>(set_s, <span class="dt">col=</span><span class="kw">as.numeric</span>(colData<span class="op">$</span>group), </span>
<span id="cb439-9"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb439-9"></a>        <span class="dt">main =</span> <span class="st">&#39;with RUVs&#39;</span>, <span class="dt">adj =</span> <span class="fl">0.5</span>,</span>
<span id="cb439-10"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb439-10"></a>        <span class="dt">ylim =</span> <span class="kw">c</span>(<span class="op">-</span><span class="fl">0.75</span>, <span class="fl">0.75</span>), <span class="dt">xlim =</span> <span class="kw">c</span>(<span class="op">-</span><span class="fl">0.75</span>, <span class="fl">0.75</span>))</span></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:ruvcompare"></span>
<img src="08-rna-seq-analysis_files/figure-html/ruvcompare-1.png" alt="PCA plots to observe the before/after effect of RUV functions." width="90%" />
<p class="caption">
FIGURE 8.22: PCA plots to observe the before/after effect of RUV functions.
</p>
</div>
<p>It looks like <code>RUVs()</code> has performed better than <code>RUVg()</code> in this case. So, let’s use count data that is processed by <code>RUVs()</code> to re-do the initial heatmap. The resulting heatmap is in Figure <a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#fig:ruvpost">8.23</a>.</p>
<div class="sourceCode" id="cb440"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb440-1"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb440-1"></a><span class="kw">library</span>(EDASeq)</span>
<span id="cb440-2"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb440-2"></a><span class="kw">library</span>(pheatmap)</span>
<span id="cb440-3"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb440-3"></a><span class="co"># extract normalized counts that are cleared from unwanted variation using RUVs</span></span>
<span id="cb440-4"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb440-4"></a>normCountData &lt;-<span class="st"> </span><span class="kw">normCounts</span>(set_s)</span>
<span id="cb440-5"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb440-5"></a>selectedGenes &lt;-<span class="st"> </span><span class="kw">names</span>(<span class="kw">sort</span>(<span class="kw">apply</span>(normCountData, <span class="dv">1</span>, var), </span>
<span id="cb440-6"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb440-6"></a>                            <span class="dt">decreasing =</span> <span class="ot">TRUE</span>))[<span class="dv">1</span><span class="op">:</span><span class="dv">500</span>]</span>
<span id="cb440-7"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb440-7"></a><span class="kw">pheatmap</span>(normCountData[selectedGenes,], </span>
<span id="cb440-8"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb440-8"></a>         <span class="dt">annotation_col =</span> colData, </span>
<span id="cb440-9"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb440-9"></a>         <span class="dt">show_rownames =</span> <span class="ot">FALSE</span>, </span>
<span id="cb440-10"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb440-10"></a>         <span class="dt">cutree_cols =</span> <span class="dv">2</span>,</span>
<span id="cb440-11"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb440-11"></a>         <span class="dt">scale =</span> <span class="st">&#39;row&#39;</span>)</span></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:ruvpost"></span>
<img src="08-rna-seq-analysis_files/figure-html/ruvpost-1.png" alt="Clustering samples using the top 500 most variable genes normalized using RUVs (k = 2)." width="55%" />
<p class="caption">
FIGURE 8.23: Clustering samples using the top 500 most variable genes normalized using RUVs (k = 2).
</p>
</div>
<p>As can be observed the replicates from different groups cluster much better with each other after processing with <code>RUVs()</code>. It is important to note that RUVs uses information from replicates to shift the expression data and it would not work in a confounding design where the replicates of case samples and replicates of the control samples are sequenced in different batches.</p>
</div>
</div>
<div id="re-run-deseq2-with-the-computed-covariates" class="section level4">
<h4><span class="header-section-number">8.3.9.4</span> Re-run DESeq2 with the computed covariates</h4>
<p>Having computed the sources of variation using <code>RUVs()</code>, we can actually integrate these variables with <code>DESeq2</code> to re-do the differential expression analysis.</p>
<div class="sourceCode" id="cb441"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb441-1"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb441-1"></a><span class="kw">library</span>(DESeq2)</span>
<span id="cb441-2"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb441-2"></a><span class="co">#set up DESeqDataSet object</span></span>
<span id="cb441-3"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb441-3"></a>dds &lt;-<span class="st"> </span><span class="kw">DESeqDataSetFromMatrix</span>(<span class="dt">countData =</span> countData,</span>
<span id="cb441-4"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb441-4"></a>                              <span class="dt">colData =</span> colData, </span>
<span id="cb441-5"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb441-5"></a>                              <span class="dt">design =</span> <span class="op">~</span><span class="st"> </span>group)</span>
<span id="cb441-6"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb441-6"></a><span class="co"># filter for low count genes</span></span>
<span id="cb441-7"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb441-7"></a>dds &lt;-<span class="st"> </span>dds[<span class="kw">rowSums</span>(DESeq2<span class="op">::</span><span class="kw">counts</span>(dds)) <span class="op">&gt;</span><span class="st"> </span><span class="dv">10</span>]</span>
<span id="cb441-8"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb441-8"></a></span>
<span id="cb441-9"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb441-9"></a><span class="co"># insert the covariates W1 and W2 computed using RUVs into DESeqDataSet object</span></span>
<span id="cb441-10"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb441-10"></a><span class="kw">colData</span>(dds) &lt;-<span class="st"> </span><span class="kw">cbind</span>(<span class="kw">colData</span>(dds), </span>
<span id="cb441-11"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb441-11"></a>                      <span class="kw">pData</span>(set_s)[<span class="kw">rownames</span>(<span class="kw">colData</span>(dds)), </span>
<span id="cb441-12"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb441-12"></a>                                   <span class="kw">grep</span>(<span class="st">&#39;W_[0-9]&#39;</span>, </span>
<span id="cb441-13"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb441-13"></a>                                        <span class="kw">colnames</span>(<span class="kw">pData</span>(set_s)))])</span>
<span id="cb441-14"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb441-14"></a></span>
<span id="cb441-15"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb441-15"></a><span class="co"># update the design formula for the DESeq analysis (save the variable of</span></span>
<span id="cb441-16"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb441-16"></a><span class="co"># interest to the last!)</span></span>
<span id="cb441-17"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb441-17"></a><span class="kw">design</span>(dds) &lt;-<span class="st"> </span><span class="er">~</span><span class="st"> </span>W_<span class="dv">1</span> <span class="op">+</span><span class="st"> </span>W_<span class="dv">2</span> <span class="op">+</span><span class="st"> </span>group </span>
<span id="cb441-18"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb441-18"></a><span class="co"># repeat the analysis </span></span>
<span id="cb441-19"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb441-19"></a>dds &lt;-<span class="st"> </span><span class="kw">DESeq</span>(dds)</span>
<span id="cb441-20"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb441-20"></a><span class="co"># extract deseq results </span></span>
<span id="cb441-21"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb441-21"></a>res &lt;-<span class="st"> </span><span class="kw">results</span>(dds, <span class="dt">contrast =</span> <span class="kw">c</span>(<span class="st">&#39;group&#39;</span>, <span class="st">&#39;CASE&#39;</span>, <span class="st">&#39;CTRL&#39;</span>))</span>
<span id="cb441-22"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#cb441-22"></a>res &lt;-<span class="st"> </span>res[<span class="kw">order</span>(res<span class="op">$</span>padj),]</span></code></pre></div>
</div>
</div>
</div>
<h3>References</h3>
<div id="refs" class="references">
<div id="ref-noauthor_babraham_nodate">
<p>Andrews. 2010. “Babraham Bioinformatics - FastQC A Quality Control Tool for High Throughput Sequence Data.” <a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/">https://www.bioinformatics.babraham.ac.uk/projects/fastqc/</a>.</p>
</div>
<div id="ref-bolger_trimmomatic:_2014">
<p>Bolger, Lohse, and Usadel. 2014. “Trimmomatic: A Flexible Trimmer for Illumina Sequence Data.” <em>Bioinformatics</em> 30 (15): 2114–20. <a href="https://doi.org/10.1093/bioinformatics/btu170">https://doi.org/10.1093/bioinformatics/btu170</a>.</p>
</div>
<div id="ref-bray_near-optimal_2016">
<p>Bray, Pimentel, Melsted, and Pachter. 2016. “Near-Optimal Probabilistic RNA-Seq Quantification.” <em>Nature Biotechnology</em> 34 (5): 525–27. <a href="https://doi.org/10.1038/nbt.3519">https://doi.org/10.1038/nbt.3519</a>.</p>
</div>
<div id="ref-dobin_star:_2013">
<p>Dobin, Davis, Schlesinger, Drenkow, Zaleski, Jha, Batut, Chaisson, and Gingeras. 2013. “STAR: Ultrafast Universal RNA-Seq Aligner.” <em>Bioinformatics</em> 29 (1): 15–21. <a href="https://doi.org/10.1093/bioinformatics/bts635">https://doi.org/10.1093/bioinformatics/bts635</a>.</p>
</div>
<div id="ref-fabregat_reactome_2018">
<p>Fabregat, Jupe, Matthews, et al. 2018. “The Reactome Pathway Knowledgebase.” <em>Nucleic Acids Research</em> 46 (D1): D649–D655. <a href="https://doi.org/10.1093/nar/gkx1132">https://doi.org/10.1093/nar/gkx1132</a>.</p>
</div>
<div id="ref-gaidatzis_quasr:_2015">
<p>Gaidatzis, Lerch, Hahne, and Stadler. 2015. “QuasR: Quantification and Annotation of Short Reads in R.” <em>Bioinformatics</em> 31 (7): 1130–2. <a href="https://doi.org/10.1093/bioinformatics/btu781">https://doi.org/10.1093/bioinformatics/btu781</a>.</p>
</div>
<div id="ref-gandolfo_rle_2018">
<p>Gandolfo, and Speed. 2018. “RLE Plots: Visualizing Unwanted Variation in High Dimensional Data.” <em>PloS One</em> 13 (2): e0191629. <a href="https://doi.org/10.1371/journal.pone.0191629">https://doi.org/10.1371/journal.pone.0191629</a>.</p>
</div>
<div id="ref-gu_complex_2016">
<p>Gu, Eils, and Schlesner. 2016a. “Complex Heatmaps Reveal Patterns and Correlations in Multidimensional Genomic Data.” <em>Bioinformatics (Oxford, England)</em> 32 (18): 2847–9. <a href="https://doi.org/10.1093/bioinformatics/btw313">https://doi.org/10.1093/bioinformatics/btw313</a>.</p>
</div>
<div id="ref-haas_novo_2013">
<p>Haas, Papanicolaou, Yassour, et al. 2013. “De Novo Transcript Sequence Reconstruction from RNA-Seq: Reference Generation and Analysis with Trinity.” <em>Nature Protocols</em> 8 (8). <a href="https://doi.org/10.1038/nprot.2013.084">https://doi.org/10.1038/nprot.2013.084</a>.</p>
</div>
<div id="ref-jiang_synthetic_2011">
<p>Jiang, Schlesinger, Davis, Zhang, Li, Salit, Gingeras, and Oliver. 2011. “Synthetic Spike-in Standards for RNA-Seq Experiments.” <em>Genome Research</em> 21 (9): 1543–51. <a href="https://doi.org/10.1101/gr.121095.111">https://doi.org/10.1101/gr.121095.111</a>.</p>
</div>
<div id="ref-kanehisa_kegg_2016">
<p>Kanehisa, Sato, Kawashima, Furumichi, and Tanabe. 2016. “KEGG as a Reference Resource for Gene and Protein Annotation.” <em>Nucleic Acids Research</em> 44 (Database issue): D457–D462. <a href="https://doi.org/10.1093/nar/gkv1070">https://doi.org/10.1093/nar/gkv1070</a>.</p>
</div>
<div id="ref-kim_hisat:_2015">
<p>Kim, Langmead, and Salzberg. 2015. “HISAT: A Fast Spliced Aligner with Low Memory Requirements.” <em>Nature Methods</em> 12 (4): 357–60. <a href="https://doi.org/10.1038/nmeth.3317">https://doi.org/10.1038/nmeth.3317</a>.</p>
</div>
<div id="ref-kim_tophat2:_2013">
<p>Kim, Pertea, Trapnell, Pimentel, Kelley, and Salzberg. 2013. “TopHat2: Accurate Alignment of Transcriptomes in the Presence of Insertions, Deletions and Gene Fusions.” <em>Genome Biology</em> 14 (4): R36. <a href="https://doi.org/10.1186/gb-2013-14-4-r36">https://doi.org/10.1186/gb-2013-14-4-r36</a>.</p>
</div>
<div id="ref-pheatmap">
<p>Kolde. 2019. <em>Pheatmap: Pretty Heatmaps</em>. <a href="https://CRAN.R-project.org/package=pheatmap">https://CRAN.R-project.org/package=pheatmap</a>.</p>
</div>
<div id="ref-leek_sva_2012">
<p>Leek, Johnson, Parker, Jaffe, and Storey. 2012. “The Sva Package for Removing Batch Effects and Other Unwanted Variation in High-Throughput Experiments.” <em>Bioinformatics</em> 28 (6): 882–83. <a href="https://doi.org/10.1093/bioinformatics/bts034">https://doi.org/10.1093/bioinformatics/bts034</a>.</p>
</div>
<div id="ref-love_moderated_2014">
<p>Love, Huber, and Anders. 2014. “Moderated Estimation of Fold Change and Dispersion for RNA-Seq Data with DESeq2.” <em>Genome Biology</em> 15 (12). <a href="https://doi.org/10.1186/s13059-014-0550-8">https://doi.org/10.1186/s13059-014-0550-8</a>.</p>
</div>
<div id="ref-luo_gage:_2009">
<p>Luo, Friedman, Shedden, Hankenson, and Woolf. 2009. “GAGE: Generally Applicable Gene Set Enrichment for Pathway Analysis.” <em>BMC Bioinformatics</em> 10 (1): 161. <a href="https://doi.org/10.1186/1471-2105-10-161">https://doi.org/10.1186/1471-2105-10-161</a>.</p>
</div>
<div id="ref-maza_comparison_2013">
<p>Maza, Frasse, Senin, Bouzayen, and Zouine. 2013. “Comparison of Normalization Methods for Differential Gene Expression Analysis in RNA-Seq Experiments: A Matter of Relative Size of Studied Transcriptomes.” <em>Communicative &amp; Integrative Biology</em> 6 (6): e25849. <a href="https://doi.org/10.4161/cib.25849">https://doi.org/10.4161/cib.25849</a>.</p>
</div>
<div id="ref-morgan_shortread:_2009">
<p>Morgan, Anders, Lawrence, Aboyoun, Pagès, and Gentleman. 2009. “ShortRead: A Bioconductor Package for Input, Quality Assessment and Exploration of High-Throughput Sequence Data.” <em>Bioinformatics</em> 25 (19): 2607–8. <a href="https://doi.org/10.1093/bioinformatics/btp450">https://doi.org/10.1093/bioinformatics/btp450</a>.</p>
</div>
<div id="ref-mortazavi_mapping_2008">
<p>Mortazavi, Williams, McCue, Schaeffer, and Wold. 2008. “Mapping and Quantifying Mammalian Transcriptomes by RNA-Seq.” <em>Nature Methods</em> 5 (7): 621–28. <a href="https://doi.org/10.1038/nmeth.1226">https://doi.org/10.1038/nmeth.1226</a>.</p>
</div>
<div id="ref-patro_salmon:_2017">
<p>Patro, Duggal, Love, Irizarry, and Kingsford. 2017. “Salmon: Fast and Bias-Aware Quantification of Transcript Expression Using Dual-Phase Inference.” <em>Nature Methods</em> 14 (4): 417–19. <a href="https://doi.org/10.1038/nmeth.4197">https://doi.org/10.1038/nmeth.4197</a>.</p>
</div>
<div id="ref-patro_sailfish_2014">
<p>Patro, Mount, and Kingsford. 2014. “Sailfish Enables Alignment-Free Isoform Quantification from RNA-Seq Reads Using Lightweight Algorithms.” <em>Nature Biotechnology</em> 32 (5): 462–64. <a href="https://doi.org/10.1038/nbt.2862">https://doi.org/10.1038/nbt.2862</a>.</p>
</div>
<div id="ref-risso_normalization_2014">
<p>Risso, Ngai, Speed, and Dudoit. 2014. “Normalization of RNA-Seq Data Using Factor Analysis of Control Genes or Samples.” <em>Nature Biotechnology</em> 32 (9): 896–902. <a href="https://doi.org/10.1038/nbt.2931">https://doi.org/10.1038/nbt.2931</a>.</p>
</div>
<div id="ref-risso_gc-content_2011">
<p>Risso, Schwartz, Sherlock, and Dudoit. 2011. “GC-Content Normalization for RNA-Seq Data.” <em>BMC Bioinformatics</em> 12 (December): 480. <a href="https://doi.org/10.1186/1471-2105-12-480">https://doi.org/10.1186/1471-2105-12-480</a>.</p>
</div>
<div id="ref-robertson_novo_2010">
<p>Robertson, Schein, Chiu, et al. 2010. “<em>De Novo</em> Assembly and Analysis of RNA-Seq Data.” <em>Nature Methods</em> 7 (11): 909–12. <a href="https://doi.org/10.1038/nmeth.1517">https://doi.org/10.1038/nmeth.1517</a>.</p>
</div>
<div id="ref-robinson_edger:_2010">
<p>Robinson, McCarthy, and Smyth. 2010. “edgeR: A Bioconductor Package for Differential Expression Analysis of Digital Gene Expression Data.” <em>Bioinformatics (Oxford, England)</em> 26 (1): 139–40. <a href="https://doi.org/10.1093/bioinformatics/btp616">https://doi.org/10.1093/bioinformatics/btp616</a>.</p>
</div>
<div id="ref-subramanian_gene_2005">
<p>Subramanian, Tamayo, Mootha, et al. 2005. “Gene Set Enrichment Analysis: A Knowledge-Based Approach for Interpreting Genome-Wide Expression Profiles.” <em>Proceedings of the National Academy of Sciences</em> 102 (43): 15545–50. <a href="https://doi.org/10.1073/pnas.0506580102">https://doi.org/10.1073/pnas.0506580102</a>.</p>
</div>
<div id="ref-trapnell_transcript_2010">
<p>Trapnell, Williams, Pertea, Mortazavi, Kwan, Baren, Salzberg, Wold, and Pachter. 2010. “Transcript Assembly and Quantification by RNA-Seq Reveals Unannotated Transcripts and Isoform Switching During Cell Differentiation.” <em>Nature Biotechnology</em> 28 (5): 511–15. <a href="https://doi.org/10.1038/nbt.1621">https://doi.org/10.1038/nbt.1621</a>.</p>
</div>
<div id="ref-wu_gmap_2016">
<p>Wu, Reeder, Lawrence, Becker, and Brauer. 2016. “GMAP and GSNAP for Genomic Sequence Alignment: Enhancements to Speed, Accuracy, and Functionality.” <em>Methods in Molecular Biology (Clifton, N.J.)</em> 1418: 283–334. <a href="https://doi.org/10.1007/978-1-4939-3578-9_15">https://doi.org/10.1007/978-1-4939-3578-9_15</a>.</p>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="methods-to-detect-gene-expression.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="other-applications-of-rna-seq.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": true,
"facebook": false,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/compgenomr/book/edit/master/08-rna-seq-analysis.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": null,
"toc": {
"collapse": "none"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
