<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>11.3 Matrix factorization methods for unsupervised multi-omics data integration | Computational Genomics with R</title>
  <meta name="description" content="A guide to computationa genomics using R. The book covers fundemental topics with practical examples for an interdisciplinery audience" />
  <meta name="generator" content="bookdown 0.19 and GitBook 2.6.7" />

  <meta property="og:title" content="11.3 Matrix factorization methods for unsupervised multi-omics data integration | Computational Genomics with R" />
  <meta property="og:type" content="book" />
  <meta property="og:url" content="https://compmgenomr.github.io/book/" />
  <meta property="og:image" content="https://compmgenomr.github.io/book/images/cover.jpg" />
  <meta property="og:description" content="A guide to computationa genomics using R. The book covers fundemental topics with practical examples for an interdisciplinery audience" />
  <meta name="github-repo" content="compgenomr/book" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="11.3 Matrix factorization methods for unsupervised multi-omics data integration | Computational Genomics with R" />
  
  <meta name="twitter:description" content="A guide to computationa genomics using R. The book covers fundemental topics with practical examples for an interdisciplinery audience" />
  <meta name="twitter:image" content="https://compmgenomr.github.io/book/images/cover.jpg" />

<meta name="author" content="Altuna Akalin" />


<meta name="date" content="2020-09-30" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="latent-variable-models-for-multi-omics-integration.html"/>
<link rel="next" href="clustering-using-latent-factors.html"/>
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />









<!-- Global site tag (gtag.js) - Google Analytics -->
<script async src="https://www.googletagmanager.com/gtag/js?id=UA-83786243-1"></script>
<script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-83786243-1');
</script>


<style type="text/css">
code.sourceCode > span { display: inline-block; line-height: 1.25; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<link rel="stylesheet" href="css/style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">Computational Genomics with R</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Preface</a><ul>
<li class="chapter" data-level="" data-path="who-is-this-book-for.html"><a href="who-is-this-book-for.html"><i class="fa fa-check"></i>Who is this book for?</a><ul>
<li class="chapter" data-level="" data-path="who-is-this-book-for.html"><a href="who-is-this-book-for.html#what-will-you-get-out-of-this"><i class="fa fa-check"></i>What will you get out of this?</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="structure-of-the-book.html"><a href="structure-of-the-book.html"><i class="fa fa-check"></i>Structure of the book</a></li>
<li class="chapter" data-level="" data-path="software-information-and-conventions.html"><a href="software-information-and-conventions.html"><i class="fa fa-check"></i>Software information and conventions</a><ul>
<li class="chapter" data-level="" data-path="software-information-and-conventions.html"><a href="software-information-and-conventions.html#assignment-operator-convention"><i class="fa fa-check"></i>Assignment operator convention</a></li>
<li class="chapter" data-level="" data-path="software-information-and-conventions.html"><a href="software-information-and-conventions.html#packages-needed-to-run-the-book-code"><i class="fa fa-check"></i>Packages needed to run the book code</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="data-for-the-book.html"><a href="data-for-the-book.html"><i class="fa fa-check"></i>Data for the book</a></li>
<li class="chapter" data-level="" data-path="exercises-in-the-book.html"><a href="exercises-in-the-book.html"><i class="fa fa-check"></i>Exercises in the book</a></li>
<li class="chapter" data-level="" data-path="reproducibility-statement.html"><a href="reproducibility-statement.html"><i class="fa fa-check"></i>Reproducibility statement</a></li>
<li class="chapter" data-level="" data-path="acknowledgements.html"><a href="acknowledgements.html"><i class="fa fa-check"></i>Acknowledgements</a></li>
<li class="chapter" data-level="" data-path="how-to-contribute.html"><a href="how-to-contribute.html"><i class="fa fa-check"></i>How to contribute</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="about-the-authors.html"><a href="about-the-authors.html"><i class="fa fa-check"></i>About the Authors</a></li>
<li class="chapter" data-level="1" data-path="intro.html"><a href="intro.html"><i class="fa fa-check"></i><b>1</b> Introduction to Genomics</a><ul>
<li class="chapter" data-level="1.1" data-path="genes-dna-and-central-dogma.html"><a href="genes-dna-and-central-dogma.html"><i class="fa fa-check"></i><b>1.1</b> Genes, DNA and central dogma</a><ul>
<li class="chapter" data-level="1.1.1" data-path="genes-dna-and-central-dogma.html"><a href="genes-dna-and-central-dogma.html#what-is-a-genome"><i class="fa fa-check"></i><b>1.1.1</b> What is a genome?</a></li>
<li class="chapter" data-level="1.1.2" data-path="genes-dna-and-central-dogma.html"><a href="genes-dna-and-central-dogma.html#what-is-a-gene"><i class="fa fa-check"></i><b>1.1.2</b> What is a gene?</a></li>
<li class="chapter" data-level="1.1.3" data-path="genes-dna-and-central-dogma.html"><a href="genes-dna-and-central-dogma.html#how-are-genes-controlled-transcriptional-and-post-transcriptional-regulation"><i class="fa fa-check"></i><b>1.1.3</b> How are genes controlled? Transcriptional and post-transcriptional regulation</a></li>
<li class="chapter" data-level="1.1.4" data-path="genes-dna-and-central-dogma.html"><a href="genes-dna-and-central-dogma.html#what-does-a-gene-look-like"><i class="fa fa-check"></i><b>1.1.4</b> What does a gene look like?</a></li>
</ul></li>
<li class="chapter" data-level="1.2" data-path="elements-of-gene-regulation.html"><a href="elements-of-gene-regulation.html"><i class="fa fa-check"></i><b>1.2</b> Elements of gene regulation</a><ul>
<li class="chapter" data-level="1.2.1" data-path="elements-of-gene-regulation.html"><a href="elements-of-gene-regulation.html#transcriptional-regulation"><i class="fa fa-check"></i><b>1.2.1</b> Transcriptional regulation</a></li>
<li class="chapter" data-level="1.2.2" data-path="elements-of-gene-regulation.html"><a href="elements-of-gene-regulation.html#post-transcriptional-regulation"><i class="fa fa-check"></i><b>1.2.2</b> Post-transcriptional regulation</a></li>
</ul></li>
<li class="chapter" data-level="1.3" data-path="shaping-the-genome-dna-mutation.html"><a href="shaping-the-genome-dna-mutation.html"><i class="fa fa-check"></i><b>1.3</b> Shaping the genome: DNA mutation</a></li>
<li class="chapter" data-level="1.4" data-path="high-throughput-experimental-methods-in-genomics.html"><a href="high-throughput-experimental-methods-in-genomics.html"><i class="fa fa-check"></i><b>1.4</b> High-throughput experimental methods in genomics</a><ul>
<li class="chapter" data-level="1.4.1" data-path="high-throughput-experimental-methods-in-genomics.html"><a href="high-throughput-experimental-methods-in-genomics.html#the-general-idea-behind-high-throughput-techniques"><i class="fa fa-check"></i><b>1.4.1</b> The general idea behind high-throughput techniques</a></li>
<li class="chapter" data-level="1.4.2" data-path="high-throughput-experimental-methods-in-genomics.html"><a href="high-throughput-experimental-methods-in-genomics.html#high-throughput-sequencing"><i class="fa fa-check"></i><b>1.4.2</b> High-throughput sequencing</a></li>
</ul></li>
<li class="chapter" data-level="1.5" data-path="visualization-and-data-repositories-for-genomics.html"><a href="visualization-and-data-repositories-for-genomics.html"><i class="fa fa-check"></i><b>1.5</b> Visualization and data repositories for genomics</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="Rintro.html"><a href="Rintro.html"><i class="fa fa-check"></i><b>2</b> Introduction to R for Genomic Data Analysis</a><ul>
<li class="chapter" data-level="2.1" data-path="steps-of-genomic-data-analysis.html"><a href="steps-of-genomic-data-analysis.html"><i class="fa fa-check"></i><b>2.1</b> Steps of (genomic) data analysis</a><ul>
<li class="chapter" data-level="2.1.1" data-path="steps-of-genomic-data-analysis.html"><a href="steps-of-genomic-data-analysis.html#data-collection"><i class="fa fa-check"></i><b>2.1.1</b> Data collection</a></li>
<li class="chapter" data-level="2.1.2" data-path="steps-of-genomic-data-analysis.html"><a href="steps-of-genomic-data-analysis.html#data-quality-check-and-cleaning"><i class="fa fa-check"></i><b>2.1.2</b> Data quality check and cleaning</a></li>
<li class="chapter" data-level="2.1.3" data-path="steps-of-genomic-data-analysis.html"><a href="steps-of-genomic-data-analysis.html#data-processing"><i class="fa fa-check"></i><b>2.1.3</b> Data processing</a></li>
<li class="chapter" data-level="2.1.4" data-path="steps-of-genomic-data-analysis.html"><a href="steps-of-genomic-data-analysis.html#exploratory-data-analysis-and-modeling"><i class="fa fa-check"></i><b>2.1.4</b> Exploratory data analysis and modeling</a></li>
<li class="chapter" data-level="2.1.5" data-path="steps-of-genomic-data-analysis.html"><a href="steps-of-genomic-data-analysis.html#visualization-and-reporting"><i class="fa fa-check"></i><b>2.1.5</b> Visualization and reporting</a></li>
<li class="chapter" data-level="2.1.6" data-path="steps-of-genomic-data-analysis.html"><a href="steps-of-genomic-data-analysis.html#why-use-r-for-genomics"><i class="fa fa-check"></i><b>2.1.6</b> Why use R for genomics ?</a></li>
</ul></li>
<li class="chapter" data-level="2.2" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html"><i class="fa fa-check"></i><b>2.2</b> Getting started with R</a><ul>
<li class="chapter" data-level="2.2.1" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#installing-packages"><i class="fa fa-check"></i><b>2.2.1</b> Installing packages</a></li>
<li class="chapter" data-level="2.2.2" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#installing-packages-in-custom-locations"><i class="fa fa-check"></i><b>2.2.2</b> Installing packages in custom locations</a></li>
<li class="chapter" data-level="2.2.3" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#getting-help-on-functions-and-packages"><i class="fa fa-check"></i><b>2.2.3</b> Getting help on functions and packages</a></li>
</ul></li>
<li class="chapter" data-level="2.3" data-path="computations-in-r.html"><a href="computations-in-r.html"><i class="fa fa-check"></i><b>2.3</b> Computations in R</a></li>
<li class="chapter" data-level="2.4" data-path="data-structures.html"><a href="data-structures.html"><i class="fa fa-check"></i><b>2.4</b> Data structures</a><ul>
<li class="chapter" data-level="2.4.1" data-path="data-structures.html"><a href="data-structures.html#vectors"><i class="fa fa-check"></i><b>2.4.1</b> Vectors</a></li>
<li class="chapter" data-level="2.4.2" data-path="data-structures.html"><a href="data-structures.html#matrices"><i class="fa fa-check"></i><b>2.4.2</b> Matrices</a></li>
<li class="chapter" data-level="2.4.3" data-path="data-structures.html"><a href="data-structures.html#data-frames"><i class="fa fa-check"></i><b>2.4.3</b> Data frames</a></li>
<li class="chapter" data-level="2.4.4" data-path="data-structures.html"><a href="data-structures.html#lists"><i class="fa fa-check"></i><b>2.4.4</b> Lists</a></li>
<li class="chapter" data-level="2.4.5" data-path="data-structures.html"><a href="data-structures.html#factors"><i class="fa fa-check"></i><b>2.4.5</b> Factors</a></li>
</ul></li>
<li class="chapter" data-level="2.5" data-path="data-types.html"><a href="data-types.html"><i class="fa fa-check"></i><b>2.5</b> Data types</a></li>
<li class="chapter" data-level="2.6" data-path="reading-and-writing-data.html"><a href="reading-and-writing-data.html"><i class="fa fa-check"></i><b>2.6</b> Reading and writing data</a><ul>
<li class="chapter" data-level="2.6.1" data-path="reading-and-writing-data.html"><a href="reading-and-writing-data.html#reading-large-files"><i class="fa fa-check"></i><b>2.6.1</b> Reading large files</a></li>
</ul></li>
<li class="chapter" data-level="2.7" data-path="plotting-in-r-with-base-graphics.html"><a href="plotting-in-r-with-base-graphics.html"><i class="fa fa-check"></i><b>2.7</b> Plotting in R with base graphics</a><ul>
<li class="chapter" data-level="2.7.1" data-path="plotting-in-r-with-base-graphics.html"><a href="plotting-in-r-with-base-graphics.html#combining-multiple-plots"><i class="fa fa-check"></i><b>2.7.1</b> Combining multiple plots</a></li>
<li class="chapter" data-level="2.7.2" data-path="plotting-in-r-with-base-graphics.html"><a href="plotting-in-r-with-base-graphics.html#saving-plots"><i class="fa fa-check"></i><b>2.7.2</b> Saving plots</a></li>
</ul></li>
<li class="chapter" data-level="2.8" data-path="plotting-in-r-with-ggplot2.html"><a href="plotting-in-r-with-ggplot2.html"><i class="fa fa-check"></i><b>2.8</b> Plotting in R with ggplot2</a><ul>
<li class="chapter" data-level="2.8.1" data-path="plotting-in-r-with-ggplot2.html"><a href="plotting-in-r-with-ggplot2.html#combining-multiple-plots-1"><i class="fa fa-check"></i><b>2.8.1</b> Combining multiple plots</a></li>
<li class="chapter" data-level="2.8.2" data-path="plotting-in-r-with-ggplot2.html"><a href="plotting-in-r-with-ggplot2.html#ggplot2-and-tidyverse"><i class="fa fa-check"></i><b>2.8.2</b> ggplot2 and tidyverse</a></li>
</ul></li>
<li class="chapter" data-level="2.9" data-path="functions-and-control-structures-for-ifelse-etc-.html"><a href="functions-and-control-structures-for-ifelse-etc-.html"><i class="fa fa-check"></i><b>2.9</b> Functions and control structures (for, if/else etc.)</a><ul>
<li class="chapter" data-level="2.9.1" data-path="functions-and-control-structures-for-ifelse-etc-.html"><a href="functions-and-control-structures-for-ifelse-etc-.html#user-defined-functions"><i class="fa fa-check"></i><b>2.9.1</b> User-defined functions</a></li>
<li class="chapter" data-level="2.9.2" data-path="functions-and-control-structures-for-ifelse-etc-.html"><a href="functions-and-control-structures-for-ifelse-etc-.html#loops-and-looping-structures-in-r"><i class="fa fa-check"></i><b>2.9.2</b> Loops and looping structures in R</a></li>
</ul></li>
<li class="chapter" data-level="2.10" data-path="exercises.html"><a href="exercises.html"><i class="fa fa-check"></i><b>2.10</b> Exercises</a><ul>
<li class="chapter" data-level="2.10.1" data-path="exercises.html"><a href="exercises.html#computations-in-r-1"><i class="fa fa-check"></i><b>2.10.1</b> Computations in R</a></li>
<li class="chapter" data-level="2.10.2" data-path="exercises.html"><a href="exercises.html#data-structures-in-r"><i class="fa fa-check"></i><b>2.10.2</b> Data structures in R</a></li>
<li class="chapter" data-level="2.10.3" data-path="exercises.html"><a href="exercises.html#reading-in-and-writing-data-out-in-r"><i class="fa fa-check"></i><b>2.10.3</b> Reading in and writing data out in R</a></li>
<li class="chapter" data-level="2.10.4" data-path="exercises.html"><a href="exercises.html#plotting-in-r"><i class="fa fa-check"></i><b>2.10.4</b> Plotting in R</a></li>
<li class="chapter" data-level="2.10.5" data-path="exercises.html"><a href="exercises.html#functions-and-control-structures-for-ifelse-etc.-1"><i class="fa fa-check"></i><b>2.10.5</b> Functions and control structures (for, if/else, etc.)</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="3" data-path="stats.html"><a href="stats.html"><i class="fa fa-check"></i><b>3</b> Statistics for Genomics</a><ul>
<li class="chapter" data-level="3.1" data-path="how-to-summarize-collection-of-data-points-the-idea-behind-statistical-distributions.html"><a href="how-to-summarize-collection-of-data-points-the-idea-behind-statistical-distributions.html"><i class="fa fa-check"></i><b>3.1</b> How to summarize collection of data points: The idea behind statistical distributions</a><ul>
<li class="chapter" data-level="3.1.1" data-path="how-to-summarize-collection-of-data-points-the-idea-behind-statistical-distributions.html"><a href="how-to-summarize-collection-of-data-points-the-idea-behind-statistical-distributions.html#describing-the-central-tendency-mean-and-median"><i class="fa fa-check"></i><b>3.1.1</b> Describing the central tendency: Mean and median</a></li>
<li class="chapter" data-level="3.1.2" data-path="how-to-summarize-collection-of-data-points-the-idea-behind-statistical-distributions.html"><a href="how-to-summarize-collection-of-data-points-the-idea-behind-statistical-distributions.html#describing-the-spread-measurements-of-variation"><i class="fa fa-check"></i><b>3.1.2</b> Describing the spread: Measurements of variation</a></li>
<li class="chapter" data-level="3.1.3" data-path="how-to-summarize-collection-of-data-points-the-idea-behind-statistical-distributions.html"><a href="how-to-summarize-collection-of-data-points-the-idea-behind-statistical-distributions.html#precision-of-estimates-confidence-intervals"><i class="fa fa-check"></i><b>3.1.3</b> Precision of estimates: Confidence intervals</a></li>
</ul></li>
<li class="chapter" data-level="3.2" data-path="how-to-test-for-differences-between-samples.html"><a href="how-to-test-for-differences-between-samples.html"><i class="fa fa-check"></i><b>3.2</b> How to test for differences between samples</a><ul>
<li class="chapter" data-level="3.2.1" data-path="how-to-test-for-differences-between-samples.html"><a href="how-to-test-for-differences-between-samples.html#randomization-based-testing-for-difference-of-the-means"><i class="fa fa-check"></i><b>3.2.1</b> Randomization-based testing for difference of the means</a></li>
<li class="chapter" data-level="3.2.2" data-path="how-to-test-for-differences-between-samples.html"><a href="how-to-test-for-differences-between-samples.html#using-t-test-for-difference-of-the-means-between-two-samples"><i class="fa fa-check"></i><b>3.2.2</b> Using t-test for difference of the means between two samples</a></li>
<li class="chapter" data-level="3.2.3" data-path="how-to-test-for-differences-between-samples.html"><a href="how-to-test-for-differences-between-samples.html#multiple-testing-correction"><i class="fa fa-check"></i><b>3.2.3</b> Multiple testing correction</a></li>
<li class="chapter" data-level="3.2.4" data-path="how-to-test-for-differences-between-samples.html"><a href="how-to-test-for-differences-between-samples.html#moderated-t-tests-using-information-from-multiple-comparisons"><i class="fa fa-check"></i><b>3.2.4</b> Moderated t-tests: Using information from multiple comparisons</a></li>
</ul></li>
<li class="chapter" data-level="3.3" data-path="relationship-between-variables-linear-models-and-correlation.html"><a href="relationship-between-variables-linear-models-and-correlation.html"><i class="fa fa-check"></i><b>3.3</b> Relationship between variables: Linear models and correlation</a><ul>
<li class="chapter" data-level="3.3.1" data-path="relationship-between-variables-linear-models-and-correlation.html"><a href="relationship-between-variables-linear-models-and-correlation.html#how-to-fit-a-line"><i class="fa fa-check"></i><b>3.3.1</b> How to fit a line</a></li>
<li class="chapter" data-level="3.3.2" data-path="relationship-between-variables-linear-models-and-correlation.html"><a href="relationship-between-variables-linear-models-and-correlation.html#how-to-estimate-the-error-of-the-coefficients"><i class="fa fa-check"></i><b>3.3.2</b> How to estimate the error of the coefficients</a></li>
<li class="chapter" data-level="3.3.3" data-path="relationship-between-variables-linear-models-and-correlation.html"><a href="relationship-between-variables-linear-models-and-correlation.html#accuracy-of-the-model"><i class="fa fa-check"></i><b>3.3.3</b> Accuracy of the model</a></li>
<li class="chapter" data-level="3.3.4" data-path="relationship-between-variables-linear-models-and-correlation.html"><a href="relationship-between-variables-linear-models-and-correlation.html#regression-with-categorical-variables"><i class="fa fa-check"></i><b>3.3.4</b> Regression with categorical variables</a></li>
<li class="chapter" data-level="3.3.5" data-path="relationship-between-variables-linear-models-and-correlation.html"><a href="relationship-between-variables-linear-models-and-correlation.html#regression-pitfalls"><i class="fa fa-check"></i><b>3.3.5</b> Regression pitfalls</a></li>
</ul></li>
<li class="chapter" data-level="3.4" data-path="exercises-1.html"><a href="exercises-1.html"><i class="fa fa-check"></i><b>3.4</b> Exercises</a><ul>
<li class="chapter" data-level="3.4.1" data-path="exercises-1.html"><a href="exercises-1.html#how-to-summarize-collection-of-data-points-the-idea-behind-statistical-distributions-1"><i class="fa fa-check"></i><b>3.4.1</b> How to summarize collection of data points: The idea behind statistical distributions</a></li>
<li class="chapter" data-level="3.4.2" data-path="exercises-1.html"><a href="exercises-1.html#how-to-test-for-differences-in-samples"><i class="fa fa-check"></i><b>3.4.2</b> How to test for differences in samples</a></li>
<li class="chapter" data-level="3.4.3" data-path="exercises-1.html"><a href="exercises-1.html#relationship-between-variables-linear-models-and-correlation-1"><i class="fa fa-check"></i><b>3.4.3</b> Relationship between variables: Linear models and correlation</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="4" data-path="unsupervisedLearning.html"><a href="unsupervisedLearning.html"><i class="fa fa-check"></i><b>4</b> Exploratory Data Analysis with Unsupervised Machine Learning</a><ul>
<li class="chapter" data-level="4.1" data-path="clustering-grouping-samples-based-on-their-similarity.html"><a href="clustering-grouping-samples-based-on-their-similarity.html"><i class="fa fa-check"></i><b>4.1</b> Clustering: Grouping samples based on their similarity</a><ul>
<li class="chapter" data-level="4.1.1" data-path="clustering-grouping-samples-based-on-their-similarity.html"><a href="clustering-grouping-samples-based-on-their-similarity.html#distance-metrics"><i class="fa fa-check"></i><b>4.1.1</b> Distance metrics</a></li>
<li class="chapter" data-level="4.1.2" data-path="clustering-grouping-samples-based-on-their-similarity.html"><a href="clustering-grouping-samples-based-on-their-similarity.html#hiearchical-clustering"><i class="fa fa-check"></i><b>4.1.2</b> Hiearchical clustering</a></li>
<li class="chapter" data-level="4.1.3" data-path="clustering-grouping-samples-based-on-their-similarity.html"><a href="clustering-grouping-samples-based-on-their-similarity.html#k-means-clustering"><i class="fa fa-check"></i><b>4.1.3</b> K-means clustering</a></li>
<li class="chapter" data-level="4.1.4" data-path="clustering-grouping-samples-based-on-their-similarity.html"><a href="clustering-grouping-samples-based-on-their-similarity.html#how-to-choose-k-the-number-of-clusters"><i class="fa fa-check"></i><b>4.1.4</b> How to choose “k”, the number of clusters</a></li>
</ul></li>
<li class="chapter" data-level="4.2" data-path="dimensionality-reduction-techniques-visualizing-complex-data-sets-in-2d.html"><a href="dimensionality-reduction-techniques-visualizing-complex-data-sets-in-2d.html"><i class="fa fa-check"></i><b>4.2</b> Dimensionality reduction techniques: Visualizing complex data sets in 2D</a><ul>
<li class="chapter" data-level="4.2.1" data-path="dimensionality-reduction-techniques-visualizing-complex-data-sets-in-2d.html"><a href="dimensionality-reduction-techniques-visualizing-complex-data-sets-in-2d.html#principal-component-analysis"><i class="fa fa-check"></i><b>4.2.1</b> Principal component analysis</a></li>
<li class="chapter" data-level="4.2.2" data-path="dimensionality-reduction-techniques-visualizing-complex-data-sets-in-2d.html"><a href="dimensionality-reduction-techniques-visualizing-complex-data-sets-in-2d.html#other-matrix-factorization-methods-for-dimensionality-reduction"><i class="fa fa-check"></i><b>4.2.2</b> Other matrix factorization methods for dimensionality reduction</a></li>
<li class="chapter" data-level="4.2.3" data-path="dimensionality-reduction-techniques-visualizing-complex-data-sets-in-2d.html"><a href="dimensionality-reduction-techniques-visualizing-complex-data-sets-in-2d.html#multi-dimensional-scaling"><i class="fa fa-check"></i><b>4.2.3</b> Multi-dimensional scaling</a></li>
<li class="chapter" data-level="4.2.4" data-path="dimensionality-reduction-techniques-visualizing-complex-data-sets-in-2d.html"><a href="dimensionality-reduction-techniques-visualizing-complex-data-sets-in-2d.html#t-distributed-stochastic-neighbor-embedding-t-sne"><i class="fa fa-check"></i><b>4.2.4</b> t-Distributed Stochastic Neighbor Embedding (t-SNE)</a></li>
</ul></li>
<li class="chapter" data-level="4.3" data-path="exercises-2.html"><a href="exercises-2.html"><i class="fa fa-check"></i><b>4.3</b> Exercises</a><ul>
<li class="chapter" data-level="4.3.1" data-path="exercises-2.html"><a href="exercises-2.html#clustering"><i class="fa fa-check"></i><b>4.3.1</b> Clustering</a></li>
<li class="chapter" data-level="4.3.2" data-path="exercises-2.html"><a href="exercises-2.html#dimension-reduction"><i class="fa fa-check"></i><b>4.3.2</b> Dimension reduction</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="5" data-path="supervisedLearning.html"><a href="supervisedLearning.html"><i class="fa fa-check"></i><b>5</b> Predictive Modeling with Supervised Machine Learning</a><ul>
<li class="chapter" data-level="5.1" data-path="how-are-machine-learning-models-fit.html"><a href="how-are-machine-learning-models-fit.html"><i class="fa fa-check"></i><b>5.1</b> How are machine learning models fit?</a><ul>
<li class="chapter" data-level="5.1.1" data-path="how-are-machine-learning-models-fit.html"><a href="how-are-machine-learning-models-fit.html#machine-learning-vs.-statistics"><i class="fa fa-check"></i><b>5.1.1</b> Machine learning vs. statistics</a></li>
</ul></li>
<li class="chapter" data-level="5.2" data-path="steps-in-supervised-machine-learning.html"><a href="steps-in-supervised-machine-learning.html"><i class="fa fa-check"></i><b>5.2</b> Steps in supervised machine learning</a></li>
<li class="chapter" data-level="5.3" data-path="use-case-disease-subtype-from-genomics-data.html"><a href="use-case-disease-subtype-from-genomics-data.html"><i class="fa fa-check"></i><b>5.3</b> Use case: Disease subtype from genomics data</a></li>
<li class="chapter" data-level="5.4" data-path="data-preprocessing.html"><a href="data-preprocessing.html"><i class="fa fa-check"></i><b>5.4</b> Data preprocessing</a><ul>
<li class="chapter" data-level="5.4.1" data-path="data-preprocessing.html"><a href="data-preprocessing.html#data-transformation"><i class="fa fa-check"></i><b>5.4.1</b> Data transformation</a></li>
<li class="chapter" data-level="5.4.2" data-path="data-preprocessing.html"><a href="data-preprocessing.html#filtering-data-and-scaling"><i class="fa fa-check"></i><b>5.4.2</b> Filtering data and scaling</a></li>
<li class="chapter" data-level="5.4.3" data-path="data-preprocessing.html"><a href="data-preprocessing.html#dealing-with-missing-values"><i class="fa fa-check"></i><b>5.4.3</b> Dealing with missing values</a></li>
</ul></li>
<li class="chapter" data-level="5.5" data-path="splitting-the-data.html"><a href="splitting-the-data.html"><i class="fa fa-check"></i><b>5.5</b> Splitting the data</a><ul>
<li class="chapter" data-level="5.5.1" data-path="splitting-the-data.html"><a href="splitting-the-data.html#holdout-test-dataset"><i class="fa fa-check"></i><b>5.5.1</b> Holdout test dataset</a></li>
<li class="chapter" data-level="5.5.2" data-path="splitting-the-data.html"><a href="splitting-the-data.html#cross-validation"><i class="fa fa-check"></i><b>5.5.2</b> Cross-validation</a></li>
<li class="chapter" data-level="5.5.3" data-path="splitting-the-data.html"><a href="splitting-the-data.html#bootstrap-resampling"><i class="fa fa-check"></i><b>5.5.3</b> Bootstrap resampling</a></li>
</ul></li>
<li class="chapter" data-level="5.6" data-path="predicting-the-subtype-with-k-nearest-neighbors.html"><a href="predicting-the-subtype-with-k-nearest-neighbors.html"><i class="fa fa-check"></i><b>5.6</b> Predicting the subtype with k-nearest neighbors</a></li>
<li class="chapter" data-level="5.7" data-path="assessing-the-performance-of-our-model.html"><a href="assessing-the-performance-of-our-model.html"><i class="fa fa-check"></i><b>5.7</b> Assessing the performance of our model</a><ul>
<li class="chapter" data-level="5.7.1" data-path="assessing-the-performance-of-our-model.html"><a href="assessing-the-performance-of-our-model.html#receiver-operating-characteristic-roc-curves"><i class="fa fa-check"></i><b>5.7.1</b> Receiver Operating Characteristic (ROC) curves</a></li>
</ul></li>
<li class="chapter" data-level="5.8" data-path="model-tuning-and-avoiding-overfitting.html"><a href="model-tuning-and-avoiding-overfitting.html"><i class="fa fa-check"></i><b>5.8</b> Model tuning and avoiding overfitting</a><ul>
<li class="chapter" data-level="5.8.1" data-path="model-tuning-and-avoiding-overfitting.html"><a href="model-tuning-and-avoiding-overfitting.html#model-complexity-and-bias-variance-trade-off"><i class="fa fa-check"></i><b>5.8.1</b> Model complexity and bias variance trade-off</a></li>
<li class="chapter" data-level="5.8.2" data-path="model-tuning-and-avoiding-overfitting.html"><a href="model-tuning-and-avoiding-overfitting.html#data-split-strategies-for-model-tuning-and-testing"><i class="fa fa-check"></i><b>5.8.2</b> Data split strategies for model tuning and testing</a></li>
</ul></li>
<li class="chapter" data-level="5.9" data-path="variable-importance.html"><a href="variable-importance.html"><i class="fa fa-check"></i><b>5.9</b> Variable importance</a></li>
<li class="chapter" data-level="5.10" data-path="how-to-deal-with-class-imbalance.html"><a href="how-to-deal-with-class-imbalance.html"><i class="fa fa-check"></i><b>5.10</b> How to deal with class imbalance</a><ul>
<li class="chapter" data-level="5.10.1" data-path="how-to-deal-with-class-imbalance.html"><a href="how-to-deal-with-class-imbalance.html#sampling-for-class-balance"><i class="fa fa-check"></i><b>5.10.1</b> Sampling for class balance</a></li>
<li class="chapter" data-level="5.10.2" data-path="how-to-deal-with-class-imbalance.html"><a href="how-to-deal-with-class-imbalance.html#altering-case-weights"><i class="fa fa-check"></i><b>5.10.2</b> Altering case weights</a></li>
<li class="chapter" data-level="5.10.3" data-path="how-to-deal-with-class-imbalance.html"><a href="how-to-deal-with-class-imbalance.html#selecting-different-classification-score-cutoffs"><i class="fa fa-check"></i><b>5.10.3</b> Selecting different classification score cutoffs</a></li>
</ul></li>
<li class="chapter" data-level="5.11" data-path="dealing-with-correlated-predictors.html"><a href="dealing-with-correlated-predictors.html"><i class="fa fa-check"></i><b>5.11</b> Dealing with correlated predictors</a></li>
<li class="chapter" data-level="5.12" data-path="trees-and-forests-random-forests-in-action.html"><a href="trees-and-forests-random-forests-in-action.html"><i class="fa fa-check"></i><b>5.12</b> Trees and forests: Random forests in action</a><ul>
<li class="chapter" data-level="5.12.1" data-path="trees-and-forests-random-forests-in-action.html"><a href="trees-and-forests-random-forests-in-action.html#decision-trees"><i class="fa fa-check"></i><b>5.12.1</b> Decision trees</a></li>
<li class="chapter" data-level="5.12.2" data-path="trees-and-forests-random-forests-in-action.html"><a href="trees-and-forests-random-forests-in-action.html#trees-to-forests"><i class="fa fa-check"></i><b>5.12.2</b> Trees to forests</a></li>
<li class="chapter" data-level="5.12.3" data-path="trees-and-forests-random-forests-in-action.html"><a href="trees-and-forests-random-forests-in-action.html#variable-importance-1"><i class="fa fa-check"></i><b>5.12.3</b> Variable importance</a></li>
</ul></li>
<li class="chapter" data-level="5.13" data-path="logistic-regression-and-regularization.html"><a href="logistic-regression-and-regularization.html"><i class="fa fa-check"></i><b>5.13</b> Logistic regression and regularization</a><ul>
<li class="chapter" data-level="5.13.1" data-path="logistic-regression-and-regularization.html"><a href="logistic-regression-and-regularization.html#regularization-in-order-to-avoid-overfitting"><i class="fa fa-check"></i><b>5.13.1</b> Regularization in order to avoid overfitting</a></li>
<li class="chapter" data-level="5.13.2" data-path="logistic-regression-and-regularization.html"><a href="logistic-regression-and-regularization.html#variable-importance-2"><i class="fa fa-check"></i><b>5.13.2</b> Variable importance</a></li>
</ul></li>
<li class="chapter" data-level="5.14" data-path="other-supervised-algorithms.html"><a href="other-supervised-algorithms.html"><i class="fa fa-check"></i><b>5.14</b> Other supervised algorithms</a><ul>
<li class="chapter" data-level="5.14.1" data-path="other-supervised-algorithms.html"><a href="other-supervised-algorithms.html#gradient-boosting"><i class="fa fa-check"></i><b>5.14.1</b> Gradient boosting</a></li>
<li class="chapter" data-level="5.14.2" data-path="other-supervised-algorithms.html"><a href="other-supervised-algorithms.html#support-vector-machines-svm"><i class="fa fa-check"></i><b>5.14.2</b> Support Vector Machines (SVM)</a></li>
<li class="chapter" data-level="5.14.3" data-path="other-supervised-algorithms.html"><a href="other-supervised-algorithms.html#neural-networks-and-deep-versions-of-it"><i class="fa fa-check"></i><b>5.14.3</b> Neural networks and deep versions of it</a></li>
<li class="chapter" data-level="5.14.4" data-path="other-supervised-algorithms.html"><a href="other-supervised-algorithms.html#ensemble-learning"><i class="fa fa-check"></i><b>5.14.4</b> Ensemble learning</a></li>
</ul></li>
<li class="chapter" data-level="5.15" data-path="predicting-continuous-variables-regression-with-machine-learning.html"><a href="predicting-continuous-variables-regression-with-machine-learning.html"><i class="fa fa-check"></i><b>5.15</b> Predicting continuous variables: Regression with machine learning</a><ul>
<li class="chapter" data-level="5.15.1" data-path="predicting-continuous-variables-regression-with-machine-learning.html"><a href="predicting-continuous-variables-regression-with-machine-learning.html#use-case-predicting-age-from-dna-methylation"><i class="fa fa-check"></i><b>5.15.1</b> Use case: Predicting age from DNA methylation</a></li>
<li class="chapter" data-level="5.15.2" data-path="predicting-continuous-variables-regression-with-machine-learning.html"><a href="predicting-continuous-variables-regression-with-machine-learning.html#reading-and-processing-the-data"><i class="fa fa-check"></i><b>5.15.2</b> Reading and processing the data</a></li>
<li class="chapter" data-level="5.15.3" data-path="predicting-continuous-variables-regression-with-machine-learning.html"><a href="predicting-continuous-variables-regression-with-machine-learning.html#running-random-forest-regression"><i class="fa fa-check"></i><b>5.15.3</b> Running random forest regression</a></li>
</ul></li>
<li class="chapter" data-level="5.16" data-path="exercises-3.html"><a href="exercises-3.html"><i class="fa fa-check"></i><b>5.16</b> Exercises</a><ul>
<li class="chapter" data-level="5.16.1" data-path="exercises-3.html"><a href="exercises-3.html#classification"><i class="fa fa-check"></i><b>5.16.1</b> Classification</a></li>
<li class="chapter" data-level="5.16.2" data-path="exercises-3.html"><a href="exercises-3.html#regression"><i class="fa fa-check"></i><b>5.16.2</b> Regression</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="6" data-path="genomicIntervals.html"><a href="genomicIntervals.html"><i class="fa fa-check"></i><b>6</b> Operations on Genomic Intervals and Genome Arithmetic</a><ul>
<li class="chapter" data-level="6.1" data-path="operations-on-genomic-intervals-with-genomicranges-package.html"><a href="operations-on-genomic-intervals-with-genomicranges-package.html"><i class="fa fa-check"></i><b>6.1</b> Operations on genomic intervals with <code>GenomicRanges</code> package</a><ul>
<li class="chapter" data-level="6.1.1" data-path="operations-on-genomic-intervals-with-genomicranges-package.html"><a href="operations-on-genomic-intervals-with-genomicranges-package.html#how-to-create-and-manipulate-a-granges-object"><i class="fa fa-check"></i><b>6.1.1</b> How to create and manipulate a GRanges object</a></li>
<li class="chapter" data-level="6.1.2" data-path="operations-on-genomic-intervals-with-genomicranges-package.html"><a href="operations-on-genomic-intervals-with-genomicranges-package.html#getting-genomic-regions-into-r-as-granges-objects"><i class="fa fa-check"></i><b>6.1.2</b> Getting genomic regions into R as GRanges objects</a></li>
<li class="chapter" data-level="6.1.3" data-path="operations-on-genomic-intervals-with-genomicranges-package.html"><a href="operations-on-genomic-intervals-with-genomicranges-package.html#finding-regions-that-dodo-not-overlap-with-another-set-of-regions"><i class="fa fa-check"></i><b>6.1.3</b> Finding regions that do/do not overlap with another set of regions</a></li>
</ul></li>
<li class="chapter" data-level="6.2" data-path="dealing-with-mapped-high-throughput-sequencing-reads.html"><a href="dealing-with-mapped-high-throughput-sequencing-reads.html"><i class="fa fa-check"></i><b>6.2</b> Dealing with mapped high-throughput sequencing reads</a><ul>
<li class="chapter" data-level="6.2.1" data-path="dealing-with-mapped-high-throughput-sequencing-reads.html"><a href="dealing-with-mapped-high-throughput-sequencing-reads.html#counting-mapped-reads-for-a-set-of-regions"><i class="fa fa-check"></i><b>6.2.1</b> Counting mapped reads for a set of regions</a></li>
</ul></li>
<li class="chapter" data-level="6.3" data-path="dealing-with-continuous-scores-over-the-genome.html"><a href="dealing-with-continuous-scores-over-the-genome.html"><i class="fa fa-check"></i><b>6.3</b> Dealing with continuous scores over the genome</a><ul>
<li class="chapter" data-level="6.3.1" data-path="dealing-with-continuous-scores-over-the-genome.html"><a href="dealing-with-continuous-scores-over-the-genome.html#extracting-subsections-of-rle-and-rlelist-objects"><i class="fa fa-check"></i><b>6.3.1</b> Extracting subsections of Rle and RleList objects</a></li>
</ul></li>
<li class="chapter" data-level="6.4" data-path="genomic-intervals-with-more-information-summarizedexperiment-class.html"><a href="genomic-intervals-with-more-information-summarizedexperiment-class.html"><i class="fa fa-check"></i><b>6.4</b> Genomic intervals with more information: SummarizedExperiment class</a><ul>
<li class="chapter" data-level="6.4.1" data-path="genomic-intervals-with-more-information-summarizedexperiment-class.html"><a href="genomic-intervals-with-more-information-summarizedexperiment-class.html#create-a-summarizedexperiment-object"><i class="fa fa-check"></i><b>6.4.1</b> Create a SummarizedExperiment object</a></li>
<li class="chapter" data-level="6.4.2" data-path="genomic-intervals-with-more-information-summarizedexperiment-class.html"><a href="genomic-intervals-with-more-information-summarizedexperiment-class.html#subset-and-manipulate-the-summarizedexperiment-object"><i class="fa fa-check"></i><b>6.4.2</b> Subset and manipulate the SummarizedExperiment object</a></li>
</ul></li>
<li class="chapter" data-level="6.5" data-path="visualizing-and-summarizing-genomic-intervals.html"><a href="visualizing-and-summarizing-genomic-intervals.html"><i class="fa fa-check"></i><b>6.5</b> Visualizing and summarizing genomic intervals</a><ul>
<li class="chapter" data-level="6.5.1" data-path="visualizing-and-summarizing-genomic-intervals.html"><a href="visualizing-and-summarizing-genomic-intervals.html#visualizing-intervals-on-a-locus-of-interest"><i class="fa fa-check"></i><b>6.5.1</b> Visualizing intervals on a locus of interest</a></li>
<li class="chapter" data-level="6.5.2" data-path="visualizing-and-summarizing-genomic-intervals.html"><a href="visualizing-and-summarizing-genomic-intervals.html#summaries-of-genomic-intervals-on-multiple-loci"><i class="fa fa-check"></i><b>6.5.2</b> Summaries of genomic intervals on multiple loci</a></li>
<li class="chapter" data-level="6.5.3" data-path="visualizing-and-summarizing-genomic-intervals.html"><a href="visualizing-and-summarizing-genomic-intervals.html#making-karyograms-and-circos-plots"><i class="fa fa-check"></i><b>6.5.3</b> Making karyograms and circos plots</a></li>
</ul></li>
<li class="chapter" data-level="6.6" data-path="exercises-4.html"><a href="exercises-4.html"><i class="fa fa-check"></i><b>6.6</b> Exercises</a><ul>
<li class="chapter" data-level="6.6.1" data-path="exercises-4.html"><a href="exercises-4.html#operations-on-genomic-intervals-with-the-genomicranges-package"><i class="fa fa-check"></i><b>6.6.1</b> Operations on genomic intervals with the <code>GenomicRanges</code> package</a></li>
<li class="chapter" data-level="6.6.2" data-path="exercises-4.html"><a href="exercises-4.html#dealing-with-mapped-high-throughput-sequencing-reads-1"><i class="fa fa-check"></i><b>6.6.2</b> Dealing with mapped high-throughput sequencing reads</a></li>
<li class="chapter" data-level="6.6.3" data-path="exercises-4.html"><a href="exercises-4.html#dealing-with-contiguous-scores-over-the-genome"><i class="fa fa-check"></i><b>6.6.3</b> Dealing with contiguous scores over the genome</a></li>
<li class="chapter" data-level="6.6.4" data-path="exercises-4.html"><a href="exercises-4.html#visualizing-and-summarizing-genomic-intervals-1"><i class="fa fa-check"></i><b>6.6.4</b> Visualizing and summarizing genomic intervals</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="7" data-path="processingReads.html"><a href="processingReads.html"><i class="fa fa-check"></i><b>7</b> Quality Check, Processing and Alignment of High-throughput Sequencing Reads</a><ul>
<li class="chapter" data-level="7.1" data-path="fasta-and-fastq-formats.html"><a href="fasta-and-fastq-formats.html"><i class="fa fa-check"></i><b>7.1</b> FASTA and FASTQ formats</a></li>
<li class="chapter" data-level="7.2" data-path="quality-check-on-sequencing-reads.html"><a href="quality-check-on-sequencing-reads.html"><i class="fa fa-check"></i><b>7.2</b> Quality check on sequencing reads</a><ul>
<li class="chapter" data-level="7.2.1" data-path="quality-check-on-sequencing-reads.html"><a href="quality-check-on-sequencing-reads.html#sequence-quality-per-basecycle"><i class="fa fa-check"></i><b>7.2.1</b> Sequence quality per base/cycle</a></li>
<li class="chapter" data-level="7.2.2" data-path="quality-check-on-sequencing-reads.html"><a href="quality-check-on-sequencing-reads.html#sequence-content-per-basecycle"><i class="fa fa-check"></i><b>7.2.2</b> Sequence content per base/cycle</a></li>
<li class="chapter" data-level="7.2.3" data-path="quality-check-on-sequencing-reads.html"><a href="quality-check-on-sequencing-reads.html#read-frequency-plot"><i class="fa fa-check"></i><b>7.2.3</b> Read frequency plot</a></li>
<li class="chapter" data-level="7.2.4" data-path="quality-check-on-sequencing-reads.html"><a href="quality-check-on-sequencing-reads.html#other-quality-metrics-and-qc-tools"><i class="fa fa-check"></i><b>7.2.4</b> Other quality metrics and QC tools</a></li>
</ul></li>
<li class="chapter" data-level="7.3" data-path="filtering-and-trimming-reads.html"><a href="filtering-and-trimming-reads.html"><i class="fa fa-check"></i><b>7.3</b> Filtering and trimming reads</a></li>
<li class="chapter" data-level="7.4" data-path="mappingaligning-reads-to-the-genome.html"><a href="mappingaligning-reads-to-the-genome.html"><i class="fa fa-check"></i><b>7.4</b> Mapping/aligning reads to the genome</a></li>
<li class="chapter" data-level="7.5" data-path="further-processing-of-aligned-reads.html"><a href="further-processing-of-aligned-reads.html"><i class="fa fa-check"></i><b>7.5</b> Further processing of aligned reads</a></li>
<li class="chapter" data-level="7.6" data-path="exercises-5.html"><a href="exercises-5.html"><i class="fa fa-check"></i><b>7.6</b> Exercises</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="rnaseqanalysis.html"><a href="rnaseqanalysis.html"><i class="fa fa-check"></i><b>8</b> RNA-seq Analysis</a><ul>
<li class="chapter" data-level="8.1" data-path="what-is-gene-expression.html"><a href="what-is-gene-expression.html"><i class="fa fa-check"></i><b>8.1</b> What is gene expression?</a></li>
<li class="chapter" data-level="8.2" data-path="methods-to-detect-gene-expression.html"><a href="methods-to-detect-gene-expression.html"><i class="fa fa-check"></i><b>8.2</b> Methods to detect gene expression</a></li>
<li class="chapter" data-level="8.3" data-path="gene-expression-analysis-using-high-throughput-sequencing-technologies.html"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html"><i class="fa fa-check"></i><b>8.3</b> Gene expression analysis using high-throughput sequencing technologies</a><ul>
<li class="chapter" data-level="8.3.1" data-path="gene-expression-analysis-using-high-throughput-sequencing-technologies.html"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#processing-raw-data"><i class="fa fa-check"></i><b>8.3.1</b> Processing raw data</a></li>
<li class="chapter" data-level="8.3.2" data-path="gene-expression-analysis-using-high-throughput-sequencing-technologies.html"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#alignment"><i class="fa fa-check"></i><b>8.3.2</b> Alignment</a></li>
<li class="chapter" data-level="8.3.3" data-path="gene-expression-analysis-using-high-throughput-sequencing-technologies.html"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#quantification"><i class="fa fa-check"></i><b>8.3.3</b> Quantification</a></li>
<li class="chapter" data-level="8.3.4" data-path="gene-expression-analysis-using-high-throughput-sequencing-technologies.html"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#within-sample-normalization-of-the-read-counts"><i class="fa fa-check"></i><b>8.3.4</b> Within sample normalization of the read counts</a></li>
<li class="chapter" data-level="8.3.5" data-path="gene-expression-analysis-using-high-throughput-sequencing-technologies.html"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#computing-different-normalization-schemes-in-r"><i class="fa fa-check"></i><b>8.3.5</b> Computing different normalization schemes in R</a></li>
<li class="chapter" data-level="8.3.6" data-path="gene-expression-analysis-using-high-throughput-sequencing-technologies.html"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#exploratory-analysis-of-the-read-count-table"><i class="fa fa-check"></i><b>8.3.6</b> Exploratory analysis of the read count table</a></li>
<li class="chapter" data-level="8.3.7" data-path="gene-expression-analysis-using-high-throughput-sequencing-technologies.html"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#differential-expression-analysis"><i class="fa fa-check"></i><b>8.3.7</b> Differential expression analysis</a></li>
<li class="chapter" data-level="8.3.8" data-path="gene-expression-analysis-using-high-throughput-sequencing-technologies.html"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#functional-enrichment-analysis"><i class="fa fa-check"></i><b>8.3.8</b> Functional enrichment analysis</a></li>
<li class="chapter" data-level="8.3.9" data-path="gene-expression-analysis-using-high-throughput-sequencing-technologies.html"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#accounting-for-additional-sources-of-variation"><i class="fa fa-check"></i><b>8.3.9</b> Accounting for additional sources of variation</a></li>
</ul></li>
<li class="chapter" data-level="8.4" data-path="other-applications-of-rna-seq.html"><a href="other-applications-of-rna-seq.html"><i class="fa fa-check"></i><b>8.4</b> Other applications of RNA-seq</a></li>
<li class="chapter" data-level="8.5" data-path="exercises-6.html"><a href="exercises-6.html"><i class="fa fa-check"></i><b>8.5</b> Exercises</a><ul>
<li class="chapter" data-level="8.5.1" data-path="exercises-6.html"><a href="exercises-6.html#exploring-the-count-tables"><i class="fa fa-check"></i><b>8.5.1</b> Exploring the count tables</a></li>
<li class="chapter" data-level="8.5.2" data-path="exercises-6.html"><a href="exercises-6.html#differential-expression-analysis-1"><i class="fa fa-check"></i><b>8.5.2</b> Differential expression analysis</a></li>
<li class="chapter" data-level="8.5.3" data-path="exercises-6.html"><a href="exercises-6.html#functional-enrichment-analysis-1"><i class="fa fa-check"></i><b>8.5.3</b> Functional enrichment analysis</a></li>
<li class="chapter" data-level="8.5.4" data-path="exercises-6.html"><a href="exercises-6.html#removing-unwanted-variation-from-the-expression-data"><i class="fa fa-check"></i><b>8.5.4</b> Removing unwanted variation from the expression data</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="9" data-path="chipseq.html"><a href="chipseq.html"><i class="fa fa-check"></i><b>9</b> ChIP-seq analysis</a><ul>
<li class="chapter" data-level="9.1" data-path="regulatory-protein-dna-interactions.html"><a href="regulatory-protein-dna-interactions.html"><i class="fa fa-check"></i><b>9.1</b> Regulatory protein-DNA interactions</a></li>
<li class="chapter" data-level="9.2" data-path="measuring-protein-dna-interactions-with-chip-seq.html"><a href="measuring-protein-dna-interactions-with-chip-seq.html"><i class="fa fa-check"></i><b>9.2</b> Measuring protein-DNA interactions with ChIP-seq</a></li>
<li class="chapter" data-level="9.3" data-path="factors-that-affect-chip-seq-experiment-and-analysis-quality.html"><a href="factors-that-affect-chip-seq-experiment-and-analysis-quality.html"><i class="fa fa-check"></i><b>9.3</b> Factors that affect ChIP-seq experiment and analysis quality</a><ul>
<li class="chapter" data-level="9.3.1" data-path="factors-that-affect-chip-seq-experiment-and-analysis-quality.html"><a href="factors-that-affect-chip-seq-experiment-and-analysis-quality.html#antibody-specificity"><i class="fa fa-check"></i><b>9.3.1</b> Antibody specificity</a></li>
<li class="chapter" data-level="9.3.2" data-path="factors-that-affect-chip-seq-experiment-and-analysis-quality.html"><a href="factors-that-affect-chip-seq-experiment-and-analysis-quality.html#sequencing-depth"><i class="fa fa-check"></i><b>9.3.2</b> Sequencing depth</a></li>
<li class="chapter" data-level="9.3.3" data-path="factors-that-affect-chip-seq-experiment-and-analysis-quality.html"><a href="factors-that-affect-chip-seq-experiment-and-analysis-quality.html#pcr-duplication"><i class="fa fa-check"></i><b>9.3.3</b> PCR duplication</a></li>
<li class="chapter" data-level="9.3.4" data-path="factors-that-affect-chip-seq-experiment-and-analysis-quality.html"><a href="factors-that-affect-chip-seq-experiment-and-analysis-quality.html#biological-replicates"><i class="fa fa-check"></i><b>9.3.4</b> Biological replicates</a></li>
<li class="chapter" data-level="9.3.5" data-path="factors-that-affect-chip-seq-experiment-and-analysis-quality.html"><a href="factors-that-affect-chip-seq-experiment-and-analysis-quality.html#control-experiments"><i class="fa fa-check"></i><b>9.3.5</b> Control experiments</a></li>
<li class="chapter" data-level="9.3.6" data-path="factors-that-affect-chip-seq-experiment-and-analysis-quality.html"><a href="factors-that-affect-chip-seq-experiment-and-analysis-quality.html#using-tagged-proteins"><i class="fa fa-check"></i><b>9.3.6</b> Using tagged proteins</a></li>
</ul></li>
<li class="chapter" data-level="9.4" data-path="pre-processing-chip-data.html"><a href="pre-processing-chip-data.html"><i class="fa fa-check"></i><b>9.4</b> Pre-processing ChIP data</a><ul>
<li class="chapter" data-level="9.4.1" data-path="pre-processing-chip-data.html"><a href="pre-processing-chip-data.html#mapping-of-chip-seq-data"><i class="fa fa-check"></i><b>9.4.1</b> Mapping of ChIP-seq data</a></li>
</ul></li>
<li class="chapter" data-level="9.5" data-path="chip-quality-control.html"><a href="chip-quality-control.html"><i class="fa fa-check"></i><b>9.5</b> ChIP quality control</a><ul>
<li class="chapter" data-level="9.5.1" data-path="chip-quality-control.html"><a href="chip-quality-control.html#the-data"><i class="fa fa-check"></i><b>9.5.1</b> The data</a></li>
<li class="chapter" data-level="9.5.2" data-path="chip-quality-control.html"><a href="chip-quality-control.html#sample-clustering"><i class="fa fa-check"></i><b>9.5.2</b> Sample clustering</a></li>
<li class="chapter" data-level="9.5.3" data-path="chip-quality-control.html"><a href="chip-quality-control.html#visualization-in-the-genome-browser"><i class="fa fa-check"></i><b>9.5.3</b> Visualization in the genome browser</a></li>
<li class="chapter" data-level="9.5.4" data-path="chip-quality-control.html"><a href="chip-quality-control.html#plus-and-minus-strand-cross-correlation"><i class="fa fa-check"></i><b>9.5.4</b> Plus and minus strand cross-correlation</a></li>
<li class="chapter" data-level="9.5.5" data-path="chip-quality-control.html"><a href="chip-quality-control.html#gc-bias-quantification"><i class="fa fa-check"></i><b>9.5.5</b> GC bias quantification</a></li>
<li class="chapter" data-level="9.5.6" data-path="chip-quality-control.html"><a href="chip-quality-control.html#sequence-read-genomic-distribution"><i class="fa fa-check"></i><b>9.5.6</b> Sequence read genomic distribution</a></li>
</ul></li>
<li class="chapter" data-level="9.6" data-path="peak-calling.html"><a href="peak-calling.html"><i class="fa fa-check"></i><b>9.6</b> Peak calling</a><ul>
<li class="chapter" data-level="9.6.1" data-path="peak-calling.html"><a href="peak-calling.html#types-of-chip-seq-experiments"><i class="fa fa-check"></i><b>9.6.1</b> Types of ChIP-seq experiments</a></li>
<li class="chapter" data-level="9.6.2" data-path="peak-calling.html"><a href="peak-calling.html#peak-calling-sharp-peaks"><i class="fa fa-check"></i><b>9.6.2</b> Peak calling: Sharp peaks</a></li>
<li class="chapter" data-level="9.6.3" data-path="peak-calling.html"><a href="peak-calling.html#peak-calling-broad-regions"><i class="fa fa-check"></i><b>9.6.3</b> Peak calling: Broad regions</a></li>
<li class="chapter" data-level="9.6.4" data-path="peak-calling.html"><a href="peak-calling.html#peak-quality-control"><i class="fa fa-check"></i><b>9.6.4</b> Peak quality control</a></li>
<li class="chapter" data-level="9.6.5" data-path="peak-calling.html"><a href="peak-calling.html#peak-annotation"><i class="fa fa-check"></i><b>9.6.5</b> Peak annotation</a></li>
</ul></li>
<li class="chapter" data-level="9.7" data-path="motif-discovery.html"><a href="motif-discovery.html"><i class="fa fa-check"></i><b>9.7</b> Motif discovery</a><ul>
<li class="chapter" data-level="9.7.1" data-path="motif-discovery.html"><a href="motif-discovery.html#motif-comparison"><i class="fa fa-check"></i><b>9.7.1</b> Motif comparison</a></li>
</ul></li>
<li class="chapter" data-level="9.8" data-path="what-to-do-next.html"><a href="what-to-do-next.html"><i class="fa fa-check"></i><b>9.8</b> What to do next?</a></li>
<li class="chapter" data-level="9.9" data-path="exercises-7.html"><a href="exercises-7.html"><i class="fa fa-check"></i><b>9.9</b> Exercises</a><ul>
<li class="chapter" data-level="9.9.1" data-path="exercises-7.html"><a href="exercises-7.html#quality-control"><i class="fa fa-check"></i><b>9.9.1</b> Quality control</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="10" data-path="bsseq.html"><a href="bsseq.html"><i class="fa fa-check"></i><b>10</b> DNA methylation analysis using bisulfite sequencing data</a><ul>
<li class="chapter" data-level="10.1" data-path="what-is-dna-methylation.html"><a href="what-is-dna-methylation.html"><i class="fa fa-check"></i><b>10.1</b> What is DNA methylation?</a><ul>
<li class="chapter" data-level="10.1.1" data-path="what-is-dna-methylation.html"><a href="what-is-dna-methylation.html#how-dna-methylation-is-set"><i class="fa fa-check"></i><b>10.1.1</b> How DNA methylation is set ?</a></li>
<li class="chapter" data-level="10.1.2" data-path="what-is-dna-methylation.html"><a href="what-is-dna-methylation.html#how-to-measure-dna-methylation-with-bisulfite-sequencing"><i class="fa fa-check"></i><b>10.1.2</b> How to measure DNA methylation with bisulfite sequencing</a></li>
</ul></li>
<li class="chapter" data-level="10.2" data-path="analyzing-dna-methylation-data.html"><a href="analyzing-dna-methylation-data.html"><i class="fa fa-check"></i><b>10.2</b> Analyzing DNA methylation data</a></li>
<li class="chapter" data-level="10.3" data-path="processing-raw-data-and-getting-data-into-r.html"><a href="processing-raw-data-and-getting-data-into-r.html"><i class="fa fa-check"></i><b>10.3</b> Processing raw data and getting data into R</a></li>
<li class="chapter" data-level="10.4" data-path="data-filtering-and-exploratory-analysis.html"><a href="data-filtering-and-exploratory-analysis.html"><i class="fa fa-check"></i><b>10.4</b> Data filtering and exploratory analysis</a><ul>
<li class="chapter" data-level="10.4.1" data-path="data-filtering-and-exploratory-analysis.html"><a href="data-filtering-and-exploratory-analysis.html#reading-methylation-call-files"><i class="fa fa-check"></i><b>10.4.1</b> Reading methylation call files</a></li>
<li class="chapter" data-level="10.4.2" data-path="data-filtering-and-exploratory-analysis.html"><a href="data-filtering-and-exploratory-analysis.html#further-quality-check"><i class="fa fa-check"></i><b>10.4.2</b> Further quality check</a></li>
<li class="chapter" data-level="10.4.3" data-path="data-filtering-and-exploratory-analysis.html"><a href="data-filtering-and-exploratory-analysis.html#merging-samples-into-a-single-table"><i class="fa fa-check"></i><b>10.4.3</b> Merging samples into a single table</a></li>
<li class="chapter" data-level="10.4.4" data-path="data-filtering-and-exploratory-analysis.html"><a href="data-filtering-and-exploratory-analysis.html#filtering-cpgs"><i class="fa fa-check"></i><b>10.4.4</b> Filtering CpGs</a></li>
<li class="chapter" data-level="10.4.5" data-path="data-filtering-and-exploratory-analysis.html"><a href="data-filtering-and-exploratory-analysis.html#clustering-samples"><i class="fa fa-check"></i><b>10.4.5</b> Clustering samples</a></li>
<li class="chapter" data-level="10.4.6" data-path="data-filtering-and-exploratory-analysis.html"><a href="data-filtering-and-exploratory-analysis.html#principal-component-analysis-1"><i class="fa fa-check"></i><b>10.4.6</b> Principal component analysis</a></li>
</ul></li>
<li class="chapter" data-level="10.5" data-path="extracting-interesting-regions-differential-methylation-and-segmentation.html"><a href="extracting-interesting-regions-differential-methylation-and-segmentation.html"><i class="fa fa-check"></i><b>10.5</b> Extracting interesting regions: Differential methylation and segmentation</a><ul>
<li class="chapter" data-level="10.5.1" data-path="extracting-interesting-regions-differential-methylation-and-segmentation.html"><a href="extracting-interesting-regions-differential-methylation-and-segmentation.html#differential-methylation"><i class="fa fa-check"></i><b>10.5.1</b> Differential methylation</a></li>
<li class="chapter" data-level="10.5.2" data-path="extracting-interesting-regions-differential-methylation-and-segmentation.html"><a href="extracting-interesting-regions-differential-methylation-and-segmentation.html#methylation-segmentation"><i class="fa fa-check"></i><b>10.5.2</b> Methylation segmentation</a></li>
<li class="chapter" data-level="10.5.3" data-path="extracting-interesting-regions-differential-methylation-and-segmentation.html"><a href="extracting-interesting-regions-differential-methylation-and-segmentation.html#working-with-large-files"><i class="fa fa-check"></i><b>10.5.3</b> Working with large files</a></li>
</ul></li>
<li class="chapter" data-level="10.6" data-path="annotation-of-dmrsdmcs-and-segments.html"><a href="annotation-of-dmrsdmcs-and-segments.html"><i class="fa fa-check"></i><b>10.6</b> Annotation of DMRs/DMCs and segments</a><ul>
<li class="chapter" data-level="10.6.1" data-path="annotation-of-dmrsdmcs-and-segments.html"><a href="annotation-of-dmrsdmcs-and-segments.html#further-annotation-with-genes-or-gene-sets"><i class="fa fa-check"></i><b>10.6.1</b> Further annotation with genes or gene sets</a></li>
</ul></li>
<li class="chapter" data-level="10.7" data-path="other-r-packages-that-can-be-used-for-methylation-analysis.html"><a href="other-r-packages-that-can-be-used-for-methylation-analysis.html"><i class="fa fa-check"></i><b>10.7</b> Other R packages that can be used for methylation analysis</a></li>
<li class="chapter" data-level="10.8" data-path="exercises-8.html"><a href="exercises-8.html"><i class="fa fa-check"></i><b>10.8</b> Exercises</a><ul>
<li class="chapter" data-level="10.8.1" data-path="exercises-8.html"><a href="exercises-8.html#differential-methylation-1"><i class="fa fa-check"></i><b>10.8.1</b> Differential methylation</a></li>
<li class="chapter" data-level="10.8.2" data-path="exercises-8.html"><a href="exercises-8.html#methylome-segmentation"><i class="fa fa-check"></i><b>10.8.2</b> Methylome segmentation</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="11" data-path="multiomics.html"><a href="multiomics.html"><i class="fa fa-check"></i><b>11</b> Multi-omics Analysis</a><ul>
<li class="chapter" data-level="11.1" data-path="use-case-multi-omics-data-from-colorectal-cancer.html"><a href="use-case-multi-omics-data-from-colorectal-cancer.html"><i class="fa fa-check"></i><b>11.1</b> Use case: Multi-omics data from colorectal cancer</a></li>
<li class="chapter" data-level="11.2" data-path="latent-variable-models-for-multi-omics-integration.html"><a href="latent-variable-models-for-multi-omics-integration.html"><i class="fa fa-check"></i><b>11.2</b> Latent variable models for multi-omics integration</a></li>
<li class="chapter" data-level="11.3" data-path="matrix-factorization-methods-for-unsupervised-multi-omics-data-integration.html"><a href="matrix-factorization-methods-for-unsupervised-multi-omics-data-integration.html"><i class="fa fa-check"></i><b>11.3</b> Matrix factorization methods for unsupervised multi-omics data integration</a><ul>
<li class="chapter" data-level="11.3.1" data-path="matrix-factorization-methods-for-unsupervised-multi-omics-data-integration.html"><a href="matrix-factorization-methods-for-unsupervised-multi-omics-data-integration.html#multiple-factor-analysis"><i class="fa fa-check"></i><b>11.3.1</b> Multiple factor analysis</a></li>
<li class="chapter" data-level="11.3.2" data-path="matrix-factorization-methods-for-unsupervised-multi-omics-data-integration.html"><a href="matrix-factorization-methods-for-unsupervised-multi-omics-data-integration.html#joint-non-negative-matrix-factorization"><i class="fa fa-check"></i><b>11.3.2</b> Joint non-negative matrix factorization</a></li>
<li class="chapter" data-level="11.3.3" data-path="matrix-factorization-methods-for-unsupervised-multi-omics-data-integration.html"><a href="matrix-factorization-methods-for-unsupervised-multi-omics-data-integration.html#icluster"><i class="fa fa-check"></i><b>11.3.3</b> iCluster</a></li>
</ul></li>
<li class="chapter" data-level="11.4" data-path="clustering-using-latent-factors.html"><a href="clustering-using-latent-factors.html"><i class="fa fa-check"></i><b>11.4</b> Clustering using latent factors</a><ul>
<li class="chapter" data-level="11.4.1" data-path="clustering-using-latent-factors.html"><a href="clustering-using-latent-factors.html#one-hot-clustering"><i class="fa fa-check"></i><b>11.4.1</b> One-hot clustering</a></li>
<li class="chapter" data-level="11.4.2" data-path="clustering-using-latent-factors.html"><a href="clustering-using-latent-factors.html#k-means-clustering-1"><i class="fa fa-check"></i><b>11.4.2</b> K-means clustering</a></li>
</ul></li>
<li class="chapter" data-level="11.5" data-path="biological-interpretation-of-latent-factors.html"><a href="biological-interpretation-of-latent-factors.html"><i class="fa fa-check"></i><b>11.5</b> Biological interpretation of latent factors</a><ul>
<li class="chapter" data-level="11.5.1" data-path="biological-interpretation-of-latent-factors.html"><a href="biological-interpretation-of-latent-factors.html#inspection-of-feature-weights-in-loading-vectors"><i class="fa fa-check"></i><b>11.5.1</b> Inspection of feature weights in loading vectors</a></li>
<li class="chapter" data-level="11.5.2" data-path="biological-interpretation-of-latent-factors.html"><a href="biological-interpretation-of-latent-factors.html#making-sense-of-factors-using-enrichment-analysis"><i class="fa fa-check"></i><b>11.5.2</b> Making sense of factors using enrichment analysis</a></li>
<li class="chapter" data-level="11.5.3" data-path="biological-interpretation-of-latent-factors.html"><a href="biological-interpretation-of-latent-factors.html#interpretation-using-additional-covariates"><i class="fa fa-check"></i><b>11.5.3</b> Interpretation using additional covariates</a></li>
</ul></li>
<li class="chapter" data-level="11.6" data-path="exercises-9.html"><a href="exercises-9.html"><i class="fa fa-check"></i><b>11.6</b> Exercises</a><ul>
<li class="chapter" data-level="11.6.1" data-path="exercises-9.html"><a href="exercises-9.html#matrix-factorization-methods"><i class="fa fa-check"></i><b>11.6.1</b> Matrix factorization methods</a></li>
<li class="chapter" data-level="11.6.2" data-path="exercises-9.html"><a href="exercises-9.html#clustering-using-latent-factors-1"><i class="fa fa-check"></i><b>11.6.2</b> Clustering using latent factors</a></li>
<li class="chapter" data-level="11.6.3" data-path="exercises-9.html"><a href="exercises-9.html#biological-interpretation-of-latent-factors-1"><i class="fa fa-check"></i><b>11.6.3</b> Biological interpretation of latent factors</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i>References</a></li>
<li class="divider"></li>
<li><a href="https://bookdown.org" target="_blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Computational Genomics with R</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="matrix-factorization-methods-for-unsupervised-multi-omics-data-integration" class="section level2">
<h2><span class="header-section-number">11.3</span> Matrix factorization methods for unsupervised multi-omics data integration</h2>
<p>Matrix factorization techniques attempt to infer a set of latent variables from the data by finding factors of a data matrix. Principal Component Analysis (introduced in Chapter <a href="unsupervisedLearning.html#unsupervisedLearning">4</a>) is a form of matrix factorization which finds factors based on the covariance structure of the data. Generally, matrix factorization methods may be formulated as</p>
<p><span class="math display">\[
X = WH,
\]</span>
where <span class="math inline">\(X\)</span> is the <em>data matrix</em>, <span class="math inline">\([M \times N]\)</span> where <span class="math inline">\(M\)</span> is the number of features (typically genes), and <span class="math inline">\(N\)</span> is the number of samples. <span class="math inline">\(W\)</span> is an <span class="math inline">\([M \times K]\)</span> <em>factors</em> matrix, and <span class="math inline">\(H\)</span> is the <span class="math inline">\([K \times N]\)</span> <em>latent variable coefficient matrix</em>. Tying this back to PCA, where <span class="math inline">\(X = U \Sigma V^T\)</span>, we may formulate the factorization in the same terms by setting <span class="math inline">\(W=U\Sigma\)</span> and <span class="math inline">\(H=V^T\)</span>. If <span class="math inline">\(K=rank(X)\)</span>, this factorization is lossless, i.e. <span class="math inline">\(X=WH\)</span>. However if we choose <span class="math inline">\(K&lt;rank(X)\)</span>, the factorization is lossy, i.e. <span class="math inline">\(X \approx WH\)</span>. In that case, matrix factorization methods normally opt to minimize the error</p>
<p><span class="math display">\[
min~\|X-WH\|.
\]</span>
As we normally seek a latent variable model with a considerably lower dimensionality than <span class="math inline">\(X\)</span>, this is the more common case.</p>
<p>The loss function we choose to minimize may be further subject to some constraints or regularization terms. Regularization has been introduced in Chapter <a href="supervisedLearning.html#supervisedLearning">5</a>. In the current context of latent factor models, a regularization term might be added to the loss function, i.e. we might choose to minimize <span class="math inline">\(min~\|X-WH\| + \lambda \|W\|^2\)</span> (this is called <span class="math inline">\(L_2\)</span>-regularization) instead of merely the reconstruction error. Adding such a term to our loss function here will push the <span class="math inline">\(W\)</span> matrix entries towards 0, in effect balancing between better reconstruction of the data and a more parsimonious model. A more parsimonious latent factor model is one with more sparsity in the latent factors. This sparsity is desirable for model interpretation, as will become evident in later sections.</p>
<div class="figure" style="text-align: center"><span id="fig:momatrixFactorization"></span>
<img src="images/matrix_factorization.png" alt="General matrix factorization framework. The data matrix on the left-hand side is decomposed into factors on the right-hand side. The equality may be an approximation as some matrix factorization methods are lossless (exact), while others are an approximation." width="75%" />
<p class="caption">
FIGURE 11.4: General matrix factorization framework. The data matrix on the left-hand side is decomposed into factors on the right-hand side. The equality may be an approximation as some matrix factorization methods are lossless (exact), while others are an approximation.
</p>
</div>
<p>In Figure <a href="matrix-factorization-methods-for-unsupervised-multi-omics-data-integration.html#fig:momatrixFactorization">11.4</a>, the <span class="math inline">\(5 \times 4\)</span> data matrix <span class="math inline">\(X\)</span> is decomposed to a 2-dimensional latent variable model.</p>
<div id="multiple-factor-analysis" class="section level3">
<h3><span class="header-section-number">11.3.1</span> Multiple factor analysis</h3>
<p>Multiple factor analysis is a natural starting point for a discussion about matrix factorization methods for integrating multiple data types. It is a straightforward extension of PCA into the domain of multiple data types.<a href="#fn2" class="footnote-ref" id="fnref2"><sup>2</sup></a></p>
<p>Figure <a href="matrix-factorization-methods-for-unsupervised-multi-omics-data-integration.html#fig:moMFA">11.5</a> sketches a naive extension of PCA to a multi-omics context.</p>
<div class="figure" style="text-align: center"><span id="fig:moMFA"></span>
<img src="images/mfa.png" alt="A naive extension of PCA to multi-omics; data matrices from different platforms are stacked, before applying PCA." width="50%" />
<p class="caption">
FIGURE 11.5: A naive extension of PCA to multi-omics; data matrices from different platforms are stacked, before applying PCA.
</p>
</div>
<p>Formally, we have
<span class="math display">\[
X = \begin{bmatrix}
           X_{1} \\
           X_{2} \\
           \vdots \\
           X_{L}
         \end{bmatrix} = WH,
\]</span>
a joint decomposition of the different data matrices (<span class="math inline">\(X_i\)</span>) into the factor matrix <span class="math inline">\(W\)</span> and the latent variable matrix <span class="math inline">\(H\)</span>. This way, we can leverage the ability of PCA to find the highest variance decomposition of the data, when the data consists of different omics types. As a reminder, PCA finds the linear combinations of the features which, when the data is projected onto them, preserve the most variance of any <span class="math inline">\(K\)</span>-dimensional space. But because measurements from different experiments have different scales, they will also have variance (and co-variance) at different scales.</p>
<p>Multiple Factor Analysis addresses this issue and achieves balance among the data types by normalizing each of the data types, before stacking them and passing them on to PCA. Formally, MFA is given by</p>
<p><span class="math display">\[
X_n = \begin{bmatrix}
           X_{1} / \lambda^{(1)}_1 \\
           X_{2} / \lambda^{(2)}_1 \\
           \vdots \\
           X_{L} / \lambda^{(L)}_1
         \end{bmatrix} = WH,
\]</span>
where <span class="math inline">\(\lambda^{(i)}_1\)</span> is the first eigenvalue of the principal component decomposition of <span class="math inline">\(X_i\)</span>.</p>
<p>Following this normalization step, we apply PCA to <span class="math inline">\(X_n\)</span>. From there on, MFA analysis is the same as PCA analysis, and we refer the reader to Chapter <a href="unsupervisedLearning.html#unsupervisedLearning">4</a> for more details.</p>
<div id="mfa-in-r" class="section level4">
<h4><span class="header-section-number">11.3.1.1</span> MFA in R</h4>
<p>MFA is available through the CRAN package <code>FactoMineR</code>. The code snippet below shows how to run it:</p>
<div class="sourceCode" id="cb607"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb607-1"><a href="matrix-factorization-methods-for-unsupervised-multi-omics-data-integration.html#cb607-1"></a><span class="co"># run the MFA function from the FactoMineR package</span></span>
<span id="cb607-2"><a href="matrix-factorization-methods-for-unsupervised-multi-omics-data-integration.html#cb607-2"></a>r.mfa &lt;-<span class="st"> </span>FactoMineR<span class="op">::</span><span class="kw">MFA</span>(</span>
<span id="cb607-3"><a href="matrix-factorization-methods-for-unsupervised-multi-omics-data-integration.html#cb607-3"></a>  <span class="kw">t</span>(<span class="kw">rbind</span>(x1,x2,x3)), <span class="co"># binding the omics types together</span></span>
<span id="cb607-4"><a href="matrix-factorization-methods-for-unsupervised-multi-omics-data-integration.html#cb607-4"></a>  <span class="kw">c</span>(<span class="kw">dim</span>(x1)[<span class="dv">1</span>], <span class="kw">dim</span>(x2)[<span class="dv">1</span>], <span class="kw">dim</span>(x3)[<span class="dv">1</span>]), <span class="co"># specifying the dimensions of each</span></span>
<span id="cb607-5"><a href="matrix-factorization-methods-for-unsupervised-multi-omics-data-integration.html#cb607-5"></a>  <span class="dt">graph=</span><span class="ot">FALSE</span>)</span></code></pre></div>
<p>Since this generates a two-dimensional factorization of the multi-omics data, we can now plot each tumor as a dot in a 2D scatter plot to see how well the MFA factors separate the cancer subtypes. The following code snippet generates Figure <a href="matrix-factorization-methods-for-unsupervised-multi-omics-data-integration.html#fig:momfascatterplot">11.6</a>:</p>
<div class="sourceCode" id="cb608"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb608-1"><a href="matrix-factorization-methods-for-unsupervised-multi-omics-data-integration.html#cb608-1"></a><span class="co"># first, extract the H and W matrices from the MFA run result</span></span>
<span id="cb608-2"><a href="matrix-factorization-methods-for-unsupervised-multi-omics-data-integration.html#cb608-2"></a>mfa.h &lt;-<span class="st"> </span>r.mfa<span class="op">$</span>global.pca<span class="op">$</span>ind<span class="op">$</span>coord</span>
<span id="cb608-3"><a href="matrix-factorization-methods-for-unsupervised-multi-omics-data-integration.html#cb608-3"></a>mfa.w &lt;-<span class="st"> </span>r.mfa<span class="op">$</span>quanti.var<span class="op">$</span>coord</span>
<span id="cb608-4"><a href="matrix-factorization-methods-for-unsupervised-multi-omics-data-integration.html#cb608-4"></a></span>
<span id="cb608-5"><a href="matrix-factorization-methods-for-unsupervised-multi-omics-data-integration.html#cb608-5"></a><span class="co"># create a dataframe with the H matrix and the CMS label</span></span>
<span id="cb608-6"><a href="matrix-factorization-methods-for-unsupervised-multi-omics-data-integration.html#cb608-6"></a>mfa_df &lt;-<span class="st"> </span><span class="kw">as.data.frame</span>(mfa.h)</span>
<span id="cb608-7"><a href="matrix-factorization-methods-for-unsupervised-multi-omics-data-integration.html#cb608-7"></a>mfa_df<span class="op">$</span>subtype &lt;-<span class="st"> </span><span class="kw">factor</span>(covariates[<span class="kw">rownames</span>(mfa_df),]<span class="op">$</span>cms_label)</span>
<span id="cb608-8"><a href="matrix-factorization-methods-for-unsupervised-multi-omics-data-integration.html#cb608-8"></a></span>
<span id="cb608-9"><a href="matrix-factorization-methods-for-unsupervised-multi-omics-data-integration.html#cb608-9"></a><span class="co"># create the plot</span></span>
<span id="cb608-10"><a href="matrix-factorization-methods-for-unsupervised-multi-omics-data-integration.html#cb608-10"></a>ggplot2<span class="op">::</span><span class="kw">ggplot</span>(mfa_df, ggplot2<span class="op">::</span><span class="kw">aes</span>(<span class="dt">x=</span>Dim<span class="fl">.1</span>, <span class="dt">y=</span>Dim<span class="fl">.2</span>, <span class="dt">color=</span>subtype)) <span class="op">+</span></span>
<span id="cb608-11"><a href="matrix-factorization-methods-for-unsupervised-multi-omics-data-integration.html#cb608-11"></a>ggplot2<span class="op">::</span><span class="kw">geom_point</span>() <span class="op">+</span><span class="st"> </span>ggplot2<span class="op">::</span><span class="kw">ggtitle</span>(<span class="st">&quot;Scatter plot of MFA&quot;</span>)</span></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:momfascatterplot"></span>
<img src="11-multiomics-analysis_files/figure-html/momfascatterplot-1.png" alt="Scatter plot of 2-dimensional MFA for multi-omics data shows separation between the subtypes." width="50%" />
<p class="caption">
FIGURE 11.6: Scatter plot of 2-dimensional MFA for multi-omics data shows separation between the subtypes.
</p>
</div>
<p>Figure <a href="matrix-factorization-methods-for-unsupervised-multi-omics-data-integration.html#fig:momfascatterplot">11.6</a> shows remarkable separation between the cancer subtypes; it is easy enough to draw a line separating the tumors to CMS subtypes with good accuracy.</p>
<p>Another way to examine the MFA factors, which is also useful for factor models with more than two components, is a heatmap, as shown in Figure <a href="matrix-factorization-methods-for-unsupervised-multi-omics-data-integration.html#fig:momfaheatmap">11.7</a>, generated by the following code snippet:</p>
<div class="sourceCode" id="cb609"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb609-1"><a href="matrix-factorization-methods-for-unsupervised-multi-omics-data-integration.html#cb609-1"></a>pheatmap<span class="op">::</span><span class="kw">pheatmap</span>(<span class="kw">t</span>(mfa.h)[<span class="dv">1</span><span class="op">:</span><span class="dv">2</span>,], <span class="dt">annotation_col =</span> anno_col,</span>
<span id="cb609-2"><a href="matrix-factorization-methods-for-unsupervised-multi-omics-data-integration.html#cb609-2"></a>                  <span class="dt">show_colnames =</span> <span class="ot">FALSE</span>,</span>
<span id="cb609-3"><a href="matrix-factorization-methods-for-unsupervised-multi-omics-data-integration.html#cb609-3"></a>                  <span class="dt">main=</span><span class="st">&quot;MFA for multi-omics integration&quot;</span>)</span></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:momfaheatmap"></span>
<img src="11-multiomics-analysis_files/figure-html/momfaheatmap-1.png" alt="A heatmap of the two MFA components shows separation between the cancer subtypes." width="50%" />
<p class="caption">
FIGURE 11.7: A heatmap of the two MFA components shows separation between the cancer subtypes.
</p>
</div>
<p>Figure <a href="matrix-factorization-methods-for-unsupervised-multi-omics-data-integration.html#fig:momfaheatmap">11.7</a> shows that indeed, when tumors are clustered and ordered using the two MFA factors we learned above, their separation into CMS clusters is nearly trivial.</p>

<div class="rmdtip">
<p><strong>Want to know more ?</strong></p>
<ul>
<li>Learn more about FactoMineR on the website: <a href="http://factominer.free.fr/" class="uri">http://factominer.free.fr/</a></li>
<li>Learn more about MFA on the Wikipedia page <a href="https://en.wikipedia.org/wiki/Multiple_factor_analysis" class="uri">https://en.wikipedia.org/wiki/Multiple_factor_analysis</a></li>
</ul>
</div>

</div>
</div>
<div id="joint-non-negative-matrix-factorization" class="section level3">
<h3><span class="header-section-number">11.3.2</span> Joint non-negative matrix factorization</h3>
<p>As introduced in Chapter <a href="unsupervisedLearning.html#unsupervisedLearning">4</a>, NMF (Non-negative Matrix Factorization) is an algorithm from 2000 that seeks to find a non-negative additive decomposition for a non-negative data matrix. It takes the familiar form <span class="math inline">\(X \approx WH\)</span>, with <span class="math inline">\(X \ge 0\)</span>, <span class="math inline">\(W \ge 0\)</span>, and <span class="math inline">\(H \ge 0\)</span>. The non-negative constraints make a lossless decomposition (i.e. <span class="math inline">\(X=WH\)</span>) generally impossible. Hence, NMF attempts to find a solution which minimizes the Frobenius norm of the reconstruction:</p>
<p><span class="math display">\[
min~\|X-WH\|_F \\
W \ge 0, \\
H \ge 0,
\]</span></p>
<p>where the Frobenius norm <span class="math inline">\(\|\cdot\|_F\)</span> is the matrix equivalent of the Euclidean distance:</p>
<p><span class="math display">\[
\|X\|_F = \sqrt{\sum_i\sum_jx_{ij}^2}.
\]</span></p>
<p>This is typically solved for <span class="math inline">\(W\)</span> and <span class="math inline">\(H\)</span> using random initializations followed by iterations of a multiplicative update rule:</p>
<p><span class="math display">\[\begin{align}
    W_{t+1} &amp;= W_t^T \frac{XH_t^T}{XH_tH_t^T} \\
    H_{t+1} &amp;= H_t \frac{W_t^TX}{W^T_tW_tX}.
\end{align}\]</span></p>
<p>Since this algorithm is guaranteed only to converge to a local minimum, it is typically run several times with random initializations, and the best result is kept.</p>
<p>In the multi-omics context, we will, as in the MFA case, wish to find a decomposition for an integrated data matrix of the form</p>
<p><span class="math display">\[
X = \begin{bmatrix}
    X_{1} \\
    X_{2} \\
    \vdots \\
    X_{L}
\end{bmatrix},
\]</span></p>
<p>with <span class="math inline">\(X_i\)</span>s denoting data from different omics platforms.</p>
<p>As NMF seeks to minimize the reconstruction error <span class="math inline">\(\|X-WH\|_F\)</span>, some care needs to be taken with regards to data normalization. Different omics platforms may produce data with different scales (i.e. real-valued gene expression quantification, binary mutation data, etc.), and so will have different baseline Frobenius norms. To address this, when doing Joint NMF, we first feature-normalize each data matrix, and then normalize by the Frobenius norm of the data matrix. Formally, we run NMF on</p>
<p><span class="math display">\[
X = \begin{bmatrix}
    X_{1}^N / \alpha_1 \\
    X_{2}^N / \alpha_2 \\
    \vdots \\
    X_{L}^N / \alpha_L
\end{bmatrix},
\]</span></p>
<p>where <span class="math inline">\(X_i^N\)</span> is the feature-normalized data matrix <span class="math inline">\(X_i^N = \frac{x^{ij}}{\sum_jx^{ij}}\)</span>, and <span class="math inline">\(\alpha_i = \|X_{i}^N\|_F\)</span>.</p>
<p>Another consideration with NMF is the non-negativity constraint. Different omics data types may have negative values, for instance, copy-number variations (CNVs) may be positive, indicating gains, or negative, indicating losses, as in Table <a href="matrix-factorization-methods-for-unsupervised-multi-omics-data-integration.html#tab:mocnvsplitcolshow1">11.4</a>. In order to turn such data into a non-negative form, we will split each feature into two features, one new feature holding all the non-negative values of the original feature, and another feature holding the absolute value of the negative ones, as in Table <a href="matrix-factorization-methods-for-unsupervised-multi-omics-data-integration.html#tab:mocnvsplitcolshow2">11.5</a>.</p>
<table>
<caption><span id="tab:mocnvsplitcolshow1">TABLE 11.4: </span>Example copy number data. Data can be both positive (amplified regions) or negative (deleted regions).</caption>
<thead>
<tr class="header">
<th></th>
<th align="right">seg1</th>
<th align="right">seg2</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>samp1</td>
<td align="right">1</td>
<td align="right">0</td>
</tr>
<tr class="even">
<td>samp2</td>
<td align="right">2</td>
<td align="right">1</td>
</tr>
<tr class="odd">
<td>samp3</td>
<td align="right">1</td>
<td align="right">-2</td>
</tr>
</tbody>
</table>
<table>
<caption><span id="tab:mocnvsplitcolshow2">TABLE 11.5: </span>Example copy number data after splitting each column into a column representing copy number gains (+) and a column representing deletions (-). This data matrix is non-negative, and thus suitable for NMF algorithms.</caption>
<thead>
<tr class="header">
<th></th>
<th align="right">seg1+</th>
<th align="right">seg1-</th>
<th align="right">seg2+</th>
<th align="right">seg2-</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>samp1</td>
<td align="right">1</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
</tr>
<tr class="even">
<td>samp2</td>
<td align="right">2</td>
<td align="right">0</td>
<td align="right">1</td>
<td align="right">0</td>
</tr>
<tr class="odd">
<td>samp3</td>
<td align="right">1</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">2</td>
</tr>
</tbody>
</table>
<div id="nmf-in-r" class="section level4">
<h4><span class="header-section-number">11.3.2.1</span> NMF in R</h4>
<p>Many NMF algorithms are available through the CRAN package <code>NMF</code>. The following code chunk demonstrates how it may be run:</p>
<div class="sourceCode" id="cb610"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb610-1"><a href="matrix-factorization-methods-for-unsupervised-multi-omics-data-integration.html#cb610-1"></a><span class="co"># Feature-normalize the data</span></span>
<span id="cb610-2"><a href="matrix-factorization-methods-for-unsupervised-multi-omics-data-integration.html#cb610-2"></a>x1.featnorm &lt;-<span class="st"> </span>x1 <span class="op">/</span><span class="st"> </span><span class="kw">rowSums</span>(x1)</span>
<span id="cb610-3"><a href="matrix-factorization-methods-for-unsupervised-multi-omics-data-integration.html#cb610-3"></a>x2.featnorm &lt;-<span class="st"> </span>x2 <span class="op">/</span><span class="st"> </span><span class="kw">rowSums</span>(x2)</span>
<span id="cb610-4"><a href="matrix-factorization-methods-for-unsupervised-multi-omics-data-integration.html#cb610-4"></a>x3.featnorm &lt;-<span class="st"> </span>x3 <span class="op">/</span><span class="st"> </span><span class="kw">rowSums</span>(x3)</span>
<span id="cb610-5"><a href="matrix-factorization-methods-for-unsupervised-multi-omics-data-integration.html#cb610-5"></a></span>
<span id="cb610-6"><a href="matrix-factorization-methods-for-unsupervised-multi-omics-data-integration.html#cb610-6"></a><span class="co"># Normalize by each omics type&#39;s frobenius norm</span></span>
<span id="cb610-7"><a href="matrix-factorization-methods-for-unsupervised-multi-omics-data-integration.html#cb610-7"></a>x1.featnorm.frobnorm &lt;-<span class="st"> </span>x1.featnorm <span class="op">/</span><span class="st"> </span><span class="kw">norm</span>(<span class="kw">as.matrix</span>(x1.featnorm), <span class="dt">type=</span><span class="st">&quot;F&quot;</span>)</span>
<span id="cb610-8"><a href="matrix-factorization-methods-for-unsupervised-multi-omics-data-integration.html#cb610-8"></a>x2.featnorm.frobnorm &lt;-<span class="st"> </span>x2.featnorm <span class="op">/</span><span class="st"> </span><span class="kw">norm</span>(<span class="kw">as.matrix</span>(x2.featnorm), <span class="dt">type=</span><span class="st">&quot;F&quot;</span>)</span>
<span id="cb610-9"><a href="matrix-factorization-methods-for-unsupervised-multi-omics-data-integration.html#cb610-9"></a>x3.featnorm.frobnorm &lt;-<span class="st"> </span>x3.featnorm <span class="op">/</span><span class="st"> </span><span class="kw">norm</span>(<span class="kw">as.matrix</span>(x3.featnorm), <span class="dt">type=</span><span class="st">&quot;F&quot;</span>)</span>
<span id="cb610-10"><a href="matrix-factorization-methods-for-unsupervised-multi-omics-data-integration.html#cb610-10"></a></span>
<span id="cb610-11"><a href="matrix-factorization-methods-for-unsupervised-multi-omics-data-integration.html#cb610-11"></a><span class="co"># Split the features of the CNV matrix into two non-negative features each</span></span>
<span id="cb610-12"><a href="matrix-factorization-methods-for-unsupervised-multi-omics-data-integration.html#cb610-12"></a>x3.featnorm.frobnorm.nonneg &lt;-<span class="st"> </span><span class="kw">t</span>(<span class="kw">split_neg_columns</span>(<span class="kw">t</span>(x3.featnorm.frobnorm)))</span>
<span id="cb610-13"><a href="matrix-factorization-methods-for-unsupervised-multi-omics-data-integration.html#cb610-13"></a></span>
<span id="cb610-14"><a href="matrix-factorization-methods-for-unsupervised-multi-omics-data-integration.html#cb610-14"></a><span class="co"># run the nmf function from the NMF package</span></span>
<span id="cb610-15"><a href="matrix-factorization-methods-for-unsupervised-multi-omics-data-integration.html#cb610-15"></a><span class="kw">require</span>(NMF)</span></code></pre></div>
<pre><code>## Warning: package &#39;NMF&#39; was built under R version 4.0.2</code></pre>
<div class="sourceCode" id="cb612"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb612-1"><a href="matrix-factorization-methods-for-unsupervised-multi-omics-data-integration.html#cb612-1"></a>r.nmf &lt;-<span class="st"> </span><span class="kw">nmf</span>(<span class="kw">t</span>(<span class="kw">rbind</span>(x1.featnorm.frobnorm,</span>
<span id="cb612-2"><a href="matrix-factorization-methods-for-unsupervised-multi-omics-data-integration.html#cb612-2"></a>                     x2.featnorm.frobnorm,</span>
<span id="cb612-3"><a href="matrix-factorization-methods-for-unsupervised-multi-omics-data-integration.html#cb612-3"></a>                     x3.featnorm.frobnorm.nonneg)),</span>
<span id="cb612-4"><a href="matrix-factorization-methods-for-unsupervised-multi-omics-data-integration.html#cb612-4"></a>             <span class="dv">2</span>,</span>
<span id="cb612-5"><a href="matrix-factorization-methods-for-unsupervised-multi-omics-data-integration.html#cb612-5"></a>             <span class="dt">method=</span><span class="st">&#39;Frobenius&#39;</span>)</span>
<span id="cb612-6"><a href="matrix-factorization-methods-for-unsupervised-multi-omics-data-integration.html#cb612-6"></a></span>
<span id="cb612-7"><a href="matrix-factorization-methods-for-unsupervised-multi-omics-data-integration.html#cb612-7"></a><span class="co"># exctract the H and W matrices from the nmf run result</span></span>
<span id="cb612-8"><a href="matrix-factorization-methods-for-unsupervised-multi-omics-data-integration.html#cb612-8"></a>nmf.h &lt;-<span class="st"> </span>NMF<span class="op">::</span><span class="kw">basis</span>(r.nmf)</span>
<span id="cb612-9"><a href="matrix-factorization-methods-for-unsupervised-multi-omics-data-integration.html#cb612-9"></a>nmf.w &lt;-<span class="st"> </span>NMF<span class="op">::</span><span class="kw">coef</span>(r.nmf)</span>
<span id="cb612-10"><a href="matrix-factorization-methods-for-unsupervised-multi-omics-data-integration.html#cb612-10"></a>nmfw &lt;-<span class="st"> </span><span class="kw">t</span>(nmf.w)</span></code></pre></div>
<p>As with MFA, we can examine how well 2-factor NMF splits tumors into subtypes by looking at the scatter plot in Figure <a href="matrix-factorization-methods-for-unsupervised-multi-omics-data-integration.html#fig:monmfscatterplot">11.8</a>, generated by the following code chunk:</p>
<div class="sourceCode" id="cb613"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb613-1"><a href="matrix-factorization-methods-for-unsupervised-multi-omics-data-integration.html#cb613-1"></a><span class="co"># create a dataframe with the H matrix and the CMS label (subtype)</span></span>
<span id="cb613-2"><a href="matrix-factorization-methods-for-unsupervised-multi-omics-data-integration.html#cb613-2"></a>nmf_df &lt;-<span class="st"> </span><span class="kw">as.data.frame</span>(nmf.h)</span>
<span id="cb613-3"><a href="matrix-factorization-methods-for-unsupervised-multi-omics-data-integration.html#cb613-3"></a><span class="kw">colnames</span>(nmf_df) &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;dim1&quot;</span>, <span class="st">&quot;dim2&quot;</span>)</span>
<span id="cb613-4"><a href="matrix-factorization-methods-for-unsupervised-multi-omics-data-integration.html#cb613-4"></a>nmf_df<span class="op">$</span>subtype &lt;-<span class="st"> </span><span class="kw">factor</span>(covariates[<span class="kw">rownames</span>(nmf_df),]<span class="op">$</span>cms_label)</span>
<span id="cb613-5"><a href="matrix-factorization-methods-for-unsupervised-multi-omics-data-integration.html#cb613-5"></a></span>
<span id="cb613-6"><a href="matrix-factorization-methods-for-unsupervised-multi-omics-data-integration.html#cb613-6"></a><span class="co"># create the scatter plot</span></span>
<span id="cb613-7"><a href="matrix-factorization-methods-for-unsupervised-multi-omics-data-integration.html#cb613-7"></a>ggplot2<span class="op">::</span><span class="kw">ggplot</span>(nmf_df, ggplot2<span class="op">::</span><span class="kw">aes</span>(<span class="dt">x=</span>dim1, <span class="dt">y=</span>dim2, <span class="dt">color=</span>subtype)) <span class="op">+</span></span>
<span id="cb613-8"><a href="matrix-factorization-methods-for-unsupervised-multi-omics-data-integration.html#cb613-8"></a>ggplot2<span class="op">::</span><span class="kw">geom_point</span>() <span class="op">+</span></span>
<span id="cb613-9"><a href="matrix-factorization-methods-for-unsupervised-multi-omics-data-integration.html#cb613-9"></a>ggplot2<span class="op">::</span><span class="kw">ggtitle</span>(<span class="st">&quot;Scatter plot of 2-component NMF for multi-omics integration&quot;</span>)</span></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:monmfscatterplot"></span>
<img src="11-multiomics-analysis_files/figure-html/monmfscatterplot-1.png" alt="NMF creates a disentangled representation of the data using two components which allow for separation between tumor sub-types CMS1 and CMS3 based on NMF factors learned from multi-omics data." width="50%" />
<p class="caption">
FIGURE 11.8: NMF creates a disentangled representation of the data using two components which allow for separation between tumor sub-types CMS1 and CMS3 based on NMF factors learned from multi-omics data.
</p>
</div>
<p>Figure <a href="matrix-factorization-methods-for-unsupervised-multi-omics-data-integration.html#fig:monmfscatterplot">11.8</a> shows an important difference between NMF and MFA (PCA). It shows the tendency of samples to lie close to the X or Y axes, that is, the tendency of each sample to be high in only one of the factors. This will be discussed more in the later section on disentangledness.</p>
<p>Again, should we choose to run NMF with more than two factors, a more useful plot might be the heatmap shown in Figure <a href="matrix-factorization-methods-for-unsupervised-multi-omics-data-integration.html#fig:monmfheatmap">11.9</a>, generated by the following code snippet:</p>
<div class="sourceCode" id="cb614"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb614-1"><a href="matrix-factorization-methods-for-unsupervised-multi-omics-data-integration.html#cb614-1"></a>pheatmap<span class="op">::</span><span class="kw">pheatmap</span>(<span class="kw">t</span>(nmf_df[,<span class="dv">1</span><span class="op">:</span><span class="dv">2</span>]),</span>
<span id="cb614-2"><a href="matrix-factorization-methods-for-unsupervised-multi-omics-data-integration.html#cb614-2"></a>                   <span class="dt">annotation_col =</span> anno_col,</span>
<span id="cb614-3"><a href="matrix-factorization-methods-for-unsupervised-multi-omics-data-integration.html#cb614-3"></a>                   <span class="dt">show_colnames=</span><span class="ot">FALSE</span>,</span>
<span id="cb614-4"><a href="matrix-factorization-methods-for-unsupervised-multi-omics-data-integration.html#cb614-4"></a>                   <span class="dt">main=</span><span class="st">&quot;Heatmap of 2-component NMF&quot;</span>)</span></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:monmfheatmap"></span>
<img src="11-multiomics-analysis_files/figure-html/monmfheatmap-1.png" alt="A heatmap of NMF factors shows separability of tumors into subtype clusters. This plot is more useful than a scatter plot when there are more than two factors." width="50%" />
<p class="caption">
FIGURE 11.9: A heatmap of NMF factors shows separability of tumors into subtype clusters. This plot is more useful than a scatter plot when there are more than two factors.
</p>
</div>

<div class="rmdtip">
<p><strong>Want to know more ?</strong></p>
<ul>
<li>Joint NMF to uncover gene regulatory networks: Zhang S., Li Q., Liu J., Zhou X. J. (2011). A novel computational framework for simultaneous integration of multiple types of genomic data to identify microRNA-gene regulatory modules. <em>Bioinformatics</em> 27, i401–i409. 10.1093/bioinformatics/btr206 <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3117336/" class="uri">https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3117336/</a></li>
<li>Joint NMF for cancer research: Zhang S., Liu C.-C., Li W., Shen H., Laird P. W., Zhou X. J. (2012). Discovery of multi-dimensional modules by integrative analysis of cancer genomic data. <em>Nucleic Acids Res.</em> 40, 9379–9391. 10.1093/nar/gks725 <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3479191/" class="uri">https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3479191/</a></li>
</ul>
</div>

</div>
</div>
<div id="icluster" class="section level3">
<h3><span class="header-section-number">11.3.3</span> iCluster</h3>
<p>iCluster takes a Bayesian approach to the latent variable model. In Bayesian statistics, we infer distributions over model parameters, rather than finding a single maximum-likelihood parameter estimate. In iCluster, we model the data as</p>
<p><span class="math display">\[
X_{(i)} = W_{(i)}Z + \epsilon_i,
\]</span></p>
<p>where <span class="math inline">\(X_{(i)}\)</span> is a data matrix from a single omics platform, <span class="math inline">\(W_{(i)}\)</span> are model parameters, <span class="math inline">\(Z\)</span> is a latent variable matrix, and is shared among the different omics platforms, and <span class="math inline">\(\epsilon_i\)</span> is a “noise” random variable, <span class="math inline">\(\epsilon \sim N(0,\Psi)\)</span>, with <span class="math inline">\(\Psi = diag(\psi_1,\dots \psi_M)\)</span> is a diagonal covariance matrix.</p>
<div class="figure" style="text-align: center"><span id="fig:moiCluster"></span>
<img src="images/icluster.png" alt="Sketch of iCluster model. Each omics datatype is decomposed to a coefficient matrix and a shared latent variable matrix, plus noise." width="75%" />
<p class="caption">
FIGURE 11.10: Sketch of iCluster model. Each omics datatype is decomposed to a coefficient matrix and a shared latent variable matrix, plus noise.
</p>
</div>
<p>Note that with this construction, the omics measurements <span class="math inline">\(X\)</span> are expected to be the same for samples with the same latent variable representation, up to Gaussian noise. Further, we assume a Gaussian prior distribution on the latent variables <span class="math inline">\(Z \sim N(0,I)\)</span>, which means we assume <span class="math inline">\(X_{(i)} \sim N \big( 0,W_{(i)} W_{(i)}^T + \Psi_{(i)} \big)\)</span>. In order to find suitable values for <span class="math inline">\(W\)</span>, <span class="math inline">\(Z\)</span>, and <span class="math inline">\(\Psi\)</span>, we can write down the multivariate normal log-likelihood function and optimize it. For a multivariate normal distribution with mean <span class="math inline">\(0\)</span> and covariance <span class="math inline">\(\Sigma\)</span>, the log-likelihood function is given by</p>
<p><span class="math display">\[
\ell = -\frac{1}{2} \bigg( \ln (|\Sigma|) + X^T \Sigma^{-1} X + k\ln (2 \pi) \bigg)
\]</span></p>
<p>(this is simply the log of the Probability Density Function of a multivariate Gaussian). For the multi-omics iCluster case, we have <span class="math inline">\(X=\big( X_{(1)}, \dots, X_{(L)} \big)^T\)</span>, <span class="math inline">\(W = \big( W_{(1)}, \dots, W_{(L)} \big)^T\)</span>, where <span class="math inline">\(X\)</span> is a multivariate normal with <span class="math inline">\(0\)</span>-mean and <span class="math inline">\(\Sigma = W W^T + \Psi\)</span> covariance. Hence, the log-likelihood function for the iCluster model is given by:</p>
<p><span class="math display">\[
\ell_{iC}(W,\Sigma) = -\frac{1}{2} \bigg( \sum_{i=1}^L \ln (|\Sigma|) + X^T\Sigma^{-1}X + p_i \ln (2 \pi) \bigg)
\]</span></p>
<p>where <span class="math inline">\(p_i\)</span> is the number of features in omics data type <span class="math inline">\(i\)</span>. Because this model has more parameters than we typically have samples, we need to push the model to use fewer parameters than it has at its disposal, by using regularization. iCluster uses Lasso regularization, which is a direct penalty on the absolute value of the parameters. I.e., instead of optimizing <span class="math inline">\(\ell_{iC}(W,\Sigma)\)</span>, we will optimize the regularized log-likelihood:</p>
<p><span class="math display">\[
\ell = \ell_{iC}(W,\Sigma) - \lambda\|W\|_1.
\]</span></p>
<p>The parameter <span class="math inline">\(\lambda\)</span> acts as a dial to weigh the trade-off between better model fits (higher log-likelihood) and a sparser model, with more <span class="math inline">\(w_{ij}\)</span>s set to <span class="math inline">\(0\)</span>, which gives models which generalize better and are more interpretable.</p>
<p>In order to solve this problem, iCluster employs the Expectation Maximization (EM) algorithm. The full details are beyond the scope of this textbook. We will introduce a short sketch instead. The intuition behind the EM algorithm is a more general case of the k-means clustering algorithm (Chapter 4). The basic <strong>EM algorithm</strong> is as follows.</p>
<ul>
<li>Initialize <span class="math inline">\(W\)</span> and <span class="math inline">\(\Psi\)</span>.</li>
<li><strong>Until convergence of <span class="math inline">\(W\)</span>, <span class="math inline">\(\Psi\)</span></strong>
<ul>
<li>E-step: Calculate the expected value of <span class="math inline">\(Z\)</span> given the current estimates of <span class="math inline">\(W\)</span> and <span class="math inline">\(\Psi\)</span> and the data <span class="math inline">\(X\)</span>.</li>
<li>M-step: Calculate maximum likelihood estimates for the parameters <span class="math inline">\(W\)</span> and <span class="math inline">\(\Psi\)</span> based on the current estimate of <span class="math inline">\(Z\)</span> and the data <span class="math inline">\(X\)</span>.</li>
</ul></li>
</ul>
<div id="icluster-extending-icluster" class="section level4">
<h4><span class="header-section-number">11.3.3.1</span> iCluster+: Extending iCluster</h4>
<p>iCluster+ is an extension of the iCluster framework, which allows for omics types to arise from distributions other than a Gaussian. While normal distributions are a good assumption for log-transformed, centered gene expression data, it is a poor model for binary mutations data, or for copy number variation data, which can typically take the values <span class="math inline">\((-2, 1, 0, 1, 2)\)</span> for heterozygous / monozygous deletions or amplifications. iCluster+ allows the different <span class="math inline">\(X\)</span>s to have different distributions:</p>
<ul>
<li>for binary mutations, <span class="math inline">\(X\)</span> is drawn from a multivariate binomial</li>
<li>for normal, continuous data, <span class="math inline">\(X\)</span> is drawn from a multivariate Gaussian</li>
<li>for copy number variations, <span class="math inline">\(X\)</span> is drawn from a multinomial</li>
<li>for count data, <span class="math inline">\(X\)</span> is drawn from a Poisson.</li>
</ul>
<p>In that way, iCluster+ allows us to explicitly model our assumptions about the distributions of our different omics data types, and leverage the strengths of Bayesian inference.</p>
<p>Both iCluster and iCluster+ make use of sophisticated Bayesian inference algorithms (EM for iCluster, Metropolis-Hastings MCMC for iCluster+), which means they do not scale up trivially. Therefore, it is recommended to filter down the features to a manageable size before inputting data to the algorithm. The exact size of “manageable” data depends on your hardware, but a rule of thumb is that dimensions in the thousands are ok, but in the tens of thousands might be too slow.</p>
</div>
<div id="running-icluster" class="section level4">
<h4><span class="header-section-number">11.3.3.2</span> Running iCluster+</h4>
<p>iCluster+ is available through the Bioconductor package <code>iClusterPlus</code>. The following code snippet demonstrates how it can be run with two components:</p>
<div class="sourceCode" id="cb615"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb615-1"><a href="matrix-factorization-methods-for-unsupervised-multi-omics-data-integration.html#cb615-1"></a><span class="co"># run the iClusterPlus function</span></span>
<span id="cb615-2"><a href="matrix-factorization-methods-for-unsupervised-multi-omics-data-integration.html#cb615-2"></a>r.icluster &lt;-<span class="st"> </span>iClusterPlus<span class="op">::</span><span class="kw">iClusterPlus</span>(</span>
<span id="cb615-3"><a href="matrix-factorization-methods-for-unsupervised-multi-omics-data-integration.html#cb615-3"></a>  <span class="kw">t</span>(x1), <span class="co"># Providing each omics type</span></span>
<span id="cb615-4"><a href="matrix-factorization-methods-for-unsupervised-multi-omics-data-integration.html#cb615-4"></a>  <span class="kw">t</span>(x2),</span>
<span id="cb615-5"><a href="matrix-factorization-methods-for-unsupervised-multi-omics-data-integration.html#cb615-5"></a>  <span class="kw">t</span>(x3),</span>
<span id="cb615-6"><a href="matrix-factorization-methods-for-unsupervised-multi-omics-data-integration.html#cb615-6"></a>  <span class="dt">type=</span><span class="kw">c</span>(<span class="st">&quot;gaussian&quot;</span>, <span class="st">&quot;binomial&quot;</span>, <span class="st">&quot;multinomial&quot;</span>), <span class="co"># Providing the distributions</span></span>
<span id="cb615-7"><a href="matrix-factorization-methods-for-unsupervised-multi-omics-data-integration.html#cb615-7"></a>  <span class="dt">K=</span><span class="dv">2</span>, <span class="co"># provide the number of factors to learn</span></span>
<span id="cb615-8"><a href="matrix-factorization-methods-for-unsupervised-multi-omics-data-integration.html#cb615-8"></a>  <span class="dt">alpha=</span><span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">1</span>,<span class="dv">1</span>), <span class="co"># as well as other model parameters</span></span>
<span id="cb615-9"><a href="matrix-factorization-methods-for-unsupervised-multi-omics-data-integration.html#cb615-9"></a>  <span class="dt">lambda=</span><span class="kw">c</span>(.<span class="dv">03</span>,.<span class="dv">03</span>,.<span class="dv">03</span>))</span>
<span id="cb615-10"><a href="matrix-factorization-methods-for-unsupervised-multi-omics-data-integration.html#cb615-10"></a></span>
<span id="cb615-11"><a href="matrix-factorization-methods-for-unsupervised-multi-omics-data-integration.html#cb615-11"></a><span class="co"># extract the H and W matrices from the run result</span></span>
<span id="cb615-12"><a href="matrix-factorization-methods-for-unsupervised-multi-omics-data-integration.html#cb615-12"></a><span class="co"># here, we refer to H as z, to keep with iCluster terminology</span></span>
<span id="cb615-13"><a href="matrix-factorization-methods-for-unsupervised-multi-omics-data-integration.html#cb615-13"></a>icluster.z &lt;-<span class="st"> </span>r.icluster<span class="op">$</span>meanZ</span>
<span id="cb615-14"><a href="matrix-factorization-methods-for-unsupervised-multi-omics-data-integration.html#cb615-14"></a><span class="kw">rownames</span>(icluster.z) &lt;-<span class="st"> </span><span class="kw">rownames</span>(covariates) <span class="co"># fix the row names</span></span>
<span id="cb615-15"><a href="matrix-factorization-methods-for-unsupervised-multi-omics-data-integration.html#cb615-15"></a>icluster.ws &lt;-<span class="st"> </span>r.icluster<span class="op">$</span>beta</span>
<span id="cb615-16"><a href="matrix-factorization-methods-for-unsupervised-multi-omics-data-integration.html#cb615-16"></a></span>
<span id="cb615-17"><a href="matrix-factorization-methods-for-unsupervised-multi-omics-data-integration.html#cb615-17"></a><span class="co"># construct a dataframe with the H matrix (z) and the cancer subtypes</span></span>
<span id="cb615-18"><a href="matrix-factorization-methods-for-unsupervised-multi-omics-data-integration.html#cb615-18"></a><span class="co"># for later plotting</span></span>
<span id="cb615-19"><a href="matrix-factorization-methods-for-unsupervised-multi-omics-data-integration.html#cb615-19"></a>icp_df &lt;-<span class="st"> </span><span class="kw">as.data.frame</span>(icluster.z)</span>
<span id="cb615-20"><a href="matrix-factorization-methods-for-unsupervised-multi-omics-data-integration.html#cb615-20"></a><span class="kw">colnames</span>(icp_df) &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;dim1&quot;</span>, <span class="st">&quot;dim2&quot;</span>)</span>
<span id="cb615-21"><a href="matrix-factorization-methods-for-unsupervised-multi-omics-data-integration.html#cb615-21"></a><span class="kw">rownames</span>(icp_df) &lt;-<span class="st"> </span><span class="kw">colnames</span>(x1)</span>
<span id="cb615-22"><a href="matrix-factorization-methods-for-unsupervised-multi-omics-data-integration.html#cb615-22"></a>icp_df<span class="op">$</span>subtype &lt;-<span class="st"> </span><span class="kw">factor</span>(covariates[<span class="kw">rownames</span>(icp_df),]<span class="op">$</span>cms_label)</span></code></pre></div>
<p>As with other methods, we examine the iCluster results by looking at the scatter plot in Figure <a href="matrix-factorization-methods-for-unsupervised-multi-omics-data-integration.html#fig:moiclusterplusscatter">11.11</a> and the heatmap in Figure <a href="matrix-factorization-methods-for-unsupervised-multi-omics-data-integration.html#fig:moiclusterplusheatmap">11.12</a>. Both figures show that iCluster learns two factors which nearly perfectly discriminate between tumors of the two subtypes.</p>
<div class="figure" style="text-align: center"><span id="fig:moiclusterplusscatter"></span>
<img src="11-multiomics-analysis_files/figure-html/moiclusterplusscatter-1.png" alt="iCluster+ learns factors which allow tumor sub-types CMS1 and CMS3 to be discriminated." width="50%" />
<p class="caption">
FIGURE 11.11: iCluster+ learns factors which allow tumor sub-types CMS1 and CMS3 to be discriminated.
</p>
</div>
<div class="figure" style="text-align: center"><span id="fig:moiclusterplusheatmap"></span>
<img src="11-multiomics-analysis_files/figure-html/moiclusterplusheatmap-1.png" alt="iCluster+ factors, shown in a heatmap, separate tumors into their subtypes well." width="50%" />
<p class="caption">
FIGURE 11.12: iCluster+ factors, shown in a heatmap, separate tumors into their subtypes well.
</p>
</div>

<div class="rmdtip">
<p><strong>Want to know more ?</strong></p>
<ul>
<li>Read the original iCluster paper: Shen R., Olshen A. B., Ladanyi M. (2009). Integrative clustering of multiple genomic data types using a joint latent variable model with application to breast and lung cancer subtype analysis. <em>Bioinformatics</em> 25, 2906–2912. 10.1093/bioinformatics/btp543 <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2800366/" class="uri">https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2800366/</a></li>
<li>Read the original iClusterPlus paper: an extension of iCluster: Shen R., Mo Q., Schultz N., Seshan V. E., Olshen A. B., Huse J., et al. (2012). Integrative subtype discovery in glioblastoma using iCluster. <em>PLoS ONE</em> 7:e35236. 10.1371/journal.pone.0035236 <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3335101/" class="uri">https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3335101/</a></li>
<li>Learn more about the LASSO for model regularization: Tibshirani, R. (1996). Regression shrinkage and selection via the lasso. <em>J. Royal. Statist. Soc B.</em>, Vol. 58, No. 1, pages 267-288: <a href="http://www-stat.stanford.edu/%7Etibs/lasso/lasso.pdf" class="uri">http://www-stat.stanford.edu/%7Etibs/lasso/lasso.pdf</a></li>
<li>Learn more about the EM algorithm: Dempster, A. P., et al. Maximum likelihood from incomplete data via the EM algorithm. <em>Journal of the Royal Statistical Society. Series B (Methodological)</em>, vol. 39, no. 1, 1977, pp. 1–38. JSTOR, JSTOR: <a href="http://www.jstor.org/stable/2984875" class="uri">http://www.jstor.org/stable/2984875</a></li>
<li>Read about MCMC algorithms: Hastings, W.K. (1970). Monte Carlo sampling methods using Markov chains and their applications. <em>Biometrika.</em> 57 (1): 97–109. <a href="doi:10.1093/biomet/57.1.97" class="uri">doi:10.1093/biomet/57.1.97</a>: <a href="https://www.jstor.org/stable/2334940" class="uri">https://www.jstor.org/stable/2334940</a></li>
</ul>
</div>

</div>
</div>
</div>
<div class="footnotes">
<hr />
<ol start="2">
<li id="fn2"><p>When dealing with categorical variables, MFA uses MCA (Multiple Correspondence Analysis). This is less relevant to biological data analysis and will not be discussed here.<a href="matrix-factorization-methods-for-unsupervised-multi-omics-data-integration.html#fnref2" class="footnote-back">↩︎</a></p></li>
</ol>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="latent-variable-models-for-multi-omics-integration.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="clustering-using-latent-factors.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": true,
"facebook": false,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/compgenomr/book/edit/master/11-multiomics-analysis.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": null,
"toc": {
"collapse": "none"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
