<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>9.6 Peak calling | Computational Genomics with R</title>
  <meta name="description" content="A guide to computationa genomics using R. The book covers fundemental topics with practical examples for an interdisciplinery audience" />
  <meta name="generator" content="bookdown 0.19 and GitBook 2.6.7" />

  <meta property="og:title" content="9.6 Peak calling | Computational Genomics with R" />
  <meta property="og:type" content="book" />
  <meta property="og:url" content="https://compmgenomr.github.io/book/" />
  <meta property="og:image" content="https://compmgenomr.github.io/book/images/cover.jpg" />
  <meta property="og:description" content="A guide to computationa genomics using R. The book covers fundemental topics with practical examples for an interdisciplinery audience" />
  <meta name="github-repo" content="compgenomr/book" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="9.6 Peak calling | Computational Genomics with R" />
  
  <meta name="twitter:description" content="A guide to computationa genomics using R. The book covers fundemental topics with practical examples for an interdisciplinery audience" />
  <meta name="twitter:image" content="https://compmgenomr.github.io/book/images/cover.jpg" />

<meta name="author" content="Altuna Akalin" />


<meta name="date" content="2020-09-30" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="chip-quality-control.html"/>
<link rel="next" href="motif-discovery.html"/>
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />









<!-- Global site tag (gtag.js) - Google Analytics -->
<script async src="https://www.googletagmanager.com/gtag/js?id=UA-83786243-1"></script>
<script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-83786243-1');
</script>


<style type="text/css">
code.sourceCode > span { display: inline-block; line-height: 1.25; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<link rel="stylesheet" href="css/style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">Computational Genomics with R</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Preface</a><ul>
<li class="chapter" data-level="" data-path="who-is-this-book-for.html"><a href="who-is-this-book-for.html"><i class="fa fa-check"></i>Who is this book for?</a><ul>
<li class="chapter" data-level="" data-path="who-is-this-book-for.html"><a href="who-is-this-book-for.html#what-will-you-get-out-of-this"><i class="fa fa-check"></i>What will you get out of this?</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="structure-of-the-book.html"><a href="structure-of-the-book.html"><i class="fa fa-check"></i>Structure of the book</a></li>
<li class="chapter" data-level="" data-path="software-information-and-conventions.html"><a href="software-information-and-conventions.html"><i class="fa fa-check"></i>Software information and conventions</a><ul>
<li class="chapter" data-level="" data-path="software-information-and-conventions.html"><a href="software-information-and-conventions.html#assignment-operator-convention"><i class="fa fa-check"></i>Assignment operator convention</a></li>
<li class="chapter" data-level="" data-path="software-information-and-conventions.html"><a href="software-information-and-conventions.html#packages-needed-to-run-the-book-code"><i class="fa fa-check"></i>Packages needed to run the book code</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="data-for-the-book.html"><a href="data-for-the-book.html"><i class="fa fa-check"></i>Data for the book</a></li>
<li class="chapter" data-level="" data-path="exercises-in-the-book.html"><a href="exercises-in-the-book.html"><i class="fa fa-check"></i>Exercises in the book</a></li>
<li class="chapter" data-level="" data-path="reproducibility-statement.html"><a href="reproducibility-statement.html"><i class="fa fa-check"></i>Reproducibility statement</a></li>
<li class="chapter" data-level="" data-path="acknowledgements.html"><a href="acknowledgements.html"><i class="fa fa-check"></i>Acknowledgements</a></li>
<li class="chapter" data-level="" data-path="how-to-contribute.html"><a href="how-to-contribute.html"><i class="fa fa-check"></i>How to contribute</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="about-the-authors.html"><a href="about-the-authors.html"><i class="fa fa-check"></i>About the Authors</a></li>
<li class="chapter" data-level="1" data-path="intro.html"><a href="intro.html"><i class="fa fa-check"></i><b>1</b> Introduction to Genomics</a><ul>
<li class="chapter" data-level="1.1" data-path="genes-dna-and-central-dogma.html"><a href="genes-dna-and-central-dogma.html"><i class="fa fa-check"></i><b>1.1</b> Genes, DNA and central dogma</a><ul>
<li class="chapter" data-level="1.1.1" data-path="genes-dna-and-central-dogma.html"><a href="genes-dna-and-central-dogma.html#what-is-a-genome"><i class="fa fa-check"></i><b>1.1.1</b> What is a genome?</a></li>
<li class="chapter" data-level="1.1.2" data-path="genes-dna-and-central-dogma.html"><a href="genes-dna-and-central-dogma.html#what-is-a-gene"><i class="fa fa-check"></i><b>1.1.2</b> What is a gene?</a></li>
<li class="chapter" data-level="1.1.3" data-path="genes-dna-and-central-dogma.html"><a href="genes-dna-and-central-dogma.html#how-are-genes-controlled-transcriptional-and-post-transcriptional-regulation"><i class="fa fa-check"></i><b>1.1.3</b> How are genes controlled? Transcriptional and post-transcriptional regulation</a></li>
<li class="chapter" data-level="1.1.4" data-path="genes-dna-and-central-dogma.html"><a href="genes-dna-and-central-dogma.html#what-does-a-gene-look-like"><i class="fa fa-check"></i><b>1.1.4</b> What does a gene look like?</a></li>
</ul></li>
<li class="chapter" data-level="1.2" data-path="elements-of-gene-regulation.html"><a href="elements-of-gene-regulation.html"><i class="fa fa-check"></i><b>1.2</b> Elements of gene regulation</a><ul>
<li class="chapter" data-level="1.2.1" data-path="elements-of-gene-regulation.html"><a href="elements-of-gene-regulation.html#transcriptional-regulation"><i class="fa fa-check"></i><b>1.2.1</b> Transcriptional regulation</a></li>
<li class="chapter" data-level="1.2.2" data-path="elements-of-gene-regulation.html"><a href="elements-of-gene-regulation.html#post-transcriptional-regulation"><i class="fa fa-check"></i><b>1.2.2</b> Post-transcriptional regulation</a></li>
</ul></li>
<li class="chapter" data-level="1.3" data-path="shaping-the-genome-dna-mutation.html"><a href="shaping-the-genome-dna-mutation.html"><i class="fa fa-check"></i><b>1.3</b> Shaping the genome: DNA mutation</a></li>
<li class="chapter" data-level="1.4" data-path="high-throughput-experimental-methods-in-genomics.html"><a href="high-throughput-experimental-methods-in-genomics.html"><i class="fa fa-check"></i><b>1.4</b> High-throughput experimental methods in genomics</a><ul>
<li class="chapter" data-level="1.4.1" data-path="high-throughput-experimental-methods-in-genomics.html"><a href="high-throughput-experimental-methods-in-genomics.html#the-general-idea-behind-high-throughput-techniques"><i class="fa fa-check"></i><b>1.4.1</b> The general idea behind high-throughput techniques</a></li>
<li class="chapter" data-level="1.4.2" data-path="high-throughput-experimental-methods-in-genomics.html"><a href="high-throughput-experimental-methods-in-genomics.html#high-throughput-sequencing"><i class="fa fa-check"></i><b>1.4.2</b> High-throughput sequencing</a></li>
</ul></li>
<li class="chapter" data-level="1.5" data-path="visualization-and-data-repositories-for-genomics.html"><a href="visualization-and-data-repositories-for-genomics.html"><i class="fa fa-check"></i><b>1.5</b> Visualization and data repositories for genomics</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="Rintro.html"><a href="Rintro.html"><i class="fa fa-check"></i><b>2</b> Introduction to R for Genomic Data Analysis</a><ul>
<li class="chapter" data-level="2.1" data-path="steps-of-genomic-data-analysis.html"><a href="steps-of-genomic-data-analysis.html"><i class="fa fa-check"></i><b>2.1</b> Steps of (genomic) data analysis</a><ul>
<li class="chapter" data-level="2.1.1" data-path="steps-of-genomic-data-analysis.html"><a href="steps-of-genomic-data-analysis.html#data-collection"><i class="fa fa-check"></i><b>2.1.1</b> Data collection</a></li>
<li class="chapter" data-level="2.1.2" data-path="steps-of-genomic-data-analysis.html"><a href="steps-of-genomic-data-analysis.html#data-quality-check-and-cleaning"><i class="fa fa-check"></i><b>2.1.2</b> Data quality check and cleaning</a></li>
<li class="chapter" data-level="2.1.3" data-path="steps-of-genomic-data-analysis.html"><a href="steps-of-genomic-data-analysis.html#data-processing"><i class="fa fa-check"></i><b>2.1.3</b> Data processing</a></li>
<li class="chapter" data-level="2.1.4" data-path="steps-of-genomic-data-analysis.html"><a href="steps-of-genomic-data-analysis.html#exploratory-data-analysis-and-modeling"><i class="fa fa-check"></i><b>2.1.4</b> Exploratory data analysis and modeling</a></li>
<li class="chapter" data-level="2.1.5" data-path="steps-of-genomic-data-analysis.html"><a href="steps-of-genomic-data-analysis.html#visualization-and-reporting"><i class="fa fa-check"></i><b>2.1.5</b> Visualization and reporting</a></li>
<li class="chapter" data-level="2.1.6" data-path="steps-of-genomic-data-analysis.html"><a href="steps-of-genomic-data-analysis.html#why-use-r-for-genomics"><i class="fa fa-check"></i><b>2.1.6</b> Why use R for genomics ?</a></li>
</ul></li>
<li class="chapter" data-level="2.2" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html"><i class="fa fa-check"></i><b>2.2</b> Getting started with R</a><ul>
<li class="chapter" data-level="2.2.1" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#installing-packages"><i class="fa fa-check"></i><b>2.2.1</b> Installing packages</a></li>
<li class="chapter" data-level="2.2.2" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#installing-packages-in-custom-locations"><i class="fa fa-check"></i><b>2.2.2</b> Installing packages in custom locations</a></li>
<li class="chapter" data-level="2.2.3" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#getting-help-on-functions-and-packages"><i class="fa fa-check"></i><b>2.2.3</b> Getting help on functions and packages</a></li>
</ul></li>
<li class="chapter" data-level="2.3" data-path="computations-in-r.html"><a href="computations-in-r.html"><i class="fa fa-check"></i><b>2.3</b> Computations in R</a></li>
<li class="chapter" data-level="2.4" data-path="data-structures.html"><a href="data-structures.html"><i class="fa fa-check"></i><b>2.4</b> Data structures</a><ul>
<li class="chapter" data-level="2.4.1" data-path="data-structures.html"><a href="data-structures.html#vectors"><i class="fa fa-check"></i><b>2.4.1</b> Vectors</a></li>
<li class="chapter" data-level="2.4.2" data-path="data-structures.html"><a href="data-structures.html#matrices"><i class="fa fa-check"></i><b>2.4.2</b> Matrices</a></li>
<li class="chapter" data-level="2.4.3" data-path="data-structures.html"><a href="data-structures.html#data-frames"><i class="fa fa-check"></i><b>2.4.3</b> Data frames</a></li>
<li class="chapter" data-level="2.4.4" data-path="data-structures.html"><a href="data-structures.html#lists"><i class="fa fa-check"></i><b>2.4.4</b> Lists</a></li>
<li class="chapter" data-level="2.4.5" data-path="data-structures.html"><a href="data-structures.html#factors"><i class="fa fa-check"></i><b>2.4.5</b> Factors</a></li>
</ul></li>
<li class="chapter" data-level="2.5" data-path="data-types.html"><a href="data-types.html"><i class="fa fa-check"></i><b>2.5</b> Data types</a></li>
<li class="chapter" data-level="2.6" data-path="reading-and-writing-data.html"><a href="reading-and-writing-data.html"><i class="fa fa-check"></i><b>2.6</b> Reading and writing data</a><ul>
<li class="chapter" data-level="2.6.1" data-path="reading-and-writing-data.html"><a href="reading-and-writing-data.html#reading-large-files"><i class="fa fa-check"></i><b>2.6.1</b> Reading large files</a></li>
</ul></li>
<li class="chapter" data-level="2.7" data-path="plotting-in-r-with-base-graphics.html"><a href="plotting-in-r-with-base-graphics.html"><i class="fa fa-check"></i><b>2.7</b> Plotting in R with base graphics</a><ul>
<li class="chapter" data-level="2.7.1" data-path="plotting-in-r-with-base-graphics.html"><a href="plotting-in-r-with-base-graphics.html#combining-multiple-plots"><i class="fa fa-check"></i><b>2.7.1</b> Combining multiple plots</a></li>
<li class="chapter" data-level="2.7.2" data-path="plotting-in-r-with-base-graphics.html"><a href="plotting-in-r-with-base-graphics.html#saving-plots"><i class="fa fa-check"></i><b>2.7.2</b> Saving plots</a></li>
</ul></li>
<li class="chapter" data-level="2.8" data-path="plotting-in-r-with-ggplot2.html"><a href="plotting-in-r-with-ggplot2.html"><i class="fa fa-check"></i><b>2.8</b> Plotting in R with ggplot2</a><ul>
<li class="chapter" data-level="2.8.1" data-path="plotting-in-r-with-ggplot2.html"><a href="plotting-in-r-with-ggplot2.html#combining-multiple-plots-1"><i class="fa fa-check"></i><b>2.8.1</b> Combining multiple plots</a></li>
<li class="chapter" data-level="2.8.2" data-path="plotting-in-r-with-ggplot2.html"><a href="plotting-in-r-with-ggplot2.html#ggplot2-and-tidyverse"><i class="fa fa-check"></i><b>2.8.2</b> ggplot2 and tidyverse</a></li>
</ul></li>
<li class="chapter" data-level="2.9" data-path="functions-and-control-structures-for-ifelse-etc-.html"><a href="functions-and-control-structures-for-ifelse-etc-.html"><i class="fa fa-check"></i><b>2.9</b> Functions and control structures (for, if/else etc.)</a><ul>
<li class="chapter" data-level="2.9.1" data-path="functions-and-control-structures-for-ifelse-etc-.html"><a href="functions-and-control-structures-for-ifelse-etc-.html#user-defined-functions"><i class="fa fa-check"></i><b>2.9.1</b> User-defined functions</a></li>
<li class="chapter" data-level="2.9.2" data-path="functions-and-control-structures-for-ifelse-etc-.html"><a href="functions-and-control-structures-for-ifelse-etc-.html#loops-and-looping-structures-in-r"><i class="fa fa-check"></i><b>2.9.2</b> Loops and looping structures in R</a></li>
</ul></li>
<li class="chapter" data-level="2.10" data-path="exercises.html"><a href="exercises.html"><i class="fa fa-check"></i><b>2.10</b> Exercises</a><ul>
<li class="chapter" data-level="2.10.1" data-path="exercises.html"><a href="exercises.html#computations-in-r-1"><i class="fa fa-check"></i><b>2.10.1</b> Computations in R</a></li>
<li class="chapter" data-level="2.10.2" data-path="exercises.html"><a href="exercises.html#data-structures-in-r"><i class="fa fa-check"></i><b>2.10.2</b> Data structures in R</a></li>
<li class="chapter" data-level="2.10.3" data-path="exercises.html"><a href="exercises.html#reading-in-and-writing-data-out-in-r"><i class="fa fa-check"></i><b>2.10.3</b> Reading in and writing data out in R</a></li>
<li class="chapter" data-level="2.10.4" data-path="exercises.html"><a href="exercises.html#plotting-in-r"><i class="fa fa-check"></i><b>2.10.4</b> Plotting in R</a></li>
<li class="chapter" data-level="2.10.5" data-path="exercises.html"><a href="exercises.html#functions-and-control-structures-for-ifelse-etc.-1"><i class="fa fa-check"></i><b>2.10.5</b> Functions and control structures (for, if/else, etc.)</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="3" data-path="stats.html"><a href="stats.html"><i class="fa fa-check"></i><b>3</b> Statistics for Genomics</a><ul>
<li class="chapter" data-level="3.1" data-path="how-to-summarize-collection-of-data-points-the-idea-behind-statistical-distributions.html"><a href="how-to-summarize-collection-of-data-points-the-idea-behind-statistical-distributions.html"><i class="fa fa-check"></i><b>3.1</b> How to summarize collection of data points: The idea behind statistical distributions</a><ul>
<li class="chapter" data-level="3.1.1" data-path="how-to-summarize-collection-of-data-points-the-idea-behind-statistical-distributions.html"><a href="how-to-summarize-collection-of-data-points-the-idea-behind-statistical-distributions.html#describing-the-central-tendency-mean-and-median"><i class="fa fa-check"></i><b>3.1.1</b> Describing the central tendency: Mean and median</a></li>
<li class="chapter" data-level="3.1.2" data-path="how-to-summarize-collection-of-data-points-the-idea-behind-statistical-distributions.html"><a href="how-to-summarize-collection-of-data-points-the-idea-behind-statistical-distributions.html#describing-the-spread-measurements-of-variation"><i class="fa fa-check"></i><b>3.1.2</b> Describing the spread: Measurements of variation</a></li>
<li class="chapter" data-level="3.1.3" data-path="how-to-summarize-collection-of-data-points-the-idea-behind-statistical-distributions.html"><a href="how-to-summarize-collection-of-data-points-the-idea-behind-statistical-distributions.html#precision-of-estimates-confidence-intervals"><i class="fa fa-check"></i><b>3.1.3</b> Precision of estimates: Confidence intervals</a></li>
</ul></li>
<li class="chapter" data-level="3.2" data-path="how-to-test-for-differences-between-samples.html"><a href="how-to-test-for-differences-between-samples.html"><i class="fa fa-check"></i><b>3.2</b> How to test for differences between samples</a><ul>
<li class="chapter" data-level="3.2.1" data-path="how-to-test-for-differences-between-samples.html"><a href="how-to-test-for-differences-between-samples.html#randomization-based-testing-for-difference-of-the-means"><i class="fa fa-check"></i><b>3.2.1</b> Randomization-based testing for difference of the means</a></li>
<li class="chapter" data-level="3.2.2" data-path="how-to-test-for-differences-between-samples.html"><a href="how-to-test-for-differences-between-samples.html#using-t-test-for-difference-of-the-means-between-two-samples"><i class="fa fa-check"></i><b>3.2.2</b> Using t-test for difference of the means between two samples</a></li>
<li class="chapter" data-level="3.2.3" data-path="how-to-test-for-differences-between-samples.html"><a href="how-to-test-for-differences-between-samples.html#multiple-testing-correction"><i class="fa fa-check"></i><b>3.2.3</b> Multiple testing correction</a></li>
<li class="chapter" data-level="3.2.4" data-path="how-to-test-for-differences-between-samples.html"><a href="how-to-test-for-differences-between-samples.html#moderated-t-tests-using-information-from-multiple-comparisons"><i class="fa fa-check"></i><b>3.2.4</b> Moderated t-tests: Using information from multiple comparisons</a></li>
</ul></li>
<li class="chapter" data-level="3.3" data-path="relationship-between-variables-linear-models-and-correlation.html"><a href="relationship-between-variables-linear-models-and-correlation.html"><i class="fa fa-check"></i><b>3.3</b> Relationship between variables: Linear models and correlation</a><ul>
<li class="chapter" data-level="3.3.1" data-path="relationship-between-variables-linear-models-and-correlation.html"><a href="relationship-between-variables-linear-models-and-correlation.html#how-to-fit-a-line"><i class="fa fa-check"></i><b>3.3.1</b> How to fit a line</a></li>
<li class="chapter" data-level="3.3.2" data-path="relationship-between-variables-linear-models-and-correlation.html"><a href="relationship-between-variables-linear-models-and-correlation.html#how-to-estimate-the-error-of-the-coefficients"><i class="fa fa-check"></i><b>3.3.2</b> How to estimate the error of the coefficients</a></li>
<li class="chapter" data-level="3.3.3" data-path="relationship-between-variables-linear-models-and-correlation.html"><a href="relationship-between-variables-linear-models-and-correlation.html#accuracy-of-the-model"><i class="fa fa-check"></i><b>3.3.3</b> Accuracy of the model</a></li>
<li class="chapter" data-level="3.3.4" data-path="relationship-between-variables-linear-models-and-correlation.html"><a href="relationship-between-variables-linear-models-and-correlation.html#regression-with-categorical-variables"><i class="fa fa-check"></i><b>3.3.4</b> Regression with categorical variables</a></li>
<li class="chapter" data-level="3.3.5" data-path="relationship-between-variables-linear-models-and-correlation.html"><a href="relationship-between-variables-linear-models-and-correlation.html#regression-pitfalls"><i class="fa fa-check"></i><b>3.3.5</b> Regression pitfalls</a></li>
</ul></li>
<li class="chapter" data-level="3.4" data-path="exercises-1.html"><a href="exercises-1.html"><i class="fa fa-check"></i><b>3.4</b> Exercises</a><ul>
<li class="chapter" data-level="3.4.1" data-path="exercises-1.html"><a href="exercises-1.html#how-to-summarize-collection-of-data-points-the-idea-behind-statistical-distributions-1"><i class="fa fa-check"></i><b>3.4.1</b> How to summarize collection of data points: The idea behind statistical distributions</a></li>
<li class="chapter" data-level="3.4.2" data-path="exercises-1.html"><a href="exercises-1.html#how-to-test-for-differences-in-samples"><i class="fa fa-check"></i><b>3.4.2</b> How to test for differences in samples</a></li>
<li class="chapter" data-level="3.4.3" data-path="exercises-1.html"><a href="exercises-1.html#relationship-between-variables-linear-models-and-correlation-1"><i class="fa fa-check"></i><b>3.4.3</b> Relationship between variables: Linear models and correlation</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="4" data-path="unsupervisedLearning.html"><a href="unsupervisedLearning.html"><i class="fa fa-check"></i><b>4</b> Exploratory Data Analysis with Unsupervised Machine Learning</a><ul>
<li class="chapter" data-level="4.1" data-path="clustering-grouping-samples-based-on-their-similarity.html"><a href="clustering-grouping-samples-based-on-their-similarity.html"><i class="fa fa-check"></i><b>4.1</b> Clustering: Grouping samples based on their similarity</a><ul>
<li class="chapter" data-level="4.1.1" data-path="clustering-grouping-samples-based-on-their-similarity.html"><a href="clustering-grouping-samples-based-on-their-similarity.html#distance-metrics"><i class="fa fa-check"></i><b>4.1.1</b> Distance metrics</a></li>
<li class="chapter" data-level="4.1.2" data-path="clustering-grouping-samples-based-on-their-similarity.html"><a href="clustering-grouping-samples-based-on-their-similarity.html#hiearchical-clustering"><i class="fa fa-check"></i><b>4.1.2</b> Hiearchical clustering</a></li>
<li class="chapter" data-level="4.1.3" data-path="clustering-grouping-samples-based-on-their-similarity.html"><a href="clustering-grouping-samples-based-on-their-similarity.html#k-means-clustering"><i class="fa fa-check"></i><b>4.1.3</b> K-means clustering</a></li>
<li class="chapter" data-level="4.1.4" data-path="clustering-grouping-samples-based-on-their-similarity.html"><a href="clustering-grouping-samples-based-on-their-similarity.html#how-to-choose-k-the-number-of-clusters"><i class="fa fa-check"></i><b>4.1.4</b> How to choose “k”, the number of clusters</a></li>
</ul></li>
<li class="chapter" data-level="4.2" data-path="dimensionality-reduction-techniques-visualizing-complex-data-sets-in-2d.html"><a href="dimensionality-reduction-techniques-visualizing-complex-data-sets-in-2d.html"><i class="fa fa-check"></i><b>4.2</b> Dimensionality reduction techniques: Visualizing complex data sets in 2D</a><ul>
<li class="chapter" data-level="4.2.1" data-path="dimensionality-reduction-techniques-visualizing-complex-data-sets-in-2d.html"><a href="dimensionality-reduction-techniques-visualizing-complex-data-sets-in-2d.html#principal-component-analysis"><i class="fa fa-check"></i><b>4.2.1</b> Principal component analysis</a></li>
<li class="chapter" data-level="4.2.2" data-path="dimensionality-reduction-techniques-visualizing-complex-data-sets-in-2d.html"><a href="dimensionality-reduction-techniques-visualizing-complex-data-sets-in-2d.html#other-matrix-factorization-methods-for-dimensionality-reduction"><i class="fa fa-check"></i><b>4.2.2</b> Other matrix factorization methods for dimensionality reduction</a></li>
<li class="chapter" data-level="4.2.3" data-path="dimensionality-reduction-techniques-visualizing-complex-data-sets-in-2d.html"><a href="dimensionality-reduction-techniques-visualizing-complex-data-sets-in-2d.html#multi-dimensional-scaling"><i class="fa fa-check"></i><b>4.2.3</b> Multi-dimensional scaling</a></li>
<li class="chapter" data-level="4.2.4" data-path="dimensionality-reduction-techniques-visualizing-complex-data-sets-in-2d.html"><a href="dimensionality-reduction-techniques-visualizing-complex-data-sets-in-2d.html#t-distributed-stochastic-neighbor-embedding-t-sne"><i class="fa fa-check"></i><b>4.2.4</b> t-Distributed Stochastic Neighbor Embedding (t-SNE)</a></li>
</ul></li>
<li class="chapter" data-level="4.3" data-path="exercises-2.html"><a href="exercises-2.html"><i class="fa fa-check"></i><b>4.3</b> Exercises</a><ul>
<li class="chapter" data-level="4.3.1" data-path="exercises-2.html"><a href="exercises-2.html#clustering"><i class="fa fa-check"></i><b>4.3.1</b> Clustering</a></li>
<li class="chapter" data-level="4.3.2" data-path="exercises-2.html"><a href="exercises-2.html#dimension-reduction"><i class="fa fa-check"></i><b>4.3.2</b> Dimension reduction</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="5" data-path="supervisedLearning.html"><a href="supervisedLearning.html"><i class="fa fa-check"></i><b>5</b> Predictive Modeling with Supervised Machine Learning</a><ul>
<li class="chapter" data-level="5.1" data-path="how-are-machine-learning-models-fit.html"><a href="how-are-machine-learning-models-fit.html"><i class="fa fa-check"></i><b>5.1</b> How are machine learning models fit?</a><ul>
<li class="chapter" data-level="5.1.1" data-path="how-are-machine-learning-models-fit.html"><a href="how-are-machine-learning-models-fit.html#machine-learning-vs.-statistics"><i class="fa fa-check"></i><b>5.1.1</b> Machine learning vs. statistics</a></li>
</ul></li>
<li class="chapter" data-level="5.2" data-path="steps-in-supervised-machine-learning.html"><a href="steps-in-supervised-machine-learning.html"><i class="fa fa-check"></i><b>5.2</b> Steps in supervised machine learning</a></li>
<li class="chapter" data-level="5.3" data-path="use-case-disease-subtype-from-genomics-data.html"><a href="use-case-disease-subtype-from-genomics-data.html"><i class="fa fa-check"></i><b>5.3</b> Use case: Disease subtype from genomics data</a></li>
<li class="chapter" data-level="5.4" data-path="data-preprocessing.html"><a href="data-preprocessing.html"><i class="fa fa-check"></i><b>5.4</b> Data preprocessing</a><ul>
<li class="chapter" data-level="5.4.1" data-path="data-preprocessing.html"><a href="data-preprocessing.html#data-transformation"><i class="fa fa-check"></i><b>5.4.1</b> Data transformation</a></li>
<li class="chapter" data-level="5.4.2" data-path="data-preprocessing.html"><a href="data-preprocessing.html#filtering-data-and-scaling"><i class="fa fa-check"></i><b>5.4.2</b> Filtering data and scaling</a></li>
<li class="chapter" data-level="5.4.3" data-path="data-preprocessing.html"><a href="data-preprocessing.html#dealing-with-missing-values"><i class="fa fa-check"></i><b>5.4.3</b> Dealing with missing values</a></li>
</ul></li>
<li class="chapter" data-level="5.5" data-path="splitting-the-data.html"><a href="splitting-the-data.html"><i class="fa fa-check"></i><b>5.5</b> Splitting the data</a><ul>
<li class="chapter" data-level="5.5.1" data-path="splitting-the-data.html"><a href="splitting-the-data.html#holdout-test-dataset"><i class="fa fa-check"></i><b>5.5.1</b> Holdout test dataset</a></li>
<li class="chapter" data-level="5.5.2" data-path="splitting-the-data.html"><a href="splitting-the-data.html#cross-validation"><i class="fa fa-check"></i><b>5.5.2</b> Cross-validation</a></li>
<li class="chapter" data-level="5.5.3" data-path="splitting-the-data.html"><a href="splitting-the-data.html#bootstrap-resampling"><i class="fa fa-check"></i><b>5.5.3</b> Bootstrap resampling</a></li>
</ul></li>
<li class="chapter" data-level="5.6" data-path="predicting-the-subtype-with-k-nearest-neighbors.html"><a href="predicting-the-subtype-with-k-nearest-neighbors.html"><i class="fa fa-check"></i><b>5.6</b> Predicting the subtype with k-nearest neighbors</a></li>
<li class="chapter" data-level="5.7" data-path="assessing-the-performance-of-our-model.html"><a href="assessing-the-performance-of-our-model.html"><i class="fa fa-check"></i><b>5.7</b> Assessing the performance of our model</a><ul>
<li class="chapter" data-level="5.7.1" data-path="assessing-the-performance-of-our-model.html"><a href="assessing-the-performance-of-our-model.html#receiver-operating-characteristic-roc-curves"><i class="fa fa-check"></i><b>5.7.1</b> Receiver Operating Characteristic (ROC) curves</a></li>
</ul></li>
<li class="chapter" data-level="5.8" data-path="model-tuning-and-avoiding-overfitting.html"><a href="model-tuning-and-avoiding-overfitting.html"><i class="fa fa-check"></i><b>5.8</b> Model tuning and avoiding overfitting</a><ul>
<li class="chapter" data-level="5.8.1" data-path="model-tuning-and-avoiding-overfitting.html"><a href="model-tuning-and-avoiding-overfitting.html#model-complexity-and-bias-variance-trade-off"><i class="fa fa-check"></i><b>5.8.1</b> Model complexity and bias variance trade-off</a></li>
<li class="chapter" data-level="5.8.2" data-path="model-tuning-and-avoiding-overfitting.html"><a href="model-tuning-and-avoiding-overfitting.html#data-split-strategies-for-model-tuning-and-testing"><i class="fa fa-check"></i><b>5.8.2</b> Data split strategies for model tuning and testing</a></li>
</ul></li>
<li class="chapter" data-level="5.9" data-path="variable-importance.html"><a href="variable-importance.html"><i class="fa fa-check"></i><b>5.9</b> Variable importance</a></li>
<li class="chapter" data-level="5.10" data-path="how-to-deal-with-class-imbalance.html"><a href="how-to-deal-with-class-imbalance.html"><i class="fa fa-check"></i><b>5.10</b> How to deal with class imbalance</a><ul>
<li class="chapter" data-level="5.10.1" data-path="how-to-deal-with-class-imbalance.html"><a href="how-to-deal-with-class-imbalance.html#sampling-for-class-balance"><i class="fa fa-check"></i><b>5.10.1</b> Sampling for class balance</a></li>
<li class="chapter" data-level="5.10.2" data-path="how-to-deal-with-class-imbalance.html"><a href="how-to-deal-with-class-imbalance.html#altering-case-weights"><i class="fa fa-check"></i><b>5.10.2</b> Altering case weights</a></li>
<li class="chapter" data-level="5.10.3" data-path="how-to-deal-with-class-imbalance.html"><a href="how-to-deal-with-class-imbalance.html#selecting-different-classification-score-cutoffs"><i class="fa fa-check"></i><b>5.10.3</b> Selecting different classification score cutoffs</a></li>
</ul></li>
<li class="chapter" data-level="5.11" data-path="dealing-with-correlated-predictors.html"><a href="dealing-with-correlated-predictors.html"><i class="fa fa-check"></i><b>5.11</b> Dealing with correlated predictors</a></li>
<li class="chapter" data-level="5.12" data-path="trees-and-forests-random-forests-in-action.html"><a href="trees-and-forests-random-forests-in-action.html"><i class="fa fa-check"></i><b>5.12</b> Trees and forests: Random forests in action</a><ul>
<li class="chapter" data-level="5.12.1" data-path="trees-and-forests-random-forests-in-action.html"><a href="trees-and-forests-random-forests-in-action.html#decision-trees"><i class="fa fa-check"></i><b>5.12.1</b> Decision trees</a></li>
<li class="chapter" data-level="5.12.2" data-path="trees-and-forests-random-forests-in-action.html"><a href="trees-and-forests-random-forests-in-action.html#trees-to-forests"><i class="fa fa-check"></i><b>5.12.2</b> Trees to forests</a></li>
<li class="chapter" data-level="5.12.3" data-path="trees-and-forests-random-forests-in-action.html"><a href="trees-and-forests-random-forests-in-action.html#variable-importance-1"><i class="fa fa-check"></i><b>5.12.3</b> Variable importance</a></li>
</ul></li>
<li class="chapter" data-level="5.13" data-path="logistic-regression-and-regularization.html"><a href="logistic-regression-and-regularization.html"><i class="fa fa-check"></i><b>5.13</b> Logistic regression and regularization</a><ul>
<li class="chapter" data-level="5.13.1" data-path="logistic-regression-and-regularization.html"><a href="logistic-regression-and-regularization.html#regularization-in-order-to-avoid-overfitting"><i class="fa fa-check"></i><b>5.13.1</b> Regularization in order to avoid overfitting</a></li>
<li class="chapter" data-level="5.13.2" data-path="logistic-regression-and-regularization.html"><a href="logistic-regression-and-regularization.html#variable-importance-2"><i class="fa fa-check"></i><b>5.13.2</b> Variable importance</a></li>
</ul></li>
<li class="chapter" data-level="5.14" data-path="other-supervised-algorithms.html"><a href="other-supervised-algorithms.html"><i class="fa fa-check"></i><b>5.14</b> Other supervised algorithms</a><ul>
<li class="chapter" data-level="5.14.1" data-path="other-supervised-algorithms.html"><a href="other-supervised-algorithms.html#gradient-boosting"><i class="fa fa-check"></i><b>5.14.1</b> Gradient boosting</a></li>
<li class="chapter" data-level="5.14.2" data-path="other-supervised-algorithms.html"><a href="other-supervised-algorithms.html#support-vector-machines-svm"><i class="fa fa-check"></i><b>5.14.2</b> Support Vector Machines (SVM)</a></li>
<li class="chapter" data-level="5.14.3" data-path="other-supervised-algorithms.html"><a href="other-supervised-algorithms.html#neural-networks-and-deep-versions-of-it"><i class="fa fa-check"></i><b>5.14.3</b> Neural networks and deep versions of it</a></li>
<li class="chapter" data-level="5.14.4" data-path="other-supervised-algorithms.html"><a href="other-supervised-algorithms.html#ensemble-learning"><i class="fa fa-check"></i><b>5.14.4</b> Ensemble learning</a></li>
</ul></li>
<li class="chapter" data-level="5.15" data-path="predicting-continuous-variables-regression-with-machine-learning.html"><a href="predicting-continuous-variables-regression-with-machine-learning.html"><i class="fa fa-check"></i><b>5.15</b> Predicting continuous variables: Regression with machine learning</a><ul>
<li class="chapter" data-level="5.15.1" data-path="predicting-continuous-variables-regression-with-machine-learning.html"><a href="predicting-continuous-variables-regression-with-machine-learning.html#use-case-predicting-age-from-dna-methylation"><i class="fa fa-check"></i><b>5.15.1</b> Use case: Predicting age from DNA methylation</a></li>
<li class="chapter" data-level="5.15.2" data-path="predicting-continuous-variables-regression-with-machine-learning.html"><a href="predicting-continuous-variables-regression-with-machine-learning.html#reading-and-processing-the-data"><i class="fa fa-check"></i><b>5.15.2</b> Reading and processing the data</a></li>
<li class="chapter" data-level="5.15.3" data-path="predicting-continuous-variables-regression-with-machine-learning.html"><a href="predicting-continuous-variables-regression-with-machine-learning.html#running-random-forest-regression"><i class="fa fa-check"></i><b>5.15.3</b> Running random forest regression</a></li>
</ul></li>
<li class="chapter" data-level="5.16" data-path="exercises-3.html"><a href="exercises-3.html"><i class="fa fa-check"></i><b>5.16</b> Exercises</a><ul>
<li class="chapter" data-level="5.16.1" data-path="exercises-3.html"><a href="exercises-3.html#classification"><i class="fa fa-check"></i><b>5.16.1</b> Classification</a></li>
<li class="chapter" data-level="5.16.2" data-path="exercises-3.html"><a href="exercises-3.html#regression"><i class="fa fa-check"></i><b>5.16.2</b> Regression</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="6" data-path="genomicIntervals.html"><a href="genomicIntervals.html"><i class="fa fa-check"></i><b>6</b> Operations on Genomic Intervals and Genome Arithmetic</a><ul>
<li class="chapter" data-level="6.1" data-path="operations-on-genomic-intervals-with-genomicranges-package.html"><a href="operations-on-genomic-intervals-with-genomicranges-package.html"><i class="fa fa-check"></i><b>6.1</b> Operations on genomic intervals with <code>GenomicRanges</code> package</a><ul>
<li class="chapter" data-level="6.1.1" data-path="operations-on-genomic-intervals-with-genomicranges-package.html"><a href="operations-on-genomic-intervals-with-genomicranges-package.html#how-to-create-and-manipulate-a-granges-object"><i class="fa fa-check"></i><b>6.1.1</b> How to create and manipulate a GRanges object</a></li>
<li class="chapter" data-level="6.1.2" data-path="operations-on-genomic-intervals-with-genomicranges-package.html"><a href="operations-on-genomic-intervals-with-genomicranges-package.html#getting-genomic-regions-into-r-as-granges-objects"><i class="fa fa-check"></i><b>6.1.2</b> Getting genomic regions into R as GRanges objects</a></li>
<li class="chapter" data-level="6.1.3" data-path="operations-on-genomic-intervals-with-genomicranges-package.html"><a href="operations-on-genomic-intervals-with-genomicranges-package.html#finding-regions-that-dodo-not-overlap-with-another-set-of-regions"><i class="fa fa-check"></i><b>6.1.3</b> Finding regions that do/do not overlap with another set of regions</a></li>
</ul></li>
<li class="chapter" data-level="6.2" data-path="dealing-with-mapped-high-throughput-sequencing-reads.html"><a href="dealing-with-mapped-high-throughput-sequencing-reads.html"><i class="fa fa-check"></i><b>6.2</b> Dealing with mapped high-throughput sequencing reads</a><ul>
<li class="chapter" data-level="6.2.1" data-path="dealing-with-mapped-high-throughput-sequencing-reads.html"><a href="dealing-with-mapped-high-throughput-sequencing-reads.html#counting-mapped-reads-for-a-set-of-regions"><i class="fa fa-check"></i><b>6.2.1</b> Counting mapped reads for a set of regions</a></li>
</ul></li>
<li class="chapter" data-level="6.3" data-path="dealing-with-continuous-scores-over-the-genome.html"><a href="dealing-with-continuous-scores-over-the-genome.html"><i class="fa fa-check"></i><b>6.3</b> Dealing with continuous scores over the genome</a><ul>
<li class="chapter" data-level="6.3.1" data-path="dealing-with-continuous-scores-over-the-genome.html"><a href="dealing-with-continuous-scores-over-the-genome.html#extracting-subsections-of-rle-and-rlelist-objects"><i class="fa fa-check"></i><b>6.3.1</b> Extracting subsections of Rle and RleList objects</a></li>
</ul></li>
<li class="chapter" data-level="6.4" data-path="genomic-intervals-with-more-information-summarizedexperiment-class.html"><a href="genomic-intervals-with-more-information-summarizedexperiment-class.html"><i class="fa fa-check"></i><b>6.4</b> Genomic intervals with more information: SummarizedExperiment class</a><ul>
<li class="chapter" data-level="6.4.1" data-path="genomic-intervals-with-more-information-summarizedexperiment-class.html"><a href="genomic-intervals-with-more-information-summarizedexperiment-class.html#create-a-summarizedexperiment-object"><i class="fa fa-check"></i><b>6.4.1</b> Create a SummarizedExperiment object</a></li>
<li class="chapter" data-level="6.4.2" data-path="genomic-intervals-with-more-information-summarizedexperiment-class.html"><a href="genomic-intervals-with-more-information-summarizedexperiment-class.html#subset-and-manipulate-the-summarizedexperiment-object"><i class="fa fa-check"></i><b>6.4.2</b> Subset and manipulate the SummarizedExperiment object</a></li>
</ul></li>
<li class="chapter" data-level="6.5" data-path="visualizing-and-summarizing-genomic-intervals.html"><a href="visualizing-and-summarizing-genomic-intervals.html"><i class="fa fa-check"></i><b>6.5</b> Visualizing and summarizing genomic intervals</a><ul>
<li class="chapter" data-level="6.5.1" data-path="visualizing-and-summarizing-genomic-intervals.html"><a href="visualizing-and-summarizing-genomic-intervals.html#visualizing-intervals-on-a-locus-of-interest"><i class="fa fa-check"></i><b>6.5.1</b> Visualizing intervals on a locus of interest</a></li>
<li class="chapter" data-level="6.5.2" data-path="visualizing-and-summarizing-genomic-intervals.html"><a href="visualizing-and-summarizing-genomic-intervals.html#summaries-of-genomic-intervals-on-multiple-loci"><i class="fa fa-check"></i><b>6.5.2</b> Summaries of genomic intervals on multiple loci</a></li>
<li class="chapter" data-level="6.5.3" data-path="visualizing-and-summarizing-genomic-intervals.html"><a href="visualizing-and-summarizing-genomic-intervals.html#making-karyograms-and-circos-plots"><i class="fa fa-check"></i><b>6.5.3</b> Making karyograms and circos plots</a></li>
</ul></li>
<li class="chapter" data-level="6.6" data-path="exercises-4.html"><a href="exercises-4.html"><i class="fa fa-check"></i><b>6.6</b> Exercises</a><ul>
<li class="chapter" data-level="6.6.1" data-path="exercises-4.html"><a href="exercises-4.html#operations-on-genomic-intervals-with-the-genomicranges-package"><i class="fa fa-check"></i><b>6.6.1</b> Operations on genomic intervals with the <code>GenomicRanges</code> package</a></li>
<li class="chapter" data-level="6.6.2" data-path="exercises-4.html"><a href="exercises-4.html#dealing-with-mapped-high-throughput-sequencing-reads-1"><i class="fa fa-check"></i><b>6.6.2</b> Dealing with mapped high-throughput sequencing reads</a></li>
<li class="chapter" data-level="6.6.3" data-path="exercises-4.html"><a href="exercises-4.html#dealing-with-contiguous-scores-over-the-genome"><i class="fa fa-check"></i><b>6.6.3</b> Dealing with contiguous scores over the genome</a></li>
<li class="chapter" data-level="6.6.4" data-path="exercises-4.html"><a href="exercises-4.html#visualizing-and-summarizing-genomic-intervals-1"><i class="fa fa-check"></i><b>6.6.4</b> Visualizing and summarizing genomic intervals</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="7" data-path="processingReads.html"><a href="processingReads.html"><i class="fa fa-check"></i><b>7</b> Quality Check, Processing and Alignment of High-throughput Sequencing Reads</a><ul>
<li class="chapter" data-level="7.1" data-path="fasta-and-fastq-formats.html"><a href="fasta-and-fastq-formats.html"><i class="fa fa-check"></i><b>7.1</b> FASTA and FASTQ formats</a></li>
<li class="chapter" data-level="7.2" data-path="quality-check-on-sequencing-reads.html"><a href="quality-check-on-sequencing-reads.html"><i class="fa fa-check"></i><b>7.2</b> Quality check on sequencing reads</a><ul>
<li class="chapter" data-level="7.2.1" data-path="quality-check-on-sequencing-reads.html"><a href="quality-check-on-sequencing-reads.html#sequence-quality-per-basecycle"><i class="fa fa-check"></i><b>7.2.1</b> Sequence quality per base/cycle</a></li>
<li class="chapter" data-level="7.2.2" data-path="quality-check-on-sequencing-reads.html"><a href="quality-check-on-sequencing-reads.html#sequence-content-per-basecycle"><i class="fa fa-check"></i><b>7.2.2</b> Sequence content per base/cycle</a></li>
<li class="chapter" data-level="7.2.3" data-path="quality-check-on-sequencing-reads.html"><a href="quality-check-on-sequencing-reads.html#read-frequency-plot"><i class="fa fa-check"></i><b>7.2.3</b> Read frequency plot</a></li>
<li class="chapter" data-level="7.2.4" data-path="quality-check-on-sequencing-reads.html"><a href="quality-check-on-sequencing-reads.html#other-quality-metrics-and-qc-tools"><i class="fa fa-check"></i><b>7.2.4</b> Other quality metrics and QC tools</a></li>
</ul></li>
<li class="chapter" data-level="7.3" data-path="filtering-and-trimming-reads.html"><a href="filtering-and-trimming-reads.html"><i class="fa fa-check"></i><b>7.3</b> Filtering and trimming reads</a></li>
<li class="chapter" data-level="7.4" data-path="mappingaligning-reads-to-the-genome.html"><a href="mappingaligning-reads-to-the-genome.html"><i class="fa fa-check"></i><b>7.4</b> Mapping/aligning reads to the genome</a></li>
<li class="chapter" data-level="7.5" data-path="further-processing-of-aligned-reads.html"><a href="further-processing-of-aligned-reads.html"><i class="fa fa-check"></i><b>7.5</b> Further processing of aligned reads</a></li>
<li class="chapter" data-level="7.6" data-path="exercises-5.html"><a href="exercises-5.html"><i class="fa fa-check"></i><b>7.6</b> Exercises</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="rnaseqanalysis.html"><a href="rnaseqanalysis.html"><i class="fa fa-check"></i><b>8</b> RNA-seq Analysis</a><ul>
<li class="chapter" data-level="8.1" data-path="what-is-gene-expression.html"><a href="what-is-gene-expression.html"><i class="fa fa-check"></i><b>8.1</b> What is gene expression?</a></li>
<li class="chapter" data-level="8.2" data-path="methods-to-detect-gene-expression.html"><a href="methods-to-detect-gene-expression.html"><i class="fa fa-check"></i><b>8.2</b> Methods to detect gene expression</a></li>
<li class="chapter" data-level="8.3" data-path="gene-expression-analysis-using-high-throughput-sequencing-technologies.html"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html"><i class="fa fa-check"></i><b>8.3</b> Gene expression analysis using high-throughput sequencing technologies</a><ul>
<li class="chapter" data-level="8.3.1" data-path="gene-expression-analysis-using-high-throughput-sequencing-technologies.html"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#processing-raw-data"><i class="fa fa-check"></i><b>8.3.1</b> Processing raw data</a></li>
<li class="chapter" data-level="8.3.2" data-path="gene-expression-analysis-using-high-throughput-sequencing-technologies.html"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#alignment"><i class="fa fa-check"></i><b>8.3.2</b> Alignment</a></li>
<li class="chapter" data-level="8.3.3" data-path="gene-expression-analysis-using-high-throughput-sequencing-technologies.html"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#quantification"><i class="fa fa-check"></i><b>8.3.3</b> Quantification</a></li>
<li class="chapter" data-level="8.3.4" data-path="gene-expression-analysis-using-high-throughput-sequencing-technologies.html"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#within-sample-normalization-of-the-read-counts"><i class="fa fa-check"></i><b>8.3.4</b> Within sample normalization of the read counts</a></li>
<li class="chapter" data-level="8.3.5" data-path="gene-expression-analysis-using-high-throughput-sequencing-technologies.html"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#computing-different-normalization-schemes-in-r"><i class="fa fa-check"></i><b>8.3.5</b> Computing different normalization schemes in R</a></li>
<li class="chapter" data-level="8.3.6" data-path="gene-expression-analysis-using-high-throughput-sequencing-technologies.html"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#exploratory-analysis-of-the-read-count-table"><i class="fa fa-check"></i><b>8.3.6</b> Exploratory analysis of the read count table</a></li>
<li class="chapter" data-level="8.3.7" data-path="gene-expression-analysis-using-high-throughput-sequencing-technologies.html"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#differential-expression-analysis"><i class="fa fa-check"></i><b>8.3.7</b> Differential expression analysis</a></li>
<li class="chapter" data-level="8.3.8" data-path="gene-expression-analysis-using-high-throughput-sequencing-technologies.html"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#functional-enrichment-analysis"><i class="fa fa-check"></i><b>8.3.8</b> Functional enrichment analysis</a></li>
<li class="chapter" data-level="8.3.9" data-path="gene-expression-analysis-using-high-throughput-sequencing-technologies.html"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#accounting-for-additional-sources-of-variation"><i class="fa fa-check"></i><b>8.3.9</b> Accounting for additional sources of variation</a></li>
</ul></li>
<li class="chapter" data-level="8.4" data-path="other-applications-of-rna-seq.html"><a href="other-applications-of-rna-seq.html"><i class="fa fa-check"></i><b>8.4</b> Other applications of RNA-seq</a></li>
<li class="chapter" data-level="8.5" data-path="exercises-6.html"><a href="exercises-6.html"><i class="fa fa-check"></i><b>8.5</b> Exercises</a><ul>
<li class="chapter" data-level="8.5.1" data-path="exercises-6.html"><a href="exercises-6.html#exploring-the-count-tables"><i class="fa fa-check"></i><b>8.5.1</b> Exploring the count tables</a></li>
<li class="chapter" data-level="8.5.2" data-path="exercises-6.html"><a href="exercises-6.html#differential-expression-analysis-1"><i class="fa fa-check"></i><b>8.5.2</b> Differential expression analysis</a></li>
<li class="chapter" data-level="8.5.3" data-path="exercises-6.html"><a href="exercises-6.html#functional-enrichment-analysis-1"><i class="fa fa-check"></i><b>8.5.3</b> Functional enrichment analysis</a></li>
<li class="chapter" data-level="8.5.4" data-path="exercises-6.html"><a href="exercises-6.html#removing-unwanted-variation-from-the-expression-data"><i class="fa fa-check"></i><b>8.5.4</b> Removing unwanted variation from the expression data</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="9" data-path="chipseq.html"><a href="chipseq.html"><i class="fa fa-check"></i><b>9</b> ChIP-seq analysis</a><ul>
<li class="chapter" data-level="9.1" data-path="regulatory-protein-dna-interactions.html"><a href="regulatory-protein-dna-interactions.html"><i class="fa fa-check"></i><b>9.1</b> Regulatory protein-DNA interactions</a></li>
<li class="chapter" data-level="9.2" data-path="measuring-protein-dna-interactions-with-chip-seq.html"><a href="measuring-protein-dna-interactions-with-chip-seq.html"><i class="fa fa-check"></i><b>9.2</b> Measuring protein-DNA interactions with ChIP-seq</a></li>
<li class="chapter" data-level="9.3" data-path="factors-that-affect-chip-seq-experiment-and-analysis-quality.html"><a href="factors-that-affect-chip-seq-experiment-and-analysis-quality.html"><i class="fa fa-check"></i><b>9.3</b> Factors that affect ChIP-seq experiment and analysis quality</a><ul>
<li class="chapter" data-level="9.3.1" data-path="factors-that-affect-chip-seq-experiment-and-analysis-quality.html"><a href="factors-that-affect-chip-seq-experiment-and-analysis-quality.html#antibody-specificity"><i class="fa fa-check"></i><b>9.3.1</b> Antibody specificity</a></li>
<li class="chapter" data-level="9.3.2" data-path="factors-that-affect-chip-seq-experiment-and-analysis-quality.html"><a href="factors-that-affect-chip-seq-experiment-and-analysis-quality.html#sequencing-depth"><i class="fa fa-check"></i><b>9.3.2</b> Sequencing depth</a></li>
<li class="chapter" data-level="9.3.3" data-path="factors-that-affect-chip-seq-experiment-and-analysis-quality.html"><a href="factors-that-affect-chip-seq-experiment-and-analysis-quality.html#pcr-duplication"><i class="fa fa-check"></i><b>9.3.3</b> PCR duplication</a></li>
<li class="chapter" data-level="9.3.4" data-path="factors-that-affect-chip-seq-experiment-and-analysis-quality.html"><a href="factors-that-affect-chip-seq-experiment-and-analysis-quality.html#biological-replicates"><i class="fa fa-check"></i><b>9.3.4</b> Biological replicates</a></li>
<li class="chapter" data-level="9.3.5" data-path="factors-that-affect-chip-seq-experiment-and-analysis-quality.html"><a href="factors-that-affect-chip-seq-experiment-and-analysis-quality.html#control-experiments"><i class="fa fa-check"></i><b>9.3.5</b> Control experiments</a></li>
<li class="chapter" data-level="9.3.6" data-path="factors-that-affect-chip-seq-experiment-and-analysis-quality.html"><a href="factors-that-affect-chip-seq-experiment-and-analysis-quality.html#using-tagged-proteins"><i class="fa fa-check"></i><b>9.3.6</b> Using tagged proteins</a></li>
</ul></li>
<li class="chapter" data-level="9.4" data-path="pre-processing-chip-data.html"><a href="pre-processing-chip-data.html"><i class="fa fa-check"></i><b>9.4</b> Pre-processing ChIP data</a><ul>
<li class="chapter" data-level="9.4.1" data-path="pre-processing-chip-data.html"><a href="pre-processing-chip-data.html#mapping-of-chip-seq-data"><i class="fa fa-check"></i><b>9.4.1</b> Mapping of ChIP-seq data</a></li>
</ul></li>
<li class="chapter" data-level="9.5" data-path="chip-quality-control.html"><a href="chip-quality-control.html"><i class="fa fa-check"></i><b>9.5</b> ChIP quality control</a><ul>
<li class="chapter" data-level="9.5.1" data-path="chip-quality-control.html"><a href="chip-quality-control.html#the-data"><i class="fa fa-check"></i><b>9.5.1</b> The data</a></li>
<li class="chapter" data-level="9.5.2" data-path="chip-quality-control.html"><a href="chip-quality-control.html#sample-clustering"><i class="fa fa-check"></i><b>9.5.2</b> Sample clustering</a></li>
<li class="chapter" data-level="9.5.3" data-path="chip-quality-control.html"><a href="chip-quality-control.html#visualization-in-the-genome-browser"><i class="fa fa-check"></i><b>9.5.3</b> Visualization in the genome browser</a></li>
<li class="chapter" data-level="9.5.4" data-path="chip-quality-control.html"><a href="chip-quality-control.html#plus-and-minus-strand-cross-correlation"><i class="fa fa-check"></i><b>9.5.4</b> Plus and minus strand cross-correlation</a></li>
<li class="chapter" data-level="9.5.5" data-path="chip-quality-control.html"><a href="chip-quality-control.html#gc-bias-quantification"><i class="fa fa-check"></i><b>9.5.5</b> GC bias quantification</a></li>
<li class="chapter" data-level="9.5.6" data-path="chip-quality-control.html"><a href="chip-quality-control.html#sequence-read-genomic-distribution"><i class="fa fa-check"></i><b>9.5.6</b> Sequence read genomic distribution</a></li>
</ul></li>
<li class="chapter" data-level="9.6" data-path="peak-calling.html"><a href="peak-calling.html"><i class="fa fa-check"></i><b>9.6</b> Peak calling</a><ul>
<li class="chapter" data-level="9.6.1" data-path="peak-calling.html"><a href="peak-calling.html#types-of-chip-seq-experiments"><i class="fa fa-check"></i><b>9.6.1</b> Types of ChIP-seq experiments</a></li>
<li class="chapter" data-level="9.6.2" data-path="peak-calling.html"><a href="peak-calling.html#peak-calling-sharp-peaks"><i class="fa fa-check"></i><b>9.6.2</b> Peak calling: Sharp peaks</a></li>
<li class="chapter" data-level="9.6.3" data-path="peak-calling.html"><a href="peak-calling.html#peak-calling-broad-regions"><i class="fa fa-check"></i><b>9.6.3</b> Peak calling: Broad regions</a></li>
<li class="chapter" data-level="9.6.4" data-path="peak-calling.html"><a href="peak-calling.html#peak-quality-control"><i class="fa fa-check"></i><b>9.6.4</b> Peak quality control</a></li>
<li class="chapter" data-level="9.6.5" data-path="peak-calling.html"><a href="peak-calling.html#peak-annotation"><i class="fa fa-check"></i><b>9.6.5</b> Peak annotation</a></li>
</ul></li>
<li class="chapter" data-level="9.7" data-path="motif-discovery.html"><a href="motif-discovery.html"><i class="fa fa-check"></i><b>9.7</b> Motif discovery</a><ul>
<li class="chapter" data-level="9.7.1" data-path="motif-discovery.html"><a href="motif-discovery.html#motif-comparison"><i class="fa fa-check"></i><b>9.7.1</b> Motif comparison</a></li>
</ul></li>
<li class="chapter" data-level="9.8" data-path="what-to-do-next.html"><a href="what-to-do-next.html"><i class="fa fa-check"></i><b>9.8</b> What to do next?</a></li>
<li class="chapter" data-level="9.9" data-path="exercises-7.html"><a href="exercises-7.html"><i class="fa fa-check"></i><b>9.9</b> Exercises</a><ul>
<li class="chapter" data-level="9.9.1" data-path="exercises-7.html"><a href="exercises-7.html#quality-control"><i class="fa fa-check"></i><b>9.9.1</b> Quality control</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="10" data-path="bsseq.html"><a href="bsseq.html"><i class="fa fa-check"></i><b>10</b> DNA methylation analysis using bisulfite sequencing data</a><ul>
<li class="chapter" data-level="10.1" data-path="what-is-dna-methylation.html"><a href="what-is-dna-methylation.html"><i class="fa fa-check"></i><b>10.1</b> What is DNA methylation?</a><ul>
<li class="chapter" data-level="10.1.1" data-path="what-is-dna-methylation.html"><a href="what-is-dna-methylation.html#how-dna-methylation-is-set"><i class="fa fa-check"></i><b>10.1.1</b> How DNA methylation is set ?</a></li>
<li class="chapter" data-level="10.1.2" data-path="what-is-dna-methylation.html"><a href="what-is-dna-methylation.html#how-to-measure-dna-methylation-with-bisulfite-sequencing"><i class="fa fa-check"></i><b>10.1.2</b> How to measure DNA methylation with bisulfite sequencing</a></li>
</ul></li>
<li class="chapter" data-level="10.2" data-path="analyzing-dna-methylation-data.html"><a href="analyzing-dna-methylation-data.html"><i class="fa fa-check"></i><b>10.2</b> Analyzing DNA methylation data</a></li>
<li class="chapter" data-level="10.3" data-path="processing-raw-data-and-getting-data-into-r.html"><a href="processing-raw-data-and-getting-data-into-r.html"><i class="fa fa-check"></i><b>10.3</b> Processing raw data and getting data into R</a></li>
<li class="chapter" data-level="10.4" data-path="data-filtering-and-exploratory-analysis.html"><a href="data-filtering-and-exploratory-analysis.html"><i class="fa fa-check"></i><b>10.4</b> Data filtering and exploratory analysis</a><ul>
<li class="chapter" data-level="10.4.1" data-path="data-filtering-and-exploratory-analysis.html"><a href="data-filtering-and-exploratory-analysis.html#reading-methylation-call-files"><i class="fa fa-check"></i><b>10.4.1</b> Reading methylation call files</a></li>
<li class="chapter" data-level="10.4.2" data-path="data-filtering-and-exploratory-analysis.html"><a href="data-filtering-and-exploratory-analysis.html#further-quality-check"><i class="fa fa-check"></i><b>10.4.2</b> Further quality check</a></li>
<li class="chapter" data-level="10.4.3" data-path="data-filtering-and-exploratory-analysis.html"><a href="data-filtering-and-exploratory-analysis.html#merging-samples-into-a-single-table"><i class="fa fa-check"></i><b>10.4.3</b> Merging samples into a single table</a></li>
<li class="chapter" data-level="10.4.4" data-path="data-filtering-and-exploratory-analysis.html"><a href="data-filtering-and-exploratory-analysis.html#filtering-cpgs"><i class="fa fa-check"></i><b>10.4.4</b> Filtering CpGs</a></li>
<li class="chapter" data-level="10.4.5" data-path="data-filtering-and-exploratory-analysis.html"><a href="data-filtering-and-exploratory-analysis.html#clustering-samples"><i class="fa fa-check"></i><b>10.4.5</b> Clustering samples</a></li>
<li class="chapter" data-level="10.4.6" data-path="data-filtering-and-exploratory-analysis.html"><a href="data-filtering-and-exploratory-analysis.html#principal-component-analysis-1"><i class="fa fa-check"></i><b>10.4.6</b> Principal component analysis</a></li>
</ul></li>
<li class="chapter" data-level="10.5" data-path="extracting-interesting-regions-differential-methylation-and-segmentation.html"><a href="extracting-interesting-regions-differential-methylation-and-segmentation.html"><i class="fa fa-check"></i><b>10.5</b> Extracting interesting regions: Differential methylation and segmentation</a><ul>
<li class="chapter" data-level="10.5.1" data-path="extracting-interesting-regions-differential-methylation-and-segmentation.html"><a href="extracting-interesting-regions-differential-methylation-and-segmentation.html#differential-methylation"><i class="fa fa-check"></i><b>10.5.1</b> Differential methylation</a></li>
<li class="chapter" data-level="10.5.2" data-path="extracting-interesting-regions-differential-methylation-and-segmentation.html"><a href="extracting-interesting-regions-differential-methylation-and-segmentation.html#methylation-segmentation"><i class="fa fa-check"></i><b>10.5.2</b> Methylation segmentation</a></li>
<li class="chapter" data-level="10.5.3" data-path="extracting-interesting-regions-differential-methylation-and-segmentation.html"><a href="extracting-interesting-regions-differential-methylation-and-segmentation.html#working-with-large-files"><i class="fa fa-check"></i><b>10.5.3</b> Working with large files</a></li>
</ul></li>
<li class="chapter" data-level="10.6" data-path="annotation-of-dmrsdmcs-and-segments.html"><a href="annotation-of-dmrsdmcs-and-segments.html"><i class="fa fa-check"></i><b>10.6</b> Annotation of DMRs/DMCs and segments</a><ul>
<li class="chapter" data-level="10.6.1" data-path="annotation-of-dmrsdmcs-and-segments.html"><a href="annotation-of-dmrsdmcs-and-segments.html#further-annotation-with-genes-or-gene-sets"><i class="fa fa-check"></i><b>10.6.1</b> Further annotation with genes or gene sets</a></li>
</ul></li>
<li class="chapter" data-level="10.7" data-path="other-r-packages-that-can-be-used-for-methylation-analysis.html"><a href="other-r-packages-that-can-be-used-for-methylation-analysis.html"><i class="fa fa-check"></i><b>10.7</b> Other R packages that can be used for methylation analysis</a></li>
<li class="chapter" data-level="10.8" data-path="exercises-8.html"><a href="exercises-8.html"><i class="fa fa-check"></i><b>10.8</b> Exercises</a><ul>
<li class="chapter" data-level="10.8.1" data-path="exercises-8.html"><a href="exercises-8.html#differential-methylation-1"><i class="fa fa-check"></i><b>10.8.1</b> Differential methylation</a></li>
<li class="chapter" data-level="10.8.2" data-path="exercises-8.html"><a href="exercises-8.html#methylome-segmentation"><i class="fa fa-check"></i><b>10.8.2</b> Methylome segmentation</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="11" data-path="multiomics.html"><a href="multiomics.html"><i class="fa fa-check"></i><b>11</b> Multi-omics Analysis</a><ul>
<li class="chapter" data-level="11.1" data-path="use-case-multi-omics-data-from-colorectal-cancer.html"><a href="use-case-multi-omics-data-from-colorectal-cancer.html"><i class="fa fa-check"></i><b>11.1</b> Use case: Multi-omics data from colorectal cancer</a></li>
<li class="chapter" data-level="11.2" data-path="latent-variable-models-for-multi-omics-integration.html"><a href="latent-variable-models-for-multi-omics-integration.html"><i class="fa fa-check"></i><b>11.2</b> Latent variable models for multi-omics integration</a></li>
<li class="chapter" data-level="11.3" data-path="matrix-factorization-methods-for-unsupervised-multi-omics-data-integration.html"><a href="matrix-factorization-methods-for-unsupervised-multi-omics-data-integration.html"><i class="fa fa-check"></i><b>11.3</b> Matrix factorization methods for unsupervised multi-omics data integration</a><ul>
<li class="chapter" data-level="11.3.1" data-path="matrix-factorization-methods-for-unsupervised-multi-omics-data-integration.html"><a href="matrix-factorization-methods-for-unsupervised-multi-omics-data-integration.html#multiple-factor-analysis"><i class="fa fa-check"></i><b>11.3.1</b> Multiple factor analysis</a></li>
<li class="chapter" data-level="11.3.2" data-path="matrix-factorization-methods-for-unsupervised-multi-omics-data-integration.html"><a href="matrix-factorization-methods-for-unsupervised-multi-omics-data-integration.html#joint-non-negative-matrix-factorization"><i class="fa fa-check"></i><b>11.3.2</b> Joint non-negative matrix factorization</a></li>
<li class="chapter" data-level="11.3.3" data-path="matrix-factorization-methods-for-unsupervised-multi-omics-data-integration.html"><a href="matrix-factorization-methods-for-unsupervised-multi-omics-data-integration.html#icluster"><i class="fa fa-check"></i><b>11.3.3</b> iCluster</a></li>
</ul></li>
<li class="chapter" data-level="11.4" data-path="clustering-using-latent-factors.html"><a href="clustering-using-latent-factors.html"><i class="fa fa-check"></i><b>11.4</b> Clustering using latent factors</a><ul>
<li class="chapter" data-level="11.4.1" data-path="clustering-using-latent-factors.html"><a href="clustering-using-latent-factors.html#one-hot-clustering"><i class="fa fa-check"></i><b>11.4.1</b> One-hot clustering</a></li>
<li class="chapter" data-level="11.4.2" data-path="clustering-using-latent-factors.html"><a href="clustering-using-latent-factors.html#k-means-clustering-1"><i class="fa fa-check"></i><b>11.4.2</b> K-means clustering</a></li>
</ul></li>
<li class="chapter" data-level="11.5" data-path="biological-interpretation-of-latent-factors.html"><a href="biological-interpretation-of-latent-factors.html"><i class="fa fa-check"></i><b>11.5</b> Biological interpretation of latent factors</a><ul>
<li class="chapter" data-level="11.5.1" data-path="biological-interpretation-of-latent-factors.html"><a href="biological-interpretation-of-latent-factors.html#inspection-of-feature-weights-in-loading-vectors"><i class="fa fa-check"></i><b>11.5.1</b> Inspection of feature weights in loading vectors</a></li>
<li class="chapter" data-level="11.5.2" data-path="biological-interpretation-of-latent-factors.html"><a href="biological-interpretation-of-latent-factors.html#making-sense-of-factors-using-enrichment-analysis"><i class="fa fa-check"></i><b>11.5.2</b> Making sense of factors using enrichment analysis</a></li>
<li class="chapter" data-level="11.5.3" data-path="biological-interpretation-of-latent-factors.html"><a href="biological-interpretation-of-latent-factors.html#interpretation-using-additional-covariates"><i class="fa fa-check"></i><b>11.5.3</b> Interpretation using additional covariates</a></li>
</ul></li>
<li class="chapter" data-level="11.6" data-path="exercises-9.html"><a href="exercises-9.html"><i class="fa fa-check"></i><b>11.6</b> Exercises</a><ul>
<li class="chapter" data-level="11.6.1" data-path="exercises-9.html"><a href="exercises-9.html#matrix-factorization-methods"><i class="fa fa-check"></i><b>11.6.1</b> Matrix factorization methods</a></li>
<li class="chapter" data-level="11.6.2" data-path="exercises-9.html"><a href="exercises-9.html#clustering-using-latent-factors-1"><i class="fa fa-check"></i><b>11.6.2</b> Clustering using latent factors</a></li>
<li class="chapter" data-level="11.6.3" data-path="exercises-9.html"><a href="exercises-9.html#biological-interpretation-of-latent-factors-1"><i class="fa fa-check"></i><b>11.6.3</b> Biological interpretation of latent factors</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i>References</a></li>
<li class="divider"></li>
<li><a href="https://bookdown.org" target="_blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Computational Genomics with R</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="peak-calling" class="section level2">
<h2><span class="header-section-number">9.6</span> Peak calling</h2>
<p></p>
<p>After we are convinced that the data is of sufficient quality, we can
proceed with the downstream analysis.
One of the first steps in the ChIP-seq analysis is peak calling.
Peak calling is a statistical procedure, which uses coverage properties
of ChIP and Input samples to find regions which are enriched due to
protein binding.
The procedure requires mapped reads, and outputs a set of regions, which
represent the putative binding locations. Each region is usually associated
with a significance score which is an indicator of enrichment.</p>
<p>For peak calling we will use the <code>normR</code> Bioconductor package.
<code>normR</code> uses a binomial mixture model, and performs simultaneous
normalization and peak finding. Due to the nature of the model, it is
quite flexible and can be used for different types of ChIP experiments.</p>
<p>One of the caveats of <code>normR</code> is that it does not inherently support
multiple biological replicates, for the same biological sample.
Therefore, the peak calling procedure needs to be done on each replicate
separately, and the peaks need to be combined in post-processing.</p>
<div id="types-of-chip-seq-experiments" class="section level3">
<h3><span class="header-section-number">9.6.1</span> Types of ChIP-seq experiments</h3>
<p>Based on the binding properties of ChIP-ped proteins, ChIP-seq
signal profiles can be divided into three classes:</p>
<ol style="list-style-type: decimal">
<li><p><strong>Sharp</strong> (point signal): A signal profile which is localized to specific
short genomic regions (up to couple of hundred base pairs)
It is usually obtained from transcription factors, or highly localized posttranslational histone modifications
(H3K4me3, which is found on gene promoters).</p></li>
<li><p><strong>Broad</strong> (wide signal): The signal covers broad genomic domains spanning up to several kilobases.
Usually produced by disperse histone modifications (H3K36me3, located
on gene bodies, or H3K23me3, which is deposited by the Polycomb complex in large genomic regions).</p></li>
<li><p><strong>Mixed</strong>: The signal consists of a mixture of sharp and broad regions.
It is produced by proteins which have dynamic behavior. Most often these are ChIP experiments
of RNA Polymerase 2.</p></li>
</ol>
<p>Different types of ChIP experiments usually require specialized analysis tools. Some peak callers are developed to specifically detect narrow peaks <span class="citation">(Zhang, Liu, Meyer, et al. <a href="#ref-zhang_2008" role="doc-biblioref">2008</a>; Xu, Handoko, Wei, et al. <a href="#ref-xu_2010" role="doc-biblioref">2010</a>; Shao, Zhang, Yuan, et al. <a href="#ref-shao_2012" role="doc-biblioref">2012</a>)</span>, while others
detect enrichment in diffuse broad regions <span class="citation">(Zang, Schones, Zeng, et al. <a href="#ref-zang_2009" role="doc-biblioref">2009</a>; Micsinai, Parisi, Strino, et al. <a href="#ref-micsinai_2012" role="doc-biblioref">2012</a>; Beck, Brandl, Boelen, et al. <a href="#ref-beck_2012" role="doc-biblioref">2012</a>; Song and Smith <a href="#ref-song_2011" role="doc-biblioref">2011</a>; Xing, Mo, Liao, et al. <a href="#ref-xing_2012" role="doc-biblioref">2012</a>)</span>,
or mixed (Polymerase 2) signals <span class="citation">(Han, Tian, Pécot, et al. <a href="#ref-han_2012" role="doc-biblioref">2012</a>)</span>.
Recent developments in peak calling methods (such as <code>normR</code>) can however accommodate
multiple types of ChIP experiments <span class="citation">(Rashid, Giresi, Ibrahim, et al. <a href="#ref-rashid_2011" role="doc-biblioref">2011</a>)</span>.
The choice of the algorithm will largely depend on the type of the wanted
results, and the peculiarities of the experimental design and execution <span class="citation">(Laajala, Raghav, Tuomela, et al. <a href="#ref-laajala_2009" role="doc-biblioref">2009</a>; Wilbanks and Facciotti <a href="#ref-wilbanks_2010" role="doc-biblioref">2010</a>)</span>.</p>
<p>If you are not certain what kind of signal profile to expect from a ChIP-seq
experiment, the best solution is to visualize the data. We will now use the data from <strong>H3K4me3</strong> (Sharp), <strong>H3K36me3</strong> (Broad), and <strong>POL2</strong> (Mixed)
ChIP experiments to show the differences in the signal profiles. We will use the bigWig files to visualize the signal profiles around a
highly expressed human gene from chromosome 21. This will give us an indication
of how the profiles for different types of ChIP experiments differ. First we select the files of interest:</p>
<div class="sourceCode" id="cb490"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb490-1"><a href="peak-calling.html#cb490-1"></a><span class="co"># set names for chip-seq bigWig files</span></span>
<span id="cb490-2"><a href="peak-calling.html#cb490-2"></a>chip_files =<span class="st"> </span><span class="kw">list</span>(</span>
<span id="cb490-3"><a href="peak-calling.html#cb490-3"></a>    <span class="dt">H3K4me3  =</span> <span class="st">&#39;GM12878_hg38_H3K4me3.chr21.bw&#39;</span>,</span>
<span id="cb490-4"><a href="peak-calling.html#cb490-4"></a></span>
<span id="cb490-5"><a href="peak-calling.html#cb490-5"></a>    <span class="dt">H3K36me3 =</span> <span class="st">&#39;GM12878_hg38_H3K36me3.chr21.bw&#39;</span>,</span>
<span id="cb490-6"><a href="peak-calling.html#cb490-6"></a></span>
<span id="cb490-7"><a href="peak-calling.html#cb490-7"></a>    <span class="dt">POL2     =</span> <span class="st">&#39;GM12878_hg38_POLR2A.chr21.bw&#39;</span></span>
<span id="cb490-8"><a href="peak-calling.html#cb490-8"></a>)</span>
<span id="cb490-9"><a href="peak-calling.html#cb490-9"></a><span class="co"># get full paths to the files</span></span>
<span id="cb490-10"><a href="peak-calling.html#cb490-10"></a>chip_files =<span class="st"> </span><span class="kw">lapply</span>(chip_files, <span class="cf">function</span>(x){</span>
<span id="cb490-11"><a href="peak-calling.html#cb490-11"></a>    <span class="kw">file.path</span>(data_path, x)</span>
<span id="cb490-12"><a href="peak-calling.html#cb490-12"></a>})</span></code></pre></div>
<p>Next we import the coverage profiles into <strong>R</strong>:</p>
<div class="sourceCode" id="cb491"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb491-1"><a href="peak-calling.html#cb491-1"></a><span class="co"># load rtracklayer</span></span>
<span id="cb491-2"><a href="peak-calling.html#cb491-2"></a><span class="kw">library</span>(rtracklayer)</span>
<span id="cb491-3"><a href="peak-calling.html#cb491-3"></a></span>
<span id="cb491-4"><a href="peak-calling.html#cb491-4"></a><span class="co"># import the ChIP bigWig files</span></span>
<span id="cb491-5"><a href="peak-calling.html#cb491-5"></a>chip_profiles =<span class="st"> </span><span class="kw">lapply</span>(chip_files, rtracklayer<span class="op">::</span>import.bw)</span></code></pre></div>
<p>We fetch the reference annotation for human chromosome 21.</p>
<div class="sourceCode" id="cb492"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb492-1"><a href="peak-calling.html#cb492-1"></a><span class="kw">library</span>(AnnotationHub)</span>
<span id="cb492-2"><a href="peak-calling.html#cb492-2"></a>hub =<span class="st"> </span><span class="kw">AnnotationHub</span>()</span>
<span id="cb492-3"><a href="peak-calling.html#cb492-3"></a>gtf =<span class="st"> </span>hub[[<span class="st">&#39;AH61126&#39;</span>]]</span>
<span id="cb492-4"><a href="peak-calling.html#cb492-4"></a></span>
<span id="cb492-5"><a href="peak-calling.html#cb492-5"></a><span class="co"># select only chromosome 21</span></span>
<span id="cb492-6"><a href="peak-calling.html#cb492-6"></a><span class="kw">seqlevels</span>(gtf, <span class="dt">pruning.mode=</span><span class="st">&#39;coarse&#39;</span>) =<span class="st"> &#39;21&#39;</span></span>
<span id="cb492-7"><a href="peak-calling.html#cb492-7"></a></span>
<span id="cb492-8"><a href="peak-calling.html#cb492-8"></a><span class="co"># extract chromosome names</span></span>
<span id="cb492-9"><a href="peak-calling.html#cb492-9"></a>ensembl_seqlevels =<span class="st"> </span><span class="kw">seqlevels</span>(gtf)</span>
<span id="cb492-10"><a href="peak-calling.html#cb492-10"></a></span>
<span id="cb492-11"><a href="peak-calling.html#cb492-11"></a><span class="co"># paste the chr prefix to the chromosome names</span></span>
<span id="cb492-12"><a href="peak-calling.html#cb492-12"></a>ucsc_seqlevels    =<span class="st"> </span><span class="kw">paste0</span>(<span class="st">&#39;chr&#39;</span>, ensembl_seqlevels)</span>
<span id="cb492-13"><a href="peak-calling.html#cb492-13"></a></span>
<span id="cb492-14"><a href="peak-calling.html#cb492-14"></a><span class="co"># replace ensembl with ucsc chromosome names</span></span>
<span id="cb492-15"><a href="peak-calling.html#cb492-15"></a><span class="kw">seqlevels</span>(gtf, <span class="dt">pruning.mode=</span><span class="st">&#39;coarse&#39;</span>) =<span class="st"> </span>ucsc_seqlevels</span></code></pre></div>
<p>To enable <code>Gviz</code> to work with genomic annotation we will convert the <code>GRanges</code>
object into a transcript database using the following function:</p>
<div class="sourceCode" id="cb493"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb493-1"><a href="peak-calling.html#cb493-1"></a><span class="co"># load the GenomicFeatures object</span></span>
<span id="cb493-2"><a href="peak-calling.html#cb493-2"></a><span class="kw">library</span>(GenomicFeatures)</span>
<span id="cb493-3"><a href="peak-calling.html#cb493-3"></a></span>
<span id="cb493-4"><a href="peak-calling.html#cb493-4"></a><span class="co"># convert the gtf annotation into a data.base</span></span>
<span id="cb493-5"><a href="peak-calling.html#cb493-5"></a>txdb         =<span class="st"> </span><span class="kw">makeTxDbFromGRanges</span>(gtf)</span></code></pre></div>
<p>And convert the transcript database into a <code>Gviz</code> track.</p>
<div class="sourceCode" id="cb494"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb494-1"><a href="peak-calling.html#cb494-1"></a><span class="co"># define the gene track object</span></span>
<span id="cb494-2"><a href="peak-calling.html#cb494-2"></a>gene_track  =<span class="st"> </span><span class="kw">GeneRegionTrack</span>(txdb, <span class="dt">chr=</span><span class="st">&#39;chr21&#39;</span>, <span class="dt">genome=</span><span class="st">&#39;hg38&#39;</span>)</span></code></pre></div>
<p>Once we have downloaded the annotation, and imported the signal profiles into <strong>R</strong> we are ready to visualize the data.
We will again use the <code>Gviz</code> library. We firstly define the coordinate system. The ideogram track which will show
the position of our current viewpoint on the chromosome, and a genome axis track, which will show the exact coordinates.</p>
<div class="sourceCode" id="cb495"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb495-1"><a href="peak-calling.html#cb495-1"></a><span class="co"># load Gviz package</span></span>
<span id="cb495-2"><a href="peak-calling.html#cb495-2"></a><span class="kw">library</span>(Gviz)</span>
<span id="cb495-3"><a href="peak-calling.html#cb495-3"></a><span class="co"># fetches the chromosome length information</span></span>
<span id="cb495-4"><a href="peak-calling.html#cb495-4"></a>hg_chrs =<span class="st"> </span><span class="kw">getChromInfoFromUCSC</span>(<span class="st">&#39;hg38&#39;</span>)</span>
<span id="cb495-5"><a href="peak-calling.html#cb495-5"></a>hg_chrs =<span class="st"> </span><span class="kw">subset</span>(hg_chrs, (<span class="kw">grepl</span>(<span class="st">&#39;chr21$&#39;</span>,chrom)))</span>
<span id="cb495-6"><a href="peak-calling.html#cb495-6"></a></span>
<span id="cb495-7"><a href="peak-calling.html#cb495-7"></a><span class="co"># convert data.frame to named vector</span></span>
<span id="cb495-8"><a href="peak-calling.html#cb495-8"></a>seqlengths =<span class="st"> </span><span class="kw">with</span>(hg_chrs, <span class="kw">setNames</span>(size, chrom))</span>
<span id="cb495-9"><a href="peak-calling.html#cb495-9"></a></span>
<span id="cb495-10"><a href="peak-calling.html#cb495-10"></a><span class="co"># constructs the ideogram track</span></span>
<span id="cb495-11"><a href="peak-calling.html#cb495-11"></a>chr_track   =<span class="st"> </span><span class="kw">IdeogramTrack</span>(</span>
<span id="cb495-12"><a href="peak-calling.html#cb495-12"></a>    <span class="dt">chromosome =</span> <span class="st">&#39;chr21&#39;</span>, </span>
<span id="cb495-13"><a href="peak-calling.html#cb495-13"></a>    <span class="dt">genome     =</span> <span class="st">&#39;hg38&#39;</span></span>
<span id="cb495-14"><a href="peak-calling.html#cb495-14"></a>)</span>
<span id="cb495-15"><a href="peak-calling.html#cb495-15"></a></span>
<span id="cb495-16"><a href="peak-calling.html#cb495-16"></a><span class="co"># constructs the coordinate system</span></span>
<span id="cb495-17"><a href="peak-calling.html#cb495-17"></a>axis =<span class="st"> </span><span class="kw">GenomeAxisTrack</span>(</span>
<span id="cb495-18"><a href="peak-calling.html#cb495-18"></a>    <span class="dt">range =</span> <span class="kw">GRanges</span>(<span class="st">&#39;chr21&#39;</span>, <span class="kw">IRanges</span>(<span class="dv">1</span>, <span class="dt">width=</span>seqlengths))</span>
<span id="cb495-19"><a href="peak-calling.html#cb495-19"></a>)</span></code></pre></div>
<p>We use a loop to convert the signal profiles into a <code>DataTrack</code> object.</p>
<div class="sourceCode" id="cb496"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb496-1"><a href="peak-calling.html#cb496-1"></a><span class="co"># use a lapply on the imported bw files to create the track objects</span></span>
<span id="cb496-2"><a href="peak-calling.html#cb496-2"></a><span class="co"># we loop over experiment names, and select the corresponding object</span></span>
<span id="cb496-3"><a href="peak-calling.html#cb496-3"></a><span class="co"># within the function</span></span>
<span id="cb496-4"><a href="peak-calling.html#cb496-4"></a>data_tracks =<span class="st"> </span><span class="kw">lapply</span>(<span class="kw">names</span>(chip_profiles), <span class="cf">function</span>(exp_name){</span>
<span id="cb496-5"><a href="peak-calling.html#cb496-5"></a>    </span>
<span id="cb496-6"><a href="peak-calling.html#cb496-6"></a>    <span class="co"># chip_profiles[[exp_name]] - selects the </span></span>
<span id="cb496-7"><a href="peak-calling.html#cb496-7"></a>    <span class="co"># proper experiment using the exp_name</span></span>
<span id="cb496-8"><a href="peak-calling.html#cb496-8"></a>    <span class="kw">DataTrack</span>(</span>
<span id="cb496-9"><a href="peak-calling.html#cb496-9"></a>        <span class="dt">range =</span> chip_profiles[[exp_name]],   </span>
<span id="cb496-10"><a href="peak-calling.html#cb496-10"></a>        <span class="dt">name  =</span> exp_name,  </span>
<span id="cb496-11"><a href="peak-calling.html#cb496-11"></a>        </span>
<span id="cb496-12"><a href="peak-calling.html#cb496-12"></a>        <span class="co"># type of the track</span></span>
<span id="cb496-13"><a href="peak-calling.html#cb496-13"></a>        <span class="dt">type  =</span> <span class="st">&#39;h&#39;</span>, </span>
<span id="cb496-14"><a href="peak-calling.html#cb496-14"></a>        </span>
<span id="cb496-15"><a href="peak-calling.html#cb496-15"></a>        <span class="co"># line width parameter</span></span>
<span id="cb496-16"><a href="peak-calling.html#cb496-16"></a>        <span class="dt">lwd   =</span> <span class="dv">5</span></span>
<span id="cb496-17"><a href="peak-calling.html#cb496-17"></a>    )</span>
<span id="cb496-18"><a href="peak-calling.html#cb496-18"></a>})</span></code></pre></div>
<p>We are finally ready to create the genome screenshot.
We will focus on an extended region around the URB1 gene.</p>
<div class="sourceCode" id="cb497"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb497-1"><a href="peak-calling.html#cb497-1"></a><span class="co"># select the start coordinate for the URB1 gene</span></span>
<span id="cb497-2"><a href="peak-calling.html#cb497-2"></a>start =<span class="st"> </span><span class="kw">min</span>(<span class="kw">start</span>(<span class="kw">subset</span>(gtf, gene_name <span class="op">==</span><span class="st"> &#39;URB1&#39;</span>)))</span>
<span id="cb497-3"><a href="peak-calling.html#cb497-3"></a></span>
<span id="cb497-4"><a href="peak-calling.html#cb497-4"></a><span class="co"># select the end coordinate for the URB1 gene</span></span>
<span id="cb497-5"><a href="peak-calling.html#cb497-5"></a>end   =<span class="st"> </span><span class="kw">max</span>(<span class="kw">end</span>(<span class="kw">subset</span>(gtf, gene_name <span class="op">==</span><span class="st"> &#39;URB1&#39;</span>)))</span>
<span id="cb497-6"><a href="peak-calling.html#cb497-6"></a></span>
<span id="cb497-7"><a href="peak-calling.html#cb497-7"></a><span class="co"># plot the signal profiles around the URB1 gene</span></span>
<span id="cb497-8"><a href="peak-calling.html#cb497-8"></a><span class="kw">plotTracks</span>(</span>
<span id="cb497-9"><a href="peak-calling.html#cb497-9"></a>    <span class="dt">trackList =</span> <span class="kw">c</span>(chr_track, axis, gene_track, data_tracks),</span>
<span id="cb497-10"><a href="peak-calling.html#cb497-10"></a>    </span>
<span id="cb497-11"><a href="peak-calling.html#cb497-11"></a>    <span class="co"># relative track sizes</span></span>
<span id="cb497-12"><a href="peak-calling.html#cb497-12"></a>    <span class="dt">sizes     =</span> <span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">1</span>,<span class="dv">1</span>,<span class="dv">1</span>,<span class="dv">1</span>,<span class="dv">1</span>), </span>
<span id="cb497-13"><a href="peak-calling.html#cb497-13"></a>    </span>
<span id="cb497-14"><a href="peak-calling.html#cb497-14"></a>    <span class="co"># background color</span></span>
<span id="cb497-15"><a href="peak-calling.html#cb497-15"></a>    <span class="dt">background.title     =</span> <span class="st">&quot;black&quot;</span>,</span>
<span id="cb497-16"><a href="peak-calling.html#cb497-16"></a>    </span>
<span id="cb497-17"><a href="peak-calling.html#cb497-17"></a>    <span class="co"># controls visualization of gene sets</span></span>
<span id="cb497-18"><a href="peak-calling.html#cb497-18"></a>    <span class="dt">collapseTranscripts  =</span> <span class="st">&quot;longest&quot;</span>, </span>
<span id="cb497-19"><a href="peak-calling.html#cb497-19"></a>    <span class="dt">transcriptAnnotation =</span> <span class="st">&quot;symbol&quot;</span>,</span>
<span id="cb497-20"><a href="peak-calling.html#cb497-20"></a>    </span>
<span id="cb497-21"><a href="peak-calling.html#cb497-21"></a>    <span class="co"># coordinates to visualize </span></span>
<span id="cb497-22"><a href="peak-calling.html#cb497-22"></a>    <span class="dt">from =</span> start <span class="op">-</span><span class="st"> </span><span class="dv">5000</span>,</span>
<span id="cb497-23"><a href="peak-calling.html#cb497-23"></a>    <span class="dt">to   =</span> end   <span class="op">+</span><span class="st"> </span><span class="dv">5000</span></span>
<span id="cb497-24"><a href="peak-calling.html#cb497-24"></a>)</span></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:chip-type-plot-gviz"></span>
<img src="09-chip-seq-analysis_files/figure-html/chip-type-plot-gviz-1.png" alt="ChIP-seq signal around the URB1 gene." width="50%" />
<p class="caption">
FIGURE 9.11: ChIP-seq signal around the URB1 gene.
</p>
</div>
<p>Figure <a href="peak-calling.html#fig:chip-type-plot-gviz">9.11</a> shows the signal profile around the URB1 gene. H3K4me3 signal profile contains a strong narrow peak on the transcription start site. H3K36me3 shows strong enrichment in the gene body, while the POL2 ChIP shows a mixed profile, with a strong peak at the TSS and an enrichment over the gene body.</p>
</div>
<div id="peak-calling-sharp-peaks" class="section level3">
<h3><span class="header-section-number">9.6.2</span> Peak calling: Sharp peaks</h3>
<p></p>
<p>We will now use the <code>normR</code> <span class="citation">(Helmuth, Li, Arrigoni, et al. <a href="#ref-helmuth_2016" role="doc-biblioref">2016</a>)</span> package for peak calling in sharp and broad peak experiments.</p>
<p>Select the input files. Since <code>normR</code> does not support the usage of biological
replicates, we will showcase the peak calling on one of the CTCF samples.</p>
<div class="sourceCode" id="cb498"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb498-1"><a href="peak-calling.html#cb498-1"></a><span class="co"># full path to the ChIP data file</span></span>
<span id="cb498-2"><a href="peak-calling.html#cb498-2"></a>chip_file    =<span class="st"> </span><span class="kw">file.path</span>(data_path, <span class="st">&#39;GM12878_hg38_CTCF_r1.chr21.bam&#39;</span>)</span>
<span id="cb498-3"><a href="peak-calling.html#cb498-3"></a></span>
<span id="cb498-4"><a href="peak-calling.html#cb498-4"></a><span class="co"># full path to the Control data file</span></span>
<span id="cb498-5"><a href="peak-calling.html#cb498-5"></a>control_file =<span class="st"> </span><span class="kw">file.path</span>(data_path, <span class="st">&#39;GM12878_hg38_Input_r5.chr21.bam&#39;</span>)</span></code></pre></div>
<p>To understand the dynamic range of enrichment, we will create a scatter plot
showing the strength of signal in the CTCF and Input.</p>
<p>Let us first count the reads in 1-kb windows, and normalize them to counts per
million sequenced reads.</p>
<div class="sourceCode" id="cb499"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb499-1"><a href="peak-calling.html#cb499-1"></a><span class="co"># as previously done, we calculate the cpm for each experiment</span></span>
<span id="cb499-2"><a href="peak-calling.html#cb499-2"></a><span class="kw">library</span>(GenomicRanges)</span>
<span id="cb499-3"><a href="peak-calling.html#cb499-3"></a><span class="kw">library</span>(GenomicAlignments)</span>
<span id="cb499-4"><a href="peak-calling.html#cb499-4"></a></span>
<span id="cb499-5"><a href="peak-calling.html#cb499-5"></a><span class="co"># select the chromosome</span></span>
<span id="cb499-6"><a href="peak-calling.html#cb499-6"></a>hg_chrs =<span class="st"> </span><span class="kw">getChromInfoFromUCSC</span>(<span class="st">&#39;hg38&#39;</span>) </span>
<span id="cb499-7"><a href="peak-calling.html#cb499-7"></a>hg_chrs =<span class="st"> </span><span class="kw">subset</span>(hg_chrs, <span class="kw">grepl</span>(<span class="st">&#39;chr21$&#39;</span>,chrom))</span>
<span id="cb499-8"><a href="peak-calling.html#cb499-8"></a></span>
<span id="cb499-9"><a href="peak-calling.html#cb499-9"></a>seqlengths =<span class="st"> </span><span class="kw">with</span>(hg_chrs, <span class="kw">setNames</span>(size, chrom))</span>
<span id="cb499-10"><a href="peak-calling.html#cb499-10"></a></span>
<span id="cb499-11"><a href="peak-calling.html#cb499-11"></a><span class="co"># define the windows</span></span>
<span id="cb499-12"><a href="peak-calling.html#cb499-12"></a>tilling_window =<span class="st"> </span><span class="kw">unlist</span>(<span class="kw">tileGenome</span>(seqlengths, <span class="dt">tilewidth=</span><span class="dv">1000</span>))</span>
<span id="cb499-13"><a href="peak-calling.html#cb499-13"></a></span>
<span id="cb499-14"><a href="peak-calling.html#cb499-14"></a><span class="co"># count the reads</span></span>
<span id="cb499-15"><a href="peak-calling.html#cb499-15"></a>counts         =<span class="st"> </span><span class="kw">summarizeOverlaps</span>(</span>
<span id="cb499-16"><a href="peak-calling.html#cb499-16"></a>    <span class="dt">features =</span> tilling_window, </span>
<span id="cb499-17"><a href="peak-calling.html#cb499-17"></a>    <span class="dt">reads    =</span> <span class="kw">c</span>(chip_file, control_file)</span>
<span id="cb499-18"><a href="peak-calling.html#cb499-18"></a>)</span>
<span id="cb499-19"><a href="peak-calling.html#cb499-19"></a></span>
<span id="cb499-20"><a href="peak-calling.html#cb499-20"></a><span class="co"># normalize read counts</span></span>
<span id="cb499-21"><a href="peak-calling.html#cb499-21"></a>counts         =<span class="st"> </span><span class="kw">assays</span>(counts)[[<span class="dv">1</span>]]</span>
<span id="cb499-22"><a href="peak-calling.html#cb499-22"></a>cpm =<span class="st"> </span><span class="kw">t</span>(<span class="kw">t</span>(counts)<span class="op">*</span>(<span class="dv">1000000</span><span class="op">/</span><span class="kw">colSums</span>(counts)))</span></code></pre></div>
<p>We can now plot the ChIP versus Input signal:</p>
<div class="sourceCode" id="cb500"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb500-1"><a href="peak-calling.html#cb500-1"></a><span class="kw">library</span>(ggplot2)</span>
<span id="cb500-2"><a href="peak-calling.html#cb500-2"></a><span class="co"># convert the matrix into a data.frame for ggplot</span></span>
<span id="cb500-3"><a href="peak-calling.html#cb500-3"></a>cpm =<span class="st"> </span><span class="kw">data.frame</span>(cpm)</span>
<span id="cb500-4"><a href="peak-calling.html#cb500-4"></a><span class="kw">ggplot</span>(</span>
<span id="cb500-5"><a href="peak-calling.html#cb500-5"></a>    <span class="dt">data =</span> cpm, </span>
<span id="cb500-6"><a href="peak-calling.html#cb500-6"></a>    <span class="kw">aes</span>(</span>
<span id="cb500-7"><a href="peak-calling.html#cb500-7"></a>        <span class="dt">x =</span> GM12878_hg38_Input_r5.chr21.bam, </span>
<span id="cb500-8"><a href="peak-calling.html#cb500-8"></a>        <span class="dt">y =</span> GM12878_hg38_CTCF_r1.chr21.bam)</span>
<span id="cb500-9"><a href="peak-calling.html#cb500-9"></a>    ) <span class="op">+</span></span>
<span id="cb500-10"><a href="peak-calling.html#cb500-10"></a><span class="st">    </span><span class="kw">geom_point</span>() <span class="op">+</span></span>
<span id="cb500-11"><a href="peak-calling.html#cb500-11"></a><span class="st">    </span><span class="kw">geom_abline</span>(<span class="dt">slope =</span> <span class="dv">1</span>) <span class="op">+</span></span>
<span id="cb500-12"><a href="peak-calling.html#cb500-12"></a><span class="st">    </span><span class="kw">theme_bw</span>() <span class="op">+</span></span>
<span id="cb500-13"><a href="peak-calling.html#cb500-13"></a><span class="st">    </span><span class="kw">theme_bw</span>() <span class="op">+</span></span>
<span id="cb500-14"><a href="peak-calling.html#cb500-14"></a><span class="st">    </span><span class="kw">scale_fill_brewer</span>(<span class="dt">palette=</span><span class="st">&#39;Set2&#39;</span>) <span class="op">+</span></span>
<span id="cb500-15"><a href="peak-calling.html#cb500-15"></a><span class="st">    </span><span class="kw">theme</span>(</span>
<span id="cb500-16"><a href="peak-calling.html#cb500-16"></a>        <span class="dt">axis.text   =</span> <span class="kw">element_text</span>(<span class="dt">size=</span><span class="dv">10</span>, <span class="dt">face=</span><span class="st">&#39;bold&#39;</span>),</span>
<span id="cb500-17"><a href="peak-calling.html#cb500-17"></a>        <span class="dt">axis.title  =</span> <span class="kw">element_text</span>(<span class="dt">size=</span><span class="dv">14</span>,<span class="dt">face=</span><span class="st">&quot;bold&quot;</span>),</span>
<span id="cb500-18"><a href="peak-calling.html#cb500-18"></a>        <span class="dt">plot.title  =</span> <span class="kw">element_text</span>(<span class="dt">hjust =</span> <span class="fl">0.5</span>),</span>
<span id="cb500-19"><a href="peak-calling.html#cb500-19"></a>        <span class="dt">axis.text.x =</span> <span class="kw">element_text</span>(<span class="dt">angle =</span> <span class="dv">90</span>, <span class="dt">hjust =</span> <span class="dv">1</span>)) <span class="op">+</span></span>
<span id="cb500-20"><a href="peak-calling.html#cb500-20"></a><span class="st">    </span><span class="kw">xlab</span>(<span class="st">&#39;Input CPM&#39;</span>) <span class="op">+</span></span>
<span id="cb500-21"><a href="peak-calling.html#cb500-21"></a><span class="st">    </span><span class="kw">ylab</span>(<span class="st">&#39;CTCF CPM&#39;</span>) <span class="op">+</span></span>
<span id="cb500-22"><a href="peak-calling.html#cb500-22"></a><span class="st">    </span><span class="kw">ggtitle</span>(<span class="st">&#39;ChIP versus Input&#39;</span>)</span></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:peak-calling-sharp-plot"></span>
<img src="09-chip-seq-analysis_files/figure-html/peak-calling-sharp-plot-1.png" alt="Comparison of CPM values between ChIP and Input experiments. Good ChIP experiments should always show enrichment." width="50%" />
<p class="caption">
FIGURE 9.12: Comparison of CPM values between ChIP and Input experiments. Good ChIP experiments should always show enrichment.
</p>
</div>
<p>Regions above the diagonal, in Figure <a href="peak-calling.html#fig:peak-calling-sharp-plot">9.12</a>, show
higher enrichment in the ChIP samples, while the regions below the diagonal
show higher enrichment in the Input samples.</p>
<p>Let us now perform for peak calling. <code>normR</code> usage is deceivingly simple; we need to provide the location ChIP and Control read files, and the genome version to the <code>enrichR()</code> function. The function will automatically create tilling windows (250bp by default), count the number of reads in each window, and fit a mixture of binomial distributions.</p>
<div class="sourceCode" id="cb501"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb501-1"><a href="peak-calling.html#cb501-1"></a><span class="kw">library</span>(normr)</span>
<span id="cb501-2"><a href="peak-calling.html#cb501-2"></a><span class="co"># peak calling using chip and control</span></span>
<span id="cb501-3"><a href="peak-calling.html#cb501-3"></a>ctcf_fit =<span class="st"> </span><span class="kw">enrichR</span>(</span>
<span id="cb501-4"><a href="peak-calling.html#cb501-4"></a>    </span>
<span id="cb501-5"><a href="peak-calling.html#cb501-5"></a>            <span class="co"># ChIP file</span></span>
<span id="cb501-6"><a href="peak-calling.html#cb501-6"></a>            <span class="dt">treatment =</span> chip_file,</span>
<span id="cb501-7"><a href="peak-calling.html#cb501-7"></a>            </span>
<span id="cb501-8"><a href="peak-calling.html#cb501-8"></a>            <span class="co"># control file</span></span>
<span id="cb501-9"><a href="peak-calling.html#cb501-9"></a>            <span class="dt">control   =</span> control_file,</span>
<span id="cb501-10"><a href="peak-calling.html#cb501-10"></a>            </span>
<span id="cb501-11"><a href="peak-calling.html#cb501-11"></a>            <span class="co"># genome version</span></span>
<span id="cb501-12"><a href="peak-calling.html#cb501-12"></a>            <span class="dt">genome    =</span> <span class="st">&quot;hg38&quot;</span>,</span>
<span id="cb501-13"><a href="peak-calling.html#cb501-13"></a>            </span>
<span id="cb501-14"><a href="peak-calling.html#cb501-14"></a>            <span class="co"># print intermediary steps during the analysis</span></span>
<span id="cb501-15"><a href="peak-calling.html#cb501-15"></a>            <span class="dt">verbose   =</span> <span class="ot">FALSE</span>)</span></code></pre></div>
<p>With the summary function we can take a look at the results:</p>
<div class="sourceCode" id="cb502"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb502-1"><a href="peak-calling.html#cb502-1"></a><span class="kw">summary</span>(ctcf_fit)</span></code></pre></div>
<pre><code>## NormRFit-class object
## 
## Type:                  &#39;enrichR&#39;
## Number of Regions:     12353090
## Number of Components:  2
## Theta* (naive bg):     0.137
## Background component B: 1
## 
## +++ Results of fit +++ 
## Mixture Proportions:
## Background       Class 1    
##     97.72%         2.28%    
## Theta:
## Background       Class 1    
##      0.103         0.695    
## 
## Bayesian Information Criterion:  539882
## 
## +++ Results of binomial test +++ 
## T-Filter threshold: 4
## Number of Regions filtered out: 12267164
## Significantly different from background B based on q-values:
## TOTAL:
##            ***        **         *         .                n.s.
## Bins         0       627       120       195        87     84897
## %        0.000     0.711     0.847     1.068     1.166    96.209
## ---
## Signif. codes:  0 &#39;***&#39; 0.001 &#39;**&#39; 0.01 &#39;*&#39; 0.05 &#39;.&#39; 0.1 &#39;  &#39; 1 &#39;n.s.&#39;</code></pre>
<p>The summary function shows that most of the regions of chromosome 21 correspond
to the background: <span class="math inline">\(97.72%\)</span>. In total we have <span class="math inline">\(1029=(627+120+195+87)\)</span> significantly enriched regions.</p>
<p>We will now extract the regions into a <code>GRanges</code> object.
The <code>getRanges()</code> function extracts the regions from the model. Using the
<code>getQvalue()</code>, and <code>getEnrichment()</code> function we assign to our regions
the statistical significance and calculated enrichment.
In order to identify only highly significant regions,
we keep only ranges where the false discovery rate (q value) is below <span class="math inline">\(0.01\)</span>.</p>
<div class="sourceCode" id="cb504"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb504-1"><a href="peak-calling.html#cb504-1"></a><span class="co"># extracts the ranges</span></span>
<span id="cb504-2"><a href="peak-calling.html#cb504-2"></a>ctcf_peaks =<span class="st"> </span><span class="kw">getRanges</span>(ctcf_fit)</span>
<span id="cb504-3"><a href="peak-calling.html#cb504-3"></a>    </span>
<span id="cb504-4"><a href="peak-calling.html#cb504-4"></a><span class="co"># annotates the ranges with the supporting p value</span></span>
<span id="cb504-5"><a href="peak-calling.html#cb504-5"></a>ctcf_peaks<span class="op">$</span>qvalue     =<span class="st"> </span><span class="kw">getQvalues</span>(ctcf_fit)</span>
<span id="cb504-6"><a href="peak-calling.html#cb504-6"></a>    </span>
<span id="cb504-7"><a href="peak-calling.html#cb504-7"></a><span class="co"># annotates the ranges with the calculated enrichment</span></span>
<span id="cb504-8"><a href="peak-calling.html#cb504-8"></a>ctcf_peaks<span class="op">$</span>enrichment =<span class="st"> </span><span class="kw">getEnrichment</span>(ctcf_fit)</span>
<span id="cb504-9"><a href="peak-calling.html#cb504-9"></a>    </span>
<span id="cb504-10"><a href="peak-calling.html#cb504-10"></a><span class="co"># selects the ranges which correspond to the enriched class</span></span>
<span id="cb504-11"><a href="peak-calling.html#cb504-11"></a>ctcf_peaks =<span class="st"> </span><span class="kw">subset</span>(ctcf_peaks, <span class="op">!</span><span class="kw">is.na</span>(component))</span>
<span id="cb504-12"><a href="peak-calling.html#cb504-12"></a>    </span>
<span id="cb504-13"><a href="peak-calling.html#cb504-13"></a><span class="co"># filter by a stringent q value threshold</span></span>
<span id="cb504-14"><a href="peak-calling.html#cb504-14"></a>ctcf_peaks =<span class="st"> </span><span class="kw">subset</span>(ctcf_peaks, qvalue <span class="op">&lt;</span><span class="st"> </span><span class="fl">0.01</span>)</span>
<span id="cb504-15"><a href="peak-calling.html#cb504-15"></a>    </span>
<span id="cb504-16"><a href="peak-calling.html#cb504-16"></a><span class="co"># order the peaks based on the q value</span></span>
<span id="cb504-17"><a href="peak-calling.html#cb504-17"></a>ctcf_peaks =<span class="st"> </span>ctcf_peaks[<span class="kw">order</span>(ctcf_peaks<span class="op">$</span>qvalue)]</span></code></pre></div>
<pre><code>## GRanges object with 724 ranges and 3 metadata columns:
##         seqnames            ranges strand | component       qvalue enrichment
##            &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt;    &lt;numeric&gt;  &lt;numeric&gt;
##     [1]    chr21 43939251-43939500      * |         1 4.69881e-140    1.37891
##     [2]    chr21 43646751-43647000      * |         1 2.52006e-137    1.42361
##     [3]    chr21 43810751-43811000      * |         1 1.86404e-121    1.30519
##     [4]    chr21 43939001-43939250      * |         1 2.10822e-121    1.19820
##     [5]    chr21 37712251-37712500      * |         1 6.35711e-118    1.70989
##     ...      ...               ...    ... .       ...          ...        ...
##   [720]    chr21 38172001-38172250      * |         1   0.00867374   0.951189
##   [721]    chr21 38806001-38806250      * |         1   0.00867374   0.951189
##   [722]    chr21 42009501-42009750      * |         1   0.00867374   0.656253
##   [723]    chr21 46153001-46153250      * |         1   0.00867374   0.951189
##   [724]    chr21 46294751-46295000      * |         1   0.00867374   0.722822
##   -------
##   seqinfo: 24 sequences from an unspecified genome</code></pre>
<p>After stringent q value filtering we are left with <span class="math inline">\(724\)</span> peaks. For the ease of downstream analysis, we will limit the sequence levels to
chromosome 21.</p>
<div class="sourceCode" id="cb506"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb506-1"><a href="peak-calling.html#cb506-1"></a><span class="kw">seqlevels</span>(ctcf_peaks, <span class="dt">pruning.mode=</span><span class="st">&#39;coarse&#39;</span>) =<span class="st"> &#39;chr21&#39;</span></span></code></pre></div>
<p>Let’s export the peaks into a .txt file which we can use the downstream in the analysis.</p>
<div class="sourceCode" id="cb507"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb507-1"><a href="peak-calling.html#cb507-1"></a><span class="co"># write the peaks loacations into a txt table</span></span>
<span id="cb507-2"><a href="peak-calling.html#cb507-2"></a><span class="kw">write.table</span>(ctcf_peaks, <span class="kw">file.path</span>(data_path, <span class="st">&#39;CTCF_peaks.txt&#39;</span>), </span>
<span id="cb507-3"><a href="peak-calling.html#cb507-3"></a>            <span class="dt">row.names=</span>F, <span class="dt">col.names=</span>T, <span class="dt">quote=</span>F, <span class="dt">sep=</span><span class="st">&#39;</span><span class="ch">\t</span><span class="st">&#39;</span>)</span></code></pre></div>
<p>We can now repeat the CTCF versus Input plot, and label significantly marked peaks. Using the count overlaps we mark which of our 1-kb regions contained significant peaks.</p>
<div class="sourceCode" id="cb508"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb508-1"><a href="peak-calling.html#cb508-1"></a><span class="co"># find enriched tilling windows</span></span>
<span id="cb508-2"><a href="peak-calling.html#cb508-2"></a>enriched_regions =<span class="st"> </span><span class="kw">countOverlaps</span>(tilling_window, ctcf_peaks) <span class="op">&gt;</span><span class="st"> </span><span class="dv">0</span></span></code></pre></div>
<div class="sourceCode" id="cb509"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb509-1"><a href="peak-calling.html#cb509-1"></a><span class="kw">library</span>(ggplot2)</span>
<span id="cb509-2"><a href="peak-calling.html#cb509-2"></a>cpm<span class="op">$</span>enriched_regions =<span class="st"> </span>enriched_regions </span>
<span id="cb509-3"><a href="peak-calling.html#cb509-3"></a></span>
<span id="cb509-4"><a href="peak-calling.html#cb509-4"></a><span class="kw">ggplot</span>(</span>
<span id="cb509-5"><a href="peak-calling.html#cb509-5"></a>    <span class="dt">data =</span> cpm, </span>
<span id="cb509-6"><a href="peak-calling.html#cb509-6"></a>    <span class="kw">aes</span>(</span>
<span id="cb509-7"><a href="peak-calling.html#cb509-7"></a>        <span class="dt">x =</span> GM12878_hg38_Input_r5.chr21.bam, </span>
<span id="cb509-8"><a href="peak-calling.html#cb509-8"></a>        <span class="dt">y =</span> GM12878_hg38_CTCF_r1.chr21.bam, </span>
<span id="cb509-9"><a href="peak-calling.html#cb509-9"></a>        <span class="dt">color =</span> enriched_regions</span>
<span id="cb509-10"><a href="peak-calling.html#cb509-10"></a>    )) <span class="op">+</span></span>
<span id="cb509-11"><a href="peak-calling.html#cb509-11"></a><span class="st">  </span><span class="kw">geom_point</span>() <span class="op">+</span></span>
<span id="cb509-12"><a href="peak-calling.html#cb509-12"></a><span class="st">  </span><span class="kw">geom_abline</span>(<span class="dt">slope =</span> <span class="dv">1</span>) <span class="op">+</span></span>
<span id="cb509-13"><a href="peak-calling.html#cb509-13"></a><span class="st">  </span><span class="kw">theme_bw</span>() <span class="op">+</span></span>
<span id="cb509-14"><a href="peak-calling.html#cb509-14"></a><span class="st">  </span><span class="kw">scale_fill_brewer</span>(<span class="dt">palette=</span><span class="st">&#39;Set2&#39;</span>) <span class="op">+</span></span>
<span id="cb509-15"><a href="peak-calling.html#cb509-15"></a><span class="st">  </span><span class="kw">theme</span>(</span>
<span id="cb509-16"><a href="peak-calling.html#cb509-16"></a>    <span class="dt">axis.text   =</span> <span class="kw">element_text</span>(<span class="dt">size=</span><span class="dv">10</span>, <span class="dt">face=</span><span class="st">&#39;bold&#39;</span>),</span>
<span id="cb509-17"><a href="peak-calling.html#cb509-17"></a>    <span class="dt">axis.title  =</span> <span class="kw">element_text</span>(<span class="dt">size=</span><span class="dv">14</span>,<span class="dt">face=</span><span class="st">&quot;bold&quot;</span>),</span>
<span id="cb509-18"><a href="peak-calling.html#cb509-18"></a>    <span class="dt">plot.title  =</span> <span class="kw">element_text</span>(<span class="dt">hjust =</span> <span class="fl">0.5</span>),</span>
<span id="cb509-19"><a href="peak-calling.html#cb509-19"></a>    <span class="dt">axis.text.x =</span> <span class="kw">element_text</span>(<span class="dt">angle =</span> <span class="dv">90</span>, <span class="dt">hjust =</span> <span class="dv">1</span>)) <span class="op">+</span></span>
<span id="cb509-20"><a href="peak-calling.html#cb509-20"></a><span class="st">  </span><span class="kw">xlab</span>(<span class="st">&#39;Input CPM&#39;</span>) <span class="op">+</span></span>
<span id="cb509-21"><a href="peak-calling.html#cb509-21"></a><span class="st">  </span><span class="kw">ylab</span>(<span class="st">&#39;CTCF CPM&#39;</span>) <span class="op">+</span></span>
<span id="cb509-22"><a href="peak-calling.html#cb509-22"></a><span class="st">  </span><span class="kw">ggtitle</span>(<span class="st">&#39;ChIP versus Input&#39;</span>) <span class="op">+</span></span>
<span id="cb509-23"><a href="peak-calling.html#cb509-23"></a><span class="st">  </span><span class="kw">scale_color_manual</span>(<span class="dt">values=</span><span class="kw">c</span>(<span class="st">&#39;gray&#39;</span>,<span class="st">&#39;red&#39;</span>))</span></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:peak-calling-sharp-peak-calling-plot"></span>
<img src="09-chip-seq-analysis_files/figure-html/peak-calling-sharp-peak-calling-plot-1.png" alt="Comparison of signal between ChIP and input samples. Red labeled dots correspond to called peaks." width="50%" />
<p class="caption">
FIGURE 9.13: Comparison of signal between ChIP and input samples. Red labeled dots correspond to called peaks.
</p>
</div>
<p>Figure <a href="peak-calling.html#fig:peak-calling-sharp-peak-calling-plot">9.13</a> shows that <code>normR</code>
identified all of the regions above the diagonal as statistically significant.
It has, however, labeled a significant number of regions below the diagonal.
Because of the sophisticated statistical model,
<code>normR</code> has greater sensitivity, and these peaks might really be enriched regions,
it is worth investigating the nature of these regions. This is left as an exercise
to the reader.</p>
<p>We can now create a genome browser screenshot around a peak region.
This will show us what kind of signal properties have contributed to the peak calling.
We would expect to see a strong, bell-shaped, enrichment in the ChIP sample, and
uniform noise in the Input sample.</p>
<p>Let us now visualize the signal around the most enriched peak. The following function takes as input a <strong>.bam</strong> file, and loads the bam into R.
It extends the reads to a size of 200 bp, and creates the coverage vector.</p>
<div class="sourceCode" id="cb510"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb510-1"><a href="peak-calling.html#cb510-1"></a><span class="co"># calculate the coverage for one bam file</span></span>
<span id="cb510-2"><a href="peak-calling.html#cb510-2"></a>calculateCoverage =<span class="st"> </span><span class="cf">function</span>(</span>
<span id="cb510-3"><a href="peak-calling.html#cb510-3"></a>  bam_file,</span>
<span id="cb510-4"><a href="peak-calling.html#cb510-4"></a>  <span class="dt">extend =</span> <span class="dv">200</span></span>
<span id="cb510-5"><a href="peak-calling.html#cb510-5"></a>){</span>
<span id="cb510-6"><a href="peak-calling.html#cb510-6"></a>  </span>
<span id="cb510-7"><a href="peak-calling.html#cb510-7"></a>  <span class="co"># load reads into R</span></span>
<span id="cb510-8"><a href="peak-calling.html#cb510-8"></a>  reads =<span class="st"> </span><span class="kw">readGAlignments</span>(bam_file)</span>
<span id="cb510-9"><a href="peak-calling.html#cb510-9"></a>  </span>
<span id="cb510-10"><a href="peak-calling.html#cb510-10"></a>  <span class="co"># convert reads into a GRanges object</span></span>
<span id="cb510-11"><a href="peak-calling.html#cb510-11"></a>  reads =<span class="st"> </span><span class="kw">granges</span>(reads)</span>
<span id="cb510-12"><a href="peak-calling.html#cb510-12"></a>  </span>
<span id="cb510-13"><a href="peak-calling.html#cb510-13"></a>  <span class="co"># resize the reads to 200bp</span></span>
<span id="cb510-14"><a href="peak-calling.html#cb510-14"></a>  reads =<span class="st"> </span><span class="kw">resize</span>(reads, <span class="dt">width=</span>extend, <span class="dt">fix=</span><span class="st">&#39;start&#39;</span>)</span>
<span id="cb510-15"><a href="peak-calling.html#cb510-15"></a>  </span>
<span id="cb510-16"><a href="peak-calling.html#cb510-16"></a>  <span class="co"># get the coverage vector</span></span>
<span id="cb510-17"><a href="peak-calling.html#cb510-17"></a>  cov   =<span class="st"> </span><span class="kw">coverage</span>(reads)</span>
<span id="cb510-18"><a href="peak-calling.html#cb510-18"></a>  </span>
<span id="cb510-19"><a href="peak-calling.html#cb510-19"></a>  <span class="co"># normalize the coverage vector to the sequencing depth</span></span>
<span id="cb510-20"><a href="peak-calling.html#cb510-20"></a>  cov =<span class="st"> </span><span class="kw">round</span>(cov <span class="op">*</span><span class="st"> </span>(<span class="dv">1000000</span><span class="op">/</span><span class="kw">length</span>(reads)),<span class="dv">2</span>)</span>
<span id="cb510-21"><a href="peak-calling.html#cb510-21"></a>  </span>
<span id="cb510-22"><a href="peak-calling.html#cb510-22"></a>  <span class="co"># convert the coverage go a GRanges object</span></span>
<span id="cb510-23"><a href="peak-calling.html#cb510-23"></a>  cov   =<span class="st"> </span><span class="kw">as</span>(cov, <span class="st">&#39;GRanges&#39;</span>)</span>
<span id="cb510-24"><a href="peak-calling.html#cb510-24"></a>  </span>
<span id="cb510-25"><a href="peak-calling.html#cb510-25"></a>  <span class="co"># keep only chromosome 21</span></span>
<span id="cb510-26"><a href="peak-calling.html#cb510-26"></a>  <span class="kw">seqlevels</span>(cov, <span class="dt">pruning.mode=</span><span class="st">&#39;coarse&#39;</span>) =<span class="st"> &#39;chr21&#39;</span></span>
<span id="cb510-27"><a href="peak-calling.html#cb510-27"></a>  <span class="kw">return</span>(cov)</span>
<span id="cb510-28"><a href="peak-calling.html#cb510-28"></a>}</span></code></pre></div>
<p>Let’s apply the function to the ChIP and input samples.</p>
<div class="sourceCode" id="cb511"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb511-1"><a href="peak-calling.html#cb511-1"></a><span class="co"># calculate coverage for the ChIP file</span></span>
<span id="cb511-2"><a href="peak-calling.html#cb511-2"></a>ctcf_cov =<span class="st"> </span><span class="kw">calculateCoverage</span>(chip_file)</span>
<span id="cb511-3"><a href="peak-calling.html#cb511-3"></a></span>
<span id="cb511-4"><a href="peak-calling.html#cb511-4"></a><span class="co"># calculate coverage for the control file</span></span>
<span id="cb511-5"><a href="peak-calling.html#cb511-5"></a>cont_cov =<span class="st"> </span><span class="kw">calculateCoverage</span>(control_file)</span></code></pre></div>
<p>Using <code>Gviz</code>, we will construct the layered tracks.
First, we layout the genome coordinates:</p>
<div class="sourceCode" id="cb512"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb512-1"><a href="peak-calling.html#cb512-1"></a><span class="co"># load Gviz and get the chromosome coordinates</span></span>
<span id="cb512-2"><a href="peak-calling.html#cb512-2"></a><span class="kw">library</span>(Gviz)</span>
<span id="cb512-3"><a href="peak-calling.html#cb512-3"></a>chr_track  =<span class="st"> </span><span class="kw">IdeogramTrack</span>(<span class="st">&#39;chr21&#39;</span>, <span class="st">&#39;hg38&#39;</span>)</span>
<span id="cb512-4"><a href="peak-calling.html#cb512-4"></a>axis       =<span class="st"> </span><span class="kw">GenomeAxisTrack</span>(</span>
<span id="cb512-5"><a href="peak-calling.html#cb512-5"></a>    <span class="dt">range =</span> <span class="kw">GRanges</span>(<span class="st">&#39;chr21&#39;</span>, <span class="kw">IRanges</span>(<span class="dv">1</span>, <span class="dt">width=</span>seqlengths))</span>
<span id="cb512-6"><a href="peak-calling.html#cb512-6"></a>)</span></code></pre></div>
<p>Then, the peak locations:</p>
<div class="sourceCode" id="cb513"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb513-1"><a href="peak-calling.html#cb513-1"></a><span class="co"># peaks track</span></span>
<span id="cb513-2"><a href="peak-calling.html#cb513-2"></a>peaks_track =<span class="st"> </span><span class="kw">AnnotationTrack</span>(ctcf_peaks, <span class="dt">name =</span> <span class="st">&quot;CTCF Peaks&quot;</span>)</span></code></pre></div>
<p>And finally, the signal files:</p>
<div class="sourceCode" id="cb514"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb514-1"><a href="peak-calling.html#cb514-1"></a>chip_track  =<span class="st"> </span><span class="kw">DataTrack</span>(</span>
<span id="cb514-2"><a href="peak-calling.html#cb514-2"></a>    <span class="dt">range =</span> ctcf_cov,   </span>
<span id="cb514-3"><a href="peak-calling.html#cb514-3"></a>    <span class="dt">name  =</span> <span class="st">&quot;CTCF&quot;</span>,  </span>
<span id="cb514-4"><a href="peak-calling.html#cb514-4"></a>    <span class="dt">type  =</span> <span class="st">&#39;h&#39;</span>, </span>
<span id="cb514-5"><a href="peak-calling.html#cb514-5"></a>    <span class="dt">lwd   =</span> <span class="dv">3</span></span>
<span id="cb514-6"><a href="peak-calling.html#cb514-6"></a>)</span>
<span id="cb514-7"><a href="peak-calling.html#cb514-7"></a></span>
<span id="cb514-8"><a href="peak-calling.html#cb514-8"></a>cont_track  =<span class="st"> </span><span class="kw">DataTrack</span>(</span>
<span id="cb514-9"><a href="peak-calling.html#cb514-9"></a>    <span class="dt">range =</span> cont_cov,   </span>
<span id="cb514-10"><a href="peak-calling.html#cb514-10"></a>    <span class="dt">name  =</span> <span class="st">&quot;Input&quot;</span>, </span>
<span id="cb514-11"><a href="peak-calling.html#cb514-11"></a>    <span class="dt">type  =</span> <span class="st">&#39;h&#39;</span>, </span>
<span id="cb514-12"><a href="peak-calling.html#cb514-12"></a>    <span class="dt">lwd=</span><span class="dv">3</span></span>
<span id="cb514-13"><a href="peak-calling.html#cb514-13"></a>)</span></code></pre></div>
<div class="sourceCode" id="cb515"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb515-1"><a href="peak-calling.html#cb515-1"></a><span class="kw">plotTracks</span>(</span>
<span id="cb515-2"><a href="peak-calling.html#cb515-2"></a>    <span class="dt">trackList =</span> <span class="kw">list</span>(chr_track, axis, peaks_track, chip_track, cont_track), </span>
<span id="cb515-3"><a href="peak-calling.html#cb515-3"></a>    <span class="dt">sizes     =</span> <span class="kw">c</span>(.<span class="dv">2</span>,.<span class="dv">5</span>,.<span class="dv">5</span>,<span class="dv">1</span>,<span class="dv">1</span>), </span>
<span id="cb515-4"><a href="peak-calling.html#cb515-4"></a>    <span class="dt">background.title =</span> <span class="st">&quot;black&quot;</span>,</span>
<span id="cb515-5"><a href="peak-calling.html#cb515-5"></a>    <span class="dt">from =</span> <span class="kw">start</span>(ctcf_peaks)[<span class="dv">1</span>] <span class="op">-</span><span class="st"> </span><span class="dv">1000</span>,</span>
<span id="cb515-6"><a href="peak-calling.html#cb515-6"></a>    <span class="dt">to   =</span> <span class="kw">end</span>(ctcf_peaks)[<span class="dv">1</span>]   <span class="op">+</span><span class="st"> </span><span class="dv">1000</span></span>
<span id="cb515-7"><a href="peak-calling.html#cb515-7"></a>)</span></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:peak-calling-signal-profile-plot"></span>
<img src="09-chip-seq-analysis_files/figure-html/peak-calling-signal-profile-plot-1.png" alt="ChIP and Input signal profile around the peak centers." width="50%" />
<p class="caption">
FIGURE 9.14: ChIP and Input signal profile around the peak centers.
</p>
</div>
<p>In Figure <a href="peak-calling.html#fig:peak-calling-signal-profile-plot">9.14</a>, the ChIP sample looks as expected.
Although the Input sample shows an enrichment,
it is important to compare the scales on both samples. The normalized ChIP signal goes up
to <span class="math inline">\(2500\)</span>, while the maximum value in the input sample is only <span class="math inline">\(60\)</span>.</p>
</div>
<div id="peak-calling-broad-regions" class="section level3">
<h3><span class="header-section-number">9.6.3</span> Peak calling: Broad regions</h3>
<p></p>
<p>We will now use <code>normR</code> to call peaks for the H3K36me3 histone modification,
which is associated with gene bodies of expressed genes. We define the ChIP and Input files:</p>
<div class="sourceCode" id="cb516"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb516-1"><a href="peak-calling.html#cb516-1"></a><span class="co"># fetch the ChIP-file for H3K36me3</span></span>
<span id="cb516-2"><a href="peak-calling.html#cb516-2"></a>chip_file    =<span class="st"> </span><span class="kw">file.path</span>(data_path, <span class="st">&#39;GM12878_hg38_H3K36me3.chr21.bam&#39;</span>)</span>
<span id="cb516-3"><a href="peak-calling.html#cb516-3"></a></span>
<span id="cb516-4"><a href="peak-calling.html#cb516-4"></a><span class="co"># fetch the corresponding input file</span></span>
<span id="cb516-5"><a href="peak-calling.html#cb516-5"></a>control_file =<span class="st"> </span><span class="kw">file.path</span>(data_path, <span class="st">&#39;GM12878_hg38_Input_r5.chr21.bam&#39;</span>)</span></code></pre></div>
<p>Because H3K36 regions span broad domains, it is necessary to increase the
tilling window size which will be used for counting.
Using the <code>countConfiguration()</code> function, we will set the tilling window size
to 5000 base pairs.</p>
<div class="sourceCode" id="cb517"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb517-1"><a href="peak-calling.html#cb517-1"></a><span class="kw">library</span>(normr)</span>
<span id="cb517-2"><a href="peak-calling.html#cb517-2"></a><span class="co"># define the window width for the counting</span></span>
<span id="cb517-3"><a href="peak-calling.html#cb517-3"></a>countConfiguration =<span class="st"> </span><span class="kw">countConfigSingleEnd</span>(<span class="dt">binsize =</span> <span class="dv">5000</span>)</span></code></pre></div>
<div class="sourceCode" id="cb518"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb518-1"><a href="peak-calling.html#cb518-1"></a><span class="co"># find broad peaks using enrichR</span></span>
<span id="cb518-2"><a href="peak-calling.html#cb518-2"></a>h3k36_fit =<span class="st"> </span><span class="kw">enrichR</span>(</span>
<span id="cb518-3"><a href="peak-calling.html#cb518-3"></a>            </span>
<span id="cb518-4"><a href="peak-calling.html#cb518-4"></a>            <span class="co"># ChIP file</span></span>
<span id="cb518-5"><a href="peak-calling.html#cb518-5"></a>            <span class="dt">treatment   =</span> chip_file,</span>
<span id="cb518-6"><a href="peak-calling.html#cb518-6"></a>            </span>
<span id="cb518-7"><a href="peak-calling.html#cb518-7"></a>            <span class="co"># control file</span></span>
<span id="cb518-8"><a href="peak-calling.html#cb518-8"></a>            <span class="dt">control     =</span> control_file,</span>
<span id="cb518-9"><a href="peak-calling.html#cb518-9"></a>            </span>
<span id="cb518-10"><a href="peak-calling.html#cb518-10"></a>            <span class="co"># genome version</span></span>
<span id="cb518-11"><a href="peak-calling.html#cb518-11"></a>            <span class="dt">genome      =</span> <span class="st">&quot;hg38&quot;</span>,</span>
<span id="cb518-12"><a href="peak-calling.html#cb518-12"></a>            <span class="dt">verbose     =</span> <span class="ot">FALSE</span>,</span>
<span id="cb518-13"><a href="peak-calling.html#cb518-13"></a>            </span>
<span id="cb518-14"><a href="peak-calling.html#cb518-14"></a>            <span class="co"># window size for counting</span></span>
<span id="cb518-15"><a href="peak-calling.html#cb518-15"></a>            <span class="dt">countConfig =</span> countConfiguration)</span></code></pre></div>
<div class="sourceCode" id="cb519"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb519-1"><a href="peak-calling.html#cb519-1"></a><span class="kw">summary</span>(h3k36_fit)</span></code></pre></div>
<pre><code>## NormRFit-class object
## 
## Type:                  &#39;enrichR&#39;
## Number of Regions:     617665
## Number of Components:  2
## Theta* (naive bg):     0.197
## Background component B: 1
## 
## +++ Results of fit +++ 
## Mixture Proportions:
## Background       Class 1    
##      85.4%         14.6%    
## Theta:
## Background       Class 1    
##      0.138         0.442    
## 
## Bayesian Information Criterion:  741525
## 
## +++ Results of binomial test +++ 
## T-Filter threshold: 5
## Number of Regions filtered out: 610736
## Significantly different from background B based on q-values:
## TOTAL:
##           ***       **        *        .              n.s.
## Bins        0     1005      314      381      237     4992
## %        0.00     9.18    12.04    15.52    17.68    45.58
## ---
## Signif. codes:  0 &#39;***&#39; 0.001 &#39;**&#39; 0.01 &#39;*&#39; 0.05 &#39;.&#39; 0.1 &#39;  &#39; 1 &#39;n.s.&#39;</code></pre>
<p>The summary function shows that we get 1937 enriched regions. We will extract enriched regions, and plot them in the same way we did for the
CTCF.</p>
<div class="sourceCode" id="cb521"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb521-1"><a href="peak-calling.html#cb521-1"></a><span class="co"># get the locations of broad peaks</span></span>
<span id="cb521-2"><a href="peak-calling.html#cb521-2"></a>h3k36_peaks            =<span class="st"> </span><span class="kw">getRanges</span>(h3k36_fit)</span>
<span id="cb521-3"><a href="peak-calling.html#cb521-3"></a></span>
<span id="cb521-4"><a href="peak-calling.html#cb521-4"></a><span class="co"># extract the qvalue and enrichment</span></span>
<span id="cb521-5"><a href="peak-calling.html#cb521-5"></a>h3k36_peaks<span class="op">$</span>qvalue     =<span class="st"> </span><span class="kw">getQvalues</span>(h3k36_fit)</span>
<span id="cb521-6"><a href="peak-calling.html#cb521-6"></a>h3k36_peaks<span class="op">$</span>enrichment =<span class="st"> </span><span class="kw">getEnrichment</span>(h3k36_fit)</span>
<span id="cb521-7"><a href="peak-calling.html#cb521-7"></a></span>
<span id="cb521-8"><a href="peak-calling.html#cb521-8"></a><span class="co"># select proper peaks</span></span>
<span id="cb521-9"><a href="peak-calling.html#cb521-9"></a>h3k36_peaks =<span class="st"> </span><span class="kw">subset</span>(h3k36_peaks, <span class="op">!</span><span class="kw">is.na</span>(component))</span>
<span id="cb521-10"><a href="peak-calling.html#cb521-10"></a>h3k36_peaks =<span class="st"> </span><span class="kw">subset</span>(h3k36_peaks, qvalue <span class="op">&lt;</span><span class="st"> </span><span class="fl">0.01</span>)</span>
<span id="cb521-11"><a href="peak-calling.html#cb521-11"></a>h3k36_peaks =<span class="st"> </span>h3k36_peaks[<span class="kw">order</span>(h3k36_peaks<span class="op">$</span>qvalue)]</span>
<span id="cb521-12"><a href="peak-calling.html#cb521-12"></a></span>
<span id="cb521-13"><a href="peak-calling.html#cb521-13"></a><span class="co"># collapse nearby enriched regions</span></span>
<span id="cb521-14"><a href="peak-calling.html#cb521-14"></a>h3k36_peaks =<span class="st"> </span><span class="kw">reduce</span>(h3k36_peaks)</span></code></pre></div>
<div class="sourceCode" id="cb522"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb522-1"><a href="peak-calling.html#cb522-1"></a><span class="co"># construct the data tracks for the H3K36me3 and Input files</span></span>
<span id="cb522-2"><a href="peak-calling.html#cb522-2"></a>h3k36_cov =<span class="st"> </span><span class="kw">calculateCoverage</span>(chip_file)</span>
<span id="cb522-3"><a href="peak-calling.html#cb522-3"></a>data_tracks =<span class="st"> </span><span class="kw">list</span>(</span>
<span id="cb522-4"><a href="peak-calling.html#cb522-4"></a>    <span class="dt">h3k36 =</span> <span class="kw">DataTrack</span>(h3k36_cov,  <span class="dt">name =</span> <span class="st">&#39;h3k36_cov&#39;</span>,  <span class="dt">type=</span><span class="st">&#39;h&#39;</span>, <span class="dt">lwd=</span><span class="dv">3</span>),</span>
<span id="cb522-5"><a href="peak-calling.html#cb522-5"></a>    <span class="dt">input =</span> <span class="kw">DataTrack</span>(cont_cov,   <span class="dt">name =</span> <span class="st">&#39;Input&#39;</span>,      <span class="dt">type=</span><span class="st">&#39;h&#39;</span>, <span class="dt">lwd=</span><span class="dv">3</span>)</span>
<span id="cb522-6"><a href="peak-calling.html#cb522-6"></a>)</span></code></pre></div>
<div class="sourceCode" id="cb523"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb523-1"><a href="peak-calling.html#cb523-1"></a><span class="co"># define the window for the visualization</span></span>
<span id="cb523-2"><a href="peak-calling.html#cb523-2"></a>start =<span class="st"> </span><span class="kw">min</span>(<span class="kw">start</span>(h3k36_peaks[<span class="dv">2</span>])) <span class="op">-</span><span class="st"> </span><span class="dv">25000</span></span>
<span id="cb523-3"><a href="peak-calling.html#cb523-3"></a>end   =<span class="st"> </span><span class="kw">max</span>(<span class="kw">end</span>(h3k36_peaks[<span class="dv">2</span>])) <span class="op">+</span><span class="st"> </span><span class="dv">25000</span></span>
<span id="cb523-4"><a href="peak-calling.html#cb523-4"></a></span>
<span id="cb523-5"><a href="peak-calling.html#cb523-5"></a><span class="co"># create the peak track</span></span>
<span id="cb523-6"><a href="peak-calling.html#cb523-6"></a>peak_track =<span class="st"> </span><span class="kw">AnnotationTrack</span>(<span class="kw">reduce</span>(h3k36_peaks), <span class="dt">name=</span><span class="st">&#39;H3K36me3&#39;</span>)</span>
<span id="cb523-7"><a href="peak-calling.html#cb523-7"></a></span>
<span id="cb523-8"><a href="peak-calling.html#cb523-8"></a><span class="co"># plots the enriched region</span></span>
<span id="cb523-9"><a href="peak-calling.html#cb523-9"></a><span class="kw">plotTracks</span>(</span>
<span id="cb523-10"><a href="peak-calling.html#cb523-10"></a>    <span class="dt">trackList =</span> <span class="kw">c</span>(chr_track, axis, gene_track, peak_track, data_tracks),</span>
<span id="cb523-11"><a href="peak-calling.html#cb523-11"></a>    <span class="dt">sizes     =</span> <span class="kw">c</span>(.<span class="dv">5</span>,.<span class="dv">5</span>,.<span class="dv">5</span>,.<span class="dv">1</span>,<span class="dv">1</span>,<span class="dv">1</span>), </span>
<span id="cb523-12"><a href="peak-calling.html#cb523-12"></a>    <span class="dt">background.title     =</span> <span class="st">&quot;black&quot;</span>,</span>
<span id="cb523-13"><a href="peak-calling.html#cb523-13"></a>    <span class="dt">collapseTranscripts  =</span> <span class="st">&quot;longest&quot;</span>, </span>
<span id="cb523-14"><a href="peak-calling.html#cb523-14"></a>    <span class="dt">transcriptAnnotation =</span> <span class="st">&quot;symbol&quot;</span>,</span>
<span id="cb523-15"><a href="peak-calling.html#cb523-15"></a>    <span class="dt">from =</span> start,</span>
<span id="cb523-16"><a href="peak-calling.html#cb523-16"></a>    <span class="dt">to   =</span> end</span>
<span id="cb523-17"><a href="peak-calling.html#cb523-17"></a>)</span></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:peak-calling-broad-gviz"></span>
<img src="09-chip-seq-analysis_files/figure-html/peak-calling-broad-gviz-1.png" alt="Visualization of H3K36me3 ChIP signal on a called broad peak." width="50%" />
<p class="caption">
FIGURE 9.15: Visualization of H3K36me3 ChIP signal on a called broad peak.
</p>
</div>
<p>Figure <a href="peak-calling.html#fig:peak-calling-broad-gviz">9.15</a> shows a highly enriched H3K36me3
region covering the gene body, as expected.</p>
</div>
<div id="peak-quality-control" class="section level3">
<h3><span class="header-section-number">9.6.4</span> Peak quality control</h3>
<p>Peak calling is not a mathematically defined procedure; it is impossible
to unambiguously define what a “peak” is. Therefore all of the peak
calling procedures use heuristics, and statistical models which have been
shown to work well in specific use cases.
After peak calling, it is always necessary to check
whether the defined peaks really are located in enriched regions, and in addition,
use prior knowledge to ascertain whether the peaks correspond to known biology.</p>
<p>Peak calling can falsely identify enriched regions if the input
sample is not sequenced to the proper depth. Because the input samples
correspond to <strong>de facto</strong> whole genome sequencing, and the ChIP procedure
enriches for a subset of the genome, it can often happen that many regions
in the genome are not sufficiently covered by the Input sample.
Such variability in the signal profile of Input samples can cause a region
to be defined as a peak, enriched in the ChIP sample, while in reality it is depleted in the
Input, due to under-sampling. For example, the figure in the previous chapter, showing
an enriched region H3K36me3 over a gene body, shows a large depletion in the Input
sample over the same region. Such depletion should be a concern and merit
further investigation.</p>
<p>The quality of enrichment can be checked by calculating the percentage of reads within peaks for both
ChIP and Input samples. ChIP samples should have a high percentage of reads in peaks,
while for the input samples, the percentage of reads should correspond to the
percentage of genome covered by peaks.</p>
<p>For transcription factor ChIP experiments, an important control is to determine whether
the peak regions contain sequences which are known to be bound
by the corresponding transcription factor - whether they contain
known transcription factor binding motifs.
Transcription factor binding motifs are sequence models which model the propensity
of binding DNA sequences.
Such sequence models can be downloaded from public databases and compared to see
whether there is a positional enrichment around our peaks.</p>
<p>We will now calculate the percentage of reads within peaks for the H3K36me3 experiment.
Subsequently, we will download the known CTCF sequence model, and compare it
to our peak regions.</p>
<div id="percentage-of-reads-in-peaks" class="section level4">
<h4><span class="header-section-number">9.6.4.1</span> Percentage of reads in peaks</h4>
<p>To calculate the reads in peaks, we will firstly extract the number of reads
in each tilling window from the <code>normR</code> produced fit object.
This is done using the <code>getCounts()</code> function.
We will then use the q-value to define which tilling windows correspond
to peaks, and count the number of reads within and outside peaks.</p>
<div class="sourceCode" id="cb524"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb524-1"><a href="peak-calling.html#cb524-1"></a><span class="co"># extract, per tilling window, counts from the fit object</span></span>
<span id="cb524-2"><a href="peak-calling.html#cb524-2"></a>h3k36_counts =<span class="st"> </span><span class="kw">data.frame</span>(<span class="kw">getCounts</span>(h3k36_fit))</span>
<span id="cb524-3"><a href="peak-calling.html#cb524-3"></a>    </span>
<span id="cb524-4"><a href="peak-calling.html#cb524-4"></a><span class="co"># change the column names of the data.frame</span></span>
<span id="cb524-5"><a href="peak-calling.html#cb524-5"></a><span class="kw">colnames</span>(h3k36_counts) =<span class="st"> </span><span class="kw">c</span>(<span class="st">&#39;Input&#39;</span>,<span class="st">&#39;H3K36me3&#39;</span>)</span>
<span id="cb524-6"><a href="peak-calling.html#cb524-6"></a>    </span>
<span id="cb524-7"><a href="peak-calling.html#cb524-7"></a><span class="co"># extract the q-value corresponding to each bin</span></span>
<span id="cb524-8"><a href="peak-calling.html#cb524-8"></a>h3k36_counts<span class="op">$</span>qvalue =<span class="st"> </span><span class="kw">getQvalues</span>(h3k36_fit)</span>
<span id="cb524-9"><a href="peak-calling.html#cb524-9"></a>    </span>
<span id="cb524-10"><a href="peak-calling.html#cb524-10"></a><span class="co"># define which regions are peaks using a q value cutoff</span></span>
<span id="cb524-11"><a href="peak-calling.html#cb524-11"></a>h3k36_counts<span class="op">$</span>enriched[<span class="kw">is.na</span>(h3k36_counts<span class="op">$</span>qvalue)]  =<span class="st"> &#39;Not Peak&#39;</span></span>
<span id="cb524-12"><a href="peak-calling.html#cb524-12"></a>h3k36_counts<span class="op">$</span>enriched[h3k36_counts<span class="op">$</span>qvalue <span class="op">&gt;</span><span class="st"> </span><span class="fl">0.05</span>]  =<span class="st"> &#39;Not Peak&#39;</span></span>
<span id="cb524-13"><a href="peak-calling.html#cb524-13"></a>h3k36_counts<span class="op">$</span>enriched[h3k36_counts<span class="op">$</span>qvalue <span class="op">&lt;=</span><span class="st"> </span><span class="fl">0.05</span>] =<span class="st"> &#39;Peak&#39;</span></span>
<span id="cb524-14"><a href="peak-calling.html#cb524-14"></a>    </span>
<span id="cb524-15"><a href="peak-calling.html#cb524-15"></a><span class="co"># remove the q value column</span></span>
<span id="cb524-16"><a href="peak-calling.html#cb524-16"></a>h3k36_counts<span class="op">$</span>qvalue =<span class="st"> </span><span class="ot">NULL</span> </span>
<span id="cb524-17"><a href="peak-calling.html#cb524-17"></a>    </span>
<span id="cb524-18"><a href="peak-calling.html#cb524-18"></a><span class="co"># reshape the data.frame into a long format</span></span>
<span id="cb524-19"><a href="peak-calling.html#cb524-19"></a>h3k36_counts_df =<span class="st"> </span>tidyr<span class="op">::</span><span class="kw">pivot_longer</span>(</span>
<span id="cb524-20"><a href="peak-calling.html#cb524-20"></a>    <span class="dt">data      =</span> h3k36_counts, </span>
<span id="cb524-21"><a href="peak-calling.html#cb524-21"></a>    <span class="dt">cols      =</span> <span class="op">-</span>enriched,</span>
<span id="cb524-22"><a href="peak-calling.html#cb524-22"></a>    <span class="dt">names_to  =</span> <span class="st">&#39;experiment&#39;</span>,</span>
<span id="cb524-23"><a href="peak-calling.html#cb524-23"></a>    <span class="dt">values_to =</span> <span class="st">&#39;counts&#39;</span></span>
<span id="cb524-24"><a href="peak-calling.html#cb524-24"></a>)</span>
<span id="cb524-25"><a href="peak-calling.html#cb524-25"></a>    </span>
<span id="cb524-26"><a href="peak-calling.html#cb524-26"></a><span class="co"># sum the number of reads in the Peak and Not Peak regions</span></span>
<span id="cb524-27"><a href="peak-calling.html#cb524-27"></a>h3k36_counts_df =<span class="st"> </span><span class="kw">group_by</span>(<span class="dt">.data =</span> h3k36_counts_df, experiment, enriched)</span>
<span id="cb524-28"><a href="peak-calling.html#cb524-28"></a>h3k36_counts_df =<span class="st"> </span><span class="kw">summarize</span>(<span class="dt">.data =</span> h3k36_counts_df, <span class="dt">num_of_reads =</span> <span class="kw">sum</span>(counts))</span>
<span id="cb524-29"><a href="peak-calling.html#cb524-29"></a>    </span>
<span id="cb524-30"><a href="peak-calling.html#cb524-30"></a><span class="co"># calculate the percentage of reads.</span></span>
<span id="cb524-31"><a href="peak-calling.html#cb524-31"></a>h3k36_counts_df       =<span class="st"> </span><span class="kw">group_by</span>(<span class="dt">.data =</span> h3k36_counts_df, experiment)</span>
<span id="cb524-32"><a href="peak-calling.html#cb524-32"></a>h3k36_counts_df       =<span class="st"> </span><span class="kw">mutate</span>(<span class="dt">.data =</span> h3k36_counts_df, <span class="dt">total=</span><span class="kw">sum</span>(num_of_reads))</span>
<span id="cb524-33"><a href="peak-calling.html#cb524-33"></a>h3k36_counts_df<span class="op">$</span>percentage =<span class="st"> </span><span class="kw">with</span>(h3k36_counts_df, <span class="kw">round</span>(num_of_reads<span class="op">/</span>total,<span class="dv">2</span>))</span></code></pre></div>
<pre><code>## # A tibble: 4 x 5
## # Groups:   experiment [2]
##   experiment enriched num_of_reads  total percentage
##   &lt;chr&gt;      &lt;chr&gt;           &lt;int&gt;  &lt;int&gt;      &lt;dbl&gt;
## 1 H3K36me3   Not Peak        67623 158616      0.43 
## 2 H3K36me3   Peak            90993 158616      0.570
## 3 Input      Not Peak       492369 648196      0.76 
## 4 Input      Peak           155827 648196      0.24</code></pre>
<p>We can now plot the percentage of reads in peaks:</p>
<div class="sourceCode" id="cb526"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb526-1"><a href="peak-calling.html#cb526-1"></a><span class="kw">ggplot</span>(</span>
<span id="cb526-2"><a href="peak-calling.html#cb526-2"></a>    <span class="dt">data =</span> h3k36_counts_df, </span>
<span id="cb526-3"><a href="peak-calling.html#cb526-3"></a>    <span class="kw">aes</span>(</span>
<span id="cb526-4"><a href="peak-calling.html#cb526-4"></a>        <span class="dt">x =</span> experiment, </span>
<span id="cb526-5"><a href="peak-calling.html#cb526-5"></a>        <span class="dt">y =</span> percentage, </span>
<span id="cb526-6"><a href="peak-calling.html#cb526-6"></a>        <span class="dt">fill =</span> enriched</span>
<span id="cb526-7"><a href="peak-calling.html#cb526-7"></a>    )) <span class="op">+</span></span>
<span id="cb526-8"><a href="peak-calling.html#cb526-8"></a><span class="st">    </span><span class="kw">geom_bar</span>(<span class="dt">stat=</span><span class="st">&#39;identity&#39;</span>, <span class="dt">position=</span><span class="st">&#39;dodge&#39;</span>) <span class="op">+</span></span>
<span id="cb526-9"><a href="peak-calling.html#cb526-9"></a><span class="st">    </span><span class="kw">theme_bw</span>() <span class="op">+</span></span>
<span id="cb526-10"><a href="peak-calling.html#cb526-10"></a><span class="st">    </span><span class="kw">theme</span>(</span>
<span id="cb526-11"><a href="peak-calling.html#cb526-11"></a>        <span class="dt">axis.text =</span> <span class="kw">element_text</span>(<span class="dt">size=</span><span class="dv">10</span>, <span class="dt">face=</span><span class="st">&#39;bold&#39;</span>),</span>
<span id="cb526-12"><a href="peak-calling.html#cb526-12"></a>        <span class="dt">axis.title =</span> <span class="kw">element_text</span>(<span class="dt">size=</span><span class="dv">12</span>,<span class="dt">face=</span><span class="st">&quot;bold&quot;</span>),</span>
<span id="cb526-13"><a href="peak-calling.html#cb526-13"></a>        <span class="dt">plot.title =</span> <span class="kw">element_text</span>(<span class="dt">hjust =</span> <span class="fl">0.5</span>)) <span class="op">+</span></span>
<span id="cb526-14"><a href="peak-calling.html#cb526-14"></a><span class="st">    </span><span class="kw">xlab</span>(<span class="st">&#39;Experiment&#39;</span>) <span class="op">+</span></span>
<span id="cb526-15"><a href="peak-calling.html#cb526-15"></a><span class="st">    </span><span class="kw">ylab</span>(<span class="st">&#39;Percetage of reads in region&#39;</span>) <span class="op">+</span></span>
<span id="cb526-16"><a href="peak-calling.html#cb526-16"></a><span class="st">    </span><span class="kw">ggtitle</span>(<span class="st">&#39;Percentage of reads in peaks for H3K36me3&#39;</span>) <span class="op">+</span></span>
<span id="cb526-17"><a href="peak-calling.html#cb526-17"></a><span class="st">    </span><span class="kw">scale_fill_manual</span>(<span class="dt">values=</span><span class="kw">c</span>(<span class="st">&#39;gray&#39;</span>,<span class="st">&#39;red&#39;</span>))</span></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:peak-quality-counts-plot"></span>
<img src="09-chip-seq-analysis_files/figure-html/peak-quality-counts-plot-1.png" alt="Percentage of ChIP reads in called peaks. Higher percentage indicates higher ChIP quality." width="50%" />
<p class="caption">
FIGURE 9.16: Percentage of ChIP reads in called peaks. Higher percentage indicates higher ChIP quality.
</p>
</div>
<p>Figure <a href="peak-calling.html#fig:peak-quality-counts-plot">9.16</a> shows that the ChIP sample is
clearly enriched in the peak regions.
The percentage of reads in peaks will depend on the quality of the antibody (strength of
enrichment), and the size of peaks which are bound by the protein of interest.
If the total size of peaks is small, relative to the genome size, we can expect that
the percentage of reads in peaks will be small.</p>
</div>
<div id="dna-motifs-on-peaks" class="section level4">
<h4><span class="header-section-number">9.6.4.2</span> DNA motifs on peaks</h4>
<p></p>
<p>Well-studied transcription factors have publicly available transcription
factor binding motifs.
If such a model is available for our transcription factor of interest, we
can use it to check the quality of our ChIP data.
Two common measures are used for this purpose:</p>
<ol style="list-style-type: decimal">
<li>Percentage of peaks containing the motif of interest.</li>
<li>Positional distribution of the motif - the distribution of motif locations should be centered on the peak centers.</li>
</ol>
<div id="representing-motifs-as-matrices" class="section level5">
<h5><span class="header-section-number">9.6.4.2.1</span> Representing motifs as matrices</h5>
<p>In order to calculate the percentage of CTCF peaks which contain a known CTCF
motif. We need to find the CTCF motif and have the computational tools to search for that motif. The DNA binding motifs can be extracted from the <code>MotifDB</code> Bioconductor
database. The <code>MotifDB</code> is an agglomeration of multiple motif databases.</p>
<div class="sourceCode" id="cb527"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb527-1"><a href="peak-calling.html#cb527-1"></a><span class="co"># load the MotifDB package</span></span>
<span id="cb527-2"><a href="peak-calling.html#cb527-2"></a><span class="kw">library</span>(MotifDb)</span>
<span id="cb527-3"><a href="peak-calling.html#cb527-3"></a></span>
<span id="cb527-4"><a href="peak-calling.html#cb527-4"></a><span class="co"># fetch the CTCF motif from the data base</span></span>
<span id="cb527-5"><a href="peak-calling.html#cb527-5"></a>motifs =<span class="st"> </span><span class="kw">query</span>(<span class="kw">query</span>(MotifDb, <span class="st">&#39;Hsapiens&#39;</span>), <span class="st">&#39;CTCF&#39;</span>)</span>
<span id="cb527-6"><a href="peak-calling.html#cb527-6"></a></span>
<span id="cb527-7"><a href="peak-calling.html#cb527-7"></a><span class="co"># show all available ctcf motifs</span></span>
<span id="cb527-8"><a href="peak-calling.html#cb527-8"></a>motifs</span></code></pre></div>
<pre><code>## MotifDb object of length 12
## | Created from downloaded public sources: 2013-Aug-30
## | 12 position frequency matrices from 8 sources:
## |        HOCOMOCOv10:    2
## | HOCOMOCOv11-core-A:    2
## |        JASPAR_2014:    1
## |        JASPAR_CORE:    1
## |       SwissRegulon:    2
## |         jaspar2016:    1
## |         jaspar2018:    2
## |          jolma2013:    1
## | 1 organism/s
## |           Hsapiens:   12
## Hsapiens-SwissRegulon-CTCFL.SwissRegulon 
## Hsapiens-SwissRegulon-CTCF.SwissRegulon 
## Hsapiens-HOCOMOCOv10-CTCFL_HUMAN.H10MO.A 
## Hsapiens-HOCOMOCOv10-CTCF_HUMAN.H10MO.A 
## Hsapiens-HOCOMOCOv11-core-A-CTCFL_HUMAN.H11MO.0.A 
## ...
## Hsapiens-JASPAR_2014-CTCF-MA0139.1 
## Hsapiens-jaspar2016-CTCF-MA0139.1 
## Hsapiens-jaspar2018-CTCF-MA0139.1 
## Hsapiens-jaspar2018-CTCFL-MA1102.1 
## Hsapiens-jolma2013-CTCF</code></pre>
<p>We will extract the CTCF from the <code>MotifDB</code> <span class="citation">(Khan, Fornes, Stigliani, et al. <a href="#ref-khan_2018" role="doc-biblioref">2018</a>)</span> database.</p>
<div class="sourceCode" id="cb529"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb529-1"><a href="peak-calling.html#cb529-1"></a><span class="co"># based on the MotifDB version, the location of the CTCF motif</span></span>
<span id="cb529-2"><a href="peak-calling.html#cb529-2"></a><span class="co"># might change, if you do not get the expected results please try</span></span>
<span id="cb529-3"><a href="peak-calling.html#cb529-3"></a><span class="co"># to subset with different indices</span></span>
<span id="cb529-4"><a href="peak-calling.html#cb529-4"></a>ctcf_motif  =<span class="st"> </span>motifs[[<span class="dv">1</span>]]</span></code></pre></div>
<p>The motifs are usually represented as matrices of 4-by-N dimensions. In the matrix, each of 4 rows correspond to one nucleotide (A, C, G, T).
The number of columns designates the width of the region bound by the transcription factor or the length of the motif that the protein recognizes.
Each element of the matrix contains the probability of observing the corresponding
nucleotide on this position.
For example, for following the CTCF matrix in Table <a href="peak-calling.html#tab:peakqualityshow">9.1</a>, the probability of observing a thymine at
the first position of the motif,<span class="math inline">\(p_{i=1,k=4}\)</span> , is 0.57 (1st column, 4th row).
Such a matrix, where each column is a probability distribution over a sequence of nucleotides,
is called a position frequency matrix (PFM). In some sources, this matrix is also called “position probability matrix (PPM)”. One way to construct such matrices is to get experimentally verified sequences that are bound by the protein of interest and then to use a motif-finding algorithm.</p>
<table>
<caption><span id="tab:peakqualityshow">TABLE 9.1: </span>Position Frequency Matrix (PFM) for the CTCF motif</caption>
<thead>
<tr class="header">
<th></th>
<th align="right">1</th>
<th align="right">2</th>
<th align="right">3</th>
<th align="right">4</th>
<th align="right">5</th>
<th align="right">6</th>
<th align="right">7</th>
<th align="right">8</th>
<th align="right">9</th>
<th align="right">10</th>
<th align="right">11</th>
<th align="right">12</th>
<th align="right">13</th>
<th align="right">14</th>
<th align="right">15</th>
<th align="right">16</th>
<th align="right">17</th>
<th align="right">18</th>
<th align="right">19</th>
<th align="right">20</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>A</td>
<td align="right">0.17</td>
<td align="right">0.23</td>
<td align="right">0.29</td>
<td align="right">0.10</td>
<td align="right">0.33</td>
<td align="right">0.06</td>
<td align="right">0.05</td>
<td align="right">0.04</td>
<td align="right">0.02</td>
<td align="right">0</td>
<td align="right">0.25</td>
<td align="right">0.00</td>
<td align="right">0</td>
<td align="right">0.05</td>
<td align="right">0.25</td>
<td align="right">0.00</td>
<td align="right">0.17</td>
<td align="right">0</td>
<td align="right">0.02</td>
<td align="right">0.19</td>
</tr>
<tr class="even">
<td>C</td>
<td align="right">0.42</td>
<td align="right">0.28</td>
<td align="right">0.30</td>
<td align="right">0.32</td>
<td align="right">0.11</td>
<td align="right">0.33</td>
<td align="right">0.56</td>
<td align="right">0.00</td>
<td align="right">0.96</td>
<td align="right">1</td>
<td align="right">0.67</td>
<td align="right">0.69</td>
<td align="right">1</td>
<td align="right">0.04</td>
<td align="right">0.07</td>
<td align="right">0.42</td>
<td align="right">0.15</td>
<td align="right">0</td>
<td align="right">0.06</td>
<td align="right">0.43</td>
</tr>
<tr class="odd">
<td>G</td>
<td align="right">0.25</td>
<td align="right">0.23</td>
<td align="right">0.26</td>
<td align="right">0.27</td>
<td align="right">0.42</td>
<td align="right">0.55</td>
<td align="right">0.05</td>
<td align="right">0.83</td>
<td align="right">0.01</td>
<td align="right">0</td>
<td align="right">0.03</td>
<td align="right">0.00</td>
<td align="right">0</td>
<td align="right">0.02</td>
<td align="right">0.53</td>
<td align="right">0.55</td>
<td align="right">0.05</td>
<td align="right">1</td>
<td align="right">0.87</td>
<td align="right">0.15</td>
</tr>
<tr class="even">
<td>T</td>
<td align="right">0.16</td>
<td align="right">0.27</td>
<td align="right">0.15</td>
<td align="right">0.31</td>
<td align="right">0.14</td>
<td align="right">0.06</td>
<td align="right">0.33</td>
<td align="right">0.13</td>
<td align="right">0.00</td>
<td align="right">0</td>
<td align="right">0.06</td>
<td align="right">0.31</td>
<td align="right">0</td>
<td align="right">0.89</td>
<td align="right">0.15</td>
<td align="right">0.03</td>
<td align="right">0.62</td>
<td align="right">0</td>
<td align="right">0.05</td>
<td align="right">0.23</td>
</tr>
</tbody>
</table>
<p>Such a matrix can be used to calculate the probability that the transcription
factor will bind to any given sequence. However, computationally, it is easier to work with summation rather than multiplication. In addition, the simple probabilistic model does not take the background probability of observing a certain base in a given position. We can correct for background base frequencies by dividing the individual probability, <span class="math inline">\(p_{i,k}\)</span> in each cell of the matrix by the background base probability for a given base, <span class="math inline">\(B_k\)</span>. We can then take the logarithm of that quantity to calculate a log-likelihood and bring everything to log-scale as follows <span class="math inline">\(Score_{i,k}=log_2(p_{i,k}/B_k)\)</span>. We can now calculate a score for any given
sequence by summing up the base-position-specific scores we obtain from the log-scaled matrix. This matrix is formally called position-specific scoring matrix (PSSM) or position-specific weight matrix (PWM). We can use this matrix to scan the genome in a sliding window manner and calculate a score for each window. Usually, a cutoff value is needed to call a motif hit. The higher the score you get from the PWM for a particular sequence, the better it is. The traditional algorithms we will use in the following sections use 80% of the maximum rescaled score you can obtain from a PWM as the default cutoff for a hit. The rescaling is simple min-max rescaling where you rescale the score by subtracting the minimum score and dividing that by <span class="math inline">\(max(PWMscore)-min(PWMscore)\)</span>. The motif scanning approach is illustrated in Figure <a href="peak-calling.html#fig:FigurePWMScanning">9.17</a>. In this example, ACACT is not considered a hit because its score only corresponds to only <span class="math inline">\(15.6\)</span> % of the rescaled maximum score.</p>

<div class="figure" style="text-align: center"><span id="fig:FigurePWMScanning"></span>
<img src="Figures/PWMScanning.png" alt="PWM scanning principle. A genomic sequence is scanned by a PWM matrix. This matrix is used to measure how likely it is that the transcription factor will bind each nucleotide in each position. Here we are looking at how likely it is that our TF will bind to the sequence ACACT. The score for this sequence is -3.6. The maximal score obtainable by the PWM is 7.2 and minimum is -5.6. After min-max rescaling, -3.6 corresponds to a 15% score and ACACT is not considered a hit." width="50%" />
<p class="caption">
FIGURE 9.17: PWM scanning principle. A genomic sequence is scanned by a PWM matrix. This matrix is used to measure how likely it is that the transcription factor will bind each nucleotide in each position. Here we are looking at how likely it is that our TF will bind to the sequence ACACT. The score for this sequence is -3.6. The maximal score obtainable by the PWM is 7.2 and minimum is -5.6. After min-max rescaling, -3.6 corresponds to a 15% score and ACACT is not considered a hit.
</p>
</div>
</div>
<div id="representing-motifs-as-sequence-logos" class="section level5">
<h5><span class="header-section-number">9.6.4.2.2</span> Representing motifs as sequence logos</h5>
<p>Using the PFM, we can calculate the information content of each position in the matrix.
The information content quantifies the contribution of each nucleotide to the
cumulative binding preference. This tells us how important each nucleotide is for the binding. It additionally allows us to visually represent the probability matrices as sequence logos.
The information content is quantified as relative entropy. It ranges from <span class="math inline">\(0\)</span>, no information,
to <span class="math inline">\(2\)</span>, maximal information. For a column in the PFM, the entropy is calculated as follows:</p>
<p><span class="math display">\[
entropy = -\sum\limits_{k=1}^n p_{i,k}\log_2(p_{i,k}) 
\]</span>
<span class="math inline">\(p_{i,k}\)</span> is the probability of observing base <span class="math inline">\(k\)</span> in the column <span class="math inline">\(i\)</span> of the PFM. In other words, <span class="math inline">\(p_{i,k}\)</span> is simply the value of the cell in the PFM. The entropy value is high when the probabilities of each base are similar and low when it is much more probable that only one base occur in a given column. The relative portion comes from the fact that we compare the entropy we calculated for a column to the maximum entropy we can obtain. If the all bases are equally likely for a position in the PFM, then we will have the maximum entropy and we compare our original entropy to that maximum entropy. The maximum entropy is simply <span class="math inline">\(log_2{n}\)</span> where <span class="math inline">\(n\)</span> is number of letters in the alphabet. In our case we have 4 letters A,C,G and T. The information content is then simply subtracting the observed entropy for a column from the maximum entropy, which translates to the following equation:</p>
<p><span class="math display">\[
IC=log_2(n)+\sum\limits_{k=1}^n p_{i,k}\log_2(p_{i,k}) 
\]</span>
The information content, <span class="math inline">\(IC\)</span>, in the preceding equation, will be high if a base has a high probability of occurrence and low if all bases are more or less equally likely to occur.</p>
<p>We can visualize the matrix by visualizing the letters weighted by their probabilities in the PFM. This approach is shown on the left-hand side of Figure <a href="peak-calling.html#fig:peak-quality-seqLogo-plot">9.18</a>. In addition, we can also calculate the information content per column to weight the probabilities. This means that the columns that have very frequent letters will be higher. This approach is shown on the right-hand side of Figure <a href="peak-calling.html#fig:peak-quality-seqLogo-plot">9.18</a>. We will use below the <code>seqLogo</code> package to visualize the CTCF motif in the two different ways we described above.</p>
<div class="figure" style="text-align: center"><span id="fig:peak-quality-seqLogo-plot"></span>
<img src="Figures/CTCF_Motif.png" alt="CTCF sequence motif visualized as a sequence logo. Y-axis ranges from zero to two, and corresponds to the amount of information each base in the motif contributes to the overall motif. The larger the letter, the greater the probability of observing just one defined base on the designated position." width="70%" />
<p class="caption">
FIGURE 9.18: CTCF sequence motif visualized as a sequence logo. Y-axis ranges from zero to two, and corresponds to the amount of information each base in the motif contributes to the overall motif. The larger the letter, the greater the probability of observing just one defined base on the designated position.
</p>
</div>
</div>
<div id="percentage-of-peaks-with-the-motif" class="section level5">
<h5><span class="header-section-number">9.6.4.2.3</span> Percentage of peaks with the motif</h5>
<p>Since we now understand how DNA motifs are used we can start annotating the CTCF peaks with the motif. First, we will extend the peak
regions to +/- 200 bp around the peak center.
Because the average fragment size is 200 bp, 400 nucleotides is the
expected variation in the position of the true binding location.</p>
<div class="sourceCode" id="cb530"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb530-1"><a href="peak-calling.html#cb530-1"></a><span class="co"># extend the peak regions</span></span>
<span id="cb530-2"><a href="peak-calling.html#cb530-2"></a>ctcf_peaks_resized =<span class="st"> </span><span class="kw">resize</span>(ctcf_peaks, <span class="dt">width =</span> <span class="dv">400</span>, <span class="dt">fix =</span> <span class="st">&#39;center&#39;</span>)</span></code></pre></div>
<p>Now we use the <code>BSgenome</code> package to
extract the sequences corresponding to the peak regions.</p>
<div class="sourceCode" id="cb531"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb531-1"><a href="peak-calling.html#cb531-1"></a><span class="co"># load the human genome sequence</span></span>
<span id="cb531-2"><a href="peak-calling.html#cb531-2"></a><span class="kw">library</span>(BSgenome.Hsapiens.UCSC.hg38)</span>
<span id="cb531-3"><a href="peak-calling.html#cb531-3"></a></span>
<span id="cb531-4"><a href="peak-calling.html#cb531-4"></a><span class="co"># extract the sequences around the peaks</span></span>
<span id="cb531-5"><a href="peak-calling.html#cb531-5"></a>seq =<span class="st"> </span><span class="kw">getSeq</span>(BSgenome.Hsapiens.UCSC.hg38, ctcf_peaks_resized)</span></code></pre></div>
<p>Once we have extracted the sequences, we can use the CTCF motif to
scan each sequence and determine the probability of CTCF binding.
For this we use the <code>TFBSTools</code> <span class="citation">(Tan and Lenhard <a href="#ref-TFBSTools" role="doc-biblioref">2016</a>)</span> package.</p>
<p>We first convert the raw probability matrix into a <code>PWMMatrix</code> object,
which can then be used for efficient scanning.</p>
<div class="sourceCode" id="cb532"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb532-1"><a href="peak-calling.html#cb532-1"></a><span class="co"># load the TFBS tools package</span></span>
<span id="cb532-2"><a href="peak-calling.html#cb532-2"></a><span class="kw">library</span>(TFBSTools)</span>
<span id="cb532-3"><a href="peak-calling.html#cb532-3"></a></span>
<span id="cb532-4"><a href="peak-calling.html#cb532-4"></a><span class="co"># convert the matrix into a PWM object</span></span>
<span id="cb532-5"><a href="peak-calling.html#cb532-5"></a>ctcf_pwm =<span class="st"> </span><span class="kw">PWMatrix</span>(</span>
<span id="cb532-6"><a href="peak-calling.html#cb532-6"></a>    <span class="dt">ID =</span> <span class="st">&#39;CTCF&#39;</span>, </span>
<span id="cb532-7"><a href="peak-calling.html#cb532-7"></a>    <span class="dt">profileMatrix =</span> ctcf_motif</span>
<span id="cb532-8"><a href="peak-calling.html#cb532-8"></a>)</span></code></pre></div>
<p>We can now use the <code>searchSeq()</code> function to scan each sequence for the motif occurrence.
Because the motif matrices are given a continuous binding score, we need to set a cutoff to
determine when a sequence contains the motif, and when it doesn’t.
The cutoff is set by determining the maximal possible score produced by the motif matrix;
a percentage of that score is then taken as the threshold value.
For example, if the best sequence would have a score of 1.4 of being bound,
then we define a threshold of 80% of 1.4, which is 1.12; and any sequence which
scores less than 1.12 would not be marked as being bound by the protein.</p>
<p>For the CTCF, we mark any peak containing a sequence with &gt; 80% of the maximal rescaled score or “relative score” as a positive hit.</p>
<pre><code>##   seqnames source feature start end absScore relScore strand   ID
## 1        1   TFBS    TFBS    44  63     11.9    0.921      - CTCF
## 2        1   TFBS    TFBS   102 121     11.0    0.839      - CTCF
## 3        2   TFBS    TFBS   151 170     11.5    0.881      + CTCF
## 4        4   TFBS    TFBS   294 313     11.9    0.921      - CTCF
## 5        4   TFBS    TFBS   352 371     11.0    0.839      - CTCF
## 6        5   TFBS    TFBS   164 183     10.9    0.831      - CTCF</code></pre>
<p>A common diagnostic plot is to graph a reverse cumulative distribution of
peak occurrences.
On the x-axis we rank the peaks, with the most highly enriched peak in the
first position, and the least enriched peak in the last position.
We then walk from the lowest to the highest ranking and measure the
percentage of peaks containing the motif.</p>
<div class="sourceCode" id="cb534"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb534-1"><a href="peak-calling.html#cb534-1"></a><span class="co"># label which peaks contain CTCF motifs</span></span>
<span id="cb534-2"><a href="peak-calling.html#cb534-2"></a>motif_hits_df =<span class="st"> </span><span class="kw">data.frame</span>(</span>
<span id="cb534-3"><a href="peak-calling.html#cb534-3"></a>  <span class="dt">peak_order     =</span> <span class="dv">1</span><span class="op">:</span><span class="kw">length</span>(ctcf_peaks)</span>
<span id="cb534-4"><a href="peak-calling.html#cb534-4"></a>)</span>
<span id="cb534-5"><a href="peak-calling.html#cb534-5"></a>motif_hits_df<span class="op">$</span>contains_motif =<span class="st"> </span>motif_hits_df<span class="op">$</span>peak_order <span class="op">%in%</span><span class="st"> </span>hits<span class="op">$</span>seqnames</span>
<span id="cb534-6"><a href="peak-calling.html#cb534-6"></a>motif_hits_df =<span class="st"> </span>motif_hits_df[<span class="kw">order</span>(<span class="op">-</span>motif_hits_df<span class="op">$</span>peak_order),]</span>
<span id="cb534-7"><a href="peak-calling.html#cb534-7"></a></span>
<span id="cb534-8"><a href="peak-calling.html#cb534-8"></a><span class="co"># calculate the percentage of peaks with motif for peaks of descending strength</span></span>
<span id="cb534-9"><a href="peak-calling.html#cb534-9"></a>motif_hits_df<span class="op">$</span>perc_peaks =<span class="st"> </span><span class="kw">with</span>(motif_hits_df, </span>
<span id="cb534-10"><a href="peak-calling.html#cb534-10"></a>                                <span class="kw">cumsum</span>(contains_motif) <span class="op">/</span><span class="st"> </span><span class="kw">max</span>(peak_order))</span>
<span id="cb534-11"><a href="peak-calling.html#cb534-11"></a>motif_hits_df<span class="op">$</span>perc_peaks =<span class="st"> </span><span class="kw">round</span>(motif_hits_df<span class="op">$</span>perc_peaks, <span class="dv">2</span>)</span></code></pre></div>
<p>We can now visualize the percentage of peaks with matching CTCF motif.</p>
<div class="sourceCode" id="cb535"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb535-1"><a href="peak-calling.html#cb535-1"></a><span class="co"># plot the cumulative distribution of motif hit percentages</span></span>
<span id="cb535-2"><a href="peak-calling.html#cb535-2"></a><span class="kw">ggplot</span>(</span>
<span id="cb535-3"><a href="peak-calling.html#cb535-3"></a>    motif_hits_df, </span>
<span id="cb535-4"><a href="peak-calling.html#cb535-4"></a>    <span class="kw">aes</span>(</span>
<span id="cb535-5"><a href="peak-calling.html#cb535-5"></a>        <span class="dt">x =</span> peak_order, </span>
<span id="cb535-6"><a href="peak-calling.html#cb535-6"></a>        <span class="dt">y =</span> perc_peaks</span>
<span id="cb535-7"><a href="peak-calling.html#cb535-7"></a>    )) <span class="op">+</span></span>
<span id="cb535-8"><a href="peak-calling.html#cb535-8"></a><span class="st">  </span><span class="kw">geom_line</span>(<span class="dt">size=</span><span class="dv">2</span>) <span class="op">+</span></span>
<span id="cb535-9"><a href="peak-calling.html#cb535-9"></a><span class="st">  </span><span class="kw">theme_bw</span>() <span class="op">+</span></span>
<span id="cb535-10"><a href="peak-calling.html#cb535-10"></a><span class="st">  </span><span class="kw">theme</span>(</span>
<span id="cb535-11"><a href="peak-calling.html#cb535-11"></a>    <span class="dt">axis.text =</span> <span class="kw">element_text</span>(<span class="dt">size=</span><span class="dv">10</span>, <span class="dt">face=</span><span class="st">&#39;bold&#39;</span>),</span>
<span id="cb535-12"><a href="peak-calling.html#cb535-12"></a>    <span class="dt">axis.title =</span> <span class="kw">element_text</span>(<span class="dt">size=</span><span class="dv">14</span>,<span class="dt">face=</span><span class="st">&quot;bold&quot;</span>),</span>
<span id="cb535-13"><a href="peak-calling.html#cb535-13"></a>    <span class="dt">plot.title =</span> <span class="kw">element_text</span>(<span class="dt">hjust =</span> <span class="fl">0.5</span>)) <span class="op">+</span></span>
<span id="cb535-14"><a href="peak-calling.html#cb535-14"></a><span class="st">  </span><span class="kw">xlab</span>(<span class="st">&#39;Peak rank&#39;</span>) <span class="op">+</span></span>
<span id="cb535-15"><a href="peak-calling.html#cb535-15"></a><span class="st">  </span><span class="kw">ylab</span>(<span class="st">&#39;Percetage of peaks with motif&#39;</span>) <span class="op">+</span></span>
<span id="cb535-16"><a href="peak-calling.html#cb535-16"></a><span class="st">  </span><span class="kw">ggtitle</span>(<span class="st">&#39;Percentage of CTCF peaks with the CTCF motif&#39;</span>)</span></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:peak-quality-scan-dist-plot"></span>
<img src="09-chip-seq-analysis_files/figure-html/peak-quality-scan-dist-plot-1.png" alt="Percentage of peaks containing the motif. Higher percentage indicates a better ChIP-experiment, and a better peak calling procedure." width="50%" />
<p class="caption">
FIGURE 9.19: Percentage of peaks containing the motif. Higher percentage indicates a better ChIP-experiment, and a better peak calling procedure.
</p>
</div>
<p>Figure <a href="peak-calling.html#fig:peak-quality-scan-dist-plot">9.19</a>
shows that, when we take all peaks into account, ~45% of
the peaks contain a CTCF motif.
This is an excellent percentage and indicates a high-quality ChIP experiment.
Our inability to locate the motif in ~50% of the sequences does not
necessarily need to be a consequence of a poor experiment; sometimes
it is a result of the molecular mechanism by which the transcription factor
binds. If a transcription factor has multiple binding modes, which are context
dependent, for example, if the transcription factor binds indirectly to
a subset of regions, through
an interacting partner, we do not have to observe a motif.</p>
</div>
</div>
<div id="motif-localization" class="section level4">
<h4><span class="header-section-number">9.6.4.3</span> Motif localization</h4>
<p>If the ChIP experiment was performed properly, we would expect the motif
to be localized just below the summit of each peak.
By plotting the motif localization around ChIP peaks, we are quantifying
the uncertainty in the peak location.</p>
<p>We will firstly resize our peaks into regions around +/−1-kb around the peak
center.</p>
<div class="sourceCode" id="cb536"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb536-1"><a href="peak-calling.html#cb536-1"></a><span class="co"># resize the region around peaks to +/- 1kb</span></span>
<span id="cb536-2"><a href="peak-calling.html#cb536-2"></a>ctcf_peaks_resized =<span class="st"> </span><span class="kw">resize</span>(ctcf_peaks, <span class="dt">width =</span> <span class="dv">2000</span>, <span class="dt">fix=</span><span class="st">&#39;center&#39;</span>)</span></code></pre></div>
<p>Now we perform the motif localization, as before.</p>
<div class="sourceCode" id="cb537"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb537-1"><a href="peak-calling.html#cb537-1"></a><span class="co"># fetch the sequence</span></span>
<span id="cb537-2"><a href="peak-calling.html#cb537-2"></a>seq =<span class="st"> </span><span class="kw">getSeq</span>(BSgenome.Hsapiens.UCSC.hg38,ctcf_peaks_resized)</span>
<span id="cb537-3"><a href="peak-calling.html#cb537-3"></a></span>
<span id="cb537-4"><a href="peak-calling.html#cb537-4"></a><span class="co"># convert the motif matrix to PWM, and scan the peaks</span></span>
<span id="cb537-5"><a href="peak-calling.html#cb537-5"></a>ctcf_pwm    =<span class="st"> </span><span class="kw">PWMatrix</span>(<span class="dt">ID =</span> <span class="st">&#39;CTCF&#39;</span>, <span class="dt">profileMatrix =</span> ctcf_motif)</span>
<span id="cb537-6"><a href="peak-calling.html#cb537-6"></a>hits =<span class="st"> </span><span class="kw">searchSeq</span>(ctcf_pwm, seq, <span class="dt">min.score=</span><span class="st">&quot;80%&quot;</span>, <span class="dt">strand=</span><span class="st">&quot;*&quot;</span>)</span>
<span id="cb537-7"><a href="peak-calling.html#cb537-7"></a>hits =<span class="st"> </span><span class="kw">as.data.frame</span>(hits)</span></code></pre></div>
<p>We now construct a plot, where the
X-axis represents the +/- 1000 nucleotides around the peak, while the
Y-axis shows the motif enrichment at each position.</p>
<div class="sourceCode" id="cb538"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb538-1"><a href="peak-calling.html#cb538-1"></a><span class="co"># set the position relative to the start</span></span>
<span id="cb538-2"><a href="peak-calling.html#cb538-2"></a>hits<span class="op">$</span>position =<span class="st"> </span>hits<span class="op">$</span>start <span class="op">-</span><span class="st"> </span><span class="dv">1000</span> </span>
<span id="cb538-3"><a href="peak-calling.html#cb538-3"></a></span>
<span id="cb538-4"><a href="peak-calling.html#cb538-4"></a><span class="co"># plot the motif hits around peaks</span></span>
<span id="cb538-5"><a href="peak-calling.html#cb538-5"></a><span class="kw">ggplot</span>(<span class="dt">data=</span>hits, <span class="kw">aes</span>(position)) <span class="op">+</span></span>
<span id="cb538-6"><a href="peak-calling.html#cb538-6"></a><span class="st">  </span><span class="kw">geom_density</span>(<span class="dt">size=</span><span class="dv">2</span>) <span class="op">+</span></span>
<span id="cb538-7"><a href="peak-calling.html#cb538-7"></a><span class="st">  </span><span class="kw">theme_bw</span>() <span class="op">+</span></span>
<span id="cb538-8"><a href="peak-calling.html#cb538-8"></a><span class="st">  </span><span class="kw">geom_vline</span>(<span class="dt">xintercept =</span> <span class="dv">0</span>, <span class="dt">linetype=</span><span class="dv">2</span>, <span class="dt">color=</span><span class="st">&#39;red&#39;</span>, <span class="dt">size=</span><span class="dv">2</span>) <span class="op">+</span></span>
<span id="cb538-9"><a href="peak-calling.html#cb538-9"></a><span class="st">  </span><span class="kw">xlab</span>(<span class="st">&#39;Position around the CTCF peaks&#39;</span>) <span class="op">+</span></span>
<span id="cb538-10"><a href="peak-calling.html#cb538-10"></a><span class="st">  </span><span class="kw">ylab</span>(<span class="st">&#39;Per position percentage</span><span class="ch">\n</span><span class="st">of motif occurence&#39;</span>) <span class="op">+</span></span>
<span id="cb538-11"><a href="peak-calling.html#cb538-11"></a><span class="st">  </span><span class="kw">theme</span>(</span>
<span id="cb538-12"><a href="peak-calling.html#cb538-12"></a>    <span class="dt">axis.text =</span> <span class="kw">element_text</span>(<span class="dt">size=</span><span class="dv">10</span>, <span class="dt">face=</span><span class="st">&#39;bold&#39;</span>),</span>
<span id="cb538-13"><a href="peak-calling.html#cb538-13"></a>    <span class="dt">axis.title =</span> <span class="kw">element_text</span>(<span class="dt">size=</span><span class="dv">14</span>,<span class="dt">face=</span><span class="st">&quot;bold&quot;</span>),</span>
<span id="cb538-14"><a href="peak-calling.html#cb538-14"></a>    <span class="dt">plot.title =</span> <span class="kw">element_text</span>(<span class="dt">hjust =</span> <span class="fl">0.5</span>))</span></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:chip-quality-motifloc-plot"></span>
<img src="09-chip-seq-analysis_files/figure-html/chip-quality-motifloc-plot-1.png" alt="Transcription factor sequence motif localization with respect to the defined binding sites." width="50%" />
<p class="caption">
FIGURE 9.20: Transcription factor sequence motif localization with respect to the defined binding sites.
</p>
</div>
<p>We can in Figure <a href="peak-calling.html#fig:chip-quality-motifloc-plot">9.20</a>, see that the bulk of motif
hits are found in a region of <span class="math inline">\(+/-\)</span> 250 bp around the peak centers.
This means that the peak calling procedure was quite precise.</p>
</div>
</div>
<div id="peak-annotation" class="section level3">
<h3><span class="header-section-number">9.6.5</span> Peak annotation</h3>
<p>As the final step of quality control we will visualize the distribution
of peaks in different functional genomic regions.
The purpose of the analysis is to check whether the location of the peaks
conforms our prior knowledge.
This analysis is equivalent to constructing distributions for reads.</p>
<p>Firstly we download the human gene models and construct the annotation hierarchy.</p>
<div class="sourceCode" id="cb539"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb539-1"><a href="peak-calling.html#cb539-1"></a><span class="co"># download the annotation</span></span>
<span id="cb539-2"><a href="peak-calling.html#cb539-2"></a>hub =<span class="st"> </span><span class="kw">AnnotationHub</span>()</span>
<span id="cb539-3"><a href="peak-calling.html#cb539-3"></a>gtf =<span class="st"> </span>hub[[<span class="st">&#39;AH61126&#39;</span>]]</span>
<span id="cb539-4"><a href="peak-calling.html#cb539-4"></a><span class="kw">seqlevels</span>(gtf, <span class="dt">pruning.mode=</span><span class="st">&#39;coarse&#39;</span>) =<span class="st"> &#39;21&#39;</span></span>
<span id="cb539-5"><a href="peak-calling.html#cb539-5"></a><span class="kw">seqlevels</span>(gtf, <span class="dt">pruning.mode=</span><span class="st">&#39;coarse&#39;</span>) =<span class="st"> </span><span class="kw">paste0</span>(<span class="st">&#39;chr&#39;</span>, <span class="kw">seqlevels</span>(gtf))</span>
<span id="cb539-6"><a href="peak-calling.html#cb539-6"></a></span>
<span id="cb539-7"><a href="peak-calling.html#cb539-7"></a><span class="co"># create the annotation hierarchy</span></span>
<span id="cb539-8"><a href="peak-calling.html#cb539-8"></a>annotation_list =<span class="st"> </span><span class="kw">GRangesList</span>(</span>
<span id="cb539-9"><a href="peak-calling.html#cb539-9"></a>  <span class="dt">tss    =</span> <span class="kw">promoters</span>(<span class="kw">subset</span>(gtf, type<span class="op">==</span><span class="st">&#39;gene&#39;</span>), <span class="dv">1000</span>, <span class="dv">1000</span>),</span>
<span id="cb539-10"><a href="peak-calling.html#cb539-10"></a>  <span class="dt">exon   =</span> <span class="kw">subset</span>(gtf, type<span class="op">==</span><span class="st">&#39;exon&#39;</span>),</span>
<span id="cb539-11"><a href="peak-calling.html#cb539-11"></a>  <span class="dt">intron =</span> <span class="kw">subset</span>(gtf, type<span class="op">==</span><span class="st">&#39;gene&#39;</span>)</span>
<span id="cb539-12"><a href="peak-calling.html#cb539-12"></a>)</span></code></pre></div>
<p>The following function finds the genomic location of each peak, annotates
the peaks using the hierarchical prioritization,
and calculates the summary statistics.</p>
<p>The function contains four major parts:</p>
<ol style="list-style-type: decimal">
<li>Creating a disjoint set of peak regions.</li>
<li>Finding the overlapping annotation for each peak.</li>
<li>Annotating each peak with the corresponding annotation class.</li>
<li>Calculating summary statistics</li>
</ol>
<div class="sourceCode" id="cb540"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb540-1"><a href="peak-calling.html#cb540-1"></a><span class="co"># function which annotates the location of each peak</span></span>
<span id="cb540-2"><a href="peak-calling.html#cb540-2"></a>annotatePeaks =<span class="st"> </span><span class="cf">function</span>(peaks, annotation_list, name){</span>
<span id="cb540-3"><a href="peak-calling.html#cb540-3"></a>  </span>
<span id="cb540-4"><a href="peak-calling.html#cb540-4"></a>  <span class="co"># ------------------------------------------------ #</span></span>
<span id="cb540-5"><a href="peak-calling.html#cb540-5"></a>  <span class="co"># 1. getting disjoint regions</span></span>
<span id="cb540-6"><a href="peak-calling.html#cb540-6"></a>  <span class="co"># collapse touching enriched regions</span></span>
<span id="cb540-7"><a href="peak-calling.html#cb540-7"></a>  peaks =<span class="st"> </span><span class="kw">reduce</span>(peaks)</span>
<span id="cb540-8"><a href="peak-calling.html#cb540-8"></a>  </span>
<span id="cb540-9"><a href="peak-calling.html#cb540-9"></a>  <span class="co"># ------------------------------------------------ #</span></span>
<span id="cb540-10"><a href="peak-calling.html#cb540-10"></a>  <span class="co"># 2. overlapping peaks and annotation</span></span>
<span id="cb540-11"><a href="peak-calling.html#cb540-11"></a>  <span class="co"># find overlaps between the peaks and annotation_list</span></span>
<span id="cb540-12"><a href="peak-calling.html#cb540-12"></a>  result =<span class="st"> </span><span class="kw">as.data.frame</span>(<span class="kw">findOverlaps</span>(peaks, annotation_list))</span>
<span id="cb540-13"><a href="peak-calling.html#cb540-13"></a>  </span>
<span id="cb540-14"><a href="peak-calling.html#cb540-14"></a>  <span class="co"># ------------------------------------------------ #</span></span>
<span id="cb540-15"><a href="peak-calling.html#cb540-15"></a>  <span class="co"># 3. annotating peaks</span></span>
<span id="cb540-16"><a href="peak-calling.html#cb540-16"></a>  <span class="co"># fetch annotation names</span></span>
<span id="cb540-17"><a href="peak-calling.html#cb540-17"></a>  result<span class="op">$</span>annotation =<span class="st"> </span><span class="kw">names</span>(annotation_list)[result<span class="op">$</span>subjectHits]</span>
<span id="cb540-18"><a href="peak-calling.html#cb540-18"></a>  </span>
<span id="cb540-19"><a href="peak-calling.html#cb540-19"></a>  <span class="co"># rank by annotation precedence</span></span>
<span id="cb540-20"><a href="peak-calling.html#cb540-20"></a>  result =<span class="st"> </span>result[<span class="kw">order</span>(result<span class="op">$</span>subjectHits),]    </span>
<span id="cb540-21"><a href="peak-calling.html#cb540-21"></a>  </span>
<span id="cb540-22"><a href="peak-calling.html#cb540-22"></a>  <span class="co"># remove overlapping annotations</span></span>
<span id="cb540-23"><a href="peak-calling.html#cb540-23"></a>  result =<span class="st"> </span><span class="kw">subset</span>(result, <span class="op">!</span><span class="kw">duplicated</span>(queryHits))</span>
<span id="cb540-24"><a href="peak-calling.html#cb540-24"></a>  </span>
<span id="cb540-25"><a href="peak-calling.html#cb540-25"></a>  <span class="co"># ------------------------------------------------ #</span></span>
<span id="cb540-26"><a href="peak-calling.html#cb540-26"></a>  <span class="co"># 4. calculating statistics</span></span>
<span id="cb540-27"><a href="peak-calling.html#cb540-27"></a>  <span class="co"># count the number of peaks in each annotation category</span></span>
<span id="cb540-28"><a href="peak-calling.html#cb540-28"></a>  result =<span class="st"> </span><span class="kw">group_by</span>(<span class="dt">.data =</span> result, annotation)</span>
<span id="cb540-29"><a href="peak-calling.html#cb540-29"></a>  result =<span class="st"> </span><span class="kw">summarise</span>(<span class="dt">.data =</span> result, <span class="dt">counts =</span> <span class="kw">length</span>(annotation))</span>
<span id="cb540-30"><a href="peak-calling.html#cb540-30"></a>  </span>
<span id="cb540-31"><a href="peak-calling.html#cb540-31"></a>  <span class="co"># fetch the number of intergenic peaks</span></span>
<span id="cb540-32"><a href="peak-calling.html#cb540-32"></a>  result =<span class="st"> </span><span class="kw">rbind</span>(result, </span>
<span id="cb540-33"><a href="peak-calling.html#cb540-33"></a>                 <span class="kw">data.frame</span>(<span class="dt">annotation =</span> <span class="st">&#39;intergenic&#39;</span>, </span>
<span id="cb540-34"><a href="peak-calling.html#cb540-34"></a>                            <span class="dt">counts     =</span> <span class="kw">length</span>(peaks) <span class="op">-</span><span class="st"> </span><span class="kw">sum</span>(result<span class="op">$</span>counts)))</span>
<span id="cb540-35"><a href="peak-calling.html#cb540-35"></a>  </span>
<span id="cb540-36"><a href="peak-calling.html#cb540-36"></a>  result<span class="op">$</span>frequency  =<span class="st"> </span><span class="kw">with</span>(result, <span class="kw">round</span>(counts<span class="op">/</span><span class="kw">sum</span>(counts),<span class="dv">2</span>))</span>
<span id="cb540-37"><a href="peak-calling.html#cb540-37"></a>  result<span class="op">$</span>experiment =<span class="st"> </span>name</span>
<span id="cb540-38"><a href="peak-calling.html#cb540-38"></a>  </span>
<span id="cb540-39"><a href="peak-calling.html#cb540-39"></a>  <span class="kw">return</span>(result)</span>
<span id="cb540-40"><a href="peak-calling.html#cb540-40"></a>}</span></code></pre></div>
<p>Using the above defined <code>annotatePeaks()</code> function we will now annotate CTCF
and H3K36me3 peaks. Firstly we create a list which contains both CTCF and H3K36me3 peaks.</p>
<div class="sourceCode" id="cb541"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb541-1"><a href="peak-calling.html#cb541-1"></a>peak_list =<span class="st"> </span><span class="kw">list</span>(</span>
<span id="cb541-2"><a href="peak-calling.html#cb541-2"></a>    <span class="dt">CTCF     =</span> ctcf_peaks, </span>
<span id="cb541-3"><a href="peak-calling.html#cb541-3"></a>    <span class="dt">H3K36me3 =</span> h3k36_peaks</span>
<span id="cb541-4"><a href="peak-calling.html#cb541-4"></a>)</span></code></pre></div>
<p>Using the <code>lapply()</code> function we apply the <code>annotatePeaks()</code> function
on each element of the list.</p>
<div class="sourceCode" id="cb542"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb542-1"><a href="peak-calling.html#cb542-1"></a><span class="co"># calculate the distribution of peaks in annotation for each experiment</span></span>
<span id="cb542-2"><a href="peak-calling.html#cb542-2"></a>annot_peaks_list =<span class="st"> </span><span class="kw">lapply</span>(<span class="kw">names</span>(peak_list), <span class="cf">function</span>(peak_name){</span>
<span id="cb542-3"><a href="peak-calling.html#cb542-3"></a>  <span class="kw">annotatePeaks</span>(peak_list[[peak_name]], annotation_list, peak_name)</span>
<span id="cb542-4"><a href="peak-calling.html#cb542-4"></a>})</span></code></pre></div>
<p>We use the <code>dplyr::bind_rows()</code> function to combine the CTCF and H3K36me3 annotation
statistics into one data frame.</p>
<div class="sourceCode" id="cb543"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb543-1"><a href="peak-calling.html#cb543-1"></a><span class="co"># combine a list of data.frames into one data.frame</span></span>
<span id="cb543-2"><a href="peak-calling.html#cb543-2"></a>annot_peaks_df =<span class="st"> </span>dplyr<span class="op">::</span><span class="kw">bind_rows</span>(annot_peaks_list)</span></code></pre></div>
<p>And visualize the results as bar plots. Resulting plot is in Figure <a href="peak-calling.html#fig:peak-annotation-plot">9.21</a>, which shows that the H3K36me3 peaks are
located preferentially in gene bodies, as expected, while the CTCF peaks are
found preferentially in introns.</p>
<div class="sourceCode" id="cb544"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb544-1"><a href="peak-calling.html#cb544-1"></a><span class="co"># plot the distribution of peaks in genomic features</span></span>
<span id="cb544-2"><a href="peak-calling.html#cb544-2"></a><span class="kw">ggplot</span>(<span class="dt">data =</span> annot_peaks_df, </span>
<span id="cb544-3"><a href="peak-calling.html#cb544-3"></a>       <span class="kw">aes</span>(</span>
<span id="cb544-4"><a href="peak-calling.html#cb544-4"></a>           <span class="dt">x    =</span> experiment, </span>
<span id="cb544-5"><a href="peak-calling.html#cb544-5"></a>           <span class="dt">y    =</span> frequency, </span>
<span id="cb544-6"><a href="peak-calling.html#cb544-6"></a>           <span class="dt">fill =</span> annotation</span>
<span id="cb544-7"><a href="peak-calling.html#cb544-7"></a>        )) <span class="op">+</span></span>
<span id="cb544-8"><a href="peak-calling.html#cb544-8"></a><span class="st">  </span><span class="kw">geom_bar</span>(<span class="dt">stat=</span><span class="st">&#39;identity&#39;</span>) <span class="op">+</span></span>
<span id="cb544-9"><a href="peak-calling.html#cb544-9"></a><span class="st">  </span><span class="kw">scale_fill_brewer</span>(<span class="dt">palette=</span><span class="st">&#39;Set2&#39;</span>) <span class="op">+</span></span>
<span id="cb544-10"><a href="peak-calling.html#cb544-10"></a><span class="st">  </span><span class="kw">theme_bw</span>()<span class="op">+</span></span>
<span id="cb544-11"><a href="peak-calling.html#cb544-11"></a><span class="st">  </span><span class="kw">theme</span>(</span>
<span id="cb544-12"><a href="peak-calling.html#cb544-12"></a>    <span class="dt">axis.text =</span> <span class="kw">element_text</span>(<span class="dt">size=</span><span class="dv">18</span>, <span class="dt">face=</span><span class="st">&#39;bold&#39;</span>),</span>
<span id="cb544-13"><a href="peak-calling.html#cb544-13"></a>    <span class="dt">axis.title =</span> <span class="kw">element_text</span>(<span class="dt">size=</span><span class="dv">14</span>,<span class="dt">face=</span><span class="st">&quot;bold&quot;</span>),</span>
<span id="cb544-14"><a href="peak-calling.html#cb544-14"></a>    <span class="dt">plot.title =</span> <span class="kw">element_text</span>(<span class="dt">hjust =</span> <span class="fl">0.5</span>))  <span class="op">+</span></span>
<span id="cb544-15"><a href="peak-calling.html#cb544-15"></a><span class="st">  </span><span class="kw">ggtitle</span>(<span class="st">&#39;Peak distribution in</span><span class="ch">\n</span><span class="st">genomic regions&#39;</span>) <span class="op">+</span></span>
<span id="cb544-16"><a href="peak-calling.html#cb544-16"></a><span class="st">  </span><span class="kw">xlab</span>(<span class="st">&#39;Experiment&#39;</span>) <span class="op">+</span></span>
<span id="cb544-17"><a href="peak-calling.html#cb544-17"></a><span class="st">  </span><span class="kw">ylab</span>(<span class="st">&#39;Frequency&#39;</span>)</span></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:peak-annotation-plot"></span>
<img src="09-chip-seq-analysis_files/figure-html/peak-annotation-plot-1.png" alt="Enrichment of transcription factor or histone modifications in functional genomic features." width="50%" />
<p class="caption">
FIGURE 9.21: Enrichment of transcription factor or histone modifications in functional genomic features.
</p>
</div>
</div>
</div>
<h3>References</h3>
<div id="refs" class="references">
<div id="ref-beck_2012">
<p>Beck, Brandl, Boelen, Unnikrishnan, Pimanda, and Wong. 2012. “Signal Analysis for Genome-Wide Maps of Histone Modifications Measured by ChIP-Seq.” <em>Bioinformatics</em> 28 (8): 1062–9. <a href="https://doi.org/10.1093/bioinformatics/bts085">https://doi.org/10.1093/bioinformatics/bts085</a>.</p>
</div>
<div id="ref-han_2012">
<p>Han, Tian, Pécot, Huang, Machiraju, and Huang. 2012. “A Signal Processing Approach for Enriched Region Detection in RNA Polymerase II ChIP-Seq Data.” <em>BMC Bioinformatics</em> 13 Suppl 2 (March): S2. <a href="https://doi.org/10.1186/1471-2105-13-S2-S2">https://doi.org/10.1186/1471-2105-13-S2-S2</a>.</p>
</div>
<div id="ref-helmuth_2016">
<p>Helmuth, Li, Arrigoni, et al. 2016. “normR: Regime Enrichment Calling for ChIP-Seq Data.” <em>bioRxiv</em>.</p>
</div>
<div id="ref-khan_2018">
<p>Khan, Fornes, Stigliani, et al. 2018. “JASPAR 2018: Update of the Open-Access Database of Transcription Factor Binding Profiles and Its Web Framework.” <em>Nucleic Acids Res</em> 46 (D1): D260–D266. <a href="https://doi.org/10.1093/nar/gkx1126">https://doi.org/10.1093/nar/gkx1126</a>.</p>
</div>
<div id="ref-laajala_2009">
<p>Laajala, Raghav, Tuomela, Lahesmaa, Aittokallio, and Elo. 2009. “A Practical Comparison of Methods for Detecting Transcription Factor Binding Sites in ChIP-Seq Experiments.” <em>BMC Genomics</em> 10 (December): 618. <a href="https://doi.org/10.1186/1471-2164-10-618">https://doi.org/10.1186/1471-2164-10-618</a>.</p>
</div>
<div id="ref-micsinai_2012">
<p>Micsinai, Parisi, Strino, Asp, Dynlacht, and Kluger. 2012. “Picking ChIP-Seq Peak Detectors for Analyzing Chromatin Modification Experiments.” <em>Nucleic Acids Res</em> 40 (9): e70. <a href="https://doi.org/10.1093/nar/gks048">https://doi.org/10.1093/nar/gks048</a>.</p>
</div>
<div id="ref-rashid_2011">
<p>Rashid, Giresi, Ibrahim, Sun, and Lieb. 2011. “ZINBA Integrates Local Covariates with DNA-Seq Data to Identify Broad and Narrow Regions of Enrichment, Even Within Amplified Genomic Regions.” <em>Genome Biol</em> 12 (7): R67. <a href="https://doi.org/10.1186/gb-2011-12-7-r67">https://doi.org/10.1186/gb-2011-12-7-r67</a>.</p>
</div>
<div id="ref-shao_2012">
<p>Shao, Zhang, Yuan, Orkin, and Waxman. 2012. “MAnorm: A Robust Model for Quantitative Comparison of ChIP-Seq Data Sets.” <em>Genome Biol</em> 13 (3): R16. <a href="https://doi.org/10.1186/gb-2012-13-3-r16">https://doi.org/10.1186/gb-2012-13-3-r16</a>.</p>
</div>
<div id="ref-song_2011">
<p>Song, and Smith. 2011. “Identifying Dispersed Epigenomic Domains from ChIP-Seq Data.” <em>Bioinformatics</em> 27 (6): 870–71. <a href="https://doi.org/10.1093/bioinformatics/btr030">https://doi.org/10.1093/bioinformatics/btr030</a>.</p>
</div>
<div id="ref-TFBSTools">
<p>Tan, and Lenhard. 2016. “TFBSTools: An R/Bioconductor Package for Transcription Factor Binding Site Analysis.” <em>Bioinformatics</em> 32 (10): 1555–6. <a href="https://doi.org/10.1093/bioinformatics/btw024">https://doi.org/10.1093/bioinformatics/btw024</a>.</p>
</div>
<div id="ref-wilbanks_2010">
<p>Wilbanks, and Facciotti. 2010. “Evaluation of Algorithm Performance in ChIP-Seq Peak Detection.” <em>PLoS ONE</em> 5 (7): e11471. <a href="https://doi.org/10.1371/journal.pone.0011471">https://doi.org/10.1371/journal.pone.0011471</a>.</p>
</div>
<div id="ref-xing_2012">
<p>Xing, Mo, Liao, and Zhang. 2012. “Genome-Wide Localization of Protein-DNA Binding and Histone Modification by a Bayesian Change-Point Method with ChIP-Seq Data.” <em>PLoS Comput Biol</em> 8 (7): e1002613. <a href="https://doi.org/10.1371/journal.pcbi.1002613">https://doi.org/10.1371/journal.pcbi.1002613</a>.</p>
</div>
<div id="ref-xu_2010">
<p>Xu, Handoko, Wei, Ye, Sheng, Wei, Lin, and Sung. 2010. “A Signal-Noise Model for Significance Analysis of ChIP-Seq with Negative Control.” <em>Bioinformatics</em> 26 (9): 1199–1204. <a href="https://doi.org/10.1093/bioinformatics/btq128">https://doi.org/10.1093/bioinformatics/btq128</a>.</p>
</div>
<div id="ref-zang_2009">
<p>Zang, Schones, Zeng, Cui, Zhao, and Peng. 2009. “A Clustering Approach for Identification of Enriched Domains from Histone Modification ChIP-Seq Data.” <em>Bioinformatics</em> 25 (15): 1952–8. <a href="https://doi.org/10.1093/bioinformatics/btp340">https://doi.org/10.1093/bioinformatics/btp340</a>.</p>
</div>
<div id="ref-zhang_2008">
<p>Zhang, Liu, Meyer, et al. 2008. “Model-Based Analysis of ChIP-Seq (MACS).” <em>Genome Biol</em> 9 (9): R137. <a href="https://doi.org/10.1186/gb-2008-9-9-r137">https://doi.org/10.1186/gb-2008-9-9-r137</a>.</p>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="chip-quality-control.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="motif-discovery.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": true,
"facebook": false,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/compgenomr/book/edit/master/09-chip-seq-analysis.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": null,
"toc": {
"collapse": "none"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
